<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>gnomad.utils.vcf &mdash; gnomad master documentation</title>
      <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../../_static/theme_overrides.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../../_static/jquery.js"></script>
        <script src="../../_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
        <script src="../../_static/doctools.js"></script>
        <script src="../../_static/sphinx_highlight.js"></script>
    <script src="../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="gnomad.utils.vep" href="vep.html" />
    <link rel="prev" title="gnomad.utils.transcript_annotation" href="transcript_annotation.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../index.html" class="icon icon-home">
            gnomad
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../getting_started.html">Getting Started</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../examples/index.html">Examples</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../index.html">API Reference</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="index.html">utils</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="annotations.html">annotations</a></li>
<li class="toctree-l3"><a class="reference internal" href="constraint.html">constraint</a></li>
<li class="toctree-l3"><a class="reference internal" href="file_utils.html">file_utils</a></li>
<li class="toctree-l3"><a class="reference internal" href="filtering.html">filtering</a></li>
<li class="toctree-l3"><a class="reference internal" href="gen_stats.html">gen_stats</a></li>
<li class="toctree-l3"><a class="reference internal" href="intervals.html">intervals</a></li>
<li class="toctree-l3"><a class="reference internal" href="liftover.html">liftover</a></li>
<li class="toctree-l3"><a class="reference internal" href="parse.html">parse</a></li>
<li class="toctree-l3"><a class="reference internal" href="plotting.html">plotting</a></li>
<li class="toctree-l3"><a class="reference internal" href="reference_genome.html">reference_genome</a></li>
<li class="toctree-l3"><a class="reference internal" href="release.html">release</a></li>
<li class="toctree-l3"><a class="reference internal" href="slack.html">slack</a></li>
<li class="toctree-l3"><a class="reference internal" href="sparse_mt.html">sparse_mt</a></li>
<li class="toctree-l3"><a class="reference internal" href="transcript_annotation.html">transcript_annotation</a></li>
<li class="toctree-l3 current"><a class="current reference internal" href="#">vcf</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#gnomad.utils.vcf.SORT_ORDER"><code class="docutils literal notranslate"><span class="pre">SORT_ORDER</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#gnomad.utils.vcf.GROUPS"><code class="docutils literal notranslate"><span class="pre">GROUPS</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#gnomad.utils.vcf.HISTS"><code class="docutils literal notranslate"><span class="pre">HISTS</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#gnomad.utils.vcf.FAF_GEN_ANC_GROUPS"><code class="docutils literal notranslate"><span class="pre">FAF_GEN_ANC_GROUPS</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#gnomad.utils.vcf.SEXES"><code class="docutils literal notranslate"><span class="pre">SEXES</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#gnomad.utils.vcf.AS_FIELDS"><code class="docutils literal notranslate"><span class="pre">AS_FIELDS</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#gnomad.utils.vcf.SITE_FIELDS"><code class="docutils literal notranslate"><span class="pre">SITE_FIELDS</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#gnomad.utils.vcf.ALLELE_TYPE_FIELDS"><code class="docutils literal notranslate"><span class="pre">ALLELE_TYPE_FIELDS</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#gnomad.utils.vcf.REGION_FLAG_FIELDS"><code class="docutils literal notranslate"><span class="pre">REGION_FLAG_FIELDS</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#gnomad.utils.vcf.JOINT_REGION_FLAG_FIELDS"><code class="docutils literal notranslate"><span class="pre">JOINT_REGION_FLAG_FIELDS</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#gnomad.utils.vcf.RF_FIELDS"><code class="docutils literal notranslate"><span class="pre">RF_FIELDS</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#gnomad.utils.vcf.AS_VQSR_FIELDS"><code class="docutils literal notranslate"><span class="pre">AS_VQSR_FIELDS</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#gnomad.utils.vcf.VQSR_FIELDS"><code class="docutils literal notranslate"><span class="pre">VQSR_FIELDS</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#gnomad.utils.vcf.INFO_DICT"><code class="docutils literal notranslate"><span class="pre">INFO_DICT</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#gnomad.utils.vcf.IN_SILICO_ANNOTATIONS_INFO_DICT"><code class="docutils literal notranslate"><span class="pre">IN_SILICO_ANNOTATIONS_INFO_DICT</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#gnomad.utils.vcf.VRS_FIELDS_DICT"><code class="docutils literal notranslate"><span class="pre">VRS_FIELDS_DICT</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#gnomad.utils.vcf.ENTRIES"><code class="docutils literal notranslate"><span class="pre">ENTRIES</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#gnomad.utils.vcf.SPARSE_ENTRIES"><code class="docutils literal notranslate"><span class="pre">SPARSE_ENTRIES</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#gnomad.utils.vcf.FORMAT_DICT"><code class="docutils literal notranslate"><span class="pre">FORMAT_DICT</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#gnomad.utils.vcf.adjust_vcf_incompatible_types"><code class="docutils literal notranslate"><span class="pre">adjust_vcf_incompatible_types()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#gnomad.utils.vcf.make_label_combos"><code class="docutils literal notranslate"><span class="pre">make_label_combos()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#gnomad.utils.vcf.index_globals"><code class="docutils literal notranslate"><span class="pre">index_globals()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#gnomad.utils.vcf.make_combo_header_text"><code class="docutils literal notranslate"><span class="pre">make_combo_header_text()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#gnomad.utils.vcf.create_label_groups"><code class="docutils literal notranslate"><span class="pre">create_label_groups()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#gnomad.utils.vcf.make_info_dict"><code class="docutils literal notranslate"><span class="pre">make_info_dict()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#gnomad.utils.vcf.add_as_info_dict"><code class="docutils literal notranslate"><span class="pre">add_as_info_dict()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#gnomad.utils.vcf.make_vcf_filter_dict"><code class="docutils literal notranslate"><span class="pre">make_vcf_filter_dict()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#gnomad.utils.vcf.make_hist_bin_edges_expr"><code class="docutils literal notranslate"><span class="pre">make_hist_bin_edges_expr()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#gnomad.utils.vcf.make_hist_dict"><code class="docutils literal notranslate"><span class="pre">make_hist_dict()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#gnomad.utils.vcf.set_xx_y_metrics_to_na"><code class="docutils literal notranslate"><span class="pre">set_xx_y_metrics_to_na()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#gnomad.utils.vcf.build_vcf_export_reference"><code class="docutils literal notranslate"><span class="pre">build_vcf_export_reference()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#gnomad.utils.vcf.rekey_new_reference"><code class="docutils literal notranslate"><span class="pre">rekey_new_reference()</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="vep.html">vep</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../resources/index.html">resources</a></li>
<li class="toctree-l2"><a class="reference internal" href="../variant_qc/index.html">variant_qc</a></li>
<li class="toctree-l2"><a class="reference internal" href="../sample_qc/index.html">sample_qc</a></li>
<li class="toctree-l2"><a class="reference internal" href="../assessment/index.html">assessment</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../resource_sources.html">Resource Sources</a></li>
<li class="toctree-l1"><a class="reference external" href="https://github.com/broadinstitute/gnomad_methods/releases">Change Log</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">gnomad</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../index.html">gnomad</a></li>
          <li class="breadcrumb-item"><a href="index.html">gnomad.utils</a></li>
      <li class="breadcrumb-item active">gnomad.utils.vcf</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../../_sources/api_reference/utils/vcf.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="gnomad-utils-vcf">
<h1>gnomad.utils.vcf<a class="headerlink" href="#gnomad-utils-vcf" title="Permalink to this heading"></a></h1>
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#gnomad.utils.vcf.SORT_ORDER" title="gnomad.utils.vcf.SORT_ORDER"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.vcf.SORT_ORDER</span></code></a></p></td>
<td><p>Order to sort subgroupings during VCF export.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#gnomad.utils.vcf.GROUPS" title="gnomad.utils.vcf.GROUPS"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.vcf.GROUPS</span></code></a></p></td>
<td><p>Group names used to generate labels for high quality genotypes and all raw genotypes.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#gnomad.utils.vcf.HISTS" title="gnomad.utils.vcf.HISTS"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.vcf.HISTS</span></code></a></p></td>
<td><p>Quality histograms used in VCF export.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#gnomad.utils.vcf.FAF_GEN_ANC_GROUPS" title="gnomad.utils.vcf.FAF_GEN_ANC_GROUPS"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.vcf.FAF_GEN_ANC_GROUPS</span></code></a></p></td>
<td><p>Genetic ancestry groups that are included in filtering allele frequency (faf) calculations.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#gnomad.utils.vcf.SEXES" title="gnomad.utils.vcf.SEXES"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.vcf.SEXES</span></code></a></p></td>
<td><p>Sample sexes used in VCF export.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#gnomad.utils.vcf.AS_FIELDS" title="gnomad.utils.vcf.AS_FIELDS"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.vcf.AS_FIELDS</span></code></a></p></td>
<td><p>Allele-specific variant annotations.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#gnomad.utils.vcf.SITE_FIELDS" title="gnomad.utils.vcf.SITE_FIELDS"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.vcf.SITE_FIELDS</span></code></a></p></td>
<td><p>Site level variant annotations.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#gnomad.utils.vcf.ALLELE_TYPE_FIELDS" title="gnomad.utils.vcf.ALLELE_TYPE_FIELDS"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.vcf.ALLELE_TYPE_FIELDS</span></code></a></p></td>
<td><p>Allele-type annotations.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#gnomad.utils.vcf.REGION_FLAG_FIELDS" title="gnomad.utils.vcf.REGION_FLAG_FIELDS"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.vcf.REGION_FLAG_FIELDS</span></code></a></p></td>
<td><p>Annotations about variant region type.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#gnomad.utils.vcf.JOINT_REGION_FLAG_FIELDS" title="gnomad.utils.vcf.JOINT_REGION_FLAG_FIELDS"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.vcf.JOINT_REGION_FLAG_FIELDS</span></code></a></p></td>
<td><p>Annotations about variant region type that are specifically created for joint dataset of exomes and genomes from gnomAD v4.1.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#gnomad.utils.vcf.RF_FIELDS" title="gnomad.utils.vcf.RF_FIELDS"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.vcf.RF_FIELDS</span></code></a></p></td>
<td><p>Annotations specific to the variant QC using a random forest model.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#gnomad.utils.vcf.AS_VQSR_FIELDS" title="gnomad.utils.vcf.AS_VQSR_FIELDS"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.vcf.AS_VQSR_FIELDS</span></code></a></p></td>
<td><p>Allele-specific VQSR annotations.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#gnomad.utils.vcf.VQSR_FIELDS" title="gnomad.utils.vcf.VQSR_FIELDS"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.vcf.VQSR_FIELDS</span></code></a></p></td>
<td><p>Annotations specific to VQSR.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#gnomad.utils.vcf.INFO_DICT" title="gnomad.utils.vcf.INFO_DICT"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.vcf.INFO_DICT</span></code></a></p></td>
<td><p>Dictionary used during VCF export to export row (variant) annotations.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#gnomad.utils.vcf.IN_SILICO_ANNOTATIONS_INFO_DICT" title="gnomad.utils.vcf.IN_SILICO_ANNOTATIONS_INFO_DICT"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.vcf.IN_SILICO_ANNOTATIONS_INFO_DICT</span></code></a></p></td>
<td><p>Dictionary with in silico score descriptions to include in the VCF INFO header.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#gnomad.utils.vcf.VRS_FIELDS_DICT" title="gnomad.utils.vcf.VRS_FIELDS_DICT"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.vcf.VRS_FIELDS_DICT</span></code></a></p></td>
<td><p>Dictionary with VRS annotations to include in the VCF INFO field and VCF header.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#gnomad.utils.vcf.ENTRIES" title="gnomad.utils.vcf.ENTRIES"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.vcf.ENTRIES</span></code></a></p></td>
<td><p>Densified entries to be selected during VCF export.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#gnomad.utils.vcf.SPARSE_ENTRIES" title="gnomad.utils.vcf.SPARSE_ENTRIES"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.vcf.SPARSE_ENTRIES</span></code></a></p></td>
<td><p>Sparse entries to be selected and densified during VCF export.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#gnomad.utils.vcf.FORMAT_DICT" title="gnomad.utils.vcf.FORMAT_DICT"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.vcf.FORMAT_DICT</span></code></a></p></td>
<td><p>Dictionary used during VCF export to export MatrixTable entries.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#gnomad.utils.vcf.adjust_vcf_incompatible_types" title="gnomad.utils.vcf.adjust_vcf_incompatible_types"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.vcf.adjust_vcf_incompatible_types</span></code></a>(ht)</p></td>
<td><p>Create a Table ready for vcf export.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#gnomad.utils.vcf.make_label_combos" title="gnomad.utils.vcf.make_label_combos"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.vcf.make_label_combos</span></code></a>(label_groups)</p></td>
<td><p>Make combinations of all possible labels for a supplied dictionary of label groups.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#gnomad.utils.vcf.index_globals" title="gnomad.utils.vcf.index_globals"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.vcf.index_globals</span></code></a>(...[, ...])</p></td>
<td><p>Create a dictionary keyed by the specified label groupings with values describing the corresponding index of each grouping entry in the meta_array annotation.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#gnomad.utils.vcf.make_combo_header_text" title="gnomad.utils.vcf.make_combo_header_text"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.vcf.make_combo_header_text</span></code></a>(...)</p></td>
<td><p>Programmatically generate text to populate the VCF header description for a given variant annotation with specific groupings and subset.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#gnomad.utils.vcf.create_label_groups" title="gnomad.utils.vcf.create_label_groups"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.vcf.create_label_groups</span></code></a>(gen_ancs)</p></td>
<td><p>Generate a list of label group dictionaries needed to populate info dictionary.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#gnomad.utils.vcf.make_info_dict" title="gnomad.utils.vcf.make_info_dict"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.vcf.make_info_dict</span></code></a>([prefix, ...])</p></td>
<td><p>Generate dictionary of Number and Description attributes of VCF INFO fields.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#gnomad.utils.vcf.add_as_info_dict" title="gnomad.utils.vcf.add_as_info_dict"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.vcf.add_as_info_dict</span></code></a>([...])</p></td>
<td><p>Update info dictionary with allele-specific terms and their descriptions.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#gnomad.utils.vcf.make_vcf_filter_dict" title="gnomad.utils.vcf.make_vcf_filter_dict"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.vcf.make_vcf_filter_dict</span></code></a>([...])</p></td>
<td><p>Generate dictionary of Number and Description attributes to be used in the VCF header, specifically for FILTER annotations.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#gnomad.utils.vcf.make_hist_bin_edges_expr" title="gnomad.utils.vcf.make_hist_bin_edges_expr"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.vcf.make_hist_bin_edges_expr</span></code></a>(ht)</p></td>
<td><p>Create dictionaries containing variant histogram annotations and their associated bin edges, formatted into a string separated by pipe delimiters.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#gnomad.utils.vcf.make_hist_dict" title="gnomad.utils.vcf.make_hist_dict"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.vcf.make_hist_dict</span></code></a>(bin_edges, adj)</p></td>
<td><p>Generate dictionary of Number and Description attributes to be used in the VCF header, specifically for histogram annotations.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#gnomad.utils.vcf.set_xx_y_metrics_to_na" title="gnomad.utils.vcf.set_xx_y_metrics_to_na"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.vcf.set_xx_y_metrics_to_na</span></code></a>(t)</p></td>
<td><p>Set AC, AN, and nhomalt chrY variant annotations for XX samples to NA (instead of 0).</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#gnomad.utils.vcf.build_vcf_export_reference" title="gnomad.utils.vcf.build_vcf_export_reference"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.vcf.build_vcf_export_reference</span></code></a>(name)</p></td>
<td><p>Create export reference based on reference genome defined by <cite>build</cite>.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#gnomad.utils.vcf.rekey_new_reference" title="gnomad.utils.vcf.rekey_new_reference"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.vcf.rekey_new_reference</span></code></a>(t, ...)</p></td>
<td><p>Re-key Table or MatrixTable with a new reference genome.</p></td>
</tr>
</tbody>
</table>
<span class="target" id="module-gnomad.utils.vcf"></span><dl class="py data">
<dt class="sig sig-object py" id="gnomad.utils.vcf.SORT_ORDER">
<span class="sig-prename descclassname"><span class="pre">gnomad.utils.vcf.</span></span><span class="sig-name descname"><span class="pre">SORT_ORDER</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">['subset',</span> <span class="pre">'downsampling',</span> <span class="pre">'grpmax',</span> <span class="pre">'popmax',</span> <span class="pre">'gen_anc',</span> <span class="pre">'pop',</span> <span class="pre">'subgrp',</span> <span class="pre">'subpop',</span> <span class="pre">'sex',</span> <span class="pre">'group']</span></em><a class="headerlink" href="#gnomad.utils.vcf.SORT_ORDER" title="Permalink to this definition"></a></dt>
<dd><p>Order to sort subgroupings during VCF export.
Ensures that INFO labels in VCF are in desired order (e.g., raw_AC_afr_XX).</p>
</dd></dl>

<dl class="py data">
<dt class="sig sig-object py" id="gnomad.utils.vcf.GROUPS">
<span class="sig-prename descclassname"><span class="pre">gnomad.utils.vcf.</span></span><span class="sig-name descname"><span class="pre">GROUPS</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">['adj',</span> <span class="pre">'raw']</span></em><a class="headerlink" href="#gnomad.utils.vcf.GROUPS" title="Permalink to this definition"></a></dt>
<dd><p>Group names used to generate labels for high quality genotypes and all raw genotypes. Used in VCF export.</p>
</dd></dl>

<dl class="py data">
<dt class="sig sig-object py" id="gnomad.utils.vcf.HISTS">
<span class="sig-prename descclassname"><span class="pre">gnomad.utils.vcf.</span></span><span class="sig-name descname"><span class="pre">HISTS</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">['gq_hist_alt',</span> <span class="pre">'gq_hist_all',</span> <span class="pre">'dp_hist_alt',</span> <span class="pre">'dp_hist_all',</span> <span class="pre">'ab_hist_alt']</span></em><a class="headerlink" href="#gnomad.utils.vcf.HISTS" title="Permalink to this definition"></a></dt>
<dd><p>Quality histograms used in VCF export.</p>
</dd></dl>

<dl class="py data">
<dt class="sig sig-object py" id="gnomad.utils.vcf.FAF_GEN_ANC_GROUPS">
<span class="sig-prename descclassname"><span class="pre">gnomad.utils.vcf.</span></span><span class="sig-name descname"><span class="pre">FAF_GEN_ANC_GROUPS</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'v3':</span> <span class="pre">['afr',</span> <span class="pre">'amr',</span> <span class="pre">'eas',</span> <span class="pre">'nfe',</span> <span class="pre">'sas'],</span> <span class="pre">'v4':</span> <span class="pre">['afr',</span> <span class="pre">'amr',</span> <span class="pre">'eas',</span> <span class="pre">'mid',</span> <span class="pre">'nfe',</span> <span class="pre">'sas']}</span></em><a class="headerlink" href="#gnomad.utils.vcf.FAF_GEN_ANC_GROUPS" title="Permalink to this definition"></a></dt>
<dd><p>Genetic ancestry groups that are included in filtering allele frequency (faf) calculations. Used in VCF export.</p>
</dd></dl>

<dl class="py data">
<dt class="sig sig-object py" id="gnomad.utils.vcf.SEXES">
<span class="sig-prename descclassname"><span class="pre">gnomad.utils.vcf.</span></span><span class="sig-name descname"><span class="pre">SEXES</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">['XX',</span> <span class="pre">'XY']</span></em><a class="headerlink" href="#gnomad.utils.vcf.SEXES" title="Permalink to this definition"></a></dt>
<dd><p>Sample sexes used in VCF export.</p>
<p>Used to stratify frequency annotations (AC, AN, AF) for each sex.
Note that sample sexes in gnomAD v3 and earlier were ‘male’ and ‘female’.</p>
</dd></dl>

<dl class="py data">
<dt class="sig sig-object py" id="gnomad.utils.vcf.AS_FIELDS">
<span class="sig-prename descclassname"><span class="pre">gnomad.utils.vcf.</span></span><span class="sig-name descname"><span class="pre">AS_FIELDS</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">['AS_FS',</span> <span class="pre">'AS_MQ',</span> <span class="pre">'AS_MQRankSum',</span> <span class="pre">'AS_pab_max',</span> <span class="pre">'AS_QUALapprox',</span> <span class="pre">'AS_QD',</span> <span class="pre">'AS_ReadPosRankSum',</span> <span class="pre">'AS_SB_TABLE',</span> <span class="pre">'AS_SOR',</span> <span class="pre">'AS_VarDP',</span> <span class="pre">'InbreedingCoeff']</span></em><a class="headerlink" href="#gnomad.utils.vcf.AS_FIELDS" title="Permalink to this definition"></a></dt>
<dd><p>Allele-specific variant annotations.</p>
</dd></dl>

<dl class="py data">
<dt class="sig sig-object py" id="gnomad.utils.vcf.SITE_FIELDS">
<span class="sig-prename descclassname"><span class="pre">gnomad.utils.vcf.</span></span><span class="sig-name descname"><span class="pre">SITE_FIELDS</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">['FS',</span> <span class="pre">'MQ',</span> <span class="pre">'MQRankSum',</span> <span class="pre">'QUALapprox',</span> <span class="pre">'QD',</span> <span class="pre">'ReadPosRankSum',</span> <span class="pre">'SB',</span> <span class="pre">'SOR',</span> <span class="pre">'VarDP']</span></em><a class="headerlink" href="#gnomad.utils.vcf.SITE_FIELDS" title="Permalink to this definition"></a></dt>
<dd><p>Site level variant annotations.</p>
</dd></dl>

<dl class="py data">
<dt class="sig sig-object py" id="gnomad.utils.vcf.ALLELE_TYPE_FIELDS">
<span class="sig-prename descclassname"><span class="pre">gnomad.utils.vcf.</span></span><span class="sig-name descname"><span class="pre">ALLELE_TYPE_FIELDS</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">['allele_type',</span> <span class="pre">'has_star',</span> <span class="pre">'n_alt_alleles',</span> <span class="pre">'original_alleles',</span> <span class="pre">'variant_type',</span> <span class="pre">'was_mixed']</span></em><a class="headerlink" href="#gnomad.utils.vcf.ALLELE_TYPE_FIELDS" title="Permalink to this definition"></a></dt>
<dd><p>Allele-type annotations.</p>
</dd></dl>

<dl class="py data">
<dt class="sig sig-object py" id="gnomad.utils.vcf.REGION_FLAG_FIELDS">
<span class="sig-prename descclassname"><span class="pre">gnomad.utils.vcf.</span></span><span class="sig-name descname"><span class="pre">REGION_FLAG_FIELDS</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">['decoy',</span> <span class="pre">'lcr',</span> <span class="pre">'nonpar',</span> <span class="pre">'non_par',</span> <span class="pre">'segdup']</span></em><a class="headerlink" href="#gnomad.utils.vcf.REGION_FLAG_FIELDS" title="Permalink to this definition"></a></dt>
<dd><p>Annotations about variant region type.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>decoy resource files do not currently exist for GRCh38/hg38.</p>
</div>
</dd></dl>

<dl class="py data">
<dt class="sig sig-object py" id="gnomad.utils.vcf.JOINT_REGION_FLAG_FIELDS">
<span class="sig-prename descclassname"><span class="pre">gnomad.utils.vcf.</span></span><span class="sig-name descname"><span class="pre">JOINT_REGION_FLAG_FIELDS</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">['fail_interval_qc',</span> <span class="pre">'outside_broad_capture_region',</span> <span class="pre">'outside_ukb_capture_region',</span> <span class="pre">'outside_broad_calling_region',</span> <span class="pre">'outside_ukb_calling_region',</span> <span class="pre">'not_called_in_exomes',</span> <span class="pre">'not_called_in_genomes']</span></em><a class="headerlink" href="#gnomad.utils.vcf.JOINT_REGION_FLAG_FIELDS" title="Permalink to this definition"></a></dt>
<dd><p>Annotations about variant region type that are specifically created for joint dataset of exomes and genomes from gnomAD v4.1.</p>
</dd></dl>

<dl class="py data">
<dt class="sig sig-object py" id="gnomad.utils.vcf.RF_FIELDS">
<span class="sig-prename descclassname"><span class="pre">gnomad.utils.vcf.</span></span><span class="sig-name descname"><span class="pre">RF_FIELDS</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">['rf_positive_label',</span> <span class="pre">'rf_negative_label',</span> <span class="pre">'rf_label',</span> <span class="pre">'rf_train',</span> <span class="pre">'rf_tp_probability']</span></em><a class="headerlink" href="#gnomad.utils.vcf.RF_FIELDS" title="Permalink to this definition"></a></dt>
<dd><p>Annotations specific to the variant QC using a random forest model.</p>
</dd></dl>

<dl class="py data">
<dt class="sig sig-object py" id="gnomad.utils.vcf.AS_VQSR_FIELDS">
<span class="sig-prename descclassname"><span class="pre">gnomad.utils.vcf.</span></span><span class="sig-name descname"><span class="pre">AS_VQSR_FIELDS</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">['AS_culprit',</span> <span class="pre">'AS_VQSLOD']</span></em><a class="headerlink" href="#gnomad.utils.vcf.AS_VQSR_FIELDS" title="Permalink to this definition"></a></dt>
<dd><p>Allele-specific VQSR annotations.</p>
</dd></dl>

<dl class="py data">
<dt class="sig sig-object py" id="gnomad.utils.vcf.VQSR_FIELDS">
<span class="sig-prename descclassname"><span class="pre">gnomad.utils.vcf.</span></span><span class="sig-name descname"><span class="pre">VQSR_FIELDS</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">['AS_culprit',</span> <span class="pre">'AS_VQSLOD',</span> <span class="pre">'NEGATIVE_TRAIN_SITE',</span> <span class="pre">'POSITIVE_TRAIN_SITE']</span></em><a class="headerlink" href="#gnomad.utils.vcf.VQSR_FIELDS" title="Permalink to this definition"></a></dt>
<dd><p>Annotations specific to VQSR.</p>
</dd></dl>

<dl class="py data">
<dt class="sig sig-object py" id="gnomad.utils.vcf.INFO_DICT">
<span class="sig-prename descclassname"><span class="pre">gnomad.utils.vcf.</span></span><span class="sig-name descname"><span class="pre">INFO_DICT</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'AS_SB_TABLE':</span> <span class="pre">{'Description':</span> <span class="pre">'Allele-specific</span> <span class="pre">forward/reverse</span> <span class="pre">read</span> <span class="pre">counts</span> <span class="pre">for</span> <span class="pre">strand</span> <span class="pre">bias</span> <span class="pre">tests',</span> <span class="pre">'Number':</span> <span class="pre">'.'},</span> <span class="pre">'AS_pab_max':</span> <span class="pre">{'Description':</span> <span class="pre">'Maximum</span> <span class="pre">p-value</span> <span class="pre">over</span> <span class="pre">callset</span> <span class="pre">for</span> <span class="pre">binomial</span> <span class="pre">test</span> <span class="pre">of</span> <span class="pre">observed</span> <span class="pre">allele</span> <span class="pre">balance</span> <span class="pre">for</span> <span class="pre">a</span> <span class="pre">heterozygous</span> <span class="pre">genotype,</span> <span class="pre">given</span> <span class="pre">expectation</span> <span class="pre">of</span> <span class="pre">0.5',</span> <span class="pre">'Number':</span> <span class="pre">'A'},</span> <span class="pre">'BaseQRankSum':</span> <span class="pre">{'Description':</span> <span class="pre">'Z-score</span> <span class="pre">from</span> <span class="pre">Wilcoxon</span> <span class="pre">rank</span> <span class="pre">sum</span> <span class="pre">test</span> <span class="pre">of</span> <span class="pre">alternate</span> <span class="pre">vs.</span> <span class="pre">reference</span> <span class="pre">base</span> <span class="pre">qualities'},</span> <span class="pre">'FS':</span> <span class="pre">{'Description':</span> <span class="pre">&quot;Phred-scaled</span> <span class="pre">p-value</span> <span class="pre">of</span> <span class="pre">Fisher's</span> <span class="pre">exact</span> <span class="pre">test</span> <span class="pre">for</span> <span class="pre">strand</span> <span class="pre">bias&quot;},</span> <span class="pre">'InbreedingCoeff':</span> <span class="pre">{'Description':</span> <span class="pre">'Inbreeding</span> <span class="pre">coefficient,</span> <span class="pre">the</span> <span class="pre">excess</span> <span class="pre">heterozygosity</span> <span class="pre">at</span> <span class="pre">a</span> <span class="pre">variant</span> <span class="pre">site,</span> <span class="pre">computed</span> <span class="pre">as</span> <span class="pre">1</span> <span class="pre">-</span> <span class="pre">(the</span> <span class="pre">number</span> <span class="pre">of</span> <span class="pre">heterozygous</span> <span class="pre">genotypes)/(the</span> <span class="pre">number</span> <span class="pre">of</span> <span class="pre">heterozygous</span> <span class="pre">genotypes</span> <span class="pre">expected</span> <span class="pre">under</span> <span class="pre">Hardy-Weinberg</span> <span class="pre">equilibrium)',</span> <span class="pre">'Number':</span> <span class="pre">'A'},</span> <span class="pre">'MQ':</span> <span class="pre">{'Description':</span> <span class="pre">'Root</span> <span class="pre">mean</span> <span class="pre">square</span> <span class="pre">of</span> <span class="pre">the</span> <span class="pre">mapping</span> <span class="pre">quality</span> <span class="pre">of</span> <span class="pre">reads</span> <span class="pre">across</span> <span class="pre">all</span> <span class="pre">samples'},</span> <span class="pre">'MQRankSum':</span> <span class="pre">{'Description':</span> <span class="pre">'Z-score</span> <span class="pre">from</span> <span class="pre">Wilcoxon</span> <span class="pre">rank</span> <span class="pre">sum</span> <span class="pre">test</span> <span class="pre">of</span> <span class="pre">alternate</span> <span class="pre">vs.</span> <span class="pre">reference</span> <span class="pre">read</span> <span class="pre">mapping</span> <span class="pre">qualities'},</span> <span class="pre">'NEGATIVE_TRAIN_SITE':</span> <span class="pre">{'Description':</span> <span class="pre">'Variant</span> <span class="pre">was</span> <span class="pre">used</span> <span class="pre">to</span> <span class="pre">build</span> <span class="pre">the</span> <span class="pre">negative</span> <span class="pre">training</span> <span class="pre">set</span> <span class="pre">of</span> <span class="pre">low-quality</span> <span class="pre">variants</span> <span class="pre">for</span> <span class="pre">VQSR'},</span> <span class="pre">'POSITIVE_TRAIN_SITE':</span> <span class="pre">{'Description':</span> <span class="pre">'Variant</span> <span class="pre">was</span> <span class="pre">used</span> <span class="pre">to</span> <span class="pre">build</span> <span class="pre">the</span> <span class="pre">positive</span> <span class="pre">training</span> <span class="pre">set</span> <span class="pre">of</span> <span class="pre">high-quality</span> <span class="pre">variants</span> <span class="pre">for</span> <span class="pre">VQSR'},</span> <span class="pre">'QD':</span> <span class="pre">{'Description':</span> <span class="pre">'Variant</span> <span class="pre">call</span> <span class="pre">confidence</span> <span class="pre">normalized</span> <span class="pre">by</span> <span class="pre">depth</span> <span class="pre">of</span> <span class="pre">sample</span> <span class="pre">reads</span> <span class="pre">supporting</span> <span class="pre">a</span> <span class="pre">variant'},</span> <span class="pre">'QUALapprox':</span> <span class="pre">{'Description':</span> <span class="pre">'Sum</span> <span class="pre">of</span> <span class="pre">PL[0]</span> <span class="pre">values;</span> <span class="pre">used</span> <span class="pre">to</span> <span class="pre">approximate</span> <span class="pre">the</span> <span class="pre">QUAL</span> <span class="pre">score',</span> <span class="pre">'Number':</span> <span class="pre">'1'},</span> <span class="pre">'ReadPosRankSum':</span> <span class="pre">{'Description':</span> <span class="pre">'Z-score</span> <span class="pre">from</span> <span class="pre">Wilcoxon</span> <span class="pre">rank</span> <span class="pre">sum</span> <span class="pre">test</span> <span class="pre">of</span> <span class="pre">alternate</span> <span class="pre">vs.</span> <span class="pre">reference</span> <span class="pre">read</span> <span class="pre">position</span> <span class="pre">bias'},</span> <span class="pre">'SOR':</span> <span class="pre">{'Description':</span> <span class="pre">'Strand</span> <span class="pre">bias</span> <span class="pre">estimated</span> <span class="pre">by</span> <span class="pre">the</span> <span class="pre">symmetric</span> <span class="pre">odds</span> <span class="pre">ratio</span> <span class="pre">test'},</span> <span class="pre">'VQSLOD':</span> <span class="pre">{'Description':</span> <span class="pre">'Log-odds</span> <span class="pre">ratio</span> <span class="pre">of</span> <span class="pre">being</span> <span class="pre">a</span> <span class="pre">true</span> <span class="pre">variant</span> <span class="pre">versus</span> <span class="pre">being</span> <span class="pre">a</span> <span class="pre">false</span> <span class="pre">positive</span> <span class="pre">under</span> <span class="pre">the</span> <span class="pre">trained</span> <span class="pre">VQSR</span> <span class="pre">Gaussian</span> <span class="pre">mixture</span> <span class="pre">model'},</span> <span class="pre">'VarDP':</span> <span class="pre">{'Description':</span> <span class="pre">'Depth</span> <span class="pre">over</span> <span class="pre">variant</span> <span class="pre">genotypes</span> <span class="pre">(does</span> <span class="pre">not</span> <span class="pre">include</span> <span class="pre">depth</span> <span class="pre">of</span> <span class="pre">reference</span> <span class="pre">samples)'},</span> <span class="pre">'allele_type':</span> <span class="pre">{'Description':</span> <span class="pre">'Allele</span> <span class="pre">type</span> <span class="pre">(snv,</span> <span class="pre">insertion,</span> <span class="pre">deletion,</span> <span class="pre">or</span> <span class="pre">mixed)'},</span> <span class="pre">'culprit':</span> <span class="pre">{'Description':</span> <span class="pre">'Worst-performing</span> <span class="pre">annotation</span> <span class="pre">in</span> <span class="pre">the</span> <span class="pre">VQSR</span> <span class="pre">Gaussian</span> <span class="pre">mixture</span> <span class="pre">model'},</span> <span class="pre">'decoy':</span> <span class="pre">{'Description':</span> <span class="pre">'Variant</span> <span class="pre">falls</span> <span class="pre">within</span> <span class="pre">a</span> <span class="pre">reference</span> <span class="pre">decoy</span> <span class="pre">region'},</span> <span class="pre">'fail_interval_qc':</span> <span class="pre">{'Description':</span> <span class="pre">'Less</span> <span class="pre">than</span> <span class="pre">85</span> <span class="pre">percent</span> <span class="pre">of</span> <span class="pre">samples</span> <span class="pre">meet</span> <span class="pre">20X</span> <span class="pre">coverage</span> <span class="pre">if</span> <span class="pre">variant</span> <span class="pre">is</span> <span class="pre">in</span> <span class="pre">autosomal</span> <span class="pre">or</span> <span class="pre">PAR</span> <span class="pre">regions</span> <span class="pre">or</span> <span class="pre">10X</span> <span class="pre">coverage</span> <span class="pre">for</span> <span class="pre">non-PAR</span> <span class="pre">regions</span> <span class="pre">of</span> <span class="pre">chromosomes</span> <span class="pre">X</span> <span class="pre">and</span> <span class="pre">Y.'},</span> <span class="pre">'has_star':</span> <span class="pre">{'Description':</span> <span class="pre">'Variant</span> <span class="pre">locus</span> <span class="pre">coincides</span> <span class="pre">with</span> <span class="pre">a</span> <span class="pre">spanning</span> <span class="pre">deletion</span> <span class="pre">(represented</span> <span class="pre">by</span> <span class="pre">a</span> <span class="pre">star)</span> <span class="pre">observed</span> <span class="pre">elsewhere</span> <span class="pre">in</span> <span class="pre">the</span> <span class="pre">callset'},</span> <span class="pre">'inbreeding_coeff':</span> <span class="pre">{'Description':</span> <span class="pre">'Inbreeding</span> <span class="pre">coefficient,</span> <span class="pre">the</span> <span class="pre">excess</span> <span class="pre">heterozygosity</span> <span class="pre">at</span> <span class="pre">a</span> <span class="pre">variant</span> <span class="pre">site,</span> <span class="pre">computed</span> <span class="pre">as</span> <span class="pre">1</span> <span class="pre">-</span> <span class="pre">(the</span> <span class="pre">number</span> <span class="pre">of</span> <span class="pre">heterozygous</span> <span class="pre">genotypes)/(the</span> <span class="pre">number</span> <span class="pre">of</span> <span class="pre">heterozygous</span> <span class="pre">genotypes</span> <span class="pre">expected</span> <span class="pre">under</span> <span class="pre">Hardy-Weinberg</span> <span class="pre">equilibrium)',</span> <span class="pre">'Number':</span> <span class="pre">'A'},</span> <span class="pre">'lcr':</span> <span class="pre">{'Description':</span> <span class="pre">'Variant</span> <span class="pre">falls</span> <span class="pre">within</span> <span class="pre">a</span> <span class="pre">low</span> <span class="pre">complexity</span> <span class="pre">region'},</span> <span class="pre">'monoallelic':</span> <span class="pre">{'Description':</span> <span class="pre">'All</span> <span class="pre">samples</span> <span class="pre">are</span> <span class="pre">homozygous</span> <span class="pre">alternate</span> <span class="pre">for</span> <span class="pre">the</span> <span class="pre">variant'},</span> <span class="pre">'n_alt_alleles':</span> <span class="pre">{'Description':</span> <span class="pre">'Total</span> <span class="pre">number</span> <span class="pre">of</span> <span class="pre">alternate</span> <span class="pre">alleles</span> <span class="pre">observed</span> <span class="pre">at</span> <span class="pre">variant</span> <span class="pre">locus',</span> <span class="pre">'Number':</span> <span class="pre">'1'},</span> <span class="pre">'negative_train_site':</span> <span class="pre">{'Description':</span> <span class="pre">'Variant</span> <span class="pre">was</span> <span class="pre">used</span> <span class="pre">to</span> <span class="pre">build</span> <span class="pre">the</span> <span class="pre">negative</span> <span class="pre">training</span> <span class="pre">set</span> <span class="pre">of</span> <span class="pre">low-quality</span> <span class="pre">variants</span> <span class="pre">for</span> <span class="pre">VQSR'},</span> <span class="pre">'non_par':</span> <span class="pre">{'Description':</span> <span class="pre">'Variant</span> <span class="pre">(on</span> <span class="pre">sex</span> <span class="pre">chromosome)</span> <span class="pre">falls</span> <span class="pre">outside</span> <span class="pre">a</span> <span class="pre">pseudoautosomal</span> <span class="pre">region'},</span> <span class="pre">'nonpar':</span> <span class="pre">{'Description':</span> <span class="pre">'Variant</span> <span class="pre">(on</span> <span class="pre">sex</span> <span class="pre">chromosome)</span> <span class="pre">falls</span> <span class="pre">outside</span> <span class="pre">a</span> <span class="pre">pseudoautosomal</span> <span class="pre">region'},</span> <span class="pre">'only_het':</span> <span class="pre">{'Description':</span> <span class="pre">'All</span> <span class="pre">samples</span> <span class="pre">are</span> <span class="pre">heterozygous</span> <span class="pre">for</span> <span class="pre">the</span> <span class="pre">variant'},</span> <span class="pre">'original_alleles':</span> <span class="pre">{'Description':</span> <span class="pre">'Alleles</span> <span class="pre">before</span> <span class="pre">splitting</span> <span class="pre">multiallelics'},</span> <span class="pre">'outside_broad_capture_region':</span> <span class="pre">{'Description':</span> <span class="pre">'Variant</span> <span class="pre">falls</span> <span class="pre">outside</span> <span class="pre">of</span> <span class="pre">Broad</span> <span class="pre">exome</span> <span class="pre">capture</span> <span class="pre">regions.'},</span> <span class="pre">'outside_ukb_capture_region':</span> <span class="pre">{'Description':</span> <span class="pre">'Variant</span> <span class="pre">falls</span> <span class="pre">outside</span> <span class="pre">of</span> <span class="pre">UK</span> <span class="pre">Biobank</span> <span class="pre">exome</span> <span class="pre">capture</span> <span class="pre">regions.'},</span> <span class="pre">'positive_train_site':</span> <span class="pre">{'Description':</span> <span class="pre">'Variant</span> <span class="pre">was</span> <span class="pre">used</span> <span class="pre">to</span> <span class="pre">build</span> <span class="pre">the</span> <span class="pre">positive</span> <span class="pre">training</span> <span class="pre">set</span> <span class="pre">of</span> <span class="pre">high-quality</span> <span class="pre">variants</span> <span class="pre">for</span> <span class="pre">VQSR'},</span> <span class="pre">'rf_label':</span> <span class="pre">{'Description':</span> <span class="pre">'Random</span> <span class="pre">forest</span> <span class="pre">training</span> <span class="pre">label'},</span> <span class="pre">'rf_negative_label':</span> <span class="pre">{'Description':</span> <span class="pre">'Variant</span> <span class="pre">was</span> <span class="pre">labelled</span> <span class="pre">as</span> <span class="pre">a</span> <span class="pre">negative</span> <span class="pre">example</span> <span class="pre">for</span> <span class="pre">training</span> <span class="pre">of</span> <span class="pre">random</span> <span class="pre">forest</span> <span class="pre">model'},</span> <span class="pre">'rf_positive_label':</span> <span class="pre">{'Description':</span> <span class="pre">'Variant</span> <span class="pre">was</span> <span class="pre">labelled</span> <span class="pre">as</span> <span class="pre">a</span> <span class="pre">positive</span> <span class="pre">example</span> <span class="pre">for</span> <span class="pre">training</span> <span class="pre">of</span> <span class="pre">random</span> <span class="pre">forest</span> <span class="pre">model'},</span> <span class="pre">'rf_tp_probability':</span> <span class="pre">{'Description':</span> <span class="pre">'Probability</span> <span class="pre">of</span> <span class="pre">a</span> <span class="pre">called</span> <span class="pre">variant</span> <span class="pre">being</span> <span class="pre">a</span> <span class="pre">true</span> <span class="pre">variant</span> <span class="pre">as</span> <span class="pre">determined</span> <span class="pre">by</span> <span class="pre">random</span> <span class="pre">forest</span> <span class="pre">model'},</span> <span class="pre">'rf_train':</span> <span class="pre">{'Description':</span> <span class="pre">'Variant</span> <span class="pre">was</span> <span class="pre">used</span> <span class="pre">in</span> <span class="pre">training</span> <span class="pre">random</span> <span class="pre">forest</span> <span class="pre">model'},</span> <span class="pre">'segdup':</span> <span class="pre">{'Description':</span> <span class="pre">'Variant</span> <span class="pre">falls</span> <span class="pre">within</span> <span class="pre">a</span> <span class="pre">segmental</span> <span class="pre">duplication</span> <span class="pre">region'},</span> <span class="pre">'sibling_singleton':</span> <span class="pre">{'Description':</span> <span class="pre">'Variant</span> <span class="pre">was</span> <span class="pre">a</span> <span class="pre">callset-wide</span> <span class="pre">doubleton</span> <span class="pre">that</span> <span class="pre">was</span> <span class="pre">present</span> <span class="pre">only</span> <span class="pre">in</span> <span class="pre">two</span> <span class="pre">siblings</span> <span class="pre">(i.e.,</span> <span class="pre">a</span> <span class="pre">singleton</span> <span class="pre">amongst</span> <span class="pre">unrelated</span> <span class="pre">samples</span> <span class="pre">in</span> <span class="pre">cohort).'},</span> <span class="pre">'transmitted_singleton':</span> <span class="pre">{'Description':</span> <span class="pre">'Variant</span> <span class="pre">was</span> <span class="pre">a</span> <span class="pre">callset-wide</span> <span class="pre">doubleton</span> <span class="pre">that</span> <span class="pre">was</span> <span class="pre">transmitted</span> <span class="pre">within</span> <span class="pre">a</span> <span class="pre">family</span> <span class="pre">from</span> <span class="pre">a</span> <span class="pre">parent</span> <span class="pre">to</span> <span class="pre">a</span> <span class="pre">child</span> <span class="pre">(i.e.,</span> <span class="pre">a</span> <span class="pre">singleton</span> <span class="pre">amongst</span> <span class="pre">unrelated</span> <span class="pre">samples</span> <span class="pre">in</span> <span class="pre">cohort)'},</span> <span class="pre">'variant_type':</span> <span class="pre">{'Description':</span> <span class="pre">'Variant</span> <span class="pre">type</span> <span class="pre">(snv,</span> <span class="pre">indel,</span> <span class="pre">multi-snv,</span> <span class="pre">multi-indel,</span> <span class="pre">or</span> <span class="pre">mixed)'},</span> <span class="pre">'was_mixed':</span> <span class="pre">{'Description':</span> <span class="pre">'Variant</span> <span class="pre">type</span> <span class="pre">was</span> <span class="pre">mixed'}}</span></em><a class="headerlink" href="#gnomad.utils.vcf.INFO_DICT" title="Permalink to this definition"></a></dt>
<dd><p>Dictionary used during VCF export to export row (variant) annotations.</p>
</dd></dl>

<dl class="py data">
<dt class="sig sig-object py" id="gnomad.utils.vcf.IN_SILICO_ANNOTATIONS_INFO_DICT">
<span class="sig-prename descclassname"><span class="pre">gnomad.utils.vcf.</span></span><span class="sig-name descname"><span class="pre">IN_SILICO_ANNOTATIONS_INFO_DICT</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'cadd_phred':</span> <span class="pre">{'Description':</span> <span class="pre">&quot;Cadd</span> <span class="pre">Phred-like</span> <span class="pre">scores</span> <span class="pre">('scaled</span> <span class="pre">C-scores')</span> <span class="pre">ranging</span> <span class="pre">from</span> <span class="pre">1</span> <span class="pre">to</span> <span class="pre">99,</span> <span class="pre">based</span> <span class="pre">on</span> <span class="pre">the</span> <span class="pre">rank</span> <span class="pre">of</span> <span class="pre">each</span> <span class="pre">variant</span> <span class="pre">relative</span> <span class="pre">to</span> <span class="pre">all</span> <span class="pre">possible</span> <span class="pre">8.6</span> <span class="pre">billion</span> <span class="pre">substitutions</span> <span class="pre">in</span> <span class="pre">the</span> <span class="pre">human</span> <span class="pre">reference</span> <span class="pre">genome.</span> <span class="pre">Larger</span> <span class="pre">values</span> <span class="pre">are</span> <span class="pre">more</span> <span class="pre">deleterious.&quot;,</span> <span class="pre">'Number':</span> <span class="pre">'1'},</span> <span class="pre">'cadd_raw_score':</span> <span class="pre">{'Description':</span> <span class="pre">&quot;Raw</span> <span class="pre">CADD</span> <span class="pre">scores</span> <span class="pre">are</span> <span class="pre">interpretable</span> <span class="pre">as</span> <span class="pre">the</span> <span class="pre">extent</span> <span class="pre">to</span> <span class="pre">which</span> <span class="pre">the</span> <span class="pre">annotation</span> <span class="pre">profile</span> <span class="pre">for</span> <span class="pre">a</span> <span class="pre">given</span> <span class="pre">variant</span> <span class="pre">suggests</span> <span class="pre">that</span> <span class="pre">the</span> <span class="pre">variant</span> <span class="pre">is</span> <span class="pre">likely</span> <span class="pre">to</span> <span class="pre">be</span> <span class="pre">'observed'</span> <span class="pre">(negative</span> <span class="pre">values)</span> <span class="pre">vs</span> <span class="pre">'simulated'</span> <span class="pre">(positive</span> <span class="pre">values).</span> <span class="pre">Larger</span> <span class="pre">values</span> <span class="pre">are</span> <span class="pre">more</span> <span class="pre">deleterious.&quot;,</span> <span class="pre">'Number':</span> <span class="pre">'1'},</span> <span class="pre">'pangolin_largest_ds':</span> <span class="pre">{'Description':</span> <span class="pre">&quot;Pangolin's</span> <span class="pre">largest</span> <span class="pre">delta</span> <span class="pre">score</span> <span class="pre">across</span> <span class="pre">2</span> <span class="pre">splicing</span> <span class="pre">consequences,</span> <span class="pre">which</span> <span class="pre">reflects</span> <span class="pre">the</span> <span class="pre">probability</span> <span class="pre">of</span> <span class="pre">the</span> <span class="pre">variant</span> <span class="pre">being</span> <span class="pre">splice-altering&quot;,</span> <span class="pre">'Number':</span> <span class="pre">'1'},</span> <span class="pre">'phylop':</span> <span class="pre">{'Description':</span> <span class="pre">'Base-wise</span> <span class="pre">conservation</span> <span class="pre">score</span> <span class="pre">across</span> <span class="pre">the</span> <span class="pre">241</span> <span class="pre">placental</span> <span class="pre">mammals</span> <span class="pre">in</span> <span class="pre">the</span> <span class="pre">Zoonomia</span> <span class="pre">project.</span> <span class="pre">Score</span> <span class="pre">ranges</span> <span class="pre">from</span> <span class="pre">-20</span> <span class="pre">to</span> <span class="pre">9.28,</span> <span class="pre">and</span> <span class="pre">reflects</span> <span class="pre">acceleration</span> <span class="pre">(faster</span> <span class="pre">evolution</span> <span class="pre">than</span> <span class="pre">expected</span> <span class="pre">under</span> <span class="pre">neutral</span> <span class="pre">drift,</span> <span class="pre">assigned</span> <span class="pre">negative</span> <span class="pre">scores)</span> <span class="pre">as</span> <span class="pre">well</span> <span class="pre">as</span> <span class="pre">conservation</span> <span class="pre">(slower</span> <span class="pre">than</span> <span class="pre">expected</span> <span class="pre">evolution,</span> <span class="pre">assigned</span> <span class="pre">positive</span> <span class="pre">scores).',</span> <span class="pre">'Number':</span> <span class="pre">'1'},</span> <span class="pre">'polyphen_max':</span> <span class="pre">{'Description':</span> <span class="pre">'Score</span> <span class="pre">that</span> <span class="pre">predicts</span> <span class="pre">the</span> <span class="pre">possible</span> <span class="pre">impact</span> <span class="pre">of</span> <span class="pre">an</span> <span class="pre">amino</span> <span class="pre">acid</span> <span class="pre">substitution</span> <span class="pre">on</span> <span class="pre">the</span> <span class="pre">structure</span> <span class="pre">and</span> <span class="pre">function</span> <span class="pre">of</span> <span class="pre">a</span> <span class="pre">human</span> <span class="pre">protein,</span> <span class="pre">ranging</span> <span class="pre">from</span> <span class="pre">0.0</span> <span class="pre">(tolerated)</span> <span class="pre">to</span> <span class="pre">1.0</span> <span class="pre">(deleterious).</span>&#160; <span class="pre">We</span> <span class="pre">prioritize</span> <span class="pre">max</span> <span class="pre">scores</span> <span class="pre">for</span> <span class="pre">MANE</span> <span class="pre">Select</span> <span class="pre">transcripts</span> <span class="pre">where</span> <span class="pre">possible</span> <span class="pre">and</span> <span class="pre">otherwise</span> <span class="pre">report</span> <span class="pre">a</span> <span class="pre">score</span> <span class="pre">for</span> <span class="pre">the</span> <span class="pre">canonical</span> <span class="pre">transcript.',</span> <span class="pre">'Number':</span> <span class="pre">'1'},</span> <span class="pre">'revel_max':</span> <span class="pre">{'Description':</span> <span class="pre">&quot;The</span> <span class="pre">maximum</span> <span class="pre">REVEL</span> <span class="pre">score</span> <span class="pre">at</span> <span class="pre">a</span> <span class="pre">site's</span> <span class="pre">MANE</span> <span class="pre">Select</span> <span class="pre">or</span> <span class="pre">canonical</span> <span class="pre">transcript.</span> <span class="pre">It's</span> <span class="pre">an</span> <span class="pre">ensemble</span> <span class="pre">score</span> <span class="pre">for</span> <span class="pre">predicting</span> <span class="pre">the</span> <span class="pre">pathogenicity</span> <span class="pre">of</span> <span class="pre">missense</span> <span class="pre">variants</span> <span class="pre">(based</span> <span class="pre">on</span> <span class="pre">13</span> <span class="pre">other</span> <span class="pre">variant</span> <span class="pre">predictors).</span> <span class="pre">Scores</span> <span class="pre">ranges</span> <span class="pre">from</span> <span class="pre">0</span> <span class="pre">to</span> <span class="pre">1.</span> <span class="pre">Variants</span> <span class="pre">with</span> <span class="pre">higher</span> <span class="pre">scores</span> <span class="pre">are</span> <span class="pre">predicted</span> <span class="pre">to</span> <span class="pre">be</span> <span class="pre">more</span> <span class="pre">likely</span> <span class="pre">to</span> <span class="pre">be</span> <span class="pre">deleterious.&quot;,</span> <span class="pre">'Number':</span> <span class="pre">'1'},</span> <span class="pre">'sift_max':</span> <span class="pre">{'Description':</span> <span class="pre">'Score</span> <span class="pre">reflecting</span> <span class="pre">the</span> <span class="pre">scaled</span> <span class="pre">probability</span> <span class="pre">of</span> <span class="pre">the</span> <span class="pre">amino</span> <span class="pre">acid</span> <span class="pre">substitution</span> <span class="pre">being</span> <span class="pre">tolerated,</span> <span class="pre">ranging</span> <span class="pre">from</span> <span class="pre">0</span> <span class="pre">to</span> <span class="pre">1.</span> <span class="pre">Scores</span> <span class="pre">below</span> <span class="pre">0.05</span> <span class="pre">are</span> <span class="pre">predicted</span> <span class="pre">to</span> <span class="pre">impact</span> <span class="pre">protein</span> <span class="pre">function.</span> <span class="pre">We</span> <span class="pre">prioritize</span> <span class="pre">max</span> <span class="pre">scores</span> <span class="pre">for</span> <span class="pre">MANE</span> <span class="pre">Select</span> <span class="pre">transcripts</span> <span class="pre">where</span> <span class="pre">possible</span> <span class="pre">and</span> <span class="pre">otherwise</span> <span class="pre">report</span> <span class="pre">a</span> <span class="pre">score</span> <span class="pre">for</span> <span class="pre">the</span> <span class="pre">canonical</span> <span class="pre">transcript.',</span> <span class="pre">'Number':</span> <span class="pre">'1'},</span> <span class="pre">'spliceai_ds_max':</span> <span class="pre">{'Description':</span> <span class="pre">&quot;Illumina's</span> <span class="pre">SpliceAI</span> <span class="pre">max</span> <span class="pre">delta</span> <span class="pre">score;</span> <span class="pre">interpreted</span> <span class="pre">as</span> <span class="pre">the</span> <span class="pre">probability</span> <span class="pre">of</span> <span class="pre">the</span> <span class="pre">variant</span> <span class="pre">being</span> <span class="pre">splice-altering.&quot;,</span> <span class="pre">'Number':</span> <span class="pre">'1'}}</span></em><a class="headerlink" href="#gnomad.utils.vcf.IN_SILICO_ANNOTATIONS_INFO_DICT" title="Permalink to this definition"></a></dt>
<dd><p>Dictionary with in silico score descriptions to include in the VCF INFO header.</p>
</dd></dl>

<dl class="py data">
<dt class="sig sig-object py" id="gnomad.utils.vcf.VRS_FIELDS_DICT">
<span class="sig-prename descclassname"><span class="pre">gnomad.utils.vcf.</span></span><span class="sig-name descname"><span class="pre">VRS_FIELDS_DICT</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'VRS_Allele_IDs':</span> <span class="pre">{'Description':</span> <span class="pre">'The</span> <span class="pre">computed</span> <span class="pre">identifiers</span> <span class="pre">for</span> <span class="pre">the</span> <span class="pre">GA4GH</span> <span class="pre">VRS</span> <span class="pre">Alleles</span> <span class="pre">corresponding</span> <span class="pre">to</span> <span class="pre">the</span> <span class="pre">values</span> <span class="pre">in</span> <span class="pre">the</span> <span class="pre">REF</span> <span class="pre">and</span> <span class="pre">ALT</span> <span class="pre">fields',</span> <span class="pre">'Number':</span> <span class="pre">'R'},</span> <span class="pre">'VRS_Ends':</span> <span class="pre">{'Description':</span> <span class="pre">'Interresidue</span> <span class="pre">coordinates</span> <span class="pre">used</span> <span class="pre">as</span> <span class="pre">the</span> <span class="pre">location</span> <span class="pre">ends</span> <span class="pre">for</span> <span class="pre">the</span> <span class="pre">GA4GH</span> <span class="pre">VRS</span> <span class="pre">Alleles</span> <span class="pre">corresponding</span> <span class="pre">to</span> <span class="pre">the</span> <span class="pre">values</span> <span class="pre">in</span> <span class="pre">the</span> <span class="pre">REF</span> <span class="pre">and</span> <span class="pre">ALT</span> <span class="pre">fields',</span> <span class="pre">'Number':</span> <span class="pre">'R'},</span> <span class="pre">'VRS_Starts':</span> <span class="pre">{'Description':</span> <span class="pre">'Interresidue</span> <span class="pre">coordinates</span> <span class="pre">used</span> <span class="pre">as</span> <span class="pre">the</span> <span class="pre">location</span> <span class="pre">starts</span> <span class="pre">for</span> <span class="pre">the</span> <span class="pre">GA4GH</span> <span class="pre">VRS</span> <span class="pre">Alleles</span> <span class="pre">corresponding</span> <span class="pre">to</span> <span class="pre">the</span> <span class="pre">values</span> <span class="pre">in</span> <span class="pre">the</span> <span class="pre">REF</span> <span class="pre">and</span> <span class="pre">ALT</span> <span class="pre">fields',</span> <span class="pre">'Number':</span> <span class="pre">'R'},</span> <span class="pre">'VRS_States':</span> <span class="pre">{'Description':</span> <span class="pre">'The</span> <span class="pre">literal</span> <span class="pre">sequence</span> <span class="pre">states</span> <span class="pre">used</span> <span class="pre">for</span> <span class="pre">the</span> <span class="pre">GA4GH</span> <span class="pre">VRS</span> <span class="pre">Alleles</span> <span class="pre">corresponding</span> <span class="pre">to</span> <span class="pre">the</span> <span class="pre">values</span> <span class="pre">in</span> <span class="pre">the</span> <span class="pre">REF</span> <span class="pre">and</span> <span class="pre">ALT</span> <span class="pre">fields',</span> <span class="pre">'Number':</span> <span class="pre">'.'}}</span></em><a class="headerlink" href="#gnomad.utils.vcf.VRS_FIELDS_DICT" title="Permalink to this definition"></a></dt>
<dd><p>Dictionary with VRS annotations to include in the VCF INFO field and VCF header.</p>
</dd></dl>

<dl class="py data">
<dt class="sig sig-object py" id="gnomad.utils.vcf.ENTRIES">
<span class="sig-prename descclassname"><span class="pre">gnomad.utils.vcf.</span></span><span class="sig-name descname"><span class="pre">ENTRIES</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">['GT',</span> <span class="pre">'GQ',</span> <span class="pre">'DP',</span> <span class="pre">'AD',</span> <span class="pre">'MIN_DP',</span> <span class="pre">'PGT',</span> <span class="pre">'PID',</span> <span class="pre">'PL',</span> <span class="pre">'SB']</span></em><a class="headerlink" href="#gnomad.utils.vcf.ENTRIES" title="Permalink to this definition"></a></dt>
<dd><p>Densified entries to be selected during VCF export.</p>
</dd></dl>

<dl class="py data">
<dt class="sig sig-object py" id="gnomad.utils.vcf.SPARSE_ENTRIES">
<span class="sig-prename descclassname"><span class="pre">gnomad.utils.vcf.</span></span><span class="sig-name descname"><span class="pre">SPARSE_ENTRIES</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">['END',</span> <span class="pre">'DP',</span> <span class="pre">'GQ',</span> <span class="pre">'LA',</span> <span class="pre">'LAD',</span> <span class="pre">'LGT',</span> <span class="pre">'LPGT',</span> <span class="pre">'LPL',</span> <span class="pre">'MIN_DP',</span> <span class="pre">'PID',</span> <span class="pre">'RGQ',</span> <span class="pre">'SB']</span></em><a class="headerlink" href="#gnomad.utils.vcf.SPARSE_ENTRIES" title="Permalink to this definition"></a></dt>
<dd><p>Sparse entries to be selected and densified during VCF export.</p>
</dd></dl>

<dl class="py data">
<dt class="sig sig-object py" id="gnomad.utils.vcf.FORMAT_DICT">
<span class="sig-prename descclassname"><span class="pre">gnomad.utils.vcf.</span></span><span class="sig-name descname"><span class="pre">FORMAT_DICT</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'AD':</span> <span class="pre">{'Description':</span> <span class="pre">'Allelic</span> <span class="pre">depths</span> <span class="pre">for</span> <span class="pre">the</span> <span class="pre">ref</span> <span class="pre">and</span> <span class="pre">alt</span> <span class="pre">alleles</span> <span class="pre">in</span> <span class="pre">the</span> <span class="pre">order</span> <span class="pre">listed',</span> <span class="pre">'Number':</span> <span class="pre">'R',</span> <span class="pre">'Type':</span> <span class="pre">'Integer'},</span> <span class="pre">'DP':</span> <span class="pre">{'Description':</span> <span class="pre">'Approximate</span> <span class="pre">read</span> <span class="pre">depth</span> <span class="pre">(reads</span> <span class="pre">with</span> <span class="pre">MQ=255</span> <span class="pre">or</span> <span class="pre">with</span> <span class="pre">bad</span> <span class="pre">mates</span> <span class="pre">are</span> <span class="pre">filtered)',</span> <span class="pre">'Number':</span> <span class="pre">'1',</span> <span class="pre">'Type':</span> <span class="pre">'Integer'},</span> <span class="pre">'GQ':</span> <span class="pre">{'Description':</span> <span class="pre">'Phred-scaled</span> <span class="pre">confidence</span> <span class="pre">that</span> <span class="pre">the</span> <span class="pre">genotype</span> <span class="pre">assignment</span> <span class="pre">is</span> <span class="pre">correct.</span> <span class="pre">Value</span> <span class="pre">is</span> <span class="pre">the</span> <span class="pre">difference</span> <span class="pre">between</span> <span class="pre">the</span> <span class="pre">second</span> <span class="pre">lowest</span> <span class="pre">PL</span> <span class="pre">and</span> <span class="pre">the</span> <span class="pre">lowest</span> <span class="pre">PL</span> <span class="pre">(always</span> <span class="pre">normalized</span> <span class="pre">to</span> <span class="pre">0).',</span> <span class="pre">'Number':</span> <span class="pre">'1',</span> <span class="pre">'Type':</span> <span class="pre">'Integer'},</span> <span class="pre">'GT':</span> <span class="pre">{'Description':</span> <span class="pre">'Genotype',</span> <span class="pre">'Number':</span> <span class="pre">'1',</span> <span class="pre">'Type':</span> <span class="pre">'String'},</span> <span class="pre">'MIN_DP':</span> <span class="pre">{'Description':</span> <span class="pre">'Minimum</span> <span class="pre">DP</span> <span class="pre">observed</span> <span class="pre">within</span> <span class="pre">the</span> <span class="pre">GVCF</span> <span class="pre">block',</span> <span class="pre">'Number':</span> <span class="pre">'1',</span> <span class="pre">'Type':</span> <span class="pre">'Integer'},</span> <span class="pre">'PGT':</span> <span class="pre">{'Description':</span> <span class="pre">'Physical</span> <span class="pre">phasing</span> <span class="pre">haplotype</span> <span class="pre">information,</span> <span class="pre">describing</span> <span class="pre">how</span> <span class="pre">the</span> <span class="pre">alternate</span> <span class="pre">alleles</span> <span class="pre">are</span> <span class="pre">phased</span> <span class="pre">in</span> <span class="pre">relation</span> <span class="pre">to</span> <span class="pre">one</span> <span class="pre">another',</span> <span class="pre">'Number':</span> <span class="pre">'1',</span> <span class="pre">'Type':</span> <span class="pre">'String'},</span> <span class="pre">'PID':</span> <span class="pre">{'Description':</span> <span class="pre">'Physical</span> <span class="pre">phasing</span> <span class="pre">ID</span> <span class="pre">information,</span> <span class="pre">where</span> <span class="pre">each</span> <span class="pre">unique</span> <span class="pre">ID</span> <span class="pre">within</span> <span class="pre">a</span> <span class="pre">given</span> <span class="pre">sample</span> <span class="pre">(but</span> <span class="pre">not</span> <span class="pre">across</span> <span class="pre">samples)</span> <span class="pre">connects</span> <span class="pre">records</span> <span class="pre">within</span> <span class="pre">a</span> <span class="pre">phasing</span> <span class="pre">group',</span> <span class="pre">'Number':</span> <span class="pre">'1',</span> <span class="pre">'Type':</span> <span class="pre">'String'},</span> <span class="pre">'PL':</span> <span class="pre">{'Description':</span> <span class="pre">'Normalized,</span> <span class="pre">phred-scaled</span> <span class="pre">likelihoods</span> <span class="pre">for</span> <span class="pre">genotypes</span> <span class="pre">as</span> <span class="pre">defined</span> <span class="pre">in</span> <span class="pre">the</span> <span class="pre">VCF</span> <span class="pre">specification',</span> <span class="pre">'Number':</span> <span class="pre">'G',</span> <span class="pre">'Type':</span> <span class="pre">'Integer'},</span> <span class="pre">'RGQ':</span> <span class="pre">{'Description':</span> <span class="pre">'Unconditional</span> <span class="pre">reference</span> <span class="pre">genotype</span> <span class="pre">confidence,</span> <span class="pre">encoded</span> <span class="pre">as</span> <span class="pre">a</span> <span class="pre">phred</span> <span class="pre">quality</span> <span class="pre">-10*log10</span> <span class="pre">p(genotype</span> <span class="pre">call</span> <span class="pre">is</span> <span class="pre">wrong)',</span> <span class="pre">'Number':</span> <span class="pre">'1',</span> <span class="pre">'Type':</span> <span class="pre">'Integer'},</span> <span class="pre">'SB':</span> <span class="pre">{'Description':</span> <span class="pre">&quot;Per-sample</span> <span class="pre">component</span> <span class="pre">statistics</span> <span class="pre">which</span> <span class="pre">comprise</span> <span class="pre">the</span> <span class="pre">Fisher's</span> <span class="pre">exact</span> <span class="pre">test</span> <span class="pre">to</span> <span class="pre">detect</span> <span class="pre">strand</span> <span class="pre">bias.</span> <span class="pre">Values</span> <span class="pre">are:</span> <span class="pre">depth</span> <span class="pre">of</span> <span class="pre">reference</span> <span class="pre">allele</span> <span class="pre">on</span> <span class="pre">forward</span> <span class="pre">strand,</span> <span class="pre">depth</span> <span class="pre">of</span> <span class="pre">reference</span> <span class="pre">allele</span> <span class="pre">on</span> <span class="pre">reverse</span> <span class="pre">strand,</span> <span class="pre">depth</span> <span class="pre">of</span> <span class="pre">alternate</span> <span class="pre">allele</span> <span class="pre">on</span> <span class="pre">forward</span> <span class="pre">strand,</span> <span class="pre">depth</span> <span class="pre">of</span> <span class="pre">alternate</span> <span class="pre">allele</span> <span class="pre">on</span> <span class="pre">reverse</span> <span class="pre">strand.&quot;,</span> <span class="pre">'Number':</span> <span class="pre">'4',</span> <span class="pre">'Type':</span> <span class="pre">'Integer'}}</span></em><a class="headerlink" href="#gnomad.utils.vcf.FORMAT_DICT" title="Permalink to this definition"></a></dt>
<dd><p>Dictionary used during VCF export to export MatrixTable entries.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gnomad.utils.vcf.adjust_vcf_incompatible_types">
<span class="sig-prename descclassname"><span class="pre">gnomad.utils.vcf.</span></span><span class="sig-name descname"><span class="pre">adjust_vcf_incompatible_types</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ht</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pipe_delimited_annotations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">['AS_QUALapprox',</span> <span class="pre">'AS_VarDP',</span> <span class="pre">'AS_MQ_DP',</span> <span class="pre">'AS_RAW_MQ',</span> <span class="pre">'AS_SB_TABLE']</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/gnomad/utils/vcf.html#adjust_vcf_incompatible_types"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gnomad.utils.vcf.adjust_vcf_incompatible_types" title="Permalink to this definition"></a></dt>
<dd><p>Create a Table ready for vcf export.</p>
<dl class="simple">
<dt>In particular, the following conversions are done:</dt><dd><ul class="simple">
<li><p>All int64 are coerced to int32</p></li>
<li><p>Fields specified by <cite>pipe_delimited_annotations</cite> are converted from arrays to pipe-delimited strings</p></li>
</ul>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ht</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2.136)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a>) – Input Table.</p></li>
<li><p><strong>pipe_delimited_annotations</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]) – List of info fields (they must be fields of the ht.info Struct).</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2.136)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Table ready for VCF export.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gnomad.utils.vcf.make_label_combos">
<span class="sig-prename descclassname"><span class="pre">gnomad.utils.vcf.</span></span><span class="sig-name descname"><span class="pre">make_label_combos</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">label_groups</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sort_order</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">['subset',</span> <span class="pre">'downsampling',</span> <span class="pre">'grpmax',</span> <span class="pre">'popmax',</span> <span class="pre">'gen_anc',</span> <span class="pre">'pop',</span> <span class="pre">'subgrp',</span> <span class="pre">'subpop',</span> <span class="pre">'sex',</span> <span class="pre">'group']</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">label_delimiter</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'_'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/gnomad/utils/vcf.html#make_label_combos"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gnomad.utils.vcf.make_label_combos" title="Permalink to this definition"></a></dt>
<dd><p>Make combinations of all possible labels for a supplied dictionary of label groups.</p>
<p>For example, if label_groups is {“sex”: [“XY”, “XX”], “gen_anc”: [“afr”, “nfe”, “amr”]},
this function will return [“afr_XY”, “afr_XX”, “nfe_XY”, “nfe_XX”, “amr_XY”, “amr_XX”]</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>label_groups</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]]) – Dictionary containing an entry for each label group, where key is the name of the grouping,
e.g. “sex” or “gen_anc”, and value is a list of all possible values for that grouping (e.g. [“XY”, “XX”] or [“afr”, “nfe”, “amr”]).</p></li>
<li><p><strong>sort_order</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]) – List containing order to sort label group combinations. Default is SORT_ORDER.</p></li>
<li><p><strong>label_delimiter</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – String to use as delimiter when making group label combinations.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]</p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>list of all possible combinations of values for the supplied label groupings.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gnomad.utils.vcf.index_globals">
<span class="sig-prename descclassname"><span class="pre">gnomad.utils.vcf.</span></span><span class="sig-name descname"><span class="pre">index_globals</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">globals_array</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">label_groups</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">label_delimiter</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'_'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/gnomad/utils/vcf.html#index_globals"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gnomad.utils.vcf.index_globals" title="Permalink to this definition"></a></dt>
<dd><p>Create a dictionary keyed by the specified label groupings with values describing the corresponding index of each grouping entry in the meta_array annotation.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>globals_array</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]]) – Ordered list containing dictionary entries describing all the grouping combinations contained in the globals_array annotation.
Keys are the grouping type (e.g., ‘group’, ‘gen_anc’, ‘sex’) and values are the grouping attribute (e.g., ‘adj’, ‘eas’, ‘XY’).</p></li>
<li><p><strong>label_groups</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]]) – Dictionary containing an entry for each label group, where key is the name of the grouping,
e.g. “sex” or “gen_anc”, and value is a list of all possible values for that grouping (e.g. [“XY”, “XX”] or [“afr”, “nfe”, “amr”])</p></li>
<li><p><strong>label_delimiter</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – String used as delimiter when making group label combinations.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>]</p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Dictionary keyed by specified label grouping combinations, with values describing the corresponding index
of each grouping entry in the globals</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gnomad.utils.vcf.make_combo_header_text">
<span class="sig-prename descclassname"><span class="pre">gnomad.utils.vcf.</span></span><span class="sig-name descname"><span class="pre">make_combo_header_text</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">preposition</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">combo_dict</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gen_anc_names</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/gnomad/utils/vcf.html#make_combo_header_text"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gnomad.utils.vcf.make_combo_header_text" title="Permalink to this definition"></a></dt>
<dd><p>Programmatically generate text to populate the VCF header description for a given variant annotation with specific groupings and subset.</p>
<p>For example, if preposition is “for”, group_types is [“group”, “gen_anc”, “sex”], and combo_fields is [“adj”, “afr”, “XX”],
this function will return the string ” for XX samples in the African-American/African genetic ancestry group”.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>preposition</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Relevant preposition to precede automatically generated text.</p></li>
<li><p><strong>combo_dict</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]) – Dict with grouping types as keys and values for grouping type as values. This function generates text for these values.
Possible grouping types are: “group”, “gen_anc”, “sex”, and “subgroup”.
Example input: {“gen_anc”: “afr”, “sex”: “XX”}</p></li>
<li><p><strong>gen_anc_names</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]) – Dict with global genetic ancestry group names (keys) and genetic ancestry group descriptions (values).</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>String with automatically generated description text for a given set of combo fields.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gnomad.utils.vcf.create_label_groups">
<span class="sig-prename descclassname"><span class="pre">gnomad.utils.vcf.</span></span><span class="sig-name descname"><span class="pre">create_label_groups</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gen_ancs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sexes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">['XX',</span> <span class="pre">'XY']</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">all_groups</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">['adj',</span> <span class="pre">'raw']</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gen_anc_sex_groups</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">['adj']</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/gnomad/utils/vcf.html#create_label_groups"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gnomad.utils.vcf.create_label_groups" title="Permalink to this definition"></a></dt>
<dd><p>Generate a list of label group dictionaries needed to populate info dictionary.</p>
<p>Label dictionaries are passed as input to <cite>make_info_dict</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>gen_ancs</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]) – List of genetic ancestry group names.</p></li>
<li><p><strong>sexes</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]) – List of sample sexes.</p></li>
<li><p><strong>all_groups</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]) – List of data types (raw, adj). Default is <cite>GROUPS</cite>, which is [“raw”, “adj”].</p></li>
<li><p><strong>gen_anc_sex_groups</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]) – List of data types (raw, adj) to populate with gen_ancs and sexes. Default is [“adj”].</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]]]</p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>List of label group dictionaries.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gnomad.utils.vcf.make_info_dict">
<span class="sig-prename descclassname"><span class="pre">gnomad.utils.vcf.</span></span><span class="sig-name descname"><span class="pre">make_info_dict</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">prefix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">suffix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix_before_metric</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gen_anc_names</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{'afr':</span> <span class="pre">'African/African-American',</span> <span class="pre">'ami':</span> <span class="pre">'Amish',</span> <span class="pre">'amr':</span> <span class="pre">'Admixed</span> <span class="pre">American',</span> <span class="pre">'asj':</span> <span class="pre">'Ashkenazi</span> <span class="pre">Jewish',</span> <span class="pre">'bgr':</span> <span class="pre">'Bulgarian</span> <span class="pre">(Eastern</span> <span class="pre">European)',</span> <span class="pre">'consanguineous':</span> <span class="pre">'South</span> <span class="pre">Asian</span> <span class="pre">(F</span> <span class="pre">&gt;</span> <span class="pre">0.05)',</span> <span class="pre">'eas':</span> <span class="pre">'East</span> <span class="pre">Asian',</span> <span class="pre">'est':</span> <span class="pre">'Estonian',</span> <span class="pre">'eur':</span> <span class="pre">'European',</span> <span class="pre">'exac':</span> <span class="pre">'ExAC',</span> <span class="pre">'fin':</span> <span class="pre">'Finnish',</span> <span class="pre">'gbr':</span> <span class="pre">'British',</span> <span class="pre">'jpn':</span> <span class="pre">'Japanese',</span> <span class="pre">'kor':</span> <span class="pre">'Korean',</span> <span class="pre">'mde':</span> <span class="pre">'Middle</span> <span class="pre">Eastern',</span> <span class="pre">'mid':</span> <span class="pre">'Middle</span> <span class="pre">Eastern',</span> <span class="pre">'nfe':</span> <span class="pre">'Non-Finnish</span> <span class="pre">European',</span> <span class="pre">'nwe':</span> <span class="pre">'North-Western</span> <span class="pre">European',</span> <span class="pre">'oea':</span> <span class="pre">'Other</span> <span class="pre">East</span> <span class="pre">Asian',</span> <span class="pre">'oeu':</span> <span class="pre">'Other</span> <span class="pre">European',</span> <span class="pre">'onf':</span> <span class="pre">'Other</span> <span class="pre">Non-Finnish</span> <span class="pre">European',</span> <span class="pre">'oth':</span> <span class="pre">'Other',</span> <span class="pre">'remaining':</span> <span class="pre">'Remaining</span> <span class="pre">individuals',</span> <span class="pre">'sas':</span> <span class="pre">'South</span> <span class="pre">Asian',</span> <span class="pre">'sas_non_consang':</span> <span class="pre">'South</span> <span class="pre">Asian</span> <span class="pre">(F</span> <span class="pre">&lt;</span> <span class="pre">0.05)',</span> <span class="pre">'seu':</span> <span class="pre">'Southern</span> <span class="pre">European',</span> <span class="pre">'sgp':</span> <span class="pre">'Singaporean',</span> <span class="pre">'swe':</span> <span class="pre">'Swedish',</span> <span class="pre">'uniform':</span> <span class="pre">'Uniform',</span> <span class="pre">'unk':</span> <span class="pre">'Unknown'}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">label_groups</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">label_delimiter</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'_'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bin_edges</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">faf</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">grpmax</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fafmax</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">callstats</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">freq_ctt</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">freq_cmh</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">freq_stat_union</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">description_text</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">age_hist_distribution</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sort_order</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">['subset',</span> <span class="pre">'downsampling',</span> <span class="pre">'grpmax',</span> <span class="pre">'popmax',</span> <span class="pre">'gen_anc',</span> <span class="pre">'pop',</span> <span class="pre">'subgrp',</span> <span class="pre">'subpop',</span> <span class="pre">'sex',</span> <span class="pre">'group']</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/gnomad/utils/vcf.html#make_info_dict"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gnomad.utils.vcf.make_info_dict" title="Permalink to this definition"></a></dt>
<dd><p>Generate dictionary of Number and Description attributes of VCF INFO fields.</p>
<p>Used to populate the INFO fields of the VCF header during export.</p>
<dl class="simple">
<dt>Creates:</dt><dd><ul class="simple">
<li><p>INFO fields for age histograms (bin freq, n_smaller, and n_larger for heterozygous and homozygous variant carriers)</p></li>
<li><p>INFO fields for grpmax AC, AN, AF, nhomalt, and grpmax genetic ancestry group</p></li>
<li><p>INFO fields for AC, AN, AF, nhomalt for each combination of sample genetic ancestry group, sex, and subgroup, both for adj and raw data</p></li>
<li><p>INFO fields for filtering allele frequency (faf) annotations</p></li>
</ul>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>prefix</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Prefix string for data, e.g. “gnomAD”. Default is empty string.</p></li>
<li><p><strong>suffix</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Suffix string for data, e.g. “gnomAD”. Default is empty string.</p></li>
<li><p><strong>prefix_before_metric</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – Whether prefix should be added before the metric (AC, AN, AF, nhomalt, faf95, faf99) in INFO field. Default is True.</p></li>
<li><p><strong>gen_anc_names</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]) – Dict with global genetic ancestry group names (keys) and genetic ancestry group descriptions (values). Default is GEN_ANC_NAMES.</p></li>
<li><p><strong>label_groups</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]]) – Dictionary containing an entry for each label group, where key is the name of the grouping,
e.g. “sex” or “gen_anc”, and value is a list of all possible values for that grouping (e.g. [“XY”, “XX”] or [“afr”, “nfe”, “amr”]).</p></li>
<li><p><strong>label_delimiter</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – String to use as delimiter when making group label combinations.</p></li>
<li><p><strong>bin_edges</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]) – Dictionary keyed by annotation type, with values that reflect the bin edges corresponding to the annotation.</p></li>
<li><p><strong>faf</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – If True, use alternate logic to auto-populate dictionary values associated with filter allele frequency annotations.</p></li>
<li><p><strong>grpmax</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – If True, use alternate logic to auto-populate dictionary values associated with grpmax annotations.</p></li>
<li><p><strong>fafmax</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – If True, use alternate logic to auto-populate dictionary values associated with fafmax annotations.</p></li>
<li><p><strong>callstats</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – If True, use alternate logic to auto-populate dictionary values associated with callstats annotations.</p></li>
<li><p><strong>freq_ctt</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – If True, use alternate logic to auto-populate dictionary values associated with frequency contingency table test (CTT) annotations.</p></li>
<li><p><strong>freq_cmh</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – If True, use alternate logic to auto-populate dictionary values associated with frequency Cochran-Mantel-Haenszel (CMH) annotations.</p></li>
<li><p><strong>freq_stat_union</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – If True, use alternate logic to auto-populate dictionary values associated with the union of the contingency table and Cochran-Mantel-Haenszel tests.</p></li>
<li><p><strong>description_text</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Optional text to append to the end of descriptions. Needs to start with a space if specified.</p></li>
<li><p><strong>age_hist_distribution</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Pipe-delimited string of overall age distribution.</p></li>
<li><p><strong>sort_order</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]) – List containing order to sort label group combinations. Default is SORT_ORDER.</p></li>
<li><p><strong>age_hist_distribution</strong> – </p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]]</p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Dictionary keyed by VCF INFO annotations, where values are dictionaries of Number and Description attributes.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gnomad.utils.vcf.add_as_info_dict">
<span class="sig-prename descclassname"><span class="pre">gnomad.utils.vcf.</span></span><span class="sig-name descname"><span class="pre">add_as_info_dict</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">info_dict</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{'AS_SB_TABLE':</span> <span class="pre">{'Description':</span> <span class="pre">'Allele-specific</span> <span class="pre">forward/reverse</span> <span class="pre">read</span> <span class="pre">counts</span> <span class="pre">for</span> <span class="pre">strand</span> <span class="pre">bias</span> <span class="pre">tests',</span> <span class="pre">'Number':</span> <span class="pre">'.'},</span> <span class="pre">'AS_pab_max':</span> <span class="pre">{'Description':</span> <span class="pre">'Maximum</span> <span class="pre">p-value</span> <span class="pre">over</span> <span class="pre">callset</span> <span class="pre">for</span> <span class="pre">binomial</span> <span class="pre">test</span> <span class="pre">of</span> <span class="pre">observed</span> <span class="pre">allele</span> <span class="pre">balance</span> <span class="pre">for</span> <span class="pre">a</span> <span class="pre">heterozygous</span> <span class="pre">genotype,</span> <span class="pre">given</span> <span class="pre">expectation</span> <span class="pre">of</span> <span class="pre">0.5',</span> <span class="pre">'Number':</span> <span class="pre">'A'},</span> <span class="pre">'BaseQRankSum':</span> <span class="pre">{'Description':</span> <span class="pre">'Z-score</span> <span class="pre">from</span> <span class="pre">Wilcoxon</span> <span class="pre">rank</span> <span class="pre">sum</span> <span class="pre">test</span> <span class="pre">of</span> <span class="pre">alternate</span> <span class="pre">vs.</span> <span class="pre">reference</span> <span class="pre">base</span> <span class="pre">qualities'},</span> <span class="pre">'FS':</span> <span class="pre">{'Description':</span> <span class="pre">&quot;Phred-scaled</span> <span class="pre">p-value</span> <span class="pre">of</span> <span class="pre">Fisher's</span> <span class="pre">exact</span> <span class="pre">test</span> <span class="pre">for</span> <span class="pre">strand</span> <span class="pre">bias&quot;},</span> <span class="pre">'InbreedingCoeff':</span> <span class="pre">{'Description':</span> <span class="pre">'Inbreeding</span> <span class="pre">coefficient,</span> <span class="pre">the</span> <span class="pre">excess</span> <span class="pre">heterozygosity</span> <span class="pre">at</span> <span class="pre">a</span> <span class="pre">variant</span> <span class="pre">site,</span> <span class="pre">computed</span> <span class="pre">as</span> <span class="pre">1</span> <span class="pre">-</span> <span class="pre">(the</span> <span class="pre">number</span> <span class="pre">of</span> <span class="pre">heterozygous</span> <span class="pre">genotypes)/(the</span> <span class="pre">number</span> <span class="pre">of</span> <span class="pre">heterozygous</span> <span class="pre">genotypes</span> <span class="pre">expected</span> <span class="pre">under</span> <span class="pre">Hardy-Weinberg</span> <span class="pre">equilibrium)',</span> <span class="pre">'Number':</span> <span class="pre">'A'},</span> <span class="pre">'MQ':</span> <span class="pre">{'Description':</span> <span class="pre">'Root</span> <span class="pre">mean</span> <span class="pre">square</span> <span class="pre">of</span> <span class="pre">the</span> <span class="pre">mapping</span> <span class="pre">quality</span> <span class="pre">of</span> <span class="pre">reads</span> <span class="pre">across</span> <span class="pre">all</span> <span class="pre">samples'},</span> <span class="pre">'MQRankSum':</span> <span class="pre">{'Description':</span> <span class="pre">'Z-score</span> <span class="pre">from</span> <span class="pre">Wilcoxon</span> <span class="pre">rank</span> <span class="pre">sum</span> <span class="pre">test</span> <span class="pre">of</span> <span class="pre">alternate</span> <span class="pre">vs.</span> <span class="pre">reference</span> <span class="pre">read</span> <span class="pre">mapping</span> <span class="pre">qualities'},</span> <span class="pre">'NEGATIVE_TRAIN_SITE':</span> <span class="pre">{'Description':</span> <span class="pre">'Variant</span> <span class="pre">was</span> <span class="pre">used</span> <span class="pre">to</span> <span class="pre">build</span> <span class="pre">the</span> <span class="pre">negative</span> <span class="pre">training</span> <span class="pre">set</span> <span class="pre">of</span> <span class="pre">low-quality</span> <span class="pre">variants</span> <span class="pre">for</span> <span class="pre">VQSR'},</span> <span class="pre">'POSITIVE_TRAIN_SITE':</span> <span class="pre">{'Description':</span> <span class="pre">'Variant</span> <span class="pre">was</span> <span class="pre">used</span> <span class="pre">to</span> <span class="pre">build</span> <span class="pre">the</span> <span class="pre">positive</span> <span class="pre">training</span> <span class="pre">set</span> <span class="pre">of</span> <span class="pre">high-quality</span> <span class="pre">variants</span> <span class="pre">for</span> <span class="pre">VQSR'},</span> <span class="pre">'QD':</span> <span class="pre">{'Description':</span> <span class="pre">'Variant</span> <span class="pre">call</span> <span class="pre">confidence</span> <span class="pre">normalized</span> <span class="pre">by</span> <span class="pre">depth</span> <span class="pre">of</span> <span class="pre">sample</span> <span class="pre">reads</span> <span class="pre">supporting</span> <span class="pre">a</span> <span class="pre">variant'},</span> <span class="pre">'QUALapprox':</span> <span class="pre">{'Description':</span> <span class="pre">'Sum</span> <span class="pre">of</span> <span class="pre">PL[0]</span> <span class="pre">values;</span> <span class="pre">used</span> <span class="pre">to</span> <span class="pre">approximate</span> <span class="pre">the</span> <span class="pre">QUAL</span> <span class="pre">score',</span> <span class="pre">'Number':</span> <span class="pre">'1'},</span> <span class="pre">'ReadPosRankSum':</span> <span class="pre">{'Description':</span> <span class="pre">'Z-score</span> <span class="pre">from</span> <span class="pre">Wilcoxon</span> <span class="pre">rank</span> <span class="pre">sum</span> <span class="pre">test</span> <span class="pre">of</span> <span class="pre">alternate</span> <span class="pre">vs.</span> <span class="pre">reference</span> <span class="pre">read</span> <span class="pre">position</span> <span class="pre">bias'},</span> <span class="pre">'SOR':</span> <span class="pre">{'Description':</span> <span class="pre">'Strand</span> <span class="pre">bias</span> <span class="pre">estimated</span> <span class="pre">by</span> <span class="pre">the</span> <span class="pre">symmetric</span> <span class="pre">odds</span> <span class="pre">ratio</span> <span class="pre">test'},</span> <span class="pre">'VQSLOD':</span> <span class="pre">{'Description':</span> <span class="pre">'Log-odds</span> <span class="pre">ratio</span> <span class="pre">of</span> <span class="pre">being</span> <span class="pre">a</span> <span class="pre">true</span> <span class="pre">variant</span> <span class="pre">versus</span> <span class="pre">being</span> <span class="pre">a</span> <span class="pre">false</span> <span class="pre">positive</span> <span class="pre">under</span> <span class="pre">the</span> <span class="pre">trained</span> <span class="pre">VQSR</span> <span class="pre">Gaussian</span> <span class="pre">mixture</span> <span class="pre">model'},</span> <span class="pre">'VarDP':</span> <span class="pre">{'Description':</span> <span class="pre">'Depth</span> <span class="pre">over</span> <span class="pre">variant</span> <span class="pre">genotypes</span> <span class="pre">(does</span> <span class="pre">not</span> <span class="pre">include</span> <span class="pre">depth</span> <span class="pre">of</span> <span class="pre">reference</span> <span class="pre">samples)'},</span> <span class="pre">'allele_type':</span> <span class="pre">{'Description':</span> <span class="pre">'Allele</span> <span class="pre">type</span> <span class="pre">(snv,</span> <span class="pre">insertion,</span> <span class="pre">deletion,</span> <span class="pre">or</span> <span class="pre">mixed)'},</span> <span class="pre">'culprit':</span> <span class="pre">{'Description':</span> <span class="pre">'Worst-performing</span> <span class="pre">annotation</span> <span class="pre">in</span> <span class="pre">the</span> <span class="pre">VQSR</span> <span class="pre">Gaussian</span> <span class="pre">mixture</span> <span class="pre">model'},</span> <span class="pre">'decoy':</span> <span class="pre">{'Description':</span> <span class="pre">'Variant</span> <span class="pre">falls</span> <span class="pre">within</span> <span class="pre">a</span> <span class="pre">reference</span> <span class="pre">decoy</span> <span class="pre">region'},</span> <span class="pre">'fail_interval_qc':</span> <span class="pre">{'Description':</span> <span class="pre">'Less</span> <span class="pre">than</span> <span class="pre">85</span> <span class="pre">percent</span> <span class="pre">of</span> <span class="pre">samples</span> <span class="pre">meet</span> <span class="pre">20X</span> <span class="pre">coverage</span> <span class="pre">if</span> <span class="pre">variant</span> <span class="pre">is</span> <span class="pre">in</span> <span class="pre">autosomal</span> <span class="pre">or</span> <span class="pre">PAR</span> <span class="pre">regions</span> <span class="pre">or</span> <span class="pre">10X</span> <span class="pre">coverage</span> <span class="pre">for</span> <span class="pre">non-PAR</span> <span class="pre">regions</span> <span class="pre">of</span> <span class="pre">chromosomes</span> <span class="pre">X</span> <span class="pre">and</span> <span class="pre">Y.'},</span> <span class="pre">'has_star':</span> <span class="pre">{'Description':</span> <span class="pre">'Variant</span> <span class="pre">locus</span> <span class="pre">coincides</span> <span class="pre">with</span> <span class="pre">a</span> <span class="pre">spanning</span> <span class="pre">deletion</span> <span class="pre">(represented</span> <span class="pre">by</span> <span class="pre">a</span> <span class="pre">star)</span> <span class="pre">observed</span> <span class="pre">elsewhere</span> <span class="pre">in</span> <span class="pre">the</span> <span class="pre">callset'},</span> <span class="pre">'inbreeding_coeff':</span> <span class="pre">{'Description':</span> <span class="pre">'Inbreeding</span> <span class="pre">coefficient,</span> <span class="pre">the</span> <span class="pre">excess</span> <span class="pre">heterozygosity</span> <span class="pre">at</span> <span class="pre">a</span> <span class="pre">variant</span> <span class="pre">site,</span> <span class="pre">computed</span> <span class="pre">as</span> <span class="pre">1</span> <span class="pre">-</span> <span class="pre">(the</span> <span class="pre">number</span> <span class="pre">of</span> <span class="pre">heterozygous</span> <span class="pre">genotypes)/(the</span> <span class="pre">number</span> <span class="pre">of</span> <span class="pre">heterozygous</span> <span class="pre">genotypes</span> <span class="pre">expected</span> <span class="pre">under</span> <span class="pre">Hardy-Weinberg</span> <span class="pre">equilibrium)',</span> <span class="pre">'Number':</span> <span class="pre">'A'},</span> <span class="pre">'lcr':</span> <span class="pre">{'Description':</span> <span class="pre">'Variant</span> <span class="pre">falls</span> <span class="pre">within</span> <span class="pre">a</span> <span class="pre">low</span> <span class="pre">complexity</span> <span class="pre">region'},</span> <span class="pre">'monoallelic':</span> <span class="pre">{'Description':</span> <span class="pre">'All</span> <span class="pre">samples</span> <span class="pre">are</span> <span class="pre">homozygous</span> <span class="pre">alternate</span> <span class="pre">for</span> <span class="pre">the</span> <span class="pre">variant'},</span> <span class="pre">'n_alt_alleles':</span> <span class="pre">{'Description':</span> <span class="pre">'Total</span> <span class="pre">number</span> <span class="pre">of</span> <span class="pre">alternate</span> <span class="pre">alleles</span> <span class="pre">observed</span> <span class="pre">at</span> <span class="pre">variant</span> <span class="pre">locus',</span> <span class="pre">'Number':</span> <span class="pre">'1'},</span> <span class="pre">'negative_train_site':</span> <span class="pre">{'Description':</span> <span class="pre">'Variant</span> <span class="pre">was</span> <span class="pre">used</span> <span class="pre">to</span> <span class="pre">build</span> <span class="pre">the</span> <span class="pre">negative</span> <span class="pre">training</span> <span class="pre">set</span> <span class="pre">of</span> <span class="pre">low-quality</span> <span class="pre">variants</span> <span class="pre">for</span> <span class="pre">VQSR'},</span> <span class="pre">'non_par':</span> <span class="pre">{'Description':</span> <span class="pre">'Variant</span> <span class="pre">(on</span> <span class="pre">sex</span> <span class="pre">chromosome)</span> <span class="pre">falls</span> <span class="pre">outside</span> <span class="pre">a</span> <span class="pre">pseudoautosomal</span> <span class="pre">region'},</span> <span class="pre">'nonpar':</span> <span class="pre">{'Description':</span> <span class="pre">'Variant</span> <span class="pre">(on</span> <span class="pre">sex</span> <span class="pre">chromosome)</span> <span class="pre">falls</span> <span class="pre">outside</span> <span class="pre">a</span> <span class="pre">pseudoautosomal</span> <span class="pre">region'},</span> <span class="pre">'only_het':</span> <span class="pre">{'Description':</span> <span class="pre">'All</span> <span class="pre">samples</span> <span class="pre">are</span> <span class="pre">heterozygous</span> <span class="pre">for</span> <span class="pre">the</span> <span class="pre">variant'},</span> <span class="pre">'original_alleles':</span> <span class="pre">{'Description':</span> <span class="pre">'Alleles</span> <span class="pre">before</span> <span class="pre">splitting</span> <span class="pre">multiallelics'},</span> <span class="pre">'outside_broad_capture_region':</span> <span class="pre">{'Description':</span> <span class="pre">'Variant</span> <span class="pre">falls</span> <span class="pre">outside</span> <span class="pre">of</span> <span class="pre">Broad</span> <span class="pre">exome</span> <span class="pre">capture</span> <span class="pre">regions.'},</span> <span class="pre">'outside_ukb_capture_region':</span> <span class="pre">{'Description':</span> <span class="pre">'Variant</span> <span class="pre">falls</span> <span class="pre">outside</span> <span class="pre">of</span> <span class="pre">UK</span> <span class="pre">Biobank</span> <span class="pre">exome</span> <span class="pre">capture</span> <span class="pre">regions.'},</span> <span class="pre">'positive_train_site':</span> <span class="pre">{'Description':</span> <span class="pre">'Variant</span> <span class="pre">was</span> <span class="pre">used</span> <span class="pre">to</span> <span class="pre">build</span> <span class="pre">the</span> <span class="pre">positive</span> <span class="pre">training</span> <span class="pre">set</span> <span class="pre">of</span> <span class="pre">high-quality</span> <span class="pre">variants</span> <span class="pre">for</span> <span class="pre">VQSR'},</span> <span class="pre">'rf_label':</span> <span class="pre">{'Description':</span> <span class="pre">'Random</span> <span class="pre">forest</span> <span class="pre">training</span> <span class="pre">label'},</span> <span class="pre">'rf_negative_label':</span> <span class="pre">{'Description':</span> <span class="pre">'Variant</span> <span class="pre">was</span> <span class="pre">labelled</span> <span class="pre">as</span> <span class="pre">a</span> <span class="pre">negative</span> <span class="pre">example</span> <span class="pre">for</span> <span class="pre">training</span> <span class="pre">of</span> <span class="pre">random</span> <span class="pre">forest</span> <span class="pre">model'},</span> <span class="pre">'rf_positive_label':</span> <span class="pre">{'Description':</span> <span class="pre">'Variant</span> <span class="pre">was</span> <span class="pre">labelled</span> <span class="pre">as</span> <span class="pre">a</span> <span class="pre">positive</span> <span class="pre">example</span> <span class="pre">for</span> <span class="pre">training</span> <span class="pre">of</span> <span class="pre">random</span> <span class="pre">forest</span> <span class="pre">model'},</span> <span class="pre">'rf_tp_probability':</span> <span class="pre">{'Description':</span> <span class="pre">'Probability</span> <span class="pre">of</span> <span class="pre">a</span> <span class="pre">called</span> <span class="pre">variant</span> <span class="pre">being</span> <span class="pre">a</span> <span class="pre">true</span> <span class="pre">variant</span> <span class="pre">as</span> <span class="pre">determined</span> <span class="pre">by</span> <span class="pre">random</span> <span class="pre">forest</span> <span class="pre">model'},</span> <span class="pre">'rf_train':</span> <span class="pre">{'Description':</span> <span class="pre">'Variant</span> <span class="pre">was</span> <span class="pre">used</span> <span class="pre">in</span> <span class="pre">training</span> <span class="pre">random</span> <span class="pre">forest</span> <span class="pre">model'},</span> <span class="pre">'segdup':</span> <span class="pre">{'Description':</span> <span class="pre">'Variant</span> <span class="pre">falls</span> <span class="pre">within</span> <span class="pre">a</span> <span class="pre">segmental</span> <span class="pre">duplication</span> <span class="pre">region'},</span> <span class="pre">'sibling_singleton':</span> <span class="pre">{'Description':</span> <span class="pre">'Variant</span> <span class="pre">was</span> <span class="pre">a</span> <span class="pre">callset-wide</span> <span class="pre">doubleton</span> <span class="pre">that</span> <span class="pre">was</span> <span class="pre">present</span> <span class="pre">only</span> <span class="pre">in</span> <span class="pre">two</span> <span class="pre">siblings</span> <span class="pre">(i.e.,</span> <span class="pre">a</span> <span class="pre">singleton</span> <span class="pre">amongst</span> <span class="pre">unrelated</span> <span class="pre">samples</span> <span class="pre">in</span> <span class="pre">cohort).'},</span> <span class="pre">'transmitted_singleton':</span> <span class="pre">{'Description':</span> <span class="pre">'Variant</span> <span class="pre">was</span> <span class="pre">a</span> <span class="pre">callset-wide</span> <span class="pre">doubleton</span> <span class="pre">that</span> <span class="pre">was</span> <span class="pre">transmitted</span> <span class="pre">within</span> <span class="pre">a</span> <span class="pre">family</span> <span class="pre">from</span> <span class="pre">a</span> <span class="pre">parent</span> <span class="pre">to</span> <span class="pre">a</span> <span class="pre">child</span> <span class="pre">(i.e.,</span> <span class="pre">a</span> <span class="pre">singleton</span> <span class="pre">amongst</span> <span class="pre">unrelated</span> <span class="pre">samples</span> <span class="pre">in</span> <span class="pre">cohort)'},</span> <span class="pre">'variant_type':</span> <span class="pre">{'Description':</span> <span class="pre">'Variant</span> <span class="pre">type</span> <span class="pre">(snv,</span> <span class="pre">indel,</span> <span class="pre">multi-snv,</span> <span class="pre">multi-indel,</span> <span class="pre">or</span> <span class="pre">mixed)'},</span> <span class="pre">'was_mixed':</span> <span class="pre">{'Description':</span> <span class="pre">'Variant</span> <span class="pre">type</span> <span class="pre">was</span> <span class="pre">mixed'}}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">as_fields</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">['AS_FS',</span> <span class="pre">'AS_MQ',</span> <span class="pre">'AS_MQRankSum',</span> <span class="pre">'AS_pab_max',</span> <span class="pre">'AS_QUALapprox',</span> <span class="pre">'AS_QD',</span> <span class="pre">'AS_ReadPosRankSum',</span> <span class="pre">'AS_SB_TABLE',</span> <span class="pre">'AS_SOR',</span> <span class="pre">'AS_VarDP',</span> <span class="pre">'InbreedingCoeff']</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/gnomad/utils/vcf.html#add_as_info_dict"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gnomad.utils.vcf.add_as_info_dict" title="Permalink to this definition"></a></dt>
<dd><p>Update info dictionary with allele-specific terms and their descriptions.</p>
<p>Used in VCF export.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>info_dict</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]]) – Dictionary containing site-level annotations and their descriptions. Default is INFO_DICT.</p></li>
<li><p><strong>as_fields</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]) – List containing allele-specific fields to be added to info_dict. Default is AS_FIELDS.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]]</p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Dictionary with allele specific annotations, their descriptions, and their VCF number field.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gnomad.utils.vcf.make_vcf_filter_dict">
<span class="sig-prename descclassname"><span class="pre">gnomad.utils.vcf.</span></span><span class="sig-name descname"><span class="pre">make_vcf_filter_dict</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">snp_cutoff</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">indel_cutoff</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">inbreeding_cutoff</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">variant_qc_filter</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'RF'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">joint</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/gnomad/utils/vcf.html#make_vcf_filter_dict"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gnomad.utils.vcf.make_vcf_filter_dict" title="Permalink to this definition"></a></dt>
<dd><p>Generate dictionary of Number and Description attributes to be used in the VCF header, specifically for FILTER annotations.</p>
<dl class="simple">
<dt>Generates descriptions for:</dt><dd><ul class="simple">
<li><p>AC0 filter</p></li>
<li><p>InbreedingCoeff filter</p></li>
<li><p>Variant QC filter (RF or AS_VQSR)</p></li>
<li><p>PASS (passed all variant filters)</p></li>
</ul>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>snp_cutoff</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>]) – Minimum SNP cutoff score from random forest model.</p></li>
<li><p><strong>indel_cutoff</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>]) – Minimum indel cutoff score from random forest model.</p></li>
<li><p><strong>inbreeding_cutoff</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>]) – Inbreeding coefficient hard cutoff.</p></li>
<li><p><strong>variant_qc_filter</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Method used for variant QC filter. One of ‘RF’ or ‘AS_VQSR’. Default is ‘RF’.</p></li>
<li><p><strong>joint</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – Whether the filter dictionary is for the joint release. Default is False.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]</p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Dictionary keyed by VCF FILTER annotations, where values are Dictionaries of Number and Description attributes.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gnomad.utils.vcf.make_hist_bin_edges_expr">
<span class="sig-prename descclassname"><span class="pre">gnomad.utils.vcf.</span></span><span class="sig-name descname"><span class="pre">make_hist_bin_edges_expr</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ht</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">hists</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">['gq_hist_alt',</span> <span class="pre">'gq_hist_all',</span> <span class="pre">'dp_hist_alt',</span> <span class="pre">'dp_hist_all',</span> <span class="pre">'ab_hist_alt']</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ann_with_hists</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">label_delimiter</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'_'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">include_age_hists</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/gnomad/utils/vcf.html#make_hist_bin_edges_expr"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gnomad.utils.vcf.make_hist_bin_edges_expr" title="Permalink to this definition"></a></dt>
<dd><p>Create dictionaries containing variant histogram annotations and their associated bin edges, formatted into a string separated by pipe delimiters.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ht</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2.136)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a>) – Table containing histogram variant annotations.</p></li>
<li><p><strong>hists</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]) – List of variant histogram annotations. Default is HISTS.</p></li>
<li><p><strong>ann_with_hists</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]) – Name of row annotation containing histogram data. In exomes or
genomes release HT, <cite>histograms</cite> is a row, but in the joint release HT, it’s
under the row of <cite>exomes</cite>, <cite>genomes</cite>, or <cite>joint</cite>.</p></li>
<li><p><strong>prefix</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Prefix text for age histogram bin edges.  Default is empty string.</p></li>
<li><p><strong>label_delimiter</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – String used as delimiter between prefix and histogram annotation.</p></li>
<li><p><strong>include_age_hists</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – Include age histogram annotations.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]</p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Dictionary keyed by histogram annotation name, with corresponding
reformatted bin edges for values.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gnomad.utils.vcf.make_hist_dict">
<span class="sig-prename descclassname"><span class="pre">gnomad.utils.vcf.</span></span><span class="sig-name descname"><span class="pre">make_hist_dict</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">bin_edges</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adj</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">hist_metric_list</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">['gq_hist_alt',</span> <span class="pre">'gq_hist_all',</span> <span class="pre">'dp_hist_alt',</span> <span class="pre">'dp_hist_all',</span> <span class="pre">'ab_hist_alt']</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">label_delimiter</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'_'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">drop_n_smaller_larger</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">suffix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">description_text</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/gnomad/utils/vcf.html#make_hist_dict"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gnomad.utils.vcf.make_hist_dict" title="Permalink to this definition"></a></dt>
<dd><p>Generate dictionary of Number and Description attributes to be used in the VCF header, specifically for histogram annotations.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>bin_edges</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]]) – Dictionary keyed by histogram annotation name, with corresponding string-reformatted bin edges for values.</p></li>
<li><p><strong>adj</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – Whether to create a header dict for raw or adj quality histograms.</p></li>
<li><p><strong>hist_metric_list</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]) – List of hists for which to build hist info dict</p></li>
<li><p><strong>label_delimiter</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – String used as delimiter in values stored in hist_metric_list.</p></li>
<li><p><strong>drop_n_smaller_larger</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – Whether to drop n_smaller and n_larger annotations from header dict. Default is False.</p></li>
<li><p><strong>prefix</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Prefix text for histogram annotations. Default is empty string.</p></li>
<li><p><strong>suffix</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Suffix text for histogram annotations. Default is empty string.</p></li>
<li><p><strong>description_text</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Optional text to append to the end of descriptions. Needs to start with a space if specified.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]</p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Dictionary keyed by VCF INFO annotations, where values are Dictionaries of Number and Description attributes.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gnomad.utils.vcf.set_xx_y_metrics_to_na">
<span class="sig-prename descclassname"><span class="pre">gnomad.utils.vcf.</span></span><span class="sig-name descname"><span class="pre">set_xx_y_metrics_to_na</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">t</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/gnomad/utils/vcf.html#set_xx_y_metrics_to_na"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gnomad.utils.vcf.set_xx_y_metrics_to_na" title="Permalink to this definition"></a></dt>
<dd><p>Set AC, AN, and nhomalt chrY variant annotations for XX samples to NA (instead of 0).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>t</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2.136)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a>, <a class="reference external" href="https://hail.is/docs/0.2/hail.MatrixTable.html#hail.MatrixTable" title="(in Hail v0.2.136)"><code class="xref py py-class docutils literal notranslate"><span class="pre">MatrixTable</span></code></a>]) – Table/MatrixTable containing XX variant annotations.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>, <a class="reference external" href="https://hail.is/docs/0.2/hail.expr.Int32Expression.html#hail.expr.Int32Expression" title="(in Hail v0.2.136)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Int32Expression</span></code></a>]</p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Dictionary with reset annotations</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gnomad.utils.vcf.build_vcf_export_reference">
<span class="sig-prename descclassname"><span class="pre">gnomad.utils.vcf.</span></span><span class="sig-name descname"><span class="pre">build_vcf_export_reference</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">build</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'GRCh38'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">keep_contigs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">['chr1',</span> <span class="pre">'chr2',</span> <span class="pre">'chr3',</span> <span class="pre">'chr4',</span> <span class="pre">'chr5',</span> <span class="pre">'chr6',</span> <span class="pre">'chr7',</span> <span class="pre">'chr8',</span> <span class="pre">'chr9',</span> <span class="pre">'chr10',</span> <span class="pre">'chr11',</span> <span class="pre">'chr12',</span> <span class="pre">'chr13',</span> <span class="pre">'chr14',</span> <span class="pre">'chr15',</span> <span class="pre">'chr16',</span> <span class="pre">'chr17',</span> <span class="pre">'chr18',</span> <span class="pre">'chr19',</span> <span class="pre">'chr20',</span> <span class="pre">'chr21',</span> <span class="pre">'chr22',</span> <span class="pre">'chrX',</span> <span class="pre">'chrY']</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">keep_chrM</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/gnomad/utils/vcf.html#build_vcf_export_reference"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gnomad.utils.vcf.build_vcf_export_reference" title="Permalink to this definition"></a></dt>
<dd><p>Create export reference based on reference genome defined by <cite>build</cite>.</p>
<p>By default this will return a new reference with all non-standard contigs eliminated. Keeps chr 1-22, Y, X, and M.</p>
<p>An example of a non-standard contig is: ##contig=&lt;ID=chr3_GL000221v1_random,length=155397,assembly=GRCh38&gt;</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>name</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Name to use for new reference.</p></li>
<li><p><strong>build</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Reference genome build to use as starting reference genome.</p></li>
<li><p><strong>keep_contigs</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]) – Contigs to keep from reference genome defined by <cite>build</cite>. Default is autosomes and sex chromosomes.</p></li>
<li><p><strong>keep_chrM</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – Whether to keep chrM. Default is True.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://hail.is/docs/0.2/genetics/hail.genetics.ReferenceGenome.html#hail.genetics.ReferenceGenome" title="(in Hail v0.2.136)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ReferenceGenome</span></code></a></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Reference genome for VCF export containing only contigs in <cite>keep_contigs</cite>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gnomad.utils.vcf.rekey_new_reference">
<span class="sig-prename descclassname"><span class="pre">gnomad.utils.vcf.</span></span><span class="sig-name descname"><span class="pre">rekey_new_reference</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">t</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/gnomad/utils/vcf.html#rekey_new_reference"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gnomad.utils.vcf.rekey_new_reference" title="Permalink to this definition"></a></dt>
<dd><p>Re-key Table or MatrixTable with a new reference genome.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>t</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2.136)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a>, <a class="reference external" href="https://hail.is/docs/0.2/hail.MatrixTable.html#hail.MatrixTable" title="(in Hail v0.2.136)"><code class="xref py py-class docutils literal notranslate"><span class="pre">MatrixTable</span></code></a>]) – Input Table/MatrixTable.</p></li>
<li><p><strong>reference</strong> (<a class="reference external" href="https://hail.is/docs/0.2/genetics/hail.genetics.ReferenceGenome.html#hail.genetics.ReferenceGenome" title="(in Hail v0.2.136)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ReferenceGenome</span></code></a>) – Reference genome to re-key with.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2.136)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a>, <a class="reference external" href="https://hail.is/docs/0.2/hail.MatrixTable.html#hail.MatrixTable" title="(in Hail v0.2.136)"><code class="xref py py-class docutils literal notranslate"><span class="pre">MatrixTable</span></code></a>]</p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Re-keyed Table/MatrixTable</p>
</dd>
</dl>
</dd></dl>

</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="transcript_annotation.html" class="btn btn-neutral float-left" title="gnomad.utils.transcript_annotation" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="vep.html" class="btn btn-neutral float-right" title="gnomad.utils.vep" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p></p>
  </div>

   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>