<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>gnomad.resources.grch38.gnomad &mdash; gnomad master documentation</title>
      <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../../../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../../../_static/theme_overrides.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../../../_static/jquery.js"></script>
        <script src="../../../_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script data-url_root="../../../" id="documentation_options" src="../../../_static/documentation_options.js"></script>
        <script src="../../../_static/doctools.js"></script>
        <script src="../../../_static/sphinx_highlight.js"></script>
    <script src="../../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="next" title="gnomad.resources.grch38.reference_data" href="reference_data.html" />
    <link rel="prev" title="gnomad.resources.grch38" href="index.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../../index.html" class="icon icon-home">
            gnomad
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../../getting_started.html">Getting Started</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../examples/index.html">Examples</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../../index.html">API Reference</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../../utils/index.html">utils</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="../index.html">resources</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="../config.html">config</a></li>
<li class="toctree-l3"><a class="reference internal" href="../grch37/index.html">grch37</a></li>
<li class="toctree-l3 current"><a class="reference internal" href="index.html">grch38</a><ul class="current">
<li class="toctree-l4 current"><a class="current reference internal" href="#">gnomad</a></li>
<li class="toctree-l4"><a class="reference internal" href="reference_data.html">reference_data</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../import_resources.html">import_resources</a></li>
<li class="toctree-l3"><a class="reference internal" href="../resource_utils.html">resource_utils</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../variant_qc/index.html">variant_qc</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../sample_qc/index.html">sample_qc</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../assessment/index.html">assessment</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../resource_sources.html">Resource Sources</a></li>
<li class="toctree-l1"><a class="reference external" href="https://github.com/broadinstitute/gnomad_methods/releases">Change Log</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">gnomad</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../../index.html">gnomad</a></li>
          <li class="breadcrumb-item"><a href="../index.html">gnomad.resources</a></li>
          <li class="breadcrumb-item"><a href="index.html">gnomad.resources.grch38</a></li>
      <li class="breadcrumb-item active">gnomad.resources.grch38.gnomad</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../../../_sources/api_reference/resources/grch38/gnomad.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="gnomad-resources-grch38-gnomad">
<h1>gnomad.resources.grch38.gnomad<a class="headerlink" href="#gnomad-resources-grch38-gnomad" title="Permalink to this heading"></a></h1>
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#gnomad.resources.grch38.gnomad.SUBSETS" title="gnomad.resources.grch38.gnomad.SUBSETS"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.resources.grch38.gnomad.SUBSETS</span></code></a></p></td>
<td><p>Order to sort subgroupings during VCF export by version.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#gnomad.resources.grch38.gnomad.GROUPS" title="gnomad.resources.grch38.gnomad.GROUPS"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.resources.grch38.gnomad.GROUPS</span></code></a></p></td>
<td><p>Group names used to generate labels for high quality genotypes and all raw genotypes.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#gnomad.resources.grch38.gnomad.SEXES" title="gnomad.resources.grch38.gnomad.SEXES"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.resources.grch38.gnomad.SEXES</span></code></a></p></td>
<td><p>Sample sexes used in VCF export.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#gnomad.resources.grch38.gnomad.GEN_ANC_GROUPS" title="gnomad.resources.grch38.gnomad.GEN_ANC_GROUPS"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.resources.grch38.gnomad.GEN_ANC_GROUPS</span></code></a></p></td>
<td><p>Genetic ancestry groups in gnomAD by version.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#gnomad.resources.grch38.gnomad.COHORTS_WITH_GEN_ANC_STORED_AS_SUBGRP" title="gnomad.resources.grch38.gnomad.COHORTS_WITH_GEN_ANC_STORED_AS_SUBGRP"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.resources.grch38.gnomad.COHORTS_WITH_GEN_ANC_STORED_AS_SUBGRP</span></code></a></p></td>
<td><p>Subsets in gnomAD v3.1 that are broken down by their known genetic ancestry subgroups instead of groups in the frequency struct.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#gnomad.resources.grch38.gnomad.TGP_GEN_ANC_GROUPS" title="gnomad.resources.grch38.gnomad.TGP_GEN_ANC_GROUPS"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.resources.grch38.gnomad.TGP_GEN_ANC_GROUPS</span></code></a></p></td>
<td><p>1000 Genomes Project (1KG/TGP) genetic ancestry subgroups.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#gnomad.resources.grch38.gnomad.HGDP_GEN_ANC_GROUPS" title="gnomad.resources.grch38.gnomad.HGDP_GEN_ANC_GROUPS"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.resources.grch38.gnomad.HGDP_GEN_ANC_GROUPS</span></code></a></p></td>
<td><p>Human Genome Diversity Project (HGDP) genetic ancestry subgroups.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#gnomad.resources.grch38.gnomad.TGP_GEN_ANC_GROUP_NAMES" title="gnomad.resources.grch38.gnomad.TGP_GEN_ANC_GROUP_NAMES"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.resources.grch38.gnomad.TGP_GEN_ANC_GROUP_NAMES</span></code></a></p></td>
<td><p>1000 Genomes Project (1KG/TGP) genetic ancestry group label map.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#gnomad.resources.grch38.gnomad.GEN_ANC_GROUPS_TO_REMOVE_FOR_GRPMAX" title="gnomad.resources.grch38.gnomad.GEN_ANC_GROUPS_TO_REMOVE_FOR_GRPMAX"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.resources.grch38.gnomad.GEN_ANC_GROUPS_TO_REMOVE_FOR_GRPMAX</span></code></a></p></td>
<td><p>Genetic ancestry groups that are removed before genetic ancestry group max calculations.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#gnomad.resources.grch38.gnomad.DOWNSAMPLINGS" title="gnomad.resources.grch38.gnomad.DOWNSAMPLINGS"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.resources.grch38.gnomad.DOWNSAMPLINGS</span></code></a></p></td>
<td><p>List of the downsampling numbers to use for frequency calculations by version.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#gnomad.resources.grch38.gnomad.public_release" title="gnomad.resources.grch38.gnomad.public_release"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.resources.grch38.gnomad.public_release</span></code></a>(...)</p></td>
<td><p>Retrieve publicly released versioned table resource.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#gnomad.resources.grch38.gnomad.coverage" title="gnomad.resources.grch38.gnomad.coverage"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.resources.grch38.gnomad.coverage</span></code></a>(...)</p></td>
<td><p>Retrieve gnomAD's coverage table by data_type.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#gnomad.resources.grch38.gnomad.all_sites_an" title="gnomad.resources.grch38.gnomad.all_sites_an"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.resources.grch38.gnomad.all_sites_an</span></code></a>(...)</p></td>
<td><p>Retrieve gnomAD's all sites allele number table by data_type.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#gnomad.resources.grch38.gnomad.coverage_tsv_path" title="gnomad.resources.grch38.gnomad.coverage_tsv_path"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.resources.grch38.gnomad.coverage_tsv_path</span></code></a>(...)</p></td>
<td><p>Retrieve gnomAD's coverage table by data_type.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#gnomad.resources.grch38.gnomad.release_vcf_path" title="gnomad.resources.grch38.gnomad.release_vcf_path"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.resources.grch38.gnomad.release_vcf_path</span></code></a>(...)</p></td>
<td><p>Publically released VCF.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#gnomad.resources.grch38.gnomad.add_grpMaxFAF95_v4" title="gnomad.resources.grch38.gnomad.add_grpMaxFAF95_v4"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.resources.grch38.gnomad.add_grpMaxFAF95_v4</span></code></a>(ht)</p></td>
<td><p>Add a grpMaxFAF95 struct with 'grpmax' and 'grpmax_gen_anc'.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#gnomad.resources.grch38.gnomad.gnomad_gks" title="gnomad.resources.grch38.gnomad.gnomad_gks"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.resources.grch38.gnomad.gnomad_gks</span></code></a>(...)</p></td>
<td><p>Perform gnomad GKS annotations on a range of variants at once.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#gnomad.resources.grch38.gnomad.pext" title="gnomad.resources.grch38.gnomad.pext"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.resources.grch38.gnomad.pext</span></code></a>([pext_type])</p></td>
<td><p>Retrieve pext table by type.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#gnomad.resources.grch38.gnomad.constraint" title="gnomad.resources.grch38.gnomad.constraint"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.resources.grch38.gnomad.constraint</span></code></a>()</p></td>
<td><p>Retrieve gene constraint Table.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#gnomad.resources.grch38.gnomad.browser_variant" title="gnomad.resources.grch38.gnomad.browser_variant"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.resources.grch38.gnomad.browser_variant</span></code></a>()</p></td>
<td><p>Retrieve browser variant table.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#gnomad.resources.grch38.gnomad.browser_gene" title="gnomad.resources.grch38.gnomad.browser_gene"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.resources.grch38.gnomad.browser_gene</span></code></a>()</p></td>
<td><p>Retrieve browser gene table.</p></td>
</tr>
</tbody>
</table>
<span class="target" id="module-gnomad.resources.grch38.gnomad"></span><dl class="py data">
<dt class="sig sig-object py" id="gnomad.resources.grch38.gnomad.SUBSETS">
<span class="sig-prename descclassname"><span class="pre">gnomad.resources.grch38.gnomad.</span></span><span class="sig-name descname"><span class="pre">SUBSETS</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'v3':</span> <span class="pre">['non_v2',</span> <span class="pre">'non_topmed',</span> <span class="pre">'non_cancer',</span> <span class="pre">'controls_and_biobanks',</span> <span class="pre">'non_neuro',</span> <span class="pre">'tgp',</span> <span class="pre">'hgdp'],</span> <span class="pre">'v4':</span> <span class="pre">['non_ukb'],</span> <span class="pre">'v5':</span> <span class="pre">['non_aou']}</span></em><a class="headerlink" href="#gnomad.resources.grch38.gnomad.SUBSETS" title="Permalink to this definition"></a></dt>
<dd><p>Order to sort subgroupings during VCF export by version.</p>
<p>Ensures that INFO labels in VCF are in desired order (e.g., tgp_raw_AC_esn_XX).</p>
</dd></dl>

<dl class="py data">
<dt class="sig sig-object py" id="gnomad.resources.grch38.gnomad.GROUPS">
<span class="sig-prename descclassname"><span class="pre">gnomad.resources.grch38.gnomad.</span></span><span class="sig-name descname"><span class="pre">GROUPS</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">['adj',</span> <span class="pre">'raw']</span></em><a class="headerlink" href="#gnomad.resources.grch38.gnomad.GROUPS" title="Permalink to this definition"></a></dt>
<dd><p>Group names used to generate labels for high quality genotypes and all raw genotypes.</p>
<p>Used in VCF export.</p>
</dd></dl>

<dl class="py data">
<dt class="sig sig-object py" id="gnomad.resources.grch38.gnomad.SEXES">
<span class="sig-prename descclassname"><span class="pre">gnomad.resources.grch38.gnomad.</span></span><span class="sig-name descname"><span class="pre">SEXES</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">['XX',</span> <span class="pre">'XY']</span></em><a class="headerlink" href="#gnomad.resources.grch38.gnomad.SEXES" title="Permalink to this definition"></a></dt>
<dd><p>Sample sexes used in VCF export.</p>
<p>Used to stratify frequency annotations (AC, AN, AF) for each sex.</p>
</dd></dl>

<dl class="py data">
<dt class="sig sig-object py" id="gnomad.resources.grch38.gnomad.GEN_ANC_GROUPS">
<span class="sig-prename descclassname"><span class="pre">gnomad.resources.grch38.gnomad.</span></span><span class="sig-name descname"><span class="pre">GEN_ANC_GROUPS</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'v3':</span> <span class="pre">{'genomes':</span> <span class="pre">['afr',</span> <span class="pre">'ami',</span> <span class="pre">'amr',</span> <span class="pre">'asj',</span> <span class="pre">'eas',</span> <span class="pre">'fin',</span> <span class="pre">'nfe',</span> <span class="pre">'oth',</span> <span class="pre">'sas',</span> <span class="pre">'mid']},</span> <span class="pre">'v4':</span> <span class="pre">{'exomes':</span> <span class="pre">['afr',</span> <span class="pre">'amr',</span> <span class="pre">'asj',</span> <span class="pre">'eas',</span> <span class="pre">'fin',</span> <span class="pre">'mid',</span> <span class="pre">'nfe',</span> <span class="pre">'remaining',</span> <span class="pre">'sas'],</span> <span class="pre">'genomes':</span> <span class="pre">['afr',</span> <span class="pre">'ami',</span> <span class="pre">'amr',</span> <span class="pre">'asj',</span> <span class="pre">'eas',</span> <span class="pre">'fin',</span> <span class="pre">'mid',</span> <span class="pre">'nfe',</span> <span class="pre">'remaining',</span> <span class="pre">'sas']},</span> <span class="pre">'v5':</span> <span class="pre">{'exomes':</span> <span class="pre">['afr',</span> <span class="pre">'amr',</span> <span class="pre">'asj',</span> <span class="pre">'eas',</span> <span class="pre">'fin',</span> <span class="pre">'mid',</span> <span class="pre">'nfe',</span> <span class="pre">'remaining',</span> <span class="pre">'sas'],</span> <span class="pre">'genomes':</span> <span class="pre">['afr',</span> <span class="pre">'ami',</span> <span class="pre">'amr',</span> <span class="pre">'asj',</span> <span class="pre">'eas',</span> <span class="pre">'fin',</span> <span class="pre">'mid',</span> <span class="pre">'nfe',</span> <span class="pre">'remaining',</span> <span class="pre">'sas']}}</span></em><a class="headerlink" href="#gnomad.resources.grch38.gnomad.GEN_ANC_GROUPS" title="Permalink to this definition"></a></dt>
<dd><p>Genetic ancestry groups in gnomAD by version.</p>
</dd></dl>

<dl class="py data">
<dt class="sig sig-object py" id="gnomad.resources.grch38.gnomad.COHORTS_WITH_GEN_ANC_STORED_AS_SUBGRP">
<span class="sig-prename descclassname"><span class="pre">gnomad.resources.grch38.gnomad.</span></span><span class="sig-name descname"><span class="pre">COHORTS_WITH_GEN_ANC_STORED_AS_SUBGRP</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">['tgp',</span> <span class="pre">'hgdp']</span></em><a class="headerlink" href="#gnomad.resources.grch38.gnomad.COHORTS_WITH_GEN_ANC_STORED_AS_SUBGRP" title="Permalink to this definition"></a></dt>
<dd><p>Subsets in gnomAD v3.1 that are broken down by their known genetic ancestry subgroups instead of groups in the frequency struct.</p>
</dd></dl>

<dl class="py data">
<dt class="sig sig-object py" id="gnomad.resources.grch38.gnomad.TGP_GEN_ANC_GROUPS">
<span class="sig-prename descclassname"><span class="pre">gnomad.resources.grch38.gnomad.</span></span><span class="sig-name descname"><span class="pre">TGP_GEN_ANC_GROUPS</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">['esn',</span> <span class="pre">'pur',</span> <span class="pre">'pjl',</span> <span class="pre">'clm',</span> <span class="pre">'jpt',</span> <span class="pre">'chb',</span> <span class="pre">'stu',</span> <span class="pre">'itu',</span> <span class="pre">'tsi',</span> <span class="pre">'mxl',</span> <span class="pre">'ceu',</span> <span class="pre">'msl',</span> <span class="pre">'yri',</span> <span class="pre">'beb',</span> <span class="pre">'fin',</span> <span class="pre">'khv',</span> <span class="pre">'cdx',</span> <span class="pre">'lwk',</span> <span class="pre">'acb',</span> <span class="pre">'asw',</span> <span class="pre">'ibs',</span> <span class="pre">'gbr',</span> <span class="pre">'pel',</span> <span class="pre">'gih',</span> <span class="pre">'chs',</span> <span class="pre">'gwd']</span></em><a class="headerlink" href="#gnomad.resources.grch38.gnomad.TGP_GEN_ANC_GROUPS" title="Permalink to this definition"></a></dt>
<dd><p>1000 Genomes Project (1KG/TGP) genetic ancestry subgroups.</p>
</dd></dl>

<dl class="py data">
<dt class="sig sig-object py" id="gnomad.resources.grch38.gnomad.HGDP_GEN_ANC_GROUPS">
<span class="sig-prename descclassname"><span class="pre">gnomad.resources.grch38.gnomad.</span></span><span class="sig-name descname"><span class="pre">HGDP_GEN_ANC_GROUPS</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">['japanese',</span> <span class="pre">'papuanhighlands',</span> <span class="pre">'papuansepik',</span> <span class="pre">'adygei',</span> <span class="pre">'orcadian',</span> <span class="pre">'biaka',</span> <span class="pre">'yakut',</span> <span class="pre">'han',</span> <span class="pre">'northernhan',</span> <span class="pre">'uygur',</span> <span class="pre">'miao',</span> <span class="pre">'mongolian',</span> <span class="pre">'balochi',</span> <span class="pre">'bedouin',</span> <span class="pre">'russian',</span> <span class="pre">'daur',</span> <span class="pre">'pima',</span> <span class="pre">'hezhen',</span> <span class="pre">'sindhi',</span> <span class="pre">'yi',</span> <span class="pre">'oroqen',</span> <span class="pre">'san',</span> <span class="pre">'tuscan',</span> <span class="pre">'tu',</span> <span class="pre">'palestinian',</span> <span class="pre">'tujia',</span> <span class="pre">'druze',</span> <span class="pre">'pathan',</span> <span class="pre">'basque',</span> <span class="pre">'makrani',</span> <span class="pre">'bergamoitalian',</span> <span class="pre">'naxi',</span> <span class="pre">'karitiana',</span> <span class="pre">'sardinian',</span> <span class="pre">'mbuti',</span> <span class="pre">'mozabite',</span> <span class="pre">'yoruba',</span> <span class="pre">'lahu',</span> <span class="pre">'dai',</span> <span class="pre">'cambodian',</span> <span class="pre">'bougainville',</span> <span class="pre">'french',</span> <span class="pre">'brahui',</span> <span class="pre">'hazara',</span> <span class="pre">'bantusouthafrica',</span> <span class="pre">'surui',</span> <span class="pre">'mandenka',</span> <span class="pre">'kalash',</span> <span class="pre">'xibo',</span> <span class="pre">'colombian',</span> <span class="pre">'bantukenya',</span> <span class="pre">'she',</span> <span class="pre">'burusho',</span> <span class="pre">'maya']</span></em><a class="headerlink" href="#gnomad.resources.grch38.gnomad.HGDP_GEN_ANC_GROUPS" title="Permalink to this definition"></a></dt>
<dd><p>Human Genome Diversity Project (HGDP) genetic ancestry subgroups.</p>
</dd></dl>

<dl class="py data">
<dt class="sig sig-object py" id="gnomad.resources.grch38.gnomad.TGP_GEN_ANC_GROUP_NAMES">
<span class="sig-prename descclassname"><span class="pre">gnomad.resources.grch38.gnomad.</span></span><span class="sig-name descname"><span class="pre">TGP_GEN_ANC_GROUP_NAMES</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'acb':</span> <span class="pre">'African</span> <span class="pre">Caribbean',</span> <span class="pre">'asw':</span> <span class="pre">'African-American',</span> <span class="pre">'beb':</span> <span class="pre">'Bengali',</span> <span class="pre">'cdx':</span> <span class="pre">'Chinese</span> <span class="pre">Dai',</span> <span class="pre">'ceu':</span> <span class="pre">'Utah</span> <span class="pre">Residents</span> <span class="pre">(European</span> <span class="pre">Ancestry)',</span> <span class="pre">'chb':</span> <span class="pre">'Han</span> <span class="pre">Chinese',</span> <span class="pre">'chs':</span> <span class="pre">'Southern</span> <span class="pre">Han</span> <span class="pre">Chinese',</span> <span class="pre">'clm':</span> <span class="pre">'Colombian',</span> <span class="pre">'esn':</span> <span class="pre">'Esan',</span> <span class="pre">'fin':</span> <span class="pre">'Finnish',</span> <span class="pre">'gbr':</span> <span class="pre">'British',</span> <span class="pre">'gih':</span> <span class="pre">'Gujarati',</span> <span class="pre">'gwd':</span> <span class="pre">'Gambian',</span> <span class="pre">'ibs':</span> <span class="pre">'Iberian',</span> <span class="pre">'itu':</span> <span class="pre">'Indian</span> <span class="pre">Telugu',</span> <span class="pre">'jpt':</span> <span class="pre">'Japanese',</span> <span class="pre">'khv':</span> <span class="pre">'Kinh',</span> <span class="pre">'lwk':</span> <span class="pre">'Luhya',</span> <span class="pre">'msl':</span> <span class="pre">'Mende',</span> <span class="pre">'mxl':</span> <span class="pre">'Mexican-American',</span> <span class="pre">'pel':</span> <span class="pre">'Peruvian',</span> <span class="pre">'pjl':</span> <span class="pre">'Punjabi',</span> <span class="pre">'pur':</span> <span class="pre">'Puerto</span> <span class="pre">Rican',</span> <span class="pre">'stu':</span> <span class="pre">'Sri</span> <span class="pre">Lankan</span> <span class="pre">Tamil',</span> <span class="pre">'tsi':</span> <span class="pre">'Toscani',</span> <span class="pre">'yri':</span> <span class="pre">'Yoruba'}</span></em><a class="headerlink" href="#gnomad.resources.grch38.gnomad.TGP_GEN_ANC_GROUP_NAMES" title="Permalink to this definition"></a></dt>
<dd><p>1000 Genomes Project (1KG/TGP) genetic ancestry group label map.</p>
</dd></dl>

<dl class="py data">
<dt class="sig sig-object py" id="gnomad.resources.grch38.gnomad.GEN_ANC_GROUPS_TO_REMOVE_FOR_GRPMAX">
<span class="sig-prename descclassname"><span class="pre">gnomad.resources.grch38.gnomad.</span></span><span class="sig-name descname"><span class="pre">GEN_ANC_GROUPS_TO_REMOVE_FOR_GRPMAX</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'v3':</span> <span class="pre">{'ami',</span> <span class="pre">'asj',</span> <span class="pre">'fin',</span> <span class="pre">'mid',</span> <span class="pre">'oth',</span> <span class="pre">'remaining'},</span> <span class="pre">'v4':</span> <span class="pre">{'ami',</span> <span class="pre">'asj',</span> <span class="pre">'fin',</span> <span class="pre">'oth',</span> <span class="pre">'remaining'},</span> <span class="pre">'v5':</span> <span class="pre">{'ami',</span> <span class="pre">'asj',</span> <span class="pre">'fin',</span> <span class="pre">'oth',</span> <span class="pre">'remaining'}}</span></em><a class="headerlink" href="#gnomad.resources.grch38.gnomad.GEN_ANC_GROUPS_TO_REMOVE_FOR_GRPMAX" title="Permalink to this definition"></a></dt>
<dd><p>Genetic ancestry groups that are removed before genetic ancestry group max calculations.</p>
</dd></dl>

<dl class="py data">
<dt class="sig sig-object py" id="gnomad.resources.grch38.gnomad.DOWNSAMPLINGS">
<span class="sig-prename descclassname"><span class="pre">gnomad.resources.grch38.gnomad.</span></span><span class="sig-name descname"><span class="pre">DOWNSAMPLINGS</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'v3':</span> <span class="pre">[10,</span> <span class="pre">20,</span> <span class="pre">50,</span> <span class="pre">100,</span> <span class="pre">200,</span> <span class="pre">500,</span> <span class="pre">1000,</span> <span class="pre">2000,</span> <span class="pre">5000,</span> <span class="pre">10000,</span> <span class="pre">15000,</span> <span class="pre">20000,</span> <span class="pre">25000,</span> <span class="pre">30000,</span> <span class="pre">40000,</span> <span class="pre">50000,</span> <span class="pre">60000,</span> <span class="pre">70000,</span> <span class="pre">75000,</span> <span class="pre">80000,</span> <span class="pre">85000,</span> <span class="pre">90000,</span> <span class="pre">95000,</span> <span class="pre">100000,</span> <span class="pre">110000,</span> <span class="pre">120000],</span> <span class="pre">'v4':</span> <span class="pre">[10,</span> <span class="pre">100,</span> <span class="pre">500,</span> <span class="pre">1000,</span> <span class="pre">2000,</span> <span class="pre">5000,</span> <span class="pre">10000,</span> <span class="pre">20000,</span> <span class="pre">50000,</span> <span class="pre">100000,</span> <span class="pre">200000,</span> <span class="pre">500000],</span> <span class="pre">'v5':</span> <span class="pre">[10000,</span> <span class="pre">59355,</span> <span class="pre">89074,</span> <span class="pre">100000,</span> <span class="pre">223733]}</span></em><a class="headerlink" href="#gnomad.resources.grch38.gnomad.DOWNSAMPLINGS" title="Permalink to this definition"></a></dt>
<dd><p>List of the downsampling numbers to use for frequency calculations by version.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gnomad.resources.grch38.gnomad.public_release">
<span class="sig-prename descclassname"><span class="pre">gnomad.resources.grch38.gnomad.</span></span><span class="sig-name descname"><span class="pre">public_release</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_type</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/gnomad/resources/grch38/gnomad.html#public_release"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gnomad.resources.grch38.gnomad.public_release" title="Permalink to this definition"></a></dt>
<dd><p>Retrieve publicly released versioned table resource.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>data_type</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – One of “exomes”, “genomes” or “joint”.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="../resource_utils.html#gnomad.resources.resource_utils.VersionedTableResource" title="gnomad.resources.resource_utils.VersionedTableResource"><code class="xref py py-class docutils literal notranslate"><span class="pre">VersionedTableResource</span></code></a></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Release Table</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gnomad.resources.grch38.gnomad.coverage">
<span class="sig-prename descclassname"><span class="pre">gnomad.resources.grch38.gnomad.</span></span><span class="sig-name descname"><span class="pre">coverage</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_type</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/gnomad/resources/grch38/gnomad.html#coverage"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gnomad.resources.grch38.gnomad.coverage" title="Permalink to this definition"></a></dt>
<dd><p>Retrieve gnomAD’s coverage table by data_type.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>data_type</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – One of “exomes” or “genomes”</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="../resource_utils.html#gnomad.resources.resource_utils.VersionedTableResource" title="gnomad.resources.resource_utils.VersionedTableResource"><code class="xref py py-class docutils literal notranslate"><span class="pre">VersionedTableResource</span></code></a></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Coverage Table</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gnomad.resources.grch38.gnomad.all_sites_an">
<span class="sig-prename descclassname"><span class="pre">gnomad.resources.grch38.gnomad.</span></span><span class="sig-name descname"><span class="pre">all_sites_an</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_type</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/gnomad/resources/grch38/gnomad.html#all_sites_an"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gnomad.resources.grch38.gnomad.all_sites_an" title="Permalink to this definition"></a></dt>
<dd><p>Retrieve gnomAD’s all sites allele number table by data_type.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>data_type</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – One of “exomes” or “genomes”</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="../resource_utils.html#gnomad.resources.resource_utils.VersionedTableResource" title="gnomad.resources.resource_utils.VersionedTableResource"><code class="xref py py-class docutils literal notranslate"><span class="pre">VersionedTableResource</span></code></a></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>All sites allele number VersionedTableResource</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gnomad.resources.grch38.gnomad.coverage_tsv_path">
<span class="sig-prename descclassname"><span class="pre">gnomad.resources.grch38.gnomad.</span></span><span class="sig-name descname"><span class="pre">coverage_tsv_path</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_type</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">version</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/gnomad/resources/grch38/gnomad.html#coverage_tsv_path"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gnomad.resources.grch38.gnomad.coverage_tsv_path" title="Permalink to this definition"></a></dt>
<dd><p>Retrieve gnomAD’s coverage table by data_type.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_type</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – One of “exomes” or “genomes”</p></li>
<li><p><strong>version</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]) – </p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Coverage Table</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gnomad.resources.grch38.gnomad.release_vcf_path">
<span class="sig-prename descclassname"><span class="pre">gnomad.resources.grch38.gnomad.</span></span><span class="sig-name descname"><span class="pre">release_vcf_path</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_type</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">version</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">contig</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/gnomad/resources/grch38/gnomad.html#release_vcf_path"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gnomad.resources.grch38.gnomad.release_vcf_path" title="Permalink to this definition"></a></dt>
<dd><p>Publically released VCF. Provide specific contig, i.e. “chr20”, to retrieve contig specific VCF.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_type</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – One of “exomes” or “genomes”</p></li>
<li><p><strong>version</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – One of the release versions of gnomAD on GRCh37</p></li>
<li><p><strong>contig</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Single contig “chr1” to “chrY”</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Path to VCF</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gnomad.resources.grch38.gnomad.add_grpMaxFAF95_v4">
<span class="sig-prename descclassname"><span class="pre">gnomad.resources.grch38.gnomad.</span></span><span class="sig-name descname"><span class="pre">add_grpMaxFAF95_v4</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ht</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/gnomad/resources/grch38/gnomad.html#add_grpMaxFAF95_v4"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gnomad.resources.grch38.gnomad.add_grpMaxFAF95_v4" title="Permalink to this definition"></a></dt>
<dd><p>Add a grpMaxFAF95 struct with ‘grpmax’ and ‘grpmax_gen_anc’.</p>
<p>Also includes a jointGrpMaxFAF95 annotation using the v4 fafmax and joint_fafmax structures.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>ht</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2.137)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a>) – Input hail table.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2.137)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Annotated hail table.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gnomad.resources.grch38.gnomad.gnomad_gks">
<span class="sig-prename descclassname"><span class="pre">gnomad.resources.grch38.gnomad.</span></span><span class="sig-name descname"><span class="pre">gnomad_gks</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">locus_interval</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">version</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">data_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'genomes'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">by_gen_anc_group</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">by_sex</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vrs_only</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">custom_ht</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">skip_checkpoint</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">skip_coverage</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">custom_coverage_ht</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/gnomad/resources/grch38/gnomad.html#gnomad_gks"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gnomad.resources.grch38.gnomad.gnomad_gks" title="Permalink to this definition"></a></dt>
<dd><p>Perform gnomad GKS annotations on a range of variants at once.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>locus_interval</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.IntervalExpression.html#hail.expr.IntervalExpression" title="(in Hail v0.2.137)"><code class="xref py py-class docutils literal notranslate"><span class="pre">IntervalExpression</span></code></a>) – Hail IntervalExpression of locus&lt;reference_genome&gt;.
e.g. hl.locus_interval(‘chr1’, 6424776, 6461367, reference_genome=”GRCh38”)</p></li>
<li><p><strong>version</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – String of version of gnomAD release to use.</p></li>
<li><p><strong>data_type</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – String of either “exomes” or “genomes” for the type of reads that are desired.</p></li>
<li><p><strong>by_gen_anc_group</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – Boolean to pass to obtain frequency information for each cohort.</p></li>
<li><p><strong>by_sex</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – Boolean to pass to return frequency information for each cohort split by chromosomal sex.</p></li>
<li><p><strong>vrs_only</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – Boolean to pass for only VRS info to be returned
(will not include allele frequency information).</p></li>
<li><p><strong>custom_ht</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2.137)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a>) – Table to use instead of what public_release() method would return for the version.</p></li>
<li><p><strong>skip_checkpoint</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – Bool to pass to skip checkpointing selected fields
(checkpointing may be desirable for large datasets by reducing data copies across the cluster).</p></li>
<li><p><strong>skip_coverage</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – Bool to pass to skip adding coverage statistics.</p></li>
<li><p><strong>custom_coverage_ht</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2.137)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a>) – Custom table to use for coverage statistics instead of the release coverage table.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>List of dictionaries containing VRS information
(and freq info split by ancestry groups and sex if desired) for specified variant.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gnomad.resources.grch38.gnomad.pext">
<span class="sig-prename descclassname"><span class="pre">gnomad.resources.grch38.gnomad.</span></span><span class="sig-name descname"><span class="pre">pext</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pext_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'base_level'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/gnomad/resources/grch38/gnomad.html#pext"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gnomad.resources.grch38.gnomad.pext" title="Permalink to this definition"></a></dt>
<dd><p>Retrieve pext table by type.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>pext_type</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – One of “base_level” or “annotation_level”. Default is “base_level”.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="../resource_utils.html#gnomad.resources.resource_utils.GnomadPublicTableResource" title="gnomad.resources.resource_utils.GnomadPublicTableResource"><code class="xref py py-class docutils literal notranslate"><span class="pre">GnomadPublicTableResource</span></code></a></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Pext Table.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gnomad.resources.grch38.gnomad.constraint">
<span class="sig-prename descclassname"><span class="pre">gnomad.resources.grch38.gnomad.</span></span><span class="sig-name descname"><span class="pre">constraint</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/gnomad/resources/grch38/gnomad.html#constraint"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gnomad.resources.grch38.gnomad.constraint" title="Permalink to this definition"></a></dt>
<dd><p>Retrieve gene constraint Table.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="../resource_utils.html#gnomad.resources.resource_utils.VersionedTableResource" title="gnomad.resources.resource_utils.VersionedTableResource"><code class="xref py py-class docutils literal notranslate"><span class="pre">VersionedTableResource</span></code></a></p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Gene constraint Table.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gnomad.resources.grch38.gnomad.browser_variant">
<span class="sig-prename descclassname"><span class="pre">gnomad.resources.grch38.gnomad.</span></span><span class="sig-name descname"><span class="pre">browser_variant</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/gnomad/resources/grch38/gnomad.html#browser_variant"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gnomad.resources.grch38.gnomad.browser_variant" title="Permalink to this definition"></a></dt>
<dd><p>Retrieve browser variant table.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="../resource_utils.html#gnomad.resources.resource_utils.VersionedTableResource" title="gnomad.resources.resource_utils.VersionedTableResource"><code class="xref py py-class docutils literal notranslate"><span class="pre">VersionedTableResource</span></code></a></p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Browser variant Table.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gnomad.resources.grch38.gnomad.browser_gene">
<span class="sig-prename descclassname"><span class="pre">gnomad.resources.grch38.gnomad.</span></span><span class="sig-name descname"><span class="pre">browser_gene</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/gnomad/resources/grch38/gnomad.html#browser_gene"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gnomad.resources.grch38.gnomad.browser_gene" title="Permalink to this definition"></a></dt>
<dd><p>Retrieve browser gene table.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="../resource_utils.html#gnomad.resources.resource_utils.GnomadPublicTableResource" title="gnomad.resources.resource_utils.GnomadPublicTableResource"><code class="xref py py-class docutils literal notranslate"><span class="pre">GnomadPublicTableResource</span></code></a></p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Browser gene Table.</p>
</dd>
</dl>
</dd></dl>

</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="index.html" class="btn btn-neutral float-left" title="gnomad.resources.grch38" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="reference_data.html" class="btn btn-neutral float-right" title="gnomad.resources.grch38.reference_data" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p></p>
  </div>

   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>