Search.setIndex({"docnames": ["api_reference/assessment/index", "api_reference/assessment/parse_validity_logs", "api_reference/assessment/summary_stats", "api_reference/assessment/validity_checks", "api_reference/index", "api_reference/resources/config", "api_reference/resources/grch37/gnomad", "api_reference/resources/grch37/gnomad_ld", "api_reference/resources/grch37/index", "api_reference/resources/grch37/reference_data", "api_reference/resources/grch38/gnomad", "api_reference/resources/grch38/index", "api_reference/resources/grch38/reference_data", "api_reference/resources/import_resources", "api_reference/resources/index", "api_reference/resources/resource_utils", "api_reference/sample_qc/ancestry", "api_reference/sample_qc/filtering", "api_reference/sample_qc/index", "api_reference/sample_qc/pipeline", "api_reference/sample_qc/platform", "api_reference/sample_qc/relatedness", "api_reference/sample_qc/sex", "api_reference/utils/annotations", "api_reference/utils/constraint", "api_reference/utils/file_utils", "api_reference/utils/filtering", "api_reference/utils/gen_stats", "api_reference/utils/index", "api_reference/utils/intervals", "api_reference/utils/liftover", "api_reference/utils/parse", "api_reference/utils/plotting", "api_reference/utils/reference_genome", "api_reference/utils/release", "api_reference/utils/slack", "api_reference/utils/sparse_mt", "api_reference/utils/transcript_annotation", "api_reference/utils/vcf", "api_reference/utils/vep", "api_reference/variant_qc/evaluation", "api_reference/variant_qc/index", "api_reference/variant_qc/ld", "api_reference/variant_qc/pipeline", "api_reference/variant_qc/random_forest", "api_reference/variant_qc/training", "examples/index", "examples/vep", "getting_started", "index", "resource_sources"], "filenames": ["api_reference/assessment/index.rst", "api_reference/assessment/parse_validity_logs.rst", "api_reference/assessment/summary_stats.rst", "api_reference/assessment/validity_checks.rst", "api_reference/index.rst", "api_reference/resources/config.rst", "api_reference/resources/grch37/gnomad.rst", "api_reference/resources/grch37/gnomad_ld.rst", "api_reference/resources/grch37/index.rst", "api_reference/resources/grch37/reference_data.rst", "api_reference/resources/grch38/gnomad.rst", "api_reference/resources/grch38/index.rst", "api_reference/resources/grch38/reference_data.rst", "api_reference/resources/import_resources.rst", "api_reference/resources/index.rst", "api_reference/resources/resource_utils.rst", "api_reference/sample_qc/ancestry.rst", "api_reference/sample_qc/filtering.rst", "api_reference/sample_qc/index.rst", "api_reference/sample_qc/pipeline.rst", "api_reference/sample_qc/platform.rst", "api_reference/sample_qc/relatedness.rst", "api_reference/sample_qc/sex.rst", "api_reference/utils/annotations.rst", "api_reference/utils/constraint.rst", "api_reference/utils/file_utils.rst", "api_reference/utils/filtering.rst", "api_reference/utils/gen_stats.rst", "api_reference/utils/index.rst", "api_reference/utils/intervals.rst", "api_reference/utils/liftover.rst", "api_reference/utils/parse.rst", "api_reference/utils/plotting.rst", "api_reference/utils/reference_genome.rst", "api_reference/utils/release.rst", "api_reference/utils/slack.rst", "api_reference/utils/sparse_mt.rst", "api_reference/utils/transcript_annotation.rst", "api_reference/utils/vcf.rst", "api_reference/utils/vep.rst", "api_reference/variant_qc/evaluation.rst", "api_reference/variant_qc/index.rst", "api_reference/variant_qc/ld.rst", "api_reference/variant_qc/pipeline.rst", "api_reference/variant_qc/random_forest.rst", "api_reference/variant_qc/training.rst", "examples/index.rst", "examples/vep.rst", "getting_started.rst", "index.rst", "resource_sources.rst"], "titles": ["gnomad.assessment", "gnomad.assessment.parse_validity_logs", "gnomad.assessment.summary_stats", "gnomad.assessment.validity_checks", "gnomad", "gnomad.resources.config", "gnomad.resources.grch37.gnomad", "gnomad.resources.grch37.gnomad_ld", "gnomad.resources.grch37", "gnomad.resources.grch37.reference_data", "gnomad.resources.grch38.gnomad", "gnomad.resources.grch38", "gnomad.resources.grch38.reference_data", "gnomad.resources.import_resources", "gnomad.resources", "gnomad.resources.resource_utils", "gnomad.sample_qc.ancestry", "gnomad.sample_qc.filtering", "gnomad.sample_qc", "gnomad.sample_qc.pipeline", "gnomad.sample_qc.platform", "gnomad.sample_qc.relatedness", "gnomad.sample_qc.sex", "gnomad.utils.annotations", "gnomad.utils.constraint", "gnomad.utils.file_utils", "gnomad.utils.filtering", "gnomad.utils.gen_stats", "gnomad.utils", "gnomad.utils.intervals", "gnomad.utils.liftover", "gnomad.utils.parse", "gnomad.utils.plotting", "gnomad.utils.reference_genome", "gnomad.utils.release", "gnomad.utils.slack", "gnomad.utils.sparse_mt", "gnomad.utils.transcript_annotation", "gnomad.utils.vcf", "gnomad.utils.vep", "gnomad.variant_qc.evaluation", "gnomad.variant_qc", "gnomad.variant_qc.ld", "gnomad.variant_qc.pipeline", "gnomad.variant_qc.random_forest", "gnomad.variant_qc.training", "Examples", "Variant Effect Predictor (VEP)", "Getting Started", "gnomad", "Resource Sources"], "terms": {"parse_validity_log": [0, 4, 49], "parse_log_fil": [0, 1], "generate_html_report": [0, 1], "summary_stat": [0, 4, 49], "freq_bin_expr": [0, 2], "get_summary_counts_dict": [0, 2], "get_summary_ac_dict": [0, 2], "get_summary_count": [0, 2], "get_an_criteria": [0, 2], "get_tx_expression_expr": [0, 2], "get_summary_stats_variant_filter_expr": [0, 2], "get_summary_stats_csq_filter_expr": [0, 2], "generate_filter_combin": [0, 2], "get_summary_stats_filter_group_meta": [0, 2], "default_generate_gene_lof_matrix": [0, 2], "get_het_hom_summary_dict": [0, 2], "default_generate_gene_lof_summari": [0, 2], "validity_check": [0, 4, 49], "custom_record_factori": [0, 3], "temp_logg": [0, 3], "generic_field_check": [0, 3], "generic_field_check_loop": [0, 3], "generate_field_check_expr": [0, 3], "make_filters_expr_dict": [0, 3], "make_group_sum_expr_dict": [0, 3], "compare_row_count": [0, 3], "summarize_variant_filt": [0, 3], "compare_subset_freq": [0, 3], "sum_group_callstat": [0, 3], "summarize_vari": [0, 3], "check_raw_and_adj_callstat": [0, 3], "check_sex_chr_metr": [0, 3], "compute_missing": [0, 3], "vcf_field_check": [0, 3], "check_global_and_row_annot_length": [0, 3], "pprint_global_ann": [0, 3], "validate_release_t": [0, 3], "count_vep_annotated_variants_per_interv": [0, 3], "check_missingness_of_struct": [0, 3], "flatten_missingness_struct": [0, 3], "unfurl_array_annot": [0, 3], "check_globals_for_retired_term": [0, 3], "script": [1, 24, 49], "pars": [1, 4, 26, 28, 49], "log": [1, 3, 15, 23, 32, 38, 49], "from": [1, 2, 3, 5, 6, 9, 12, 13, 15, 16, 17, 19, 21, 22, 23, 24, 25, 26, 30, 31, 32, 33, 36, 37, 38, 39, 40, 42, 43, 44, 45, 47, 48, 50], "valid": [1, 3], "check": [1, 3, 15, 23, 24, 25, 39, 44], "output": [1, 2, 3, 15, 16, 17, 21, 22, 23, 24, 25, 32, 36, 39, 44], "log_fil": 1, "sourc": [1, 2, 3, 5, 6, 7, 9, 10, 12, 13, 15, 16, 17, 19, 20, 21, 22, 23, 24, 25, 26, 27, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 42, 43, 44, 45, 49], "file": [1, 15, 23, 25, 26, 30, 32, 35, 38, 39, 44], "categor": [1, 24], "messag": [1, 3, 25, 35], "format": [1, 16, 19, 21, 23, 24, 26, 31, 37, 38, 39, 48], "extract": [1, 21, 24, 37, 44], "function": [1, 2, 3, 15, 16, 19, 21, 22, 23, 24, 26, 30, 31, 36, 37, 38, 39, 40, 43, 44, 45, 49, 50], "name": [1, 2, 3, 5, 10, 13, 15, 17, 19, 23, 24, 25, 26, 34, 36, 37, 38, 39, 43, 44, 47, 48, 50], "paramet": [1, 2, 3, 6, 7, 9, 10, 12, 13, 15, 16, 17, 19, 20, 21, 22, 23, 24, 25, 26, 27, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 42, 43, 44, 45], "path": [1, 6, 10, 13, 15, 17, 19, 23, 25, 26, 30, 32, 35, 39, 43, 44, 47], "contain": [1, 2, 3, 6, 10, 15, 16, 17, 20, 21, 22, 23, 24, 25, 26, 30, 31, 32, 33, 34, 36, 37, 38, 39, 40, 43, 44, 45, 49], "python": [1, 23, 25, 48], "info": [1, 3, 10, 19, 21, 23, 36, 38, 39, 43, 44], "warn": [1, 15, 17, 42], "statement": 1, "return": [1, 2, 3, 5, 6, 7, 9, 10, 12, 13, 15, 16, 17, 19, 20, 21, 22, 23, 24, 25, 26, 27, 29, 30, 31, 32, 33, 34, 36, 37, 38, 39, 40, 42, 43, 44, 45], "list": [1, 2, 3, 10, 13, 15, 16, 17, 19, 20, 21, 23, 24, 25, 26, 29, 31, 32, 33, 34, 36, 37, 38, 39, 43, 44, 50], "tupl": [1, 2, 13, 16, 17, 20, 21, 22, 23, 24, 25, 26, 30, 36, 37, 42, 43, 44], "logger": [1, 3], "inform": [1, 2, 10, 15, 17, 23, 24, 25, 37, 38, 42, 44, 48], "describ": [1, 3, 12, 21, 23, 24, 38, 48], "statu": [1, 3, 21, 23, 24, 26], "categori": [1, 2, 21, 22, 24], "associ": [1, 3, 23, 33, 38], "tabl": [1, 2, 3, 6, 9, 10, 12, 15, 16, 17, 19, 20, 21, 22, 23, 24, 25, 26, 30, 32, 33, 36, 38, 39, 40, 42, 43, 44, 45, 47, 48], "relev": [1, 3, 17, 23, 24, 26, 38], "parsed_log": 1, "output_fil": 1, "gener": [1, 2, 3, 10, 13, 15, 16, 21, 23, 24, 29, 30, 31, 36, 37, 38, 39, 43], "an": [1, 2, 3, 10, 15, 16, 17, 19, 20, 21, 22, 23, 24, 25, 26, 27, 29, 32, 33, 35, 36, 37, 38, 39, 40, 43, 44, 45, 48], "html": [1, 21], "report": [1, 21, 38], "sortabl": 1, "filter": [1, 2, 3, 4, 18, 19, 21, 23, 24, 28, 33, 34, 36, 37, 38, 39, 42, 43, 49], "column": [1, 3, 16, 21, 23, 32, 36, 37, 40, 43, 44], "expand": [1, 16], "result": [1, 3, 16, 17, 20, 21, 23, 32, 36, 39, 43, 44], "freq_expr": [2, 23, 24], "index": [2, 3, 17, 21, 23, 24, 26, 30, 36, 38, 39, 42, 44], "0": [2, 3, 9, 12, 15, 16, 17, 19, 20, 21, 23, 24, 26, 30, 32, 34, 37, 38, 39, 40, 42, 43, 44, 45, 48], "ac_cutoff": 2, "ac0": [2, 3, 38], "1": [2, 5, 6, 9, 10, 12, 16, 17, 19, 20, 21, 23, 24, 25, 26, 27, 31, 32, 36, 37, 38, 40, 42, 43, 45, 48], "singleton": [2, 23, 24, 38, 40, 43], "2": [2, 9, 12, 16, 17, 19, 21, 23, 24, 26, 37, 38, 40, 43, 44, 48], "doubleton": [2, 23, 38], "af_cutoff": 2, "0001": [2, 23], "01": [2, 23], "001": [2, 19, 23, 24, 44], "10": [2, 10, 12, 16, 17, 19, 20, 21, 23, 32, 36, 37, 38, 39], "upper_af": 2, "95": [2, 23, 24], "frequenc": [2, 3, 10, 16, 19, 21, 23, 24, 26, 34, 38], "string": [2, 3, 10, 13, 15, 16, 21, 22, 23, 24, 25, 26, 31, 34, 38, 39, 42], "annot": [2, 3, 4, 9, 10, 12, 15, 16, 17, 19, 20, 21, 22, 24, 26, 28, 30, 34, 36, 37, 38, 39, 40, 42, 43, 44, 45, 47, 49], "base": [2, 15, 16, 19, 20, 21, 22, 23, 24, 26, 29, 33, 34, 36, 37, 38, 39, 40, 43, 44], "input": [2, 3, 10, 13, 16, 17, 19, 20, 21, 22, 23, 24, 25, 26, 27, 29, 30, 33, 36, 38, 39, 40, 42, 43, 44, 45], "ac": [2, 3, 10, 21, 23, 24, 26, 36, 38, 42, 43], "af": [2, 3, 10, 19, 23, 24, 26, 38], "default": [2, 3, 5, 6, 9, 10, 12, 15, 16, 17, 19, 20, 21, 22, 23, 24, 25, 26, 27, 32, 33, 34, 36, 37, 38, 39, 43, 44, 50], "i": [2, 3, 5, 6, 9, 10, 12, 13, 15, 16, 17, 19, 20, 21, 22, 23, 24, 25, 26, 31, 32, 33, 34, 36, 37, 38, 39, 40, 43, 44, 45, 47, 50], "becaus": [2, 17, 21, 23, 24, 37, 40], "assum": [2, 17, 19, 21, 22, 23, 24, 26, 27, 29, 30, 31, 39], "wa": [2, 13, 21, 23, 24, 38, 39, 43, 44, 48], "calcul": [2, 3, 10, 16, 17, 19, 21, 23, 24, 25, 27, 36, 37, 38, 40], "annotate_freq": [2, 23, 24, 28], "all": [2, 3, 10, 13, 15, 16, 17, 20, 21, 22, 23, 24, 25, 26, 29, 32, 33, 36, 37, 38, 39, 40, 42, 43, 44, 45, 50], "gen": [2, 23], "anc": [2, 23], "group": [2, 3, 10, 11, 16, 17, 21, 22, 23, 24, 26, 28, 34, 36, 37, 38, 40, 42, 43, 49], "adj": [2, 3, 10, 19, 23, 24, 26, 34, 38, 43], "genotyp": [2, 10, 17, 19, 21, 22, 23, 24, 26, 36, 38, 40], "onli": [2, 3, 10, 13, 16, 17, 19, 20, 21, 22, 23, 24, 25, 26, 33, 36, 37, 38, 39, 40, 43, 44, 47], "union": [2, 3, 5, 9, 12, 16, 17, 19, 21, 22, 23, 24, 25, 26, 29, 30, 31, 32, 33, 35, 36, 37, 38, 39], "structexpress": [2, 3, 17, 21, 22, 23, 24, 26, 27, 30, 31, 36, 37, 39], "arrayexpress": [2, 16, 21, 23, 24, 26, 27, 30, 36, 39], "arrai": [2, 3, 9, 12, 16, 23, 24, 26, 27, 29, 33, 34, 36, 37, 38, 39], "struct": [2, 3, 10, 17, 19, 22, 23, 24, 26, 27, 30, 31, 32, 36, 37, 38, 39, 43, 44], "int": [2, 3, 6, 13, 16, 17, 19, 20, 21, 22, 23, 24, 25, 26, 29, 31, 32, 34, 36, 37, 38, 40, 43, 44], "which": [2, 3, 13, 16, 17, 19, 21, 22, 23, 24, 26, 30, 36, 37, 38, 39, 40, 42, 43, 47], "us": [2, 3, 5, 10, 15, 16, 17, 19, 20, 21, 22, 23, 24, 25, 26, 30, 32, 34, 35, 36, 37, 38, 39, 40, 43, 44, 45, 47, 48, 50], "option": [2, 3, 9, 10, 12, 13, 15, 16, 17, 19, 20, 21, 22, 23, 24, 25, 26, 31, 33, 34, 35, 36, 37, 38, 39, 40, 43, 44, 45], "str": [2, 3, 5, 6, 7, 10, 13, 15, 16, 17, 19, 20, 21, 22, 23, 24, 25, 26, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 42, 43, 44], "cutoff": [2, 3, 12, 19, 21, 22, 24, 38, 39], "bin": [2, 3, 23, 32, 36, 38, 40, 43], "float": [2, 3, 16, 17, 19, 20, 21, 22, 23, 24, 25, 26, 27, 32, 38, 39, 43, 44, 45], "none": [2, 3, 5, 9, 10, 12, 13, 15, 16, 17, 19, 20, 21, 22, 23, 24, 25, 26, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 42, 43, 44, 45], "upper": [2, 3, 17, 21, 22, 24, 32], "type": [2, 3, 5, 6, 7, 9, 10, 12, 13, 15, 16, 17, 19, 20, 21, 22, 23, 24, 25, 26, 27, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 43, 44, 45], "stringexpress": [2, 3, 21, 22, 23, 24, 36, 37, 39], "locus_expr": [2, 21, 22, 23, 26], "allele_expr": 2, "lof_expr": 2, "no_lof_flags_expr": 2, "most_severe_csq_expr": 2, "prefix_str": 2, "dictionari": [2, 3, 10, 15, 17, 21, 23, 24, 26, 34, 36, 38, 39, 40, 43, 44], "count": [2, 3, 21, 23, 24, 26, 32, 38, 40, 43, 44], "multipl": [2, 3, 16, 21, 25, 26, 27, 32, 39, 40, 50], "variant": [2, 3, 6, 10, 19, 21, 23, 24, 25, 26, 27, 31, 36, 37, 38, 39, 40, 42, 43, 44, 45, 46, 48, 49], "ar": [2, 3, 9, 10, 12, 13, 15, 16, 17, 19, 20, 21, 22, 23, 24, 25, 26, 29, 32, 33, 34, 36, 37, 38, 39, 40, 43, 44, 45, 48], "number": [2, 3, 10, 15, 16, 17, 19, 20, 21, 22, 23, 24, 25, 26, 29, 36, 37, 38, 40, 42, 43, 44, 49], "indel": [2, 9, 12, 17, 21, 23, 24, 27, 36, 38, 40, 43, 45], "snv": [2, 6, 17, 19, 21, 23, 27, 38, 40, 43], "lof": [2, 24, 37, 39], "pass": [2, 3, 10, 15, 16, 19, 23, 26, 36, 38, 39, 42, 43], "lofte": [2, 24, 39], "without": [2, 21, 23, 24, 26], "ani": [2, 3, 9, 15, 16, 17, 21, 23, 24, 25, 26, 36, 39, 40, 43, 44], "flg": 2, "other": [2, 3, 15, 16, 17, 19, 20, 23, 24, 26, 32, 38, 39, 47], "splice": [2, 37, 38, 39], "o": [2, 39], "fail": [2, 3, 21, 25, 30], "missens": [2, 24, 38], "synonym": [2, 24, 39], "autosom": [2, 12, 16, 17, 20, 21, 23, 24, 26, 27, 36, 38, 43], "allosom": 2, "two": [2, 3, 6, 15, 21, 22, 23, 24, 37, 38, 42, 43], "multi": [2, 17, 19, 21, 23, 27, 36, 38, 45], "allel": [2, 3, 10, 16, 17, 19, 20, 21, 23, 24, 26, 27, 30, 31, 33, 34, 36, 37, 38, 39, 40, 42, 43, 45], "have": [2, 3, 16, 17, 20, 21, 23, 24, 25, 36, 39, 40, 43], "been": [2, 21, 23, 24, 39, 40], "split": [2, 10, 19, 21, 22, 23, 24, 26, 36, 37, 38], "locusexpress": [2, 21, 22, 23, 26, 30, 33], "booleanexpress": [2, 3, 17, 21, 23, 24, 26, 36, 39, 40, 43, 45], "indic": [2, 3, 7, 17, 21, 23, 24, 25, 34, 36, 37, 39, 43, 44, 45], "whether": [2, 3, 15, 16, 17, 19, 21, 23, 24, 25, 26, 30, 32, 36, 37, 38, 39, 40, 44], "ha": [2, 3, 17, 19, 20, 21, 22, 23, 24, 26, 31, 36, 37, 39, 40, 44, 45], "flag": [2, 3, 23, 24, 26, 39, 47, 48], "most": [2, 23, 26, 36, 39, 47, 48], "sever": [2, 24, 39, 47], "consequ": [2, 26, 37, 38, 39, 47], "desir": [2, 3, 10, 20, 21, 23, 25, 26, 30, 36, 38, 45], "prefix": [2, 3, 13, 17, 22, 35, 38, 44], "empti": [2, 3, 6, 23, 24, 32, 38, 39, 44], "dict": [2, 3, 13, 15, 17, 21, 23, 24, 26, 32, 34, 36, 38, 40, 43, 44], "int64express": [2, 3, 23, 24], "per": [2, 3, 23, 24, 26, 36, 37, 38, 39, 43], "ac_expr": 2, "total": [2, 3, 23, 24, 29, 36, 38], "ht": [2, 3, 9, 10, 12, 15, 16, 17, 20, 21, 22, 23, 24, 25, 26, 30, 32, 36, 37, 38, 39, 40, 43, 44, 45, 48], "freq_field": 2, "freq": [2, 3, 10, 23, 24, 26, 34, 38], "filter_field": 2, "filter_decoi": [2, 19, 26], "fals": [2, 3, 10, 15, 17, 19, 21, 23, 24, 25, 26, 29, 32, 33, 36, 37, 38, 39, 40, 43, 44, 45], "canonical_onli": 2, "true": [2, 3, 15, 16, 17, 19, 20, 21, 23, 24, 25, 26, 30, 32, 33, 36, 37, 38, 39, 40, 43, 45, 47], "mane_select_onli": 2, "summari": [2, 3, 39], "across": [2, 3, 6, 10, 23, 24, 30, 36, 37, 38], "includ": [2, 3, 9, 10, 12, 15, 17, 19, 21, 22, 23, 24, 26, 33, 35, 36, 37, 38, 39, 43, 44, 47], "also": [2, 3, 10, 15, 16, 17, 19, 21, 23, 24, 26, 39, 44, 45], "": [2, 3, 5, 6, 10, 15, 17, 19, 21, 22, 23, 24, 26, 29, 31, 35, 36, 37, 38, 39, 40, 43, 44, 47, 50], "global": [2, 3, 6, 21, 23, 24, 34, 36, 37, 38, 39, 40, 43, 44], "befor": [2, 10, 19, 21, 23, 24, 25, 36, 37, 38, 40, 43], "out": [2, 17, 19, 21, 23, 24, 25, 36, 37, 39, 43, 45], "low": [2, 21, 24, 26, 36, 38, 39], "confid": [2, 9, 12, 21, 24, 26, 38, 40, 43], "region": [2, 3, 19, 21, 23, 24, 26, 37, 38, 39], "canon": [2, 24, 38, 39], "transcript": [2, 6, 15, 24, 26, 37, 38, 39], "mane": [2, 38, 39], "select": [2, 10, 13, 15, 17, 23, 24, 25, 38, 39], "vep": [2, 3, 4, 23, 24, 26, 28, 37, 42, 46, 49], "multiallel": [2, 21, 24, 38], "bi": [2, 17, 19, 20, 21, 23, 26, 27, 40, 43], "field": [2, 3, 10, 15, 17, 19, 20, 21, 22, 23, 24, 26, 36, 37, 38, 39, 40, 43, 44, 47], "bool": [2, 3, 10, 15, 16, 17, 19, 20, 21, 23, 24, 25, 26, 29, 30, 32, 33, 36, 37, 38, 39, 40, 43, 44], "decoi": [2, 19, 23, 26, 38], "aggreg": [2, 3, 17, 23, 24, 32, 36, 37, 40, 43], "mt": [2, 15, 16, 19, 20, 21, 22, 23, 25, 26, 32, 36, 37, 39, 40, 43, 44], "samples_by_sex": 2, "meta_root": 2, "meta": [2, 3, 23, 24, 26], "sex_field": 2, "sex_imput": 2, "sex_karyotyp": [2, 19, 22], "xy_str": [2, 22, 23], "xy": [2, 3, 10, 19, 21, 22, 23, 34, 36, 38], "xx_str": [2, 22], "xx": [2, 3, 10, 19, 21, 22, 23, 26, 34, 36, 38], "freq_index": 2, "an_proportion_cutoff": 2, "8": [2, 16, 19, 25, 37, 38], "criteria": [2, 19, 21, 23, 24, 26, 39], "sampl": [2, 3, 10, 16, 17, 19, 20, 21, 22, 23, 24, 26, 34, 36, 37, 38, 40, 43, 44, 45], "proxi": 2, "call": [2, 3, 16, 17, 19, 20, 21, 23, 36, 38, 39], "rate": [2, 19, 20, 24, 39], "matrixt": [2, 3, 15, 16, 17, 19, 20, 21, 22, 23, 24, 25, 26, 30, 32, 36, 37, 38, 39, 40, 43, 44], "valu": [2, 3, 5, 9, 13, 15, 16, 17, 19, 20, 21, 22, 23, 24, 25, 26, 27, 32, 34, 36, 37, 38, 39, 40, 42, 44], "each": [2, 3, 10, 16, 17, 20, 21, 23, 24, 25, 26, 29, 32, 34, 36, 37, 38, 39, 40, 43, 44, 47], "sex": [2, 3, 4, 10, 11, 17, 18, 19, 21, 23, 26, 27, 28, 34, 36, 38, 49], "kei": [2, 3, 13, 16, 17, 20, 21, 23, 24, 26, 34, 36, 37, 38, 39, 40], "metadata": [2, 3, 23, 24, 26, 32, 34], "assign": [2, 16, 20, 22, 24, 38, 40], "defualt": 2, "mark": 2, "proport": [2, 6, 16, 24, 37], "key_expr": 2, "tx_ht": [2, 37], "csq_expr": [2, 39], "gene_field": 2, "ensg": 2, "csq_field": [2, 39], "csq": [2, 39], "tx_struct": 2, "tx_annot": 2, "pull": [2, 24, 37], "appropri": 2, "express": [2, 3, 6, 9, 12, 15, 16, 17, 21, 22, 23, 24, 26, 27, 30, 33, 36, 37, 39, 40, 43, 44, 45], "given": [2, 3, 7, 16, 19, 21, 22, 23, 24, 27, 33, 36, 37, 38, 39, 42, 44, 45], "specif": [2, 3, 6, 10, 15, 17, 21, 23, 24, 26, 32, 36, 38, 39, 40, 44, 47], "locu": [2, 3, 10, 20, 21, 22, 23, 26, 30, 31, 33, 36, 37, 38, 42], "provid": [2, 3, 5, 6, 9, 10, 12, 16, 17, 21, 23, 24, 26, 31, 32, 33, 36, 40, 45, 50], "both": [2, 3, 21, 22, 23, 24, 25, 26, 30, 36, 38, 40, 45], "row_key_expr": 2, "search": 2, "gene": [2, 3, 6, 10, 15, 24, 26, 37, 39, 42], "id": [2, 15, 16, 17, 21, 23, 24, 25, 26, 37, 38], "most_severe_consequ": [2, 24, 37, 39], "float64express": [2, 21, 23, 24, 27], "t": [2, 3, 15, 17, 19, 21, 23, 24, 25, 26, 30, 32, 36, 37, 38, 39, 42, 44], "filter_lcr": [2, 19, 26], "filter_expr": [2, 24, 26], "grpmax_expr": [2, 23, 28], "max_af": [2, 24], "max_grpmax": 2, "min_an_proport": 2, "stat": [2, 3, 19, 21, 22, 23, 24, 25, 27, 36, 43, 45], "The": [2, 3, 5, 9, 12, 13, 15, 16, 17, 20, 21, 22, 23, 24, 25, 26, 27, 31, 32, 34, 36, 37, 38, 39, 40, 43, 44, 45, 50], "possibl": [2, 3, 23, 24, 33, 38], "no_lcr": 2, "variant_qc_pass": 2, "grpmax": [2, 3, 10, 34, 38], "min_an": 2, "an_proport": 2, "setexpress": [2, 21, 26], "max": [2, 10, 21, 23, 38, 43], "maximum": [2, 13, 21, 23, 24, 26, 36, 38, 44], "can": [2, 3, 6, 13, 16, 17, 20, 21, 22, 23, 24, 25, 26, 29, 32, 33, 36, 39, 40, 43, 50], "singl": [2, 6, 10, 16, 17, 19, 21, 22, 25, 26, 29, 36, 37, 39, 40], "genet": [2, 3, 6, 10, 16, 21, 23, 24, 26, 34, 38, 42], "ancestri": [2, 3, 4, 6, 10, 18, 23, 24, 26, 34, 38, 42, 49], "minimum": [2, 16, 19, 21, 23, 24, 26, 37, 38], "lof_csq_set": [2, 28, 39], "lof_label_set": 2, "lof_no_flag": 2, "lof_any_flag": 2, "additional_csq_set": 2, "additional_csq": 2, "requir": [2, 15, 19, 21, 23, 24, 25, 26, 30, 36, 37, 39, 40, 48], "request": [2, 3, 15, 23, 24, 25, 26, 37, 40, 47, 50], "no_lof_flag": [2, 39], "loftee_no_flag": 2, "loftee_with_flag": 2, "loftee_label": [2, 28, 39], "lof_label": 2, "csq_set": 2, "set": [2, 3, 5, 13, 15, 16, 17, 19, 20, 21, 22, 23, 24, 25, 26, 29, 33, 34, 36, 37, 38, 39, 40, 43, 44, 45, 50], "thi": [2, 3, 9, 12, 13, 15, 16, 17, 19, 20, 21, 22, 23, 24, 25, 26, 29, 30, 31, 33, 36, 37, 38, 39, 40, 42, 43, 44, 45, 47, 48, 49, 50], "label": [2, 3, 10, 16, 21, 23, 24, 32, 34, 38, 39, 44, 45], "addit": [2, 15, 16, 19, 21, 23, 24, 26, 32, 36, 37, 39, 40, 43, 44, 48], "keep": [2, 19, 21, 23, 24, 26, 36, 37, 38, 39, 44], "combo": [2, 38], "combo_opt": 2, "combin": [2, 3, 17, 21, 22, 23, 24, 25, 26, 34, 36, 37, 38, 39, 40, 43], "exampl": [2, 3, 13, 21, 23, 24, 26, 27, 32, 34, 35, 37, 38, 40, 43, 44, 45, 48, 49, 50], "pass_filt": 2, "captur": [2, 24, 38], "ukb": 2, "broad": [2, 36, 38], "suppli": [2, 3, 16, 19, 22, 23, 24, 38, 39], "all_sum_stat_filt": 2, "common_filter_combo": 2, "common_filter_overrid": 2, "lof_filter_combo": 2, "lof_filter_overrid": 2, "filter_key_renam": 2, "variou": [2, 3], "statist": [2, 10, 23, 24, 36, 38, 39, 43], "It": [2, 22, 23, 24, 36, 37, 38, 39], "ensur": [2, 10, 21, 23, 38], "common": [2, 21, 30], "loss": [2, 24, 39], "variant_qc": [2, 4, 49], "remov": [2, 3, 10, 21, 23, 24, 25, 26, 32, 36, 39], "lead": [2, 37], "renam": 2, "after": [2, 17, 23, 24, 35, 36, 39, 43, 44], "creat": [2, 17, 19, 21, 22, 23, 24, 26, 31, 33, 34, 36, 37, 38, 39, 40, 43, 44, 47, 48], "follow": [2, 3, 9, 12, 13, 17, 19, 21, 22, 23, 24, 25, 26, 36, 37, 38, 39, 40, 43, 45], "lof_csq": 2, "stop_gain": [2, 37, 39], "loftee_hc": 2, "loftee_flag": 2, "hc": [2, 39], "with_flag": 2, "overrid": 2, "lof_combo": 2, "final": [2, 22, 40], "high_expression_cutoff": 2, "9": [2, 15, 16, 37, 38], "low_expression_cutoff": 2, "missense_vari": [2, 37, 39], "synonymous_vari": [2, 37, 39], "all_transcript": 2, "filter_an": 2, "filter_to_rar": 2, "pre_lofte": 2, "frameshift_vari": [2, 37, 39], "splice_acceptor_vari": [2, 37, 39], "splice_donor_vari": [2, 37, 39], "remove_ultra_common": 2, "matrix": [2, 7, 21, 23, 36, 42, 43, 44], "metric": [2, 3, 17, 21, 23, 24, 38], "level": [2, 9, 12, 15, 23, 24, 36, 37, 38, 39], "mean": [2, 9, 19, 21, 22, 23, 24, 36, 37, 38, 39, 50], "consid": [2, 3, 20, 21, 23, 24, 26], "highli": [2, 21, 23], "lowli": 2, "instead": [2, 10, 17, 19, 22, 23, 24, 26, 36, 37, 38, 39, 50], "just": [2, 26], "rare": 2, "5": [2, 3, 12, 16, 19, 21, 22, 23, 24, 36, 37, 38, 43, 44], "ultra": 2, "defined_sites_expr": 2, "num_homs_expr": 2, "num_hets_expr": 2, "gen_anc_expr": [2, 23], "site": [2, 3, 6, 9, 10, 12, 15, 19, 20, 23, 24, 36, 38, 40, 43, 47], "defin": [2, 3, 17, 21, 23, 24, 26, 36, 37, 38, 39, 40, 43, 45, 50], "heterozyg": [2, 3, 21, 24, 38], "homozyg": [2, 3, 19, 21, 22, 24, 38], "collapse_indel": 2, "tx": 2, "gen_anc_field": 2, "gen_anc": [2, 3, 16, 23, 24, 26, 32, 34, 38, 42], "filter_lofte": 2, "p": [2, 19, 21, 23, 32, 38], "haplotyp": [2, 38], "carri": 2, "put": 2, "plof": [2, 24], "observ": [2, 19, 21, 23, 24, 26, 38], "expect": [2, 3, 15, 17, 19, 21, 23, 24, 26, 36, 37, 38, 43], "oe": [2, 24], "ratio": [2, 21, 23, 24, 38, 43, 45], "abov": [2, 19, 21, 22, 23, 24, 26, 36], "were": [2, 12, 21, 23, 24, 26, 27, 38, 43, 44], "If": [2, 3, 5, 15, 16, 17, 19, 20, 21, 22, 23, 24, 25, 26, 29, 31, 32, 33, 35, 36, 37, 38, 39, 40, 43, 44, 45, 50], "row": [2, 3, 19, 21, 23, 24, 26, 32, 36, 37, 38, 39, 40, 43, 44, 45, 47], "collaps": [2, 24], "data": [2, 3, 5, 6, 9, 10, 12, 15, 16, 17, 19, 21, 23, 26, 27, 30, 32, 34, 35, 36, 37, 38, 39, 40, 43, 44, 47, 48, 50], "top": [2, 3, 37, 39], "het": [2, 19, 21, 22, 23, 36], "hom": [2, 19, 22, 23], "suffix": [3, 21, 38], "custom": [3, 10, 49], "logrecord": 3, "factori": 3, "append": [3, 21, 38], "function_nam": 3, "temporarili": 3, "modifi": [3, 23, 24, 39], "check_descript": 3, "display_field": 3, "cond_expr": 3, "verbos": 3, "show_percent_sit": 3, "n_fail": 3, "ht_count": 3, "logic": [3, 38], "condit": [3, 21], "involv": 3, "hail": [3, 5, 9, 10, 12, 15, 16, 17, 21, 22, 23, 24, 25, 30, 31, 32, 33, 39, 40, 43, 44, 47, 48, 49, 50], "when": [3, 15, 16, 17, 19, 20, 21, 22, 23, 24, 25, 26, 34, 36, 37, 38, 39, 47, 48], "absent": 3, "print": [3, 25, 44], "stdout": 3, "displai": [3, 26, 32], "percent": [3, 25, 38], "either": [3, 10, 16, 21, 23, 24, 25, 26, 50], "previous": [3, 16, 21, 23], "comput": [3, 16, 17, 19, 20, 21, 23, 24, 27, 36, 38, 40, 43, 44], "match": [3, 20, 30, 37, 39, 44], "otherwis": [3, 5, 16, 21, 23, 24, 29, 31, 38, 50], "opposit": 3, "should": [3, 16, 17, 20, 21, 23, 24, 25, 26, 32, 33, 36, 37, 38, 39, 40, 43, 44, 45], "never": 3, "same": [3, 17, 20, 21, 23, 24, 25, 32, 36, 39, 40, 44], "e": [3, 6, 10, 13, 15, 16, 17, 20, 21, 23, 24, 26, 31, 34, 36, 38, 39, 40, 44], "g": [3, 10, 15, 17, 20, 21, 23, 24, 25, 26, 30, 31, 32, 34, 36, 38, 39, 40, 44, 50], "instanc": [3, 17], "where": [3, 16, 21, 23, 24, 26, 33, 34, 36, 37, 38, 39, 40, 44], "raw": [3, 10, 21, 23, 24, 34, 38, 43], "less": [3, 9, 17, 23, 24, 38, 39], "than": [3, 9, 16, 17, 21, 22, 23, 24, 25, 26, 38, 39, 45], "greater": [3, 9, 17, 24, 25, 26, 45], "equal": [3, 9, 23, 24, 26, 39, 40, 43], "being": [3, 19, 21, 23, 24, 38, 50], "case": [3, 15, 16, 21, 23, 26, 36, 40], "failur": [3, 25], "troubleshoot": 3, "purpos": [3, 17], "refer": [3, 17, 19, 21, 23, 24, 25, 26, 30, 31, 33, 36, 37, 38, 39, 42, 47, 48, 49], "show": [3, 23], "percentag": 3, "obtain": [3, 10, 24, 45], "within": [3, 17, 21, 23, 24, 32, 36, 38, 39, 40, 42], "perform": [3, 10, 21, 23, 24, 36, 38, 39, 40, 43], "initi": [3, 25], "field_check_expr": 3, "loop": [3, 21], "through": [3, 23, 50], "allow": [3, 37, 50], "onc": [3, 10, 21, 23, 25], "oppos": 3, "dure": [3, 10, 17, 38, 50], "everi": [3, 36], "whose": [3, 21], "sum": [3, 21, 23, 24, 36, 37, 38], "left_expr": 3, "right_expr": 3, "oper": [3, 23, 26], "comparison": 3, "while": [3, 39], "handl": [3, 23, 37], "miss": [3, 15, 21, 22, 23, 24, 26, 32, 37, 39, 44], "retur": 3, "one": [3, 15, 16, 17, 19, 21, 22, 23, 24, 26, 30, 32, 33, 36, 37, 38, 39, 40, 43, 50], "numericexpress": [3, 9, 12, 17, 19, 21, 22, 23, 24, 26, 27, 36, 37, 40], "left": [3, 26], "right": [3, 26], "extra_filter_check": 3, "variant_filter_field": 3, "rf": [3, 16, 38, 43, 44], "make": [3, 25, 34, 38, 43], "measur": [3, 39], "under": [3, 21, 23, 24, 37, 38], "vari": [3, 24], "interest": [3, 25, 39], "fraction": [3, 19, 22, 23, 36], "due": 3, "inbreed": [3, 19, 23, 38], "coeffici": [3, 19, 21, 23, 38], "examin": [3, 24], "extra": [3, 24, 39], "filtrat": [3, 43], "vqsr": [3, 36, 38, 43], "as_vqsr": [3, 38], "subset": [3, 10, 11, 17, 23, 24, 26, 34, 36, 38, 40, 45], "label_group": [3, 38], "sort_ord": [3, 28, 34, 38], "downsampl": [3, 10, 11, 23, 24, 26, 34, 38], "popmax": [3, 34, 38], "pop": [3, 7, 24, 34, 38], "subgrp": [3, 6, 26, 34, 38], "subpop": [3, 34, 38], "delimit": [3, 15, 34, 38, 39], "_": [3, 34, 38], "metric_first_field": 3, "nhomalt": [3, 38], "specifi": [3, 10, 17, 19, 20, 21, 23, 24, 25, 26, 32, 36, 37, 38, 39, 43, 44], "compar": [3, 12, 40], "version": [3, 6, 9, 10, 12, 13, 15, 16, 21, 23, 24, 26, 36, 37, 39, 48], "For": [3, 12, 15, 17, 21, 23, 24, 25, 26, 34, 37, 38, 39, 40, 43, 48, 50], "subset1": 3, "consist": [3, 36], "gen_anc1": 3, "gen_anc2": 3, "gen_anc3": 3, "entri": [3, 17, 19, 20, 22, 23, 26, 28, 34, 36, 37, 38], "afr": [3, 10, 23, 24, 26, 34, 38], "nfe": [3, 6, 10, 23, 24, 34, 38, 42], "amr": [3, 10, 34, 38], "order": [3, 10, 19, 21, 23, 24, 26, 37, 38, 39, 40, 43, 44], "sort": [3, 10, 29, 34, 38, 39], "preced": [3, 38], "hgdp": [3, 10], "ht1": 3, "ht2": 3, "first": [3, 19, 21, 23, 24, 25, 36, 39, 43], "second": [3, 21, 23, 24, 36, 38], "problematic_region": 3, "lcr": [3, 19, 23, 26, 38], "segdup": [3, 19, 23, 26, 38], "nonpar": [3, 23, 38], "single_filter_count": 3, "site_gt_check_expr": 3, "n_row": 3, "50": [3, 10, 17, 36, 39], "n_col": 3, "140": [3, 32], "summar": [3, 37], "problemat": [3, 19, 23], "run": [3, 5, 16, 17, 19, 20, 21, 23, 24, 35, 43, 44, 50], "explod": [3, 21, 24, 39], "give": [3, 17, 21, 23, 45], "supplement": 3, "boolean": [3, 10, 17, 21, 23, 24, 25, 26, 40], "typic": [3, 15, 16, 19, 21, 23, 24, 25, 36], "how": [3, 21, 26, 38], "mani": [3, 21], "monoallel": [3, 38], "100": [3, 10, 13, 16, 21, 23, 24, 36, 37, 40, 43], "callset": [3, 24, 38, 40], "eg": 3, "overal": [3, 38], "non_v2": [3, 10], "between": [3, 6, 9, 21, 24, 26, 37, 38, 39, 40, 42, 43], "entir": 3, "gen_anc_group": [3, 10, 11, 23, 34], "asj": [3, 10, 34, 38], "ea": [3, 6, 10, 34, 38], "fin": [3, 10, 23, 34, 38], "mid": [3, 10, 21, 34, 38], "remain": [3, 10, 21, 34, 38], "sa": [3, 10, 34, 38], "additional_subsets_and_gen_anc_group": 3, "gen_anc_label_nam": 3, "present": [3, 13, 16, 21, 23, 24, 25, 31, 34, 36, 38, 39, 40, 43], "test": [3, 19, 23, 38, 43, 44], "current_major_releas": [3, 34], "exom": [3, 6, 10, 19, 20, 24, 25, 26, 30, 34, 38, 48], "callstat": [3, 23, 38], "doe": [3, 23, 26, 38, 45], "store": [3, 13, 15, 16, 17, 19, 23, 36, 38, 39, 44, 47], "denot": [3, 15], "expected_contig": 3, "get": [3, 5, 6, 7, 19, 21, 22, 23, 24, 25, 36, 37, 39, 40, 42, 43, 49], "chromosom": [3, 10, 17, 19, 21, 22, 23, 24, 26, 27, 31, 36, 38], "contig": [3, 6, 10, 26, 29, 31, 33, 36, 38, 40], "found": [3, 23, 24, 26, 34, 44], "unexpect": 3, "nhomalt_metr": 3, "neg": [3, 21, 23, 24, 38, 39, 43, 45], "correspond": [3, 9, 12, 17, 21, 23, 24, 34, 36, 37, 38, 39, 42, 45], "must": [3, 9, 12, 15, 19, 21, 22, 23, 24, 26, 36, 38, 39, 40, 43, 47, 50], "altern": [3, 19, 21, 22, 23, 24, 31, 36, 38, 50], "info_metr": 3, "That": 3, "chry": [3, 10, 12, 23, 24, 38], "na": [3, 15, 23, 38, 44], "non": [3, 12, 15, 17, 20, 21, 22, 23, 24, 26, 27, 36, 37, 38, 44], "par": [3, 12, 21, 22, 23, 24, 26, 38], "chrx": [3, 12, 23, 24, 38], "non_info_metr": 3, "n_site": 3, "missingness_threshold": 3, "amount": [3, 24, 36], "missing": [3, 23], "sdout": 3, "exce": 3, "minu": 3, "header_dict": 3, "row_annot": 3, "entry_annot": 3, "hist": [3, 23, 28, 32, 38], "gq_hist_alt": [3, 38], "gq_hist_al": [3, 38], "dp_hist_alt": [3, 38], "dp_hist_al": [3, 38], "ab_hist_alt": [3, 38], "vcf": [3, 4, 6, 10, 15, 17, 23, 25, 28, 34, 36, 39, 47, 49], "descript": [3, 38, 39], "header": [3, 25, 26, 38, 39], "export": [3, 10, 23, 25, 34, 38, 39], "histogram": [3, 23, 32, 38], "row_to_globals_check": 3, "check_all_row": 3, "length": [3, 23, 24, 29, 38, 43], "pretti": 3, "sum_metr": 3, "sample_sum_sets_and_gen_anc_group": 3, "summarize_variants_check": 3, "filters_check": 3, "raw_adj_check": 3, "subset_freq_check": 3, "samples_sum_check": 3, "sex_chr_check": 3, "missingness_check": 3, "pprint_glob": 3, "check_all_rows_in_row_to_global_check": 3, "batteri": 3, "differ": [3, 17, 21, 23, 38, 39, 50], "partit": [3, 16, 19, 24, 25, 36], "outlier": [3, 17, 24, 26], "subgroup": [3, 10, 23, 26, 38], "add": [3, 10, 23, 24, 26, 29, 30, 33, 37, 39, 40, 43, 47], "up": [3, 34, 47, 48], "supergroup": 3, "main": [3, 13, 14, 21, 23], "regardless": 3, "violat": 3, "method": [3, 10, 15, 17, 21, 23, 36, 38], "check_sex_chr_metricss": 3, "vep_ht": [3, 42], "interval_ht": [3, 36], "interv": [3, 4, 19, 20, 24, 25, 26, 28, 31, 33, 36, 37, 39, 40, 43, 44, 49], "transcript_consequ": [3, 24, 37, 39], "biotyp": [3, 39], "determin": [3, 5, 12, 17, 19, 21, 22, 23, 24, 36, 38, 40, 43, 45, 50], "protein": [3, 24, 26, 37, 38, 39], "code": [3, 21, 23, 24, 26, 35, 37, 39], "gene_stable_id": 3, "ensembl": [3, 38, 39, 47], "releas": [3, 4, 6, 10, 17, 28, 37, 38, 39, 49], "ad": [3, 10, 21, 22, 23, 24, 26, 29, 30, 33, 37, 38, 39, 40, 43, 44], "n_total_vari": 3, "n_pcg_variant": 3, "struct_expr": 3, "recurs": [3, 23], "standalon": 3, "nest": [3, 23, 24, 37], "element": [3, 16, 23, 24, 25, 26, 34, 37, 39, 40], "map": [3, 10, 23, 30, 32, 36, 38], "repres": [3, 22, 23, 24, 36, 38], "missingness_struct": 3, "flatten": [3, 21], "evalu": [3, 4, 21, 26, 41, 43, 49], "array_meta_dict": 3, "freq_meta": [3, 23, 24, 26, 34, 36], "sorted_kei": 3, "unfurl": 3, "content": [3, 12, 35, 44], "retired_term": 3, "oth": [3, 10, 16, 38], "popul": [3, 7, 21, 23, 38], "see": [3, 23, 25, 26, 37, 39, 40, 48, 50], "retir": 3, "term": [3, 21, 24, 38, 39], "util": [4, 43, 47, 49], "constraint": [4, 6, 8, 10, 11, 28, 39, 49], "file_util": [4, 28, 49], "gen_stat": [4, 28, 49], "liftov": [4, 6, 8, 12, 28, 49], "plot": [4, 28, 49], "reference_genom": [4, 10, 25, 26, 28, 31, 42, 47, 49], "slack": [4, 28, 49], "sparse_mt": [4, 28, 49], "transcript_annot": [4, 28, 49], "resourc": [4, 23, 26, 37, 38, 39, 48, 49], "config": [4, 14, 39, 49, 50], "grch37": [4, 10, 12, 13, 14, 26, 30, 31, 33, 39, 42, 47, 48, 49], "grch38": [4, 14, 19, 25, 26, 30, 31, 33, 38, 39, 48, 49], "import_resourc": [4, 14, 15, 49], "resource_util": [4, 14, 49], "ld": [4, 7, 19, 41, 49], "pipelin": [4, 18, 24, 25, 41, 44, 49], "random_forest": [4, 41, 49], "train": [4, 16, 38, 41, 43, 44, 49], "sample_qc": [4, 49], "platform": [4, 18, 21, 23, 24, 36, 49], "related": [4, 18, 43, 49], "assess": [4, 49], "configur": [5, 39, 47, 49], "load": [5, 6, 16, 21, 30, 37, 44, 50], "class": [5, 15, 35], "gnomadpublicresourcesourc": [5, 14, 50], "modul": [5, 13, 21, 24, 31, 37, 43], "qualnam": 5, "start": [5, 23, 24, 26, 29, 31, 32, 37, 38, 39, 47, 49, 50], "boundari": [5, 23, 32, 36, 39], "public": [5, 6, 10, 15, 48, 50], "google_cloud_public_dataset": 5, "googl": [5, 25, 47, 48, 50], "cloud": [5, 25, 48, 50], "dataset": [5, 10, 17, 21, 23, 24, 25, 26, 34, 36, 38, 47, 50], "registry_of_open_data_on_aw": 5, "registri": 5, "open": 5, "aw": 5, "get_default_public_resource_sourc": [5, 14], "gnomad_default_public_resource_sourc": [5, 50], "environ": [5, 39, 49], "variabl": [5, 13, 21, 39, 50], "spark": [5, 44, 50], "public_releas": [6, 8, 10, 11, 48], "data_typ": [6, 10, 30], "retriev": [6, 10, 23], "publicli": [6, 10], "One": [6, 9, 10, 12, 23, 26, 30, 37, 38], "genom": [6, 9, 10, 12, 24, 25, 26, 30, 31, 33, 36, 37, 38, 39, 42, 43, 47, 48, 49], "versionedtableresourc": [6, 10, 14, 15, 39], "coverag": [6, 8, 10, 11, 19, 24, 36, 38], "38": 6, "v2": [6, 30, 48], "public_pca_load": [6, 8], "tableresourc": [6, 14, 15], "pca": [6, 16, 20], "gnomadpublictableresourc": [6, 7, 10, 14, 15], "release_vcf_path": [6, 8, 10, 11], "20": [6, 10, 12, 19, 21, 23, 36, 38, 39], "y": [6, 19, 21, 22, 23, 24, 26, 32, 36, 38], "pext": [6, 8, 10, 11, 37], "pext_typ": [6, 10], "base_level": [6, 10], "annotation_level": [6, 10], "mnv": [6, 8], "distanc": [6, 17, 39], "multinucleotid": 6, "browser_gen": [6, 8, 10, 11], "browser": [6, 10, 37], "ld_matrix": [7, 8], "gnomadpublicblockmatrixresourc": [7, 14, 15], "ld_index": [7, 8, 42], "ld_score": [7, 8], "score": [7, 9, 12, 16, 17, 20, 24, 26, 38, 39, 40, 43], "gnomad_ld": [8, 14], "reference_data": [8, 11, 14, 26], "get_truth_ht": [8, 9, 11, 12], "transform_grch37_methyl": [8, 9, 24], "latest": [9, 12], "truth": [9, 12, 40, 43], "hapmap": [9, 12, 43], "kgp_omni": [9, 12], "1000": [9, 10, 12, 23, 26, 31, 39, 43], "intersect": [9, 12], "onni": [9, 12], "5m": [9, 12], "kgp_phase_1_hc": [9, 12], "high": [9, 10, 12, 19, 21, 23, 24, 26, 38, 39, 40, 43], "genonm": [9, 12], "mill": [9, 12, 43], "devin": [9, 12], "A": [9, 12, 13, 15, 16, 17, 20, 21, 23, 24, 25, 26, 36, 37, 38, 39, 42, 43, 44, 45], "popular": [9, 12], "methylation_expr": [9, 12, 24], "transform": [9, 12, 20, 23, 24, 36], "methyl": [9, 12, 24], "scale": [9, 12, 21, 23, 24, 27, 36, 38], "rang": [9, 10, 12, 24, 26, 36, 38], "we": [9, 21, 24, 36, 38, 40], "convert": [9, 16, 22, 23, 24, 35, 37, 38], "6": [9, 12, 19, 22, 37, 38], "v3": [10, 38, 48], "non_topm": 10, "non_canc": 10, "controls_and_biobank": 10, "non_neuro": 10, "tgp": 10, "v4": [10, 24, 38, 39], "non_ukb": [10, 34], "v5": 10, "non_aou": 10, "tgp_raw_ac_esn_xx": 10, "qualiti": [10, 21, 23, 24, 36, 38], "stratifi": [10, 21, 23, 24, 36, 38, 40, 43], "ami": [10, 38], "cohorts_with_gen_anc_stored_as_subgrp": [10, 11], "broken": [10, 17, 44], "down": [10, 44, 48], "known": [10, 16, 19, 44], "tgp_gen_anc_group": [10, 11], "esn": 10, "pur": 10, "pjl": 10, "clm": 10, "jpt": 10, "chb": 10, "stu": 10, "itu": 10, "tsi": 10, "mxl": 10, "ceu": 10, "msl": 10, "yri": 10, "beb": 10, "khv": 10, "cdx": 10, "lwk": 10, "acb": 10, "asw": 10, "ib": 10, "gbr": [10, 38], "pel": 10, "gih": 10, "ch": 10, "gwd": 10, "project": [10, 15, 16, 23, 25, 38, 47, 49, 50], "1kg": 10, "hgdp_gen_anc_group": [10, 11], "japanes": [10, 38], "papuanhighland": 10, "papuansepik": 10, "adygei": 10, "orcadian": 10, "biaka": 10, "yakut": 10, "han": [10, 26], "northernhan": 10, "uygur": 10, "miao": 10, "mongolian": 10, "balochi": 10, "bedouin": 10, "russian": 10, "daur": 10, "pima": 10, "hezhen": 10, "sindhi": 10, "yi": 10, "oroqen": 10, "san": 10, "tuscan": 10, "tu": 10, "palestinian": 10, "tujia": 10, "druze": 10, "pathan": 10, "basqu": 10, "makrani": 10, "bergamoitalian": 10, "naxi": 10, "karitiana": 10, "sardinian": 10, "mbuti": 10, "mozabit": 10, "yoruba": 10, "lahu": 10, "dai": 10, "cambodian": 10, "bougainvil": 10, "french": 10, "brahui": 10, "hazara": 10, "bantusouthafrica": 10, "surui": 10, "mandenka": 10, "kalash": 10, "xibo": 10, "colombian": 10, "bantukenya": 10, "she": 10, "burusho": 10, "maya": 10, "human": [10, 24, 38], "divers": 10, "tgp_gen_anc_group_nam": [10, 11], "african": [10, 23, 38], "caribbean": 10, "american": [10, 23, 38], "bengali": 10, "chines": 10, "utah": 10, "resid": 10, "european": [10, 38], "southern": [10, 38], "esan": 10, "finnish": [10, 38], "british": [10, 38], "gujarati": 10, "gambian": 10, "iberian": 10, "indian": 10, "telugu": 10, "kinh": 10, "luhya": 10, "mend": 10, "mexican": 10, "peruvian": 10, "punjabi": 10, "puerto": 10, "rican": 10, "sri": 10, "lankan": 10, "tamil": 10, "toscani": 10, "gen_anc_groups_to_remove_for_grpmax": [10, 11], "200": [10, 23, 36], "500": [10, 17, 20, 39, 43, 44], "2000": [10, 26, 31], "5000": [10, 24, 36], "10000": [10, 24], "15000": 10, "20000": 10, "25000": 10, "30000": 10, "40000": 10, "50000": 10, "60000": 10, "70000": 10, "75000": 10, "80000": 10, "85000": 10, "90000": 10, "95000": 10, "100000": 10, "110000": 10, "120000": 10, "200000": 10, "500000": 10, "59355": 10, "89074": 10, "223733": 10, "joint": [10, 38], "all_sites_an": [10, 11], "coverage_tsv_path": [10, 11], "chr20": [10, 19, 36, 38], "chr1": [10, 26, 31, 37, 38], "add_grpmaxfaf95_v4": [10, 11], "grpmaxfaf95": 10, "grpmax_gen_anc": 10, "jointgrpmaxfaf95": 10, "fafmax": [10, 38], "joint_fafmax": 10, "structur": [10, 23, 38], "gnomad_gk": [10, 11], "locus_interv": 10, "by_gen_anc_group": 10, "by_sex": [10, 23], "vrs_onli": 10, "custom_ht": 10, "skip_checkpoint": 10, "skip_coverag": 10, "custom_coverage_ht": 10, "gk": [10, 23], "intervalexpress": [10, 25, 26, 29, 31, 33], "hl": [10, 15, 17, 19, 21, 23, 24, 26, 32, 36, 37, 39, 42, 43, 47], "6424776": 10, "6461367": 10, "read": [10, 15, 23, 25, 36, 38, 47, 50], "cohort": [10, 23, 34, 38], "vr": [10, 23, 38], "what": [10, 24, 43], "would": [10, 21, 24, 34, 37, 44], "skip": [10, 23, 24], "checkpoint": [10, 19, 40], "mai": [10, 24, 26, 39, 48], "larg": [10, 17, 21, 23, 26, 36, 42], "reduc": 10, "copi": [10, 23], "cluster": [10, 20, 39, 47, 48, 50], "browser_vari": [10, 11], "transform_grch38_methyl": [11, 12, 24], "15": [12, 21, 36], "chen": 12, "et": [12, 21, 24], "al": [12, 21, 24], "http": [12, 21, 23, 25, 37], "www": 12, "biorxiv": 12, "org": [12, 23], "1101": 12, "2022": [12, 48], "03": 12, "485034v2": 12, "full": [12, 15, 21, 23, 24, 40], "12": [12, 24], "directli": 12, "analyz": 12, "separ": [12, 16, 17, 23, 24, 27, 38, 40], "rel": [12, 21, 22, 38, 44], "translat": [12, 39, 49], "correl": [12, 42], "propos": 12, "3": [12, 21, 23, 24, 26, 32, 37, 40, 44], "4": [12, 16, 17, 21, 23, 34, 37, 38, 40], "get_module_importable_resourc": [13, 14], "take": [13, 16, 21, 23, 24, 26, 32, 43], "import": [13, 15, 26, 43, 44, 47, 48, 50], "import_func": [13, 15], "produc": [13, 21, 23], "resource_nam": 13, "grch37_resourc": 13, "baseresourc": [13, 14, 15], "get_resources_descript": [13, 14], "width": [13, 32], "along": [13, 21], "thei": [13, 21, 23, 25, 27, 36, 37, 38, 40], "line": [13, 21, 25, 32, 37], "arg": 13, "gnomad_public_bucket": [14, 15], "matrixtableresourc": [14, 15], "variantdatasetresourc": [14, 15], "pedigreeresourc": [14, 15], "blockmatrixresourc": [14, 15], "expressionresourc": [14, 15], "baseversionedresourc": [14, 15], "versionedmatrixtableresourc": [14, 15], "versionedvariantdatasetresourc": [14, 15], "versionedpedigreeresourc": [14, 15], "versionedblockmatrixresourc": [14, 15], "resourcenotavail": [14, 15], "gnomadpublicresourc": [14, 15], "gnomadpublicmatrixtableresourc": [14, 15], "gnomadpublicpedigreeresourc": [14, 15], "dataexcept": [14, 15], "import_sites_vcf": [14, 15], "import_gencod": [14, 15], "pai": [15, 25, 47, 50], "bucket": [15, 25, 32, 47, 50], "stage": [15, 36, 44], "legaci": 15, "readm": 15, "text": [15, 21, 25, 35, 38], "product": 15, "team": 15, "write": [15, 17, 25], "sync": 15, "gcp": 15, "import_arg": 15, "abstract": 15, "need": [15, 17, 23, 24, 36, 38, 40], "kept": [15, 19, 21, 26, 36, 37], "track": 15, "callabl": [15, 16, 23, 24, 26, 32, 36, 37, 39], "kwarg": [15, 26, 32, 36, 37, 39], "expected_file_extens": 15, "extens": [15, 25], "doesn": [15, 19, 25, 30, 36], "end": [15, 23, 24, 25, 26, 29, 31, 32, 36, 38, 39], "properti": 15, "overwrit": [15, 25, 44], "its": [15, 23, 40], "exist": [15, 19, 24, 25, 38, 44], "destin": [15, 30], "underli": 15, "accept": 15, "depend": [15, 16, 21, 24], "force_import": 15, "read_arg": 15, "forc": 15, "even": [15, 17, 23], "alreadi": [15, 23, 24, 25, 29, 30, 39], "argument": [15, 23, 26, 33, 36, 37, 44, 48], "read_tabl": 15, "noth": [15, 44], "read_matrix_t": 15, "overwritten": 15, "variantdataset": [15, 17, 19, 21, 23, 25, 26, 36], "vd": [15, 19, 25, 26, 36], "read_vd": 15, "quant_pheno": 15, "pedigre": [15, 21], "fam": 15, "ped": 15, "phenotyp": 15, "interpret": [15, 26, 38, 39], "quantit": 15, "regex": 15, "control": [15, 36], "numer": [15, 17, 21, 23, 26, 32, 44], "treat": [15, 23, 36, 39], "famili": [15, 21, 38, 43], "import_fam": [15, 21], "current": [15, 19, 22, 23, 26, 38, 39], "implement": [15, 19, 23], "blockmatrix": [15, 42], "bm": [15, 25, 42], "he": [15, 25], "experiment": [15, 21, 43], "read_express": 15, "write_express": 15, "default_vers": 15, "attribut": [15, 38], "those": [15, 16, 19, 21, 23, 24, 36, 39, 45], "In": [15, 16, 19, 23, 24, 38, 40, 43], "resource_class": 15, "alia": 15, "param": [15, 26], "except": [15, 25, 32, 35], "rais": [15, 17, 21, 23, 25, 35], "avail": [15, 23, 50], "is_resource_avail": 15, "publish": 15, "gtf_path": 15, "include_vers": 15, "gencod": [15, 24, 26, 37], "gtf": 15, "gene_id_vers": [15, 24], "transcript_id_vers": [15, 24], "pc_project": [16, 18], "loadings_ht": [16, 20], "loading_loc": 16, "af_loc": 16, "pca_af": 16, "pre": [16, 50], "pc": [16, 17, 20, 21], "locat": [16, 24, 38, 44], "apply_onnx_classification_model": [16, 18], "data_pd": 16, "fit": [16, 24], "appli": [16, 19, 21, 23, 24, 26, 36, 39, 43, 44], "onnx": 16, "classif": 16, "model": [16, 19, 22, 24, 38, 43, 44], "panda": 16, "datafram": [16, 32, 44], "classifi": 16, "modelproto": 16, "ndarrai": 16, "probabl": [16, 21, 24, 38, 39, 44], "apply_sklearn_classification_model": [16, 18], "sklearn": 16, "convert_sklearn_rf_to_onnx": [16, 18], "target_opset": 16, "random": [16, 21, 38, 40, 43, 44], "forest": [16, 38, 43, 44], "target": [16, 24], "opset": 16, "assign_genetic_ancestry_pc": [16, 18], "gen_anc_pca_scor": 16, "pc_col": 16, "known_col": 16, "known_label": 16, "seed": 16, "42": [16, 21], "prop_train": 16, "n_estim": 16, "min_prob": 16, "output_col": 16, "missing_label": 16, "pc_expr": 16, "convert_model_func": 16, "apply_model_func": 16, "n_partit": [16, 19, 36], "As": 16, "hwe_normalized_pca": 16, "integ": [16, 23, 24, 38], "you": [16, 23, 37, 39, 48], "expand_pd_array_col": 16, "pc1": 16, "pc2": 16, "pc4": 16, "pc5": 16, "previou": 16, "randomforestclassifi": [16, 44], "tree": [16, 17, 43, 44], "belong": [16, 23, 37], "smaller": [16, 21], "convers": [16, 38, 44], "work": [16, 17, 21, 39], "repartit": [16, 19], "infer": [16, 19, 21, 23, 30, 31], "imput": [16, 19, 20, 36, 44], "run_pca_with_rel": [16, 18], "qc_mt": 16, "related_samples_to_drop": 16, "additional_samples_to_drop": 16, "n_pc": [16, 17, 20], "autosomes_onli": [16, 20, 43], "exclud": [16, 19, 21, 23, 24, 33, 36, 37, 45], "relat": [16, 21, 38, 43], "space": [16, 21, 38], "qc": [16, 17, 19, 38, 40, 43], "drop": [16, 21, 36, 38, 44], "analysi": [16, 24], "eigenvalu": [16, 20], "compute_qc_metrics_residu": [17, 18], "pc_score": 17, "qc_metric": 17, "use_pc_squar": 17, "regression_sample_inclusion_expr": 17, "strata": [17, 21, 23, 26, 36, 44], "residu": 17, "regress": [17, 24], "alwai": [17, 38], "arraynumericexpress": [17, 23, 24, 27, 36], "stratif": [17, 21, 23, 43, 44], "These": [17, 21, 22, 26, 39], "discret": 17, "compute_stratified_metrics_filt": [17, 18], "lower_threshold": 17, "upper_threshold": 17, "metric_threshold": 17, "filter_nam": 17, "qc_metrics_filt": 17, "comparison_sample_expr": 17, "median": [17, 24, 27, 36, 37, 44], "mad": [17, 27], "lower": [17, 21, 22, 23, 24, 32, 39], "threshold": [17, 19, 20, 21, 23, 24], "expr": [17, 23, 40], "critic": 17, "more": [17, 21, 22, 23, 24, 25, 26, 36, 37, 38, 39, 40, 48], "collectionexpress": 17, "well": [17, 23, 24, 30, 36, 38], "determine_nearest_neighbor": [17, 18], "compute_stratified_sample_qc": [17, 18], "mtd": [17, 19, 21, 26, 36], "tmp_ht_prefix": 17, "gt_col": 17, "merg": [17, 23, 27, 29], "overlap": [17, 20, 27, 29, 36], "v": [17, 21, 36, 38, 40], "intermedi": [17, 40], "recommend": [17, 22, 25, 36], "larger": [17, 21, 38], "gt": [17, 19, 21, 23, 26, 36, 38, 40], "strat": 17, "merge_sample_qc_expr": [17, 18], "sample_qc_expr": 17, "note": [17, 19, 21, 22, 23, 24, 27, 36, 38, 39, 40, 47, 48], "regard": 17, "dp_stat": 17, "gq_stat": 17, "n": [17, 21, 32, 33, 43, 44], "stdev": [17, 22], "n_call": [17, 19], "veri": 17, "standard": [17, 19, 21, 22, 24, 36, 38], "gatk": [17, 19, 23, 36], "essenti": 17, "dp": [17, 19, 21, 23, 36, 38], "gq": [17, 19, 21, 23, 38], "do": [17, 21, 23, 25, 36, 38], "assumpt": [17, 21], "some": [17, 21, 23, 24, 35, 36, 45], "shouldn": 17, "matter": 17, "too": [17, 19, 21, 26, 43], "much": [17, 21, 25], "scores_expr": 17, "n_neighbor": 17, "n_job": 17, "add_neighbor_dist": 17, "distance_metr": 17, "euclidean": 17, "use_approxim": 17, "n_tree": 17, "nearest": 17, "neighbor": 17, "limit": 17, "nearest_neighbor": 17, "nearest_neighbor_dist": 17, "identifi": [17, 21, 35, 38, 39, 47], "thread": 17, "find": [17, 39], "cpu": 17, "head": 17, "node": 17, "scikit": 17, "learn": 17, "cityblock": 17, "cosin": 17, "haversin": 17, "l1": 17, "l2": 17, "manhattan": 17, "annoi": 17, "angular": 17, "ham": 17, "dot": 17, "packag": [17, 47, 48, 49, 50], "approxim": [17, 21, 36, 38, 40, 44], "nearestneighbor": 17, "faster": [17, 26, 38, 39], "000": 17, "approach": [17, 21], "build": [17, 23, 24, 26, 30, 31, 37, 38, 39], "time": [17, 21, 24, 36, 48], "affect": [17, 21], "size": [17, 19, 23, 24, 25, 32, 34, 36, 37, 43], "accur": [17, 36], "filter_rows_for_qc": [18, 19], "get_qc_mt": [18, 19], "infer_sex_karyotyp": [18, 19], "annotate_sex": [18, 19], "compute_callrate_mt": [18, 20], "run_platform_pca": [18, 20], "assign_platform_from_pc": [18, 20], "unrel": [18, 21, 38, 43], "second_degree_rel": [18, 21], "parent_child": [18, 21], "sibl": [18, 21, 38, 43], "duplicate_or_twin": [18, 21], "ambiguous_relationship": [18, 21], "get_duplicated_sampl": [18, 21], "get_duplicated_samples_ht": [18, 21], "explode_duplicate_samples_ht": [18, 21], "get_relationship_expr": [18, 21], "get_slope_int_relationship_expr": [18, 21], "infer_famili": [18, 21], "create_fake_pedigre": [18, 21], "compute_related_samples_to_drop": [18, 21], "filter_to_trio": [18, 21], "generate_trio_stats_expr": [18, 21, 43], "generate_sib_stats_expr": [18, 21, 43], "calculate_de_novo_post_prob": [18, 21], "default_get_de_novo_expr": [18, 21], "adjusted_sex_ploidy_expr": [18, 22, 23], "adjust_sex_ploidi": [18, 22], "gaussian_mixture_model_karyotype_assign": [18, 22], "get_ploidy_cutoff": [18, 19, 22], "get_chr_x_hom_alt_cutoff": [18, 22], "get_sex_expr": [18, 19, 22], "min_af": 19, "min_callr": 19, "99": [19, 21, 23, 38], "min_inbreeding_coeff_threshold": 19, "min_hardy_weinberg_threshold": 19, "1e": [19, 23], "08": [19, 21, 23], "apply_hard_filt": 19, "bi_allelic_onli": [19, 20, 43], "snv_onli": 19, "sites_callr": 19, "site_inbreeding_coeff": 19, "callrat": [19, 20], "taken": [19, 24, 26], "coeff": 19, "mq": [19, 36, 38, 43], "f": [19, 22, 23, 38, 43], "qd": [19, 38, 43], "best": 19, "practic": 19, "hard": [19, 21, 38], "Not": 19, "hw": 19, "gakt": 19, "60": [19, 24, 43], "30": [19, 23, 24, 36, 43], "adj_onli": 19, "ld_r2": 19, "filter_segdup": [19, 26], "filter_exome_low_coverage_region": [19, 26], "high_conf_region": [19, 26], "checkpoint_path": [19, 40], "block_siz": 19, "readi": [19, 38, 44], "outsid": [19, 26, 32, 38], "maf": 19, "ab": [19, 21, 23], "prune": [19, 45], "r2": 19, "segment": [19, 38], "duplic": [19, 21, 38], "persist": 19, "written": 19, "reread": 19, "new": [19, 25, 38, 39], "block": [19, 23, 35, 36, 38, 42], "ploidy_ht": 19, "f_stat_cutoff": [19, 22], "use_gaussian_mixture_model": 19, "normal_ploidy_cutoff": [19, 22], "aneuploidy_cutoff": [19, 22], "chr_x_frac_hom_alt_expr": [19, 22], "normal_chr_x_hom_alt_cutoff": 19, "x_karyotyp": [19, 22], "y_karyotyp": [19, 22], "x": [19, 21, 22, 23, 24, 26, 36, 38], "ploidi": [19, 21, 22, 23, 36], "karyotyp": [19, 21, 22, 23, 36], "By": [19, 24, 26, 36, 37, 38, 39, 43], "roughli": [19, 22], "gaussian": [19, 22, 38], "mixtur": [19, 22, 38], "divid": [19, 22, 40], "below": [19, 21, 22, 24, 26, 38, 48], "deviat": [19, 22, 24, 27], "aneuploidi": [19, 22, 23], "alt": [19, 22, 23, 24, 31, 38], "is_spars": 19, "excluded_interv": [19, 33], "included_interv": 19, "normalization_contig": [19, 36], "sites_ht": [19, 36], "aaf_expr": 19, "gt_expr": [19, 22, 23, 26], "aaf_threshold": 19, "variants_only_x_ploidi": 19, "variants_only_y_ploidi": 19, "variants_filter_lcr": 19, "variants_filter_segdup": 19, "variants_filter_decoi": 19, "variants_snv_onli": 19, "coverage_mt": 19, "compute_x_frac_variants_hom_alt": 19, "compute_fstat": 19, "infer_karyotyp": 19, "heterozygos": [19, 23, 38], "_mean_dp": 19, "float32": 19, "over": [19, 21, 24, 25, 30, 36, 37, 38], "chrx_mean_dp": 19, "chry_mean_dp": 19, "chrx_ploidi": [19, 22], "chry_ploidi": [19, 22], "f_stat": [19, 22], "float64": [19, 23, 24, 37], "impute_sex": 19, "int64": [19, 38], "expected_hom": 19, "homozygot": [19, 24], "observed_hom": 19, "spars": [19, 21, 26, 36, 38, 43], "normal": [19, 23, 27, 32, 36, 38], "prior": [19, 21, 23, 24, 30, 36], "depth": [19, 21, 23, 36, 38, 43, 44], "estim": [19, 24, 37, 38], "nucleotid": 19, "precomput": [19, 47], "intervals_ht": 20, "sequenc": [20, 30, 33, 38], "interval_info": 20, "callrate_mt": 20, "binarization_threshold": 20, "25": [20, 21, 36], "binzarization_threshold": 20, "scores_ht": 20, "platform_pca_scores_ht": 20, "pc_scores_ann": 20, "hdbscan_min_cluster_s": 20, "hdbscan_min_sampl": 20, "hbdscan": 20, "hdbscan": 20, "min_cluster_s": 20, "smallest": 20, "n_sampl": 20, "min_sampl": [20, 37], "qc_platform": 20, "represent": [21, 39], "pair": [21, 24, 26, 29, 39, 42, 43], "individu": [21, 23, 38, 43], "2nd": [21, 43], "degre": 21, "user": [21, 35, 37], "parent": [21, 38], "child": [21, 38], "twin": 21, "who": 21, "ident": 21, "mz": 21, "ambigu": [21, 23], "relationship": [21, 43], "kinship": 21, "ibd": 21, "biolog": 21, "relationship_ht": 21, "i_col": [21, 43], "j_col": [21, 43], "j": [21, 43], "rel_col": 21, "ouput": 21, "pc_relat": 21, "1st": [21, 43], "duplicated_sampl": 21, "samples_rankings_ht": 21, "rank_ann": 21, "rank": [21, 38, 40], "better": [21, 36, 39], "samples_ranking_ht": 21, "dups_ht": 21, "so": [21, 22, 23, 24, 25, 39], "dup_filt": 21, "kin_expr": 21, "ibd0_expr": 21, "ibd1_expr": 21, "ibd2_expr": 21, "first_degree_kin_threshold": 21, "19": 21, "second_degree_min_kin": 21, "ibd0_0_max": 21, "025": 21, "ibd0_25_threshold": 21, "425": 21, "ibd1_0_threshold": 21, "ibd1_50_threshold": 21, "275": 21, "75": 21, "ibd1_100_min": 21, "ibd2_0_max": 21, "125": 21, "ibd2_25_threshold": 21, "ibd2_100_threshold": 21, "kin": 21, "ibdo": 21, "ibd1": 21, "ibd2": 21, "doc": 21, "highlight": 21, "idb2": 21, "min": [21, 43], "ibd0": 21, "share": [21, 35, 43], "corner": 21, "00": 21, "constant": [21, 26, 39, 43], "y_expr": 21, "parent_child_max_i": 21, "second_degree_sibling_lower_cutoff_slop": 21, "second_degree_sibling_lower_cutoff_intercept": 21, "second_degree_upper_sibling_lower_cutoff_slop": 21, "second_degree_upper_sibling_lower_cutoff_intercept": 21, "duplicate_twin_min_kin": 21, "duplicate_twin_ibd1_min": 21, "duplicate_twin_ibd1_max": 21, "slope": [21, 24], "intercept": [21, 24], "axi": [21, 32], "independ": [21, 23], "ibs0": 21, "ibs2": 21, "No": [21, 23], "els": [21, 39], "second_degree_sibling_lower_cutoff": 21, "second_degree_upper_sibling_lower_cutoff": 21, "met": [21, 24, 26, 39], "08838835": 21, "bycroft": 21, "2018": 21, "theoret": 21, "distribut": [21, 23, 24, 27, 36, 38], "duplicate_samples_ht": 21, "relationship_col": [21, 43], "trio": [21, 43], "ok": [21, 43], "member": 21, "sib": 21, "complet": [21, 23, 24], "father": 21, "mother": 21, "is_xx": 21, "TO": 21, "won": 21, "exactli": 21, "contatin": 21, "sample_list": 21, "exclude_real_proband": 21, "max_tri": 21, "real_pedigre": 21, "sample_list_stratif": 21, "made": 21, "children": 21, "fake": 21, "proband": 21, "real": 21, "ones": [21, 23, 33], "cannot": 21, "try": 21, "bail": 21, "prevent": 21, "infinit": 21, "w": 21, "r": [21, 38, 42], "pick": 21, "chosen": [21, 39], "relatedness_ht": [21, 43], "rank_ht": 21, "kin_threshold": 21, "filtered_sampl": 21, "min_related_hard_filt": 21, "keep_sampl": 21, "keep_samples_when_rel": 21, "maxim": [21, 24], "prefer": [21, 24], "appear": [21, 36], "error": [21, 25, 26, 44], "among": 21, "don": [21, 24, 37], "fam_ht": 21, "filter_col": 21, "howev": [21, 26], "filter_sampl": 21, "trio_mt": 21, "transmitted_strata": 21, "de_novo_strata": 21, "ac_strata": 21, "proband_is_female_expr": 21, "wise": [21, 23, 38], "transmiss": 21, "transmit": [21, 38, 43], "de": [21, 43], "novo": [21, 43], "mutat": [21, 24], "dens": [21, 36, 43], "deal": [21, 43], "densifi": [21, 36, 38, 43], "trio_matrix": [21, 43], "proband_is_xx_expr": 21, "dnm": 21, "sib_ht": 21, "is_femal": 21, "stata": 21, "col": [21, 23], "anoth": [21, 23, 30, 38], "proband_pl_expr": 21, "father_pl_expr": 21, "mother_pl_expr": 21, "diploid_expr": [21, 23], "hemi_x_expr": [21, 23], "hemi_y_expr": [21, 23], "freq_prior_expr": 21, "min_pop_prior": 21, "3333333333333333e": 21, "06": 21, "de_novo_prior": 21, "3333333333333334e": 21, "posterior": 21, "p_dn": 21, "likelihood": [21, 38], "adapt": 21, "kaitlin": 21, "samocha": 21, "caller": [21, 23], "de_novo": 21, "neither": [21, 24], "explicitli": 21, "document": [21, 23, 48], "hemizyg": [21, 23], "To": [21, 47, 50], "address": [21, 48], "equat": 21, "docstr": 21, "event": 21, "truli": 21, "heterozygot": 21, "p_": 21, "dn": 21, "frac": 21, "least": [21, 23, 24, 26, 36], "bayesian": 21, "theorem": 21, "denomin": 21, "yield": 21, "cdot": 21, "hom_ref": 21, "adjust": [21, 22, 23, 24, 36, 39], "why": [21, 24], "formula": 21, "hom_alt": 21, "literatur": 21, "7": [21, 37], "freq_prior": 21, "phred": [21, 23, 27, 36, 38], "diploid": [21, 23], "3e7": 21, "alleles_expr": [21, 23], "proband_expr": 21, "father_expr": 21, "mother_expr": 21, "is_xx_expr": [21, 23], "min_dp_ratio": 21, "min_gq": 21, "min_proband_ab": 21, "max_parent_ab": 21, "05": [21, 23, 24, 38], "min_de_novo_p": 21, "high_conf_dp_ratio": 21, "dp_threshold_snp": 21, "high_med_conf_ab": 21, "low_conf_ab": 21, "high_conf_p": 21, "med_conf_p": 21, "dr": 21, "OR": 21, "medium": 21, "slightli": 21, "origin": [21, 26, 30, 40, 43, 48], "design": [21, 39], "fill": [21, 23, 32], "gap": 21, "meet": [21, 38], "unclassifi": 21, "simplifi": 21, "github": [21, 23], "com": [21, 23, 37], "ksamocha": 21, "de_novo_script": 21, "ignore_in_sample_allele_frequ": 21, "mode": [21, 32], "might": 21, "pl": [21, 38], "balanc": [21, 23, 38], "our": 21, "intention": 21, "save": [21, 32, 44], "storag": [21, 25], "reason": [21, 26], "wai": [21, 23, 26, 36, 39], "them": [21, 23, 36, 39], "or_els": 21, "biallel": [21, 40, 43], "snp": [21, 24, 38], "karyotype_expr": [22, 23], "xy_karyotype_str": [22, 23], "xx_karyotype_str": [22, 23], "haploid": [22, 23], "callexpress": [22, 23, 40], "sex_expr": [22, 23], "point": [22, 24, 25], "chang": [22, 23, 24, 37, 49], "fix": [22, 24], "sex_ht": 22, "chrx_ploidy_expr": 22, "chry_ploidy_expr": 22, "karyotype_output_prefix": 22, "gmm": 22, "compon": [22, 38], "_x_karyotyp": 22, "karyotype_output_prefix_y_karyotyp": 22, "_karyotyp": 22, "_y_karyotyp": 22, "yy": 22, "group_by_expr": 22, "x_ploidy_cutoff": 22, "y_ploidy_cutoff": 22, "doubl": [22, 23], "tripl": 22, "cutoff_stdev": 22, "chr_x_ploidi": 22, "chr_y_ploidi": 22, "chr_x_frac_hom_alt_cutoff": 22, "x0": 22, "annotate_with_ht": [23, 28], "annotation_ht": 23, "annotate_col": 23, "filter_miss": 23, "join": [23, 24, 25, 34, 36, 44], "gen_anc_groups_to_exclud": 23, "gen_anc_label": 23, "about": [23, 38, 44, 45], "highest": 23, "int32": [23, 36, 38], "homozygote_count": [23, 26], "project_max_expr": [23, 28], "project_expr": 23, "n_project": 23, "largest": [23, 38], "Will": 23, "itself": 23, "ti": [23, 45, 47], "guarante": 23, "projectmax": 23, "faf_expr": [23, 28], "faf_threshold": 23, "faf": [23, 34, 38], "cardiodb": 23, "allelefrequencyapp": 23, "agg": [23, 32, 37], "call_stat": 23, "bottleneck": 23, "consanguin": [23, 38], "gen_anc_faf_max_expr": [23, 28], "faf_meta": [23, 34], "faf95_max": 23, "faf95_max_gen_anc": 23, "faf99_max": 23, "faf99_max_gen_anc": 23, "faf95": [23, 38], "faf99": [23, 38], "qual_hist_expr": [23, 28], "gq_expr": 23, "dp_expr": 23, "ad_expr": 23, "adj_expr": 23, "ab_expr": 23, "split_adj_and_raw": 23, "here": [23, 24], "age_hists_expr": [23, 28], "age_expr": 23, "lowest_boundari": 23, "highest_boundari": 23, "80": 23, "n_bin": [23, 40, 43], "ag": [23, 38], "lowest": [23, 38], "younger": 23, "n_smaller": [23, 38], "older": 23, "n_larger": [23, 38], "age_hist_het": 23, "age_hist_hom": 23, "get_lowqual_expr": [23, 28], "qual_approx_expr": 23, "snv_phred_threshold": 23, "snv_phred_het_prior": 23, "indel_phred_threshold": 23, "indel_phred_het_prior": 23, "39": 23, "lowqual": [23, 36], "unsplit": 23, "qualapprox": [23, 36, 38], "as_qualapprox": [23, 38], "stringent": 23, "mix": [23, 38], "thu": 23, "certain": [23, 37, 39], "emiss": 23, "similar": [23, 24], "arraynumer": 23, "get_annotations_hist": [23, 28], "annotations_hist": 23, "log10_annot": 23, "json": [23, 25, 39, 44], "meric": 23, "create_frequency_bins_expr": [23, 28], "prepar": [23, 30, 37, 39], "qual": [23, 36, 38], "00005": 23, "0002": 23, "0005": 23, "002": 23, "005": 23, "02": 23, "annotate_and_index_source_mt_for_sex_ploidi": [23, 28], "downstream": [23, 26], "optim": 23, "get_is_haploid_expr": [23, 28], "in_non_par": 23, "x_nonpar": 23, "y_par": [23, 24], "y_nonpar": 23, "get_gq_dp_adj_expr": [23, 28], "adj_gq": 23, "adj_dp": 23, "haploid_adj_dp": 23, "account": [23, 26], "int32express": [23, 24, 36, 38], "get_het_ab_adj_expr": [23, 28], "adj_ab": 23, "get_adj_expr": [23, 28], "annotate_adj": [23, 28], "add_variant_typ": [23, 28], "alt_allel": 23, "variant_typ": [23, 24, 38], "n_alt_allel": [23, 38], "annotate_allele_info": [23, 28], "allele_info": 23, "has_star": [23, 38], "star": [23, 26, 38], "allele_typ": [23, 38], "insert": [23, 38, 43], "delet": [23, 38, 43], "was_mix": [23, 38], "nonsplit_allel": 23, "annotation_type_is_numer": [23, 28], "annotation_type_in_vcf_info": [23, 28], "nativ": 23, "aren": 23, "bi_allelic_site_inbreeding_expr": [23, 28], "callstats_expr": 23, "inbreedingcoeff": [23, 38], "softwar": 23, "broadinstitut": 23, "articl": [23, 37], "php": 23, "8032": 23, "place": 23, "float32express": 23, "fs_from_sb": [23, 28], "sb": [23, 36, 38], "min_cell_count": 23, "min_count": 23, "min_p_valu": 23, "320": 23, "fisher": [23, 38], "strand": [23, 24, 38, 39], "side": [23, 29], "exact": [23, 38], "behavior": 23, "truncat": [23, 24, 32, 39], "chi": 23, "squar": [23, 24, 38], "cell": [23, 37], "four": [23, 40], "forward": [23, 38], "revers": [23, 24, 38], "ref": [23, 24, 31, 33, 36, 38, 39], "fwd": 23, "rev": 23, "dimension": 23, "blob": 23, "master": 23, "src": 23, "java": 23, "hellbend": 23, "tool": [23, 32], "walker": 23, "fisherstrand": 23, "sq": 23, "fet": 23, "null": [23, 24], "sor_from_sb": [23, 28], "sor": [23, 38], "symmetr": [23, 38], "odd": [23, 38], "strandoddsratio": 23, "pab_max_expr": [23, 28], "la_expr": 23, "n_alleles_expr": 23, "binomi": [23, 38], "pab": 23, "local": [23, 25, 36, 39], "bi_allelic_expr": [23, 28], "unphase_call_expr": [23, 28], "call_expr": 23, "unphas": 23, "phase": [23, 38, 43], "region_flag_expr": [23, 28], "non_par": [23, 38], "prob_region": 23, "region_flag": 23, "hg38": [23, 38], "self": 23, "chain": [23, 30], "yet": 23, "loci": [23, 26], "occur": [23, 39], "pseudoautosom": [23, 38], "missing_callstats_expr": [23, 28], "set_xx_y_metrics_to_na_expr": [23, 28], "freq_meta_expr": [23, 24], "freq_index_dict_expr": 23, "freq_index_dict": 23, "dictexpress": [23, 24], "hemi_expr": [23, 28], "male_str": 23, "merge_array_express": [23, 28], "set_negatives_to_zero": 23, "struct_field": 23, "count_arrai": 23, "etc": [23, 24, 40], "arr1": 23, "arr2": 23, "meta1": 23, "meta2": 23, "diff": 23, "subtract": 23, "primari": 23, "valueerror": 23, "descriptor": 23, "merge_freq_arrai": [23, 28], "farrai": 23, "fmeta": 23, "freq1": 23, "freq2": 23, "fmeta1": 23, "fmeta2": 23, "merge_histogram": [23, 28], "fashion": 23, "bin_edg": [23, 38], "bin_freq": 23, "edg": [23, 38], "subgrp_expr": 23, "additional_strata_expr": 23, "downsampling_expr": 23, "ds_gen_anc_count": 23, "entry_agg_func": [23, 36], "annotate_mt": 23, "freq_meta_sample_count": [23, 24, 26], "support": [23, 25, 33, 38, 39, 44, 45], "age_bin": 23, "replac": [23, 37, 44], "enough": 23, "randomli": 23, "higher": [23, 36, 37, 38, 39], "global_idx": 23, "gen_anc_idx": 23, "annotate_downsampl": [23, 28], "adj_sampl": 23, "build_freq_stratification_list": [23, 28], "primarili": 23, "generate_freq_group_membership_arrai": [23, 28], "strata_expr": [23, 26, 36], "remove_zero_sample_group": 23, "no_raw_group": 23, "group_membership": [23, 36], "pleas": [23, 48], "compute_freq_by_strata": [23, 28], "select_field": 23, "group_membership_includes_raw_group": 23, "like": [23, 26, 38, 39], "still": [23, 45], "membership": [23, 36], "agg_by_strata": [23, 28], "group_membership_ht": [23, 36], "entry_agg_group_membership": [23, 36], "update_structured_annot": [23, 28], "annotation_update_expr": 23, "annotation_update_label": 23, "updat": [23, 38, 39, 48], "fill_missing_key_combin": [23, 28], "fill_valu": 23, "key_valu": 23, "uniqu": [23, 38], "collect": [23, 24, 39], "unless": [23, 24, 36], "want": [23, 24, 37, 39, 40], "happen": 23, "parallel": 23, "key1": 23, "key2": 23, "b": 23, "tstruct": [23, 44], "tstr": 23, "tint32": 23, "filled_ht": 23, "missing_struct_expr": [23, 28], "dtype": 23, "add_gks_vr": [23, 28], "input_locu": 23, "input_vr": 23, "ga4gh": [23, 38], "conform": 23, "add_gks_va": [23, 28], "input_struct": 23, "label_nam": 23, "label_vers": 23, "gen_anc_groups_dict": 23, "va": 23, "schema": [23, 37], "subcohort": 23, "mean_depth": 23, "ancillaryresult": 23, "gks_va_freq_dict": 23, "focusallel": 23, "shorten": 23, "breakdown": 23, "get_copy_state_by_sex": [23, 28], "state": [23, 24, 25, 38], "karotyp": 23, "xxy": 23, "xyi": 23, "check_annotation_missing": [23, 28], "high_missingness_threshold": 23, "remove_missing_field": 23, "include_col_annot": 23, "high_missingness_field": 23, "exceed": 23, "completely_missing_field": 23, "usag": [23, 32, 35, 36, 37], "potenti": 23, "coverage_cutoff": [24, 28], "differenti": 24, "calibr": 24, "variat": [24, 37], "get_mu_annotation_expr": [24, 28], "mutation_ht": 24, "mu_expr": 24, "mu_snp": 24, "context": [24, 26, 39], "methylation_level": 24, "avoid": 24, "unnecessari": 24, "shuffl": [24, 30, 40], "small": [24, 37], "annotate_with_mu": [24, 28], "mu_annot": 24, "count_variants_by_group": [24, 28], "count_singleton": 24, "count_downsampl": 24, "additional_group": 24, "partition_hint": 24, "omit_methyl": 24, "use_table_group_bi": 24, "singleton_expr": 24, "detail": [24, 26, 37, 39, 48], "variant_count": 24, "trinucleotid": 24, "featur": [24, 26, 39, 43, 44], "exome_coverag": 24, "omit": 24, "singleton_express": 24, "singleton_count": 24, "get_downsampling_freq_indic": [24, 28], "variant_qu": 24, "genetic_ancestry_label": 24, "downsampling_counts_expr": [24, 28], "explode_downsamplings_o": [24, 28], "downsampling_meta": 24, "syn": 24, "mi": 24, "gen_anc_exp": 24, "annotate_mutation_typ": [24, 28], "context_length": 24, "num_scan_context_length": 24, "cpg": 24, "transit": [24, 43], "mutation_typ": 24, "transvers": 24, "mutation_type_model": 24, "repo": 24, "trimer_from_heptam": [24, 28], "trim": 24, "heptam": 24, "trimer": 24, "collapse_strand": [24, 28], "dedupl": [24, 39], "dna": 24, "complement": 24, "was_flip": 24, "flip": 24, "transform_methylation_level": [24, 28], "methylation_cutoff": 24, "assemble_constraint_context_ht": [24, 28], "methylation_ht": 24, "gerp_ht": 24, "coverage_ht": 24, "an_ht": 24, "freq_ht": 24, "filter_ht": 24, "transformation_func": 24, "assembl": 24, "necessari": [24, 37], "was_split": [24, 40], "add_most_severe_csq_to_tc_within_vep_root": [24, 28, 39], "gerp": [24, 26, 39], "uk": [24, 38], "biobank": [24, 38], "an_glob": 24, "strata_sample_count": 24, "strata_meta": 24, "freq_glob": 24, "build_model": [24, 28], "coverage_expr": 24, "weight": 24, "high_cov_definit": 24, "upper_cov_cutoff": 24, "skip_coverage_model": 24, "log10_coverag": 24, "plateau": 24, "plateau_model": 24, "linear": [24, 27], "against": [24, 32], "substitut": [24, 33, 38], "remaind": 24, "observed_vari": 24, "observed_": 24, "possible_vari": 24, "built": [24, 26], "factor": [24, 27], "correct": [24, 38], "log10": [24, 38], "high_coverage_scale_factor": 24, "get_proportion_observed_by_coverag": 24, "downsampling_counts_": 24, "cover": 24, "build_plateau_model": [24, 28], "cpg_expr": 24, "mu_snp_expr": 24, "observed_variants_expr": 24, "possible_variants_expr": 24, "gen_ancs_observed_variants_array_expr": 24, "predict": [24, 26, 38, 39, 44], "build_coverage_model": [24, 28], "low_coverage_oe_expr": 24, "get_all_gen_anc_length": [24, 28], "obs_expr": 24, "get_constraint_grouping_expr": [24, 28], "vep_annotation_expr": 24, "include_transcript_group": 24, "include_canonical_group": 24, "include_mane_select_group": 24, "classic": 24, "polyphen": [24, 37, 39], "polyphen_predict": 24, "gene_symbol": [24, 37, 39], "insid": [24, 39], "transcript_id": [24, 26, 37], "mane_select": [24, 39], "actual": 24, "annotate_exploded_vep_for_constraint_group": [24, 28], "vep_annot": [24, 37, 39], "worst_csq_by_gen": [24, 37, 39, 47], "ignor": [24, 37, 44], "compute_expected_vari": [24, 28], "plateau_models_expr": 24, "cov_corr_expr": 24, "oe_aggregation_expr": [24, 28], "exclude_mu_sum": 24, "ob": 24, "mu": 24, "exp": 24, "gen_anc_ob": 24, "expected_vari": 24, "expected_variants_": 24, "compute_pli": [24, 28], "exp_expr": 24, "expected_valu": 24, "min_diff_converg": 24, "pli": 24, "exac": [24, 38], "paper": 24, "lek": 24, "m": [24, 38], "karczewski": 24, "k": [24, 27], "minikel": 24, "706": 24, "natur": 24, "536": 24, "285": 24, "291": 24, "2016": 24, "intoler": 24, "em": 24, "algorithm": 24, "three": [24, 40, 43], "respect": [24, 26, 32], "sensit": 24, "toler": [24, 38], "recess": 24, "rec": 24, "haploinsuffici": 24, "li": 24, "deplet": 24, "blekhman": 24, "clingen": 24, "dosag": 24, "supplementari": 24, "463": 24, "deriv": 24, "empir": 24, "diseas": 24, "089": 24, "fall": [24, 32, 37, 38, 39], "pnull": 24, "unconstrain": 24, "prec": 24, "iter": [24, 35], "converg": 24, "oe_confidence_interv": [24, 28], "alpha": [24, 32], "around": [24, 32], "densiti": [24, 32], "poisson": 24, "dpoi": 24, "lambda": 24, "lamb": 24, "extend": [24, 36, 43], "bound": 24, "cumul": [24, 26, 32, 36], "signific": [24, 26], "calculate_raw_z_scor": [24, 28], "sign": 24, "z": [24, 38], "posit": [24, 29, 30, 31, 36, 38, 39, 43, 45], "had": 24, "fewer": 24, "get_constraint_flag": [24, 28], "raw_z_expr": 24, "raw_z_lower_threshold": 24, "raw_z_upper_threshold": 24, "flag_postfix": 24, "no_exp_": 24, "zero": [24, 26], "outlier_": 24, "postfix": 24, "calculate_raw_z_score_sd": [24, 28], "flag_expr": 24, "mirror_neg_raw_z": 24, "multipli": 24, "mirror": 24, "add_gencode_transcript_annot": [24, 28], "gencode_ht": [24, 26, 37], "transcript_typ": [24, 26], "start_posit": 24, "end_posit": 24, "remove_y_par": 24, "cds_length": 24, "num_coding_exon": 24, "strip": 24, "file_exist": [25, 28], "fname": 25, "parquet": 25, "_success": 25, "check_file_exists_raise_error": [25, 28], "error_if_exist": 25, "error_if_not_exist": 25, "error_if_exists_msg": 25, "error_if_not_exists_msg": 25, "therefor": [25, 26], "write_temp_gc": [25, 28], "gcs_path": 25, "temp_path": 25, "select_primitives_from_ht": [25, 28], "primit": 25, "particularli": 25, "get_file_stat": [25, 28], "url": [25, 50], "project_id": 25, "md5": 25, "read_list_data": [25, 28], "input_file_path": 25, "gz": [25, 30, 47], "compress": 25, "repartition_for_join": [25, 28], "ht_path": 25, "new_partition_perc": 25, "via": 25, "_interv": 25, "effici": 25, "discuss": 25, "room": 25, "improv": 25, "2278": 25, "increas": [25, 36], "create_vd": [25, 28], "output_path": 25, "vdse": 25, "gvcf": [25, 36, 38], "save_path": 25, "use_genome_default_interv": 25, "use_exome_default_interv": 25, "gvcf_batch_siz": 25, "directori": [25, 32], "temporari": 25, "life": 25, "cycl": 25, "polici": 25, "restart": 25, "combiner_plan": 25, "creation": [25, 50], "filter_to_adj": [26, 28], "filter_by_frequ": [26, 28], "direct": 26, "allele_count": 26, "At": [26, 36], "filter_row": 26, "combine_funct": [26, 28], "func_list": 26, "operator_func": 26, "iand": 26, "low_conf_regions_expr": [26, 28], "filter_telomeres_and_centromer": 26, "telomer": 26, "centromer": 26, "restrict": 26, "filter_low_conf_region": [26, 28], "keyword": [26, 36], "filter_to_autosom": [26, 28], "add_filters_expr": [26, 28], "current_filt": 26, "subset_samples_and_vari": [26, 28], "sample_path": 26, "table_kei": 26, "remove_dead_allel": 26, "filter_to_clinvar_pathogen": [26, 28], "clnrevstat_field": 26, "clnrevstat": 26, "clnsig_field": 26, "clnsig": 26, "clnsigconf_field": 26, "clnsigconf": 26, "remove_no_assert": 26, "remove_conflict": 26, "clinvar": [26, 43], "pathogen": [26, 38], "clinvar_ht": 26, "could": [26, 40], "review": 26, "clinic": 26, "signifc": 26, "conflict": 26, "undefin": 26, "assert": 26, "filter_gencode_ht": [26, 28], "protein_cod": [26, 39], "by_gene_symbol": 26, "cd": [26, 37], "symbol": [26, 37, 39, 42], "filter_by_interv": [26, 28], "padding_bp": [26, 29], "max_collect_interv": 26, "3000": 26, "pad": [26, 29], "filter_interv": [26, 36], "caus": [26, 39], "memori": [26, 36], "filter_by_gencode_interv": [26, 28], "filter_to_gencode_cd": [26, 28], "filter_gencode_to_cd": 26, "analys": [26, 37], "step": [26, 48], "retain": [26, 36], "remove_fields_from_const": [26, 28], "fields_to_remov": 26, "filter_x_nonpar": [26, 28], "filter_y_nonpar": [26, 28], "filter_by_numeric_expr_rang": [26, 28], "filter_rang": 26, "keep_between": 26, "inclus": 26, "themselv": 26, "filter_for_mu": [26, 28], "gerp_lower_cutoff": 26, "9885": 26, "gerp_upper_cutoff": 26, "6607": 26, "precalcul": 26, "5th": [26, 34], "95th": 26, "percentil": 26, "intron": [26, 39], "intergen": 26, "split_vds_by_strata": [26, 28], "vdss": 26, "variant_data": [26, 36], "filter_arrays_by_meta": [26, 28], "meta_expr": 26, "meta_indexed_expr": 26, "items_to_filt": 26, "combine_oper": 26, "exact_match": 26, "item": [26, 40], "papuan": 26, "index_glob": [26, 28, 38], "often": [26, 39], "to_phr": [27, 28], "linear_expr": 27, "from_phr": [27, 28], "phred_score_expr": 27, "get_median_and_mad_expr": [27, 28], "metric_expr": 27, "4826": 27, "absolut": 27, "merge_stats_counters_expr": [27, 28], "counter": 27, "sort_interv": [28, 29], "union_interv": [28, 29], "interval_length": [28, 29], "pad_interv": [28, 29], "grch37_to_grch38_chain": [28, 30], "grch38_to_grch37_chain": [28, 30], "get_liftover_genom": [28, 30], "liftover_expr": [28, 30], "default_lift_data": [28, 30], "liftover_using_gnomad_map": [28, 30], "parse_vari": [28, 31], "parse_locus_interv": [28, 31], "new_show": [28, 32], "plot_hail_hist": [28, 32], "plot_multi_hail_hist": [28, 32], "plot_hail_hist_cumul": [28, 32], "plot_hail_hist_both": [28, 32], "set_font_s": [28, 32], "linear_and_log_tab": [28, 32], "plot_hail_file_metadata": [28, 32], "scale_file_s": [28, 32], "get_rows_data": [28, 32], "pair_plot": [28, 32], "get_reference_ht": [28, 33, 36], "add_reference_sequ": [28, 33], "get_reference_genom": [28, 33], "make_faf_index_dict": [28, 34], "make_freq_index_dict": [28, 34], "make_freq_index_dict_from_meta": [28, 34], "slackclient": [28, 35], "slack_notif": [28, 35], "compute_last_ref_block_end": [28, 36], "densify_sit": [28, 36], "get_as_info_expr": [28, 36], "get_site_info_expr": [28, 36], "default_compute_info": [28, 36], "split_info_annot": [28, 36], "split_lowqual_annot": [28, 36], "impute_sex_ploidi": [28, 36], "densify_all_reference_sit": [28, 36], "compute_stats_per_ref_sit": [28, 36], "get_coverage_agg_func": [28, 36], "compute_coverage_stat": [28, 36], "get_allele_number_agg_func": [28, 36], "compute_allele_number_per_ref_sit": [28, 36], "filter_ref_block": [28, 36], "reproductive_tissu": [28, 37], "cell_lin": [28, 37], "get_tissues_to_exclud": [28, 37], "summarize_transcript_express": [28, 37], "get_expression_proport": [28, 37], "filter_expression_ht_by_tissu": [28, 37], "tissue_expression_ht_to_arrai": [28, 37], "tx_filter_variants_by_csq": [28, 37], "tx_annotate_vari": [28, 37], "tx_aggregate_vari": [28, 37], "perform_tx_annotation_pipelin": [28, 37], "clean_tissue_name_for_brows": [28, 37], "create_tx_annotation_by_region": [28, 37], "faf_gen_anc_group": [28, 38], "as_field": [28, 38], "site_field": [28, 38], "allele_type_field": [28, 38], "region_flag_field": [28, 38], "joint_region_flag_field": [28, 38], "rf_field": [28, 38], "as_vqsr_field": [28, 38], "vqsr_field": [28, 38], "info_dict": [28, 38], "in_silico_annotations_info_dict": [28, 38], "vrs_fields_dict": [28, 38], "sparse_entri": [28, 38], "format_dict": [28, 38], "adjust_vcf_incompatible_typ": [28, 38], "make_label_combo": [28, 38], "make_combo_header_text": [28, 38], "create_label_group": [28, 38], "make_info_dict": [28, 38], "add_as_info_dict": [28, 38], "make_vcf_filter_dict": [28, 38], "make_hist_bin_edges_expr": [28, 38], "make_hist_dict": [28, 38], "set_xx_y_metrics_to_na": [28, 38], "build_vcf_export_refer": [28, 38], "rekey_new_refer": [28, 38], "current_vep_vers": [28, 39], "csq_code": [28, 39], "csq_splice": [28, 39], "possible_ref": [28, 39], "vep_config_path": [28, 39], "vep_csq_field": [28, 39], "vep_csq_head": [28, 39], "get_vep_csq_head": [28, 39], "get_vep_help": [28, 39], "get_vep_context": [28, 39], "vep_or_lookup_vep": [28, 39, 47], "get_most_severe_consequence_expr": [28, 39], "add_most_severe_consequence_to_consequ": [28, 39], "process_consequ": [28, 39, 47], "filter_vep_to_canonical_transcript": [28, 39], "filter_vep_to_mane_select_transcript": [28, 39], "filter_vep_to_synonymous_vari": [28, 39], "filter_vep_to_gene_list": [28, 39], "vep_struct_to_csq": [28, 39], "get_most_severe_consequence_for_summari": [28, 39], "filter_vep_transcript_csq": [28, 39], "filter_vep_transcript_csqs_expr": [28, 39], "explode_by_vep_annot": [28, 39], "get_loftee_end_trunc_filter_expr": [28, 39], "update_loftee_end_trunc_filt": [28, 39], "is_sort": 29, "grch37_to_grch38": 30, "lift": 30, "grch38_to_grch37": 30, "referencegenom": [30, 33, 38], "destination_refer": 30, "coordin": [30, 38], "original_locu": 30, "original_allel": [30, 38], "locus_fail_liftov": 30, "ref_allele_mismatch": 30, "remove_failed_sit": 30, "establish": 30, "variant_str": 31, "chr": [31, 38], "40": [32, 36], "hist_data": 32, "titl": 32, "fill_color": 32, "033649": 32, "outlier_fill_color": 32, "036564": 32, "line_color": 32, "hover_mod": 32, "mous": 32, "hide_zero": 32, "come": 32, "straight": 32, "color": 32, "bar": 32, "hover": 32, "vline": 32, "hline": 32, "figur": 32, "group_bi": 32, "len": [32, 36], "line_width": 32, "font_siz": 32, "12pt": 32, "plot_func": 32, "tab": 32, "t_path": 32, "grid": 32, "Or": 32, "unord": 32, "file_s": 32, "rows_fil": 32, "label_col": [32, 44], "pan": 32, "box_zoom": 32, "reset": [32, 38], "wheel_zoom": 32, "box_select": 32, "lasso_select": 32, "help": [32, 39, 48], "tooltip_col": 32, "diagon": 32, "scatter": 32, "color_dict": 32, "rgb": 32, "hex": 32, "tooltip": 32, "add_all_substitut": 33, "filter_n": 33, "obscur": 33, "slow": 33, "unknown": [33, 38], "fasta": 33, "add_sequ": 33, "label_delimit": [34, 38], "look": [34, 40], "access": [34, 50], "nfe_xx": [34, 38], "token": 35, "api": [35, 49], "client": 35, "send_fil": 35, "filenam": 35, "txt": 35, "filetyp": 35, "comment": 35, "send": 35, "channel": 35, "upload": 35, "send_messag": 35, "icon_emoji": 35, "long": 35, "snippet": 35, "emoji": 35, "icon": 35, "notif": 35, "wrap": 35, "stack": 35, "trace": 35, "usernam": 35, "run_analysi": 35, "upstream": 36, "sinc": 36, "beyond": 36, "last_end_posit": 36, "last_end_positions_ht": 36, "semi_join_row": 36, "minim": 36, "furthest": 36, "tag": 36, "semi": 36, "few": 36, "sum_agg_field": 36, "int32_sum_agg_field": 36, "vardp": [36, 38], "median_agg_field": 36, "readposranksum": [36, 38], "mqranksum": [36, 38], "array_sum_agg_field": 36, "raw_mqanddp": 36, "alt_alleles_range_array_field": 36, "alt_alleles_range_arrai": 36, "treat_fields_as_allele_specif": 36, "retain_cdf": 36, "cdf_k": 36, "as_sb_tabl": [36, 38], "accord": 36, "nomenclatur": 36, "raw_mq": 36, "mq_dp": 36, "gvcf_info": 36, "prioriti": 36, "clash": 36, "Then": [36, 39], "cdf": 36, "later": [36, 39, 44, 48], "accuraci": [36, 44], "tradeoff": 36, "AS": 36, "site_annot": 36, "as_annot": 36, "quasi_as_annot": 36, "lowqual_indel_phred_het_prior": 36, "ac_filter_group": 36, "nonref": 36, "naive_coalesc": 36, "10k": 36, "scienc": 36, "info_expr": 36, "a_index": 36, "split_multi": 36, "split_multi_ht": 36, "lowqual_expr": 36, "particular": [36, 38], "excluded_calling_interv": 36, "included_calling_interv": 36, "chr_x": 36, "chr_y": 36, "use_only_vari": 36, "reference_ht": 36, "row_key_field": 36, "entry_keep_field": 36, "densif": 36, "row_keep_field": 36, "sex_karyotype_field": 36, "lgt": [36, 38], "dp_field": 36, "max_cov_bin": 36, "coverage_over_x_bin": 36, "gt_field": 36, "awar": 37, "cervix_ectocervix": 37, "cervix_endocervix": 37, "fallopiantub": 37, "ovari": 37, "prostat": 37, "testi": 37, "uteru": 37, "vagina": 37, "reproduct": 37, "tissu": 37, "gtex": 37, "cells_ebv_transformedlymphocyt": 37, "cells_transformedfibroblast": 37, "cells_culturedfibroblast": 37, "reach": 37, "satur": 37, "stabl": 37, "studi": 37, "bmcgenom": 37, "biomedcentr": 37, "1186": 37, "s12864": 37, "021": 37, "08125": 37, "transcript_expression_expr": 37, "transcript_tpm": 37, "tissue_expr": 37, "summary_agg_func": 37, "transcript_express": 37, "expression_proport": 37, "gene_id": [37, 39], "tissue_1": 37, "tissue_2": 37, "quantif": 37, "tissues_to_keep": 37, "tissues_to_exclud": 37, "annotations_to_extract": 37, "tissue_express": 37, "filter_to_cd": 37, "filter_to_gen": 37, "match_by_gene_symbol": [37, 39], "filter_to_csq": 37, "ignore_spl": 37, "filter_to_protein_cod": 37, "vep_root": [37, 39], "include_polyphen_priorit": 37, "further": 37, "amino_acid": [37, 39], "stop_retained_vari": [37, 39], "And": 37, "root": [37, 38, 39], "priorit": [37, 38], "process": [37, 47], "preprocess": 37, "tissues_to_exclude_from_mean": 37, "worst_csq_for_vari": [37, 47], "worst_csq_by_gene_canon": 37, "worst_csq_for_variant_canon": 37, "worst": [37, 38, 39], "additional_group_bi": 37, "lof_flag": [37, 39], "transcript_abl": [37, 39], "stop_lost": [37, 39], "start_lost": [37, 39], "initiator_codon_vari": [37, 39], "transcript_amplif": [37, 39], "inframe_insert": [37, 39], "inframe_delet": [37, 39], "protein_altering_vari": [37, 39], "splice_donor_5th_base_vari": [37, 39], "splice_region_vari": [37, 39], "splice_donor_region_vari": [37, 39], "splice_polypyrimidine_tract_vari": [37, 39], "incomplete_terminal_codon_vari": [37, 39], "start_retained_vari": [37, 39], "coding_sequence_vari": [37, 39], "coding_transcript_vari": [37, 39], "stop": [37, 48], "tissue_nam": 37, "clean": 37, "compat": 37, "uppercas": 37, "letter": 37, "lowercas": 37, "underscor": 37, "word": 37, "addition": 37, "organ": 37, "exp_prop_mean": 37, "tissue1": 37, "tissue2": 37, "gene1": 37, "gene2": 37, "chrom": 37, "raw_ac_afr_xx": 38, "earlier": 38, "male": 38, "femal": 38, "as_f": 38, "as_mq": 38, "as_mqranksum": 38, "as_pab_max": 38, "as_qd": 38, "as_readposranksum": 38, "as_sor": 38, "as_vardp": 38, "fail_interval_qc": 38, "outside_broad_capture_region": 38, "outside_ukb_capture_region": 38, "outside_broad_calling_region": 38, "outside_ukb_calling_region": 38, "not_called_in_exom": 38, "not_called_in_genom": 38, "rf_positive_label": 38, "rf_negative_label": 38, "rf_label": [38, 43, 44], "rf_train": [38, 43], "rf_tp_probabl": 38, "as_culprit": 38, "as_vqslod": 38, "negative_train_sit": [38, 43], "positive_train_sit": [38, 43], "bia": 38, "baseqranksum": 38, "wilcoxon": 38, "excess": 38, "hardi": 38, "weinberg": 38, "equilibrium": 38, "vqslod": 38, "versu": 38, "culprit": 38, "85": 38, "20x": 38, "10x": 38, "coincid": 38, "span": 38, "elsewher": 38, "inbreeding_coeff": 38, "complex": 38, "only_het": 38, "sibling_singleton": 38, "wide": 38, "amongst": 38, "transmitted_singleton": 38, "cadd_phr": 38, "cadd": 38, "c": 38, "billion": 38, "deleteri": 38, "cadd_raw_scor": 38, "extent": 38, "profil": [38, 39], "suggest": 38, "simul": 38, "pangolin_largest_d": 38, "pangolin": 38, "delta": 38, "reflect": 38, "alter": 38, "phylop": 38, "conserv": [38, 39], "241": 38, "placent": 38, "mammal": 38, "zoonomia": 38, "28": 38, "acceler": 38, "evolut": 38, "neutral": 38, "drift": 38, "slower": 38, "polyphen_max": 38, "impact": [38, 39], "amino": 38, "acid": 38, "revel_max": 38, "revel": 38, "13": 38, "predictor": [38, 46, 49], "sift_max": 38, "spliceai_ds_max": 38, "illumina": 38, "spliceai": 38, "silico": 38, "vrs_allele_id": 38, "vrs_end": 38, "interresidu": 38, "vrs_length": 38, "referencelengthexpress": 38, "vrs_repeatsubunitlength": 38, "repeatsubunitlength": 38, "vrs_start": 38, "vrs_state": 38, "liter": 38, "min_dp": 38, "pgt": 38, "pid": 38, "la": 38, "lad": 38, "lpgt": 38, "lpl": 38, "rgq": 38, "255": 38, "bad": 38, "mate": 38, "physic": 38, "connect": [38, 48], "record": 38, "uncondit": 38, "encod": 38, "wrong": 38, "compris": 38, "detect": 38, "pipe_delimited_annot": 38, "as_mq_dp": 38, "as_raw_mq": 38, "done": [38, 40, 43, 44], "coerc": 38, "pipe": 38, "afr_xi": 38, "afr_xx": 38, "nfe_xi": 38, "amr_xi": 38, "amr_xx": 38, "globals_arrai": 38, "meta_arrai": 38, "preposit": 38, "combo_dict": 38, "gen_anc_nam": 38, "programmat": 38, "group_typ": 38, "combo_field": 38, "automat": [38, 43], "all_group": 38, "gen_anc_sex_group": 38, "prefix_before_metr": 38, "amish": 38, "admix": 38, "ashkenazi": 38, "jewish": 38, "bgr": 38, "bulgarian": 38, "eastern": 38, "south": 38, "asian": 38, "east": 38, "est": 38, "estonian": 38, "eur": 38, "jpn": 38, "kor": 38, "korean": 38, "mde": 38, "middl": 38, "nwe": 38, "north": 38, "western": 38, "oea": 38, "oeu": 38, "onf": 38, "sas_non_consang": 38, "seu": 38, "sgp": 38, "singaporean": 38, "swe": 38, "swedish": 38, "uniform": 38, "unk": 38, "freq_ctt": 38, "freq_cmh": 38, "freq_stat_union": 38, "description_text": 38, "age_hist_distribut": 38, "carrier": 38, "auto": 38, "conting": 38, "ctt": 38, "cochran": 38, "mantel": 38, "haenszel": 38, "cmh": 38, "snp_cutoff": 38, "indel_cutoff": 38, "inbreeding_cutoff": 38, "variant_qc_filt": 38, "ann_with_hist": 38, "include_age_hist": 38, "reformat": 38, "hist_metric_list": 38, "drop_n_smaller_larg": 38, "keep_contig": 38, "chr2": 38, "chr3": 38, "chr4": 38, "chr5": 38, "chr6": 38, "chr7": 38, "chr8": 38, "chr9": 38, "chr10": 38, "chr11": 38, "chr12": 38, "chr13": 38, "chr14": 38, "chr15": 38, "chr16": 38, "chr17": 38, "chr18": 38, "chr19": 38, "chr21": 38, "chr22": 38, "keep_chrm": 38, "elimin": 38, "22": 38, "chr3_gl000221v1_random": 38, "155397": 38, "assembli": 38, "chrm": 38, "re": [38, 40], "105": 39, "vep_data": 39, "gcloud": 39, "101": 39, "feature_typ": 39, "exon": 39, "hgvsc": 39, "hgvsp": 39, "cdna_posit": 39, "cds_posit": 39, "protein_posit": 39, "codon": 39, "allele_num": 39, "variant_class": 39, "minimis": 39, "symbol_sourc": 39, "hgnc_id": 39, "tsl": 39, "appri": 39, "ccd": 39, "ensp": 39, "swissprot": 39, "trembl": 39, "uniparc": 39, "gene_pheno": 39, "sift": 39, "domain": 39, "hgvs_offset": 39, "motif_nam": 39, "motif_po": 39, "high_inf_po": 39, "motif_score_chang": 39, "lof_filt": 39, "lof_info": 39, "mane_plus_clin": 39, "uniprot_isoform": 39, "mirna": 39, "pubm": 39, "transcription_factor": 39, "115": 39, "vep_vers": 39, "lc": 39, "dataproc": [39, 47, 48, 50], "hailctl": [39, 47, 48, 50], "command": [39, 48], "vep_config_uri": 39, "default_refer": 39, "vepe": 39, "reference_vep_ht": 39, "lookup": 39, "databas": 39, "confirm": 39, "referenc": 39, "suitabl": 39, "grab": 39, "vep_config": 39, "vep_context": 39, "csq_order": 39, "intron_vari": 39, "tc": 39, "most_severe_csq_field": 39, "consequence_term": 39, "penalize_flag": 39, "has_polyphen": 39, "any_lof": 39, "csq_score": 39, "deduct": 39, "NO": 39, "everyth": 39, "penal": 39, "filter_empty_csq": 39, "upstream_gene_vari": 39, "downstream_gene_vari": 39, "vep_expr": 39, "has_polyphen_sift": 39, "flexibl": 39, "counterpart": 39, "Their": 39, "usual": 39, "mature_mirna_vari": 39, "5_prime_utr_vari": 39, "3_prime_utr_vari": 39, "non_coding_transcript_exon_vari": 39, "non_coding_exon_vari": 39, "nmd_transcript_vari": 39, "non_coding_transcript_vari": 39, "nc_transcript_vari": 39, "tfbs_ablat": 39, "tfbs_amplif": 39, "tf_binding_site_vari": 39, "regulatory_region_abl": 39, "regulatory_region_amplif": 39, "feature_elong": 39, "regulatory_region_vari": 39, "feature_trunc": 39, "intergenic_vari": 39, "sequence_vari": 39, "most_severe_csq": 39, "ensembl_onli": 39, "refseq": 39, "emsembl": 39, "keep_csq": 39, "keep_gen": 39, "additional_filtering_criteria": 39, "gerp_dist_cutoff": 39, "end_trunc": 39, "gerp_dist": 39, "50_bp_rule": 39, "evolutionari": 39, "NOT": 39, "constrain": 39, "last": 39, "nonsens": 39, "mediat": 39, "decai": 39, "nmd": 39, "escap": 39, "compute_ranked_bin": [40, 41, 43], "score_expr": 40, "bin_expr": 40, "compute_snv_indel_separ": 40, "desc": 40, "mutual": 40, "exclus": 40, "var1": [40, 42], "var2": [40, 42], "var3": 40, "var4": 40, "biallelic_bin": 40, "singleton_bin": 40, "descend": 40, "compute_grouped_binned_ht": [40, 41, 43], "bin_ht": 40, "create_binned_ht": [40, 41, 43], "bin_id": 40, "bi_allel": 40, "score_bin_agg": [40, 41, 43], "_parent": [40, 43], "groupedt": [40, 43], "compute_binned_truth_sample_concord": [40, 41], "binned_score_ht": 40, "add_bin": 40, "concord": 40, "tp": [40, 43, 45], "fp": [40, 43, 45], "fn": 40, "truth_gt": 40, "create_truth_sample_ht": [40, 41], "truth_mt": 40, "high_confidence_intervals_ht": 40, "add_rank": [40, 41], "subrank_expr": 40, "subrank": 40, "rank_variant_count": 40, "form": 40, "name_of_subrank": 40, "subrank_filtering_expr": 40, "get_r_human_read": [41, 42], "get_r_for_pair_of_vari": [41, 42], "get_r_within_gene_in_gen_anc": [41, 42], "get_r_within_gen": [41, 42], "generate_trio_stat": [41, 43], "generate_sib_stat": [41, 43], "train_rf_model": [41, 43], "run_rf_test": [41, 44], "check_ht_fields_for_spark": [41, 44], "get_columns_quantil": [41, 44], "median_impute_featur": [41, 44], "ht_to_rf_df": [41, 44], "get_features_import": [41, 44], "get_label": [41, 44], "test_model": [41, 44], "apply_rf_model": [41, 44], "save_model": [41, 44], "load_model": [41, 44], "train_rf": [41, 43, 44], "get_rf_run": [41, 44], "get_run_data": [41, 44], "pretty_print_run": [41, 44], "sample_training_exampl": [41, 45], "ref_genom": 42, "get_ld_matrix": 42, "get_ld_index": 42, "parse_locu": 42, "10146": 42, "10151": 42, "ta": 42, "01789767935482124": 42, "tlocu": 42, "tarrai": 42, "quadrat": 42, "exercis": 42, "caution": 42, "get_gnomad_public_data": 42, "get_gene_interv": 42, "fast": 42, "add_substrat": 43, "meant": 43, "wrapper": [43, 44], "ac_raw": 43, "sub": 43, "fam_stats_ht": 43, "easili": 43, "binned_ht": 43, "grouped_binned_ht": 43, "agg_ht": 43, "ac_qc_samples_unrelated_raw": 43, "fail_hard_filt": 43, "truth_ht": 43, "omni": 43, "kgp_phase1_hc": 43, "n_de_novos_adj": 43, "n_de_novos_raw": 43, "n_transmitted_raw": 43, "n_untransmitted_raw": 43, "min_scor": 43, "minimun": 43, "max_scor": 43, "maiximum": 43, "n_in": 43, "n_del": 43, "n_ti": 43, "n_tv": 43, "trnasvers": 43, "n_1bp_indel": 43, "n_mod3bp_indel": 43, "divis": 43, "n_singleton": 43, "n_vqsr_pos_train": 43, "n_vqsr_neg_train": 43, "n_clinvar": 43, "n_de_novos_singleton_adj": 43, "n_de_novo_singleton": 43, "unfilt": 43, "n_de_novo": 43, "n_trans_singleton": 43, "n_untrans_singleton": 43, "untransmit": 43, "n_omni": 43, "n_mill": 43, "n_hapmap": 43, "n_kgp_phase1_hc": 43, "rf_featur": 43, "tp_expr": [43, 45], "fp_expr": [43, 45], "fp_to_tp": [43, 45], "num_tre": [43, 44], "max_depth": [43, 44], "test_expr": [43, 45], "withheld": [43, 44], "rf_test": 43, "features_import": [43, 44], "test_result": [43, 44], "maxmimum": 43, "hold": 43, "pipelinemodel": [43, 44], "tmp": 44, "dummi": 44, "quantil": 44, "relative_error": 44, "approx_median": 44, "feature_imput": 44, "features_median": 44, "variants_by_strata": 44, "basic": 44, "back": 44, "rf_pipelin": 44, "rf_index": 44, "assembler_index": 44, "vectorassembl": 44, "stringindex": 44, "rf_model": 44, "prediction_col_nam": 44, "rf_predict": 44, "confus": 44, "probability_col_nam": 44, "rf_probabl": 44, "out_path": 44, "input_path": 44, "rf_json_fp": 44, "wasn": 44, "input_arg": 44, "test_interv": 44, "hash": 44, "tv": 45, "effect": [46, 49], "gnomad": [47, 48, 50], "bill": [47, 50], "charg": [47, 50], "d": 47, "import_vcf": 47, "drop_sampl": 47, "drastic": 47, "speed": 47, "worst_consequence_term": 47, "short": 48, "guid": 48, "date": 48, "septemb": 48, "2024": 48, "ip": 48, "131": 48, "instal": 48, "pip": 48, "jupyt": 48, "notebook": 48, "gnomad_v2_exom": 48, "exomes_ht": 48, "gnomad_v2_genom": 48, "genomes_ht": 48, "gnomad_v3_genom": 48, "shut": 48, "finish": 48, "program": 50, "gc": 50, "gnomad_public_resource_configur": 50, "transfer": 50, "your": 50, "my": 50}, "objects": {"gnomad.assessment": [[1, 0, 0, "-", "parse_validity_logs"], [2, 0, 0, "-", "summary_stats"], [3, 0, 0, "-", "validity_checks"]], "gnomad.assessment.parse_validity_logs": [[1, 1, 1, "", "generate_html_report"], [1, 1, 1, "", "parse_log_file"]], "gnomad.assessment.summary_stats": [[2, 1, 1, "", "default_generate_gene_lof_matrix"], [2, 1, 1, "", "default_generate_gene_lof_summary"], [2, 1, 1, "", "freq_bin_expr"], [2, 1, 1, "", "generate_filter_combinations"], [2, 1, 1, "", "get_an_criteria"], [2, 1, 1, "", "get_het_hom_summary_dict"], [2, 1, 1, "", "get_summary_ac_dict"], [2, 1, 1, "", "get_summary_counts"], [2, 1, 1, "", "get_summary_counts_dict"], [2, 1, 1, "", "get_summary_stats_csq_filter_expr"], [2, 1, 1, "", "get_summary_stats_filter_group_meta"], [2, 1, 1, "", "get_summary_stats_variant_filter_expr"], [2, 1, 1, "", "get_tx_expression_expr"]], "gnomad.assessment.validity_checks": [[3, 1, 1, "", "check_global_and_row_annot_lengths"], [3, 1, 1, "", "check_globals_for_retired_terms"], [3, 1, 1, "", "check_missingness_of_struct"], [3, 1, 1, "", "check_raw_and_adj_callstats"], [3, 1, 1, "", "check_sex_chr_metrics"], [3, 1, 1, "", "compare_row_counts"], [3, 1, 1, "", "compare_subset_freqs"], [3, 1, 1, "", "compute_missingness"], [3, 1, 1, "", "count_vep_annotated_variants_per_interval"], [3, 1, 1, "", "custom_record_factory"], [3, 1, 1, "", "flatten_missingness_struct"], [3, 1, 1, "", "generate_field_check_expr"], [3, 1, 1, "", "generic_field_check"], [3, 1, 1, "", "generic_field_check_loop"], [3, 1, 1, "", "make_filters_expr_dict"], [3, 1, 1, "", "make_group_sum_expr_dict"], [3, 1, 1, "", "pprint_global_anns"], [3, 1, 1, "", "sum_group_callstats"], [3, 1, 1, "", "summarize_variant_filters"], [3, 1, 1, "", "summarize_variants"], [3, 1, 1, "", "temp_logger"], [3, 1, 1, "", "unfurl_array_annotations"], [3, 1, 1, "", "validate_release_t"], [3, 1, 1, "", "vcf_field_check"]], "gnomad.resources": [[5, 0, 0, "-", "config"], [13, 0, 0, "-", "import_resources"], [15, 0, 0, "-", "resource_utils"]], "gnomad.resources.config": [[5, 2, 1, "", "GnomadPublicResourceSource"], [5, 1, 1, "", "get_default_public_resource_source"]], "gnomad.resources.config.GnomadPublicResourceSource": [[5, 3, 1, "", "GNOMAD"], [5, 3, 1, "", "GOOGLE_CLOUD_PUBLIC_DATASETS"], [5, 3, 1, "", "REGISTRY_OF_OPEN_DATA_ON_AWS"]], "gnomad.resources.grch37": [[6, 0, 0, "-", "gnomad"], [7, 0, 0, "-", "gnomad_ld"], [9, 0, 0, "-", "reference_data"]], "gnomad.resources.grch37.gnomad": [[6, 1, 1, "", "browser_gene"], [6, 1, 1, "", "constraint"], [6, 1, 1, "", "coverage"], [6, 1, 1, "", "liftover"], [6, 1, 1, "", "mnv"], [6, 1, 1, "", "pext"], [6, 1, 1, "", "public_pca_loadings"], [6, 1, 1, "", "public_release"], [6, 1, 1, "", "release_vcf_path"]], "gnomad.resources.grch37.gnomad_ld": [[7, 1, 1, "", "ld_index"], [7, 1, 1, "", "ld_matrix"], [7, 1, 1, "", "ld_scores"]], "gnomad.resources.grch37.reference_data": [[9, 1, 1, "", "get_truth_ht"], [9, 1, 1, "", "transform_grch37_methylation"]], "gnomad.resources.grch38": [[10, 0, 0, "-", "gnomad"], [12, 0, 0, "-", "reference_data"]], "gnomad.resources.grch38.gnomad": [[10, 4, 1, "", "COHORTS_WITH_GEN_ANC_STORED_AS_SUBGRP"], [10, 4, 1, "", "DOWNSAMPLINGS"], [10, 4, 1, "", "GEN_ANC_GROUPS"], [10, 4, 1, "", "GEN_ANC_GROUPS_TO_REMOVE_FOR_GRPMAX"], [10, 4, 1, "", "GROUPS"], [10, 4, 1, "", "HGDP_GEN_ANC_GROUPS"], [10, 4, 1, "", "SEXES"], [10, 4, 1, "", "SUBSETS"], [10, 4, 1, "", "TGP_GEN_ANC_GROUPS"], [10, 4, 1, "", "TGP_GEN_ANC_GROUP_NAMES"], [10, 1, 1, "", "add_grpMaxFAF95_v4"], [10, 1, 1, "", "all_sites_an"], [10, 1, 1, "", "browser_gene"], [10, 1, 1, "", "browser_variant"], [10, 1, 1, "", "constraint"], [10, 1, 1, "", "coverage"], [10, 1, 1, "", "coverage_tsv_path"], [10, 1, 1, "", "gnomad_gks"], [10, 1, 1, "", "pext"], [10, 1, 1, "", "public_release"], [10, 1, 1, "", "release_vcf_path"]], "gnomad.resources.grch38.reference_data": [[12, 1, 1, "", "get_truth_ht"], [12, 1, 1, "", "transform_grch38_methylation"]], "gnomad.resources.import_resources": [[13, 1, 1, "", "get_module_importable_resources"], [13, 1, 1, "", "get_resources_descriptions"], [13, 1, 1, "", "main"]], "gnomad.resources.resource_utils": [[15, 2, 1, "", "BaseResource"], [15, 2, 1, "", "BaseVersionedResource"], [15, 2, 1, "", "BlockMatrixResource"], [15, 7, 1, "", "DataException"], [15, 2, 1, "", "ExpressionResource"], [15, 4, 1, "", "GNOMAD_PUBLIC_BUCKETS"], [15, 2, 1, "", "GnomadPublicBlockMatrixResource"], [15, 2, 1, "", "GnomadPublicMatrixTableResource"], [15, 2, 1, "", "GnomadPublicPedigreeResource"], [15, 2, 1, "", "GnomadPublicResource"], [15, 2, 1, "", "GnomadPublicTableResource"], [15, 2, 1, "", "MatrixTableResource"], [15, 2, 1, "", "PedigreeResource"], [15, 7, 1, "", "ResourceNotAvailable"], [15, 2, 1, "", "TableResource"], [15, 2, 1, "", "VariantDatasetResource"], [15, 2, 1, "", "VersionedBlockMatrixResource"], [15, 2, 1, "", "VersionedMatrixTableResource"], [15, 2, 1, "", "VersionedPedigreeResource"], [15, 2, 1, "", "VersionedTableResource"], [15, 2, 1, "", "VersionedVariantDatasetResource"], [15, 1, 1, "", "import_gencode"], [15, 1, 1, "", "import_sites_vcf"]], "gnomad.resources.resource_utils.BaseResource": [[15, 3, 1, "", "expected_file_extensions"], [15, 5, 1, "", "import_resource"], [15, 6, 1, "", "path"]], "gnomad.resources.resource_utils.BaseVersionedResource": [[15, 3, 1, "", "default_version"], [15, 3, 1, "", "resource_class"], [15, 3, 1, "", "versions"]], "gnomad.resources.resource_utils.BlockMatrixResource": [[15, 5, 1, "", "bm"], [15, 3, 1, "", "expected_file_extensions"], [15, 5, 1, "", "import_resource"]], "gnomad.resources.resource_utils.ExpressionResource": [[15, 3, 1, "", "expected_file_extensions"], [15, 5, 1, "", "he"], [15, 5, 1, "", "import_resource"]], "gnomad.resources.resource_utils.GnomadPublicBlockMatrixResource": [[15, 5, 1, "", "bm"]], "gnomad.resources.resource_utils.GnomadPublicMatrixTableResource": [[15, 5, 1, "", "mt"]], "gnomad.resources.resource_utils.GnomadPublicPedigreeResource": [[15, 5, 1, "", "ht"], [15, 5, 1, "", "pedigree"]], "gnomad.resources.resource_utils.GnomadPublicResource": [[15, 5, 1, "", "is_resource_available"]], "gnomad.resources.resource_utils.GnomadPublicTableResource": [[15, 5, 1, "", "ht"]], "gnomad.resources.resource_utils.MatrixTableResource": [[15, 3, 1, "", "expected_file_extensions"], [15, 5, 1, "", "import_resource"], [15, 5, 1, "", "mt"]], "gnomad.resources.resource_utils.PedigreeResource": [[15, 3, 1, "", "expected_file_extensions"], [15, 5, 1, "", "ht"], [15, 5, 1, "", "import_resource"], [15, 5, 1, "", "pedigree"]], "gnomad.resources.resource_utils.TableResource": [[15, 3, 1, "", "expected_file_extensions"], [15, 5, 1, "", "ht"], [15, 5, 1, "", "import_resource"]], "gnomad.resources.resource_utils.VariantDatasetResource": [[15, 3, 1, "", "expected_file_extensions"], [15, 5, 1, "", "import_resource"], [15, 5, 1, "", "vds"]], "gnomad.resources.resource_utils.VersionedBlockMatrixResource": [[15, 3, 1, "", "resource_class"]], "gnomad.resources.resource_utils.VersionedMatrixTableResource": [[15, 3, 1, "", "resource_class"]], "gnomad.resources.resource_utils.VersionedPedigreeResource": [[15, 3, 1, "", "resource_class"]], "gnomad.resources.resource_utils.VersionedTableResource": [[15, 3, 1, "", "resource_class"]], "gnomad.resources.resource_utils.VersionedVariantDatasetResource": [[15, 3, 1, "", "resource_class"]], "gnomad.sample_qc": [[16, 0, 0, "-", "ancestry"], [17, 0, 0, "-", "filtering"], [19, 0, 0, "-", "pipeline"], [20, 0, 0, "-", "platform"], [21, 0, 0, "-", "relatedness"], [22, 0, 0, "-", "sex"]], "gnomad.sample_qc.ancestry": [[16, 1, 1, "", "apply_onnx_classification_model"], [16, 1, 1, "", "apply_sklearn_classification_model"], [16, 1, 1, "", "assign_genetic_ancestry_pcs"], [16, 1, 1, "", "convert_sklearn_rf_to_onnx"], [16, 1, 1, "", "pc_project"], [16, 1, 1, "", "run_pca_with_relateds"]], "gnomad.sample_qc.filtering": [[17, 1, 1, "", "compute_qc_metrics_residuals"], [17, 1, 1, "", "compute_stratified_metrics_filter"], [17, 1, 1, "", "compute_stratified_sample_qc"], [17, 1, 1, "", "determine_nearest_neighbors"], [17, 1, 1, "", "merge_sample_qc_expr"]], "gnomad.sample_qc.pipeline": [[19, 1, 1, "", "annotate_sex"], [19, 1, 1, "", "filter_rows_for_qc"], [19, 1, 1, "", "get_qc_mt"], [19, 1, 1, "", "infer_sex_karyotype"]], "gnomad.sample_qc.platform": [[20, 1, 1, "", "assign_platform_from_pcs"], [20, 1, 1, "", "compute_callrate_mt"], [20, 1, 1, "", "run_platform_pca"]], "gnomad.sample_qc.relatedness": [[21, 4, 1, "", "AMBIGUOUS_RELATIONSHIP"], [21, 4, 1, "", "DUPLICATE_OR_TWINS"], [21, 4, 1, "", "PARENT_CHILD"], [21, 4, 1, "", "SECOND_DEGREE_RELATIVES"], [21, 4, 1, "", "SIBLINGS"], [21, 4, 1, "", "UNRELATED"], [21, 1, 1, "", "calculate_de_novo_post_prob"], [21, 1, 1, "", "compute_related_samples_to_drop"], [21, 1, 1, "", "create_fake_pedigree"], [21, 1, 1, "", "default_get_de_novo_expr"], [21, 1, 1, "", "explode_duplicate_samples_ht"], [21, 1, 1, "", "filter_to_trios"], [21, 1, 1, "", "generate_sib_stats_expr"], [21, 1, 1, "", "generate_trio_stats_expr"], [21, 1, 1, "", "get_duplicated_samples"], [21, 1, 1, "", "get_duplicated_samples_ht"], [21, 1, 1, "", "get_relationship_expr"], [21, 1, 1, "", "get_slope_int_relationship_expr"], [21, 1, 1, "", "infer_families"]], "gnomad.sample_qc.sex": [[22, 1, 1, "", "adjust_sex_ploidy"], [22, 1, 1, "", "adjusted_sex_ploidy_expr"], [22, 1, 1, "", "gaussian_mixture_model_karyotype_assignment"], [22, 1, 1, "", "get_chr_x_hom_alt_cutoffs"], [22, 1, 1, "", "get_ploidy_cutoffs"], [22, 1, 1, "", "get_sex_expr"]], "gnomad.utils": [[23, 0, 0, "-", "annotations"], [24, 0, 0, "-", "constraint"], [25, 0, 0, "-", "file_utils"], [26, 0, 0, "-", "filtering"], [27, 0, 0, "-", "gen_stats"], [29, 0, 0, "-", "intervals"], [30, 0, 0, "-", "liftover"], [31, 0, 0, "-", "parse"], [32, 0, 0, "-", "plotting"], [33, 0, 0, "-", "reference_genome"], [34, 0, 0, "-", "release"], [35, 0, 0, "-", "slack"], [36, 0, 0, "-", "sparse_mt"], [37, 0, 0, "-", "transcript_annotation"], [38, 0, 0, "-", "vcf"], [39, 0, 0, "-", "vep"]], "gnomad.utils.annotations": [[23, 1, 1, "", "add_gks_va"], [23, 1, 1, "", "add_gks_vrs"], [23, 1, 1, "", "add_variant_type"], [23, 1, 1, "", "age_hists_expr"], [23, 1, 1, "", "agg_by_strata"], [23, 1, 1, "", "annotate_adj"], [23, 1, 1, "", "annotate_allele_info"], [23, 1, 1, "", "annotate_and_index_source_mt_for_sex_ploidy"], [23, 1, 1, "", "annotate_downsamplings"], [23, 1, 1, "", "annotate_freq"], [23, 1, 1, "", "annotate_with_ht"], [23, 1, 1, "", "annotation_type_in_vcf_info"], [23, 1, 1, "", "annotation_type_is_numeric"], [23, 1, 1, "", "bi_allelic_expr"], [23, 1, 1, "", "bi_allelic_site_inbreeding_expr"], [23, 1, 1, "", "build_freq_stratification_list"], [23, 1, 1, "", "check_annotation_missingness"], [23, 1, 1, "", "compute_freq_by_strata"], [23, 1, 1, "", "create_frequency_bins_expr"], [23, 1, 1, "", "faf_expr"], [23, 1, 1, "", "fill_missing_key_combinations"], [23, 1, 1, "", "fs_from_sb"], [23, 1, 1, "", "gen_anc_faf_max_expr"], [23, 1, 1, "", "generate_freq_group_membership_array"], [23, 1, 1, "", "get_adj_expr"], [23, 1, 1, "", "get_annotations_hists"], [23, 1, 1, "", "get_copy_state_by_sex"], [23, 1, 1, "", "get_gq_dp_adj_expr"], [23, 1, 1, "", "get_het_ab_adj_expr"], [23, 1, 1, "", "get_is_haploid_expr"], [23, 1, 1, "", "get_lowqual_expr"], [23, 1, 1, "", "grpmax_expr"], [23, 1, 1, "", "hemi_expr"], [23, 1, 1, "", "merge_array_expressions"], [23, 1, 1, "", "merge_freq_arrays"], [23, 1, 1, "", "merge_histograms"], [23, 1, 1, "", "missing_callstats_expr"], [23, 1, 1, "", "missing_struct_expr"], [23, 1, 1, "", "pab_max_expr"], [23, 1, 1, "", "project_max_expr"], [23, 1, 1, "", "qual_hist_expr"], [23, 1, 1, "", "region_flag_expr"], [23, 1, 1, "", "set_xx_y_metrics_to_na_expr"], [23, 1, 1, "", "sor_from_sb"], [23, 1, 1, "", "unphase_call_expr"], [23, 1, 1, "", "update_structured_annotations"]], "gnomad.utils.constraint": [[24, 4, 1, "", "COVERAGE_CUTOFF"], [24, 1, 1, "", "add_gencode_transcript_annotations"], [24, 1, 1, "", "annotate_exploded_vep_for_constraint_groupings"], [24, 1, 1, "", "annotate_mutation_type"], [24, 1, 1, "", "annotate_with_mu"], [24, 1, 1, "", "assemble_constraint_context_ht"], [24, 1, 1, "", "build_coverage_model"], [24, 1, 1, "", "build_models"], [24, 1, 1, "", "build_plateau_models"], [24, 1, 1, "", "calculate_raw_z_score"], [24, 1, 1, "", "calculate_raw_z_score_sd"], [24, 1, 1, "", "collapse_strand"], [24, 1, 1, "", "compute_expected_variants"], [24, 1, 1, "", "compute_pli"], [24, 1, 1, "", "count_variants_by_group"], [24, 1, 1, "", "downsampling_counts_expr"], [24, 1, 1, "", "explode_downsamplings_oe"], [24, 1, 1, "", "get_all_gen_anc_lengths"], [24, 1, 1, "", "get_constraint_flags"], [24, 1, 1, "", "get_constraint_grouping_expr"], [24, 1, 1, "", "get_downsampling_freq_indices"], [24, 1, 1, "", "get_mu_annotation_expr"], [24, 1, 1, "", "oe_aggregation_expr"], [24, 1, 1, "", "oe_confidence_interval"], [24, 1, 1, "", "transform_methylation_level"], [24, 1, 1, "", "trimer_from_heptamer"]], "gnomad.utils.file_utils": [[25, 1, 1, "", "check_file_exists_raise_error"], [25, 1, 1, "", "create_vds"], [25, 1, 1, "", "file_exists"], [25, 1, 1, "", "get_file_stats"], [25, 1, 1, "", "read_list_data"], [25, 1, 1, "", "repartition_for_join"], [25, 1, 1, "", "select_primitives_from_ht"], [25, 1, 1, "", "write_temp_gcs"]], "gnomad.utils.filtering": [[26, 1, 1, "", "add_filters_expr"], [26, 1, 1, "", "combine_functions"], [26, 1, 1, "", "filter_arrays_by_meta"], [26, 1, 1, "", "filter_by_frequency"], [26, 1, 1, "", "filter_by_gencode_intervals"], [26, 1, 1, "", "filter_by_intervals"], [26, 1, 1, "", "filter_by_numeric_expr_range"], [26, 1, 1, "", "filter_for_mu"], [26, 1, 1, "", "filter_gencode_ht"], [26, 1, 1, "", "filter_low_conf_regions"], [26, 1, 1, "", "filter_to_adj"], [26, 1, 1, "", "filter_to_autosomes"], [26, 1, 1, "", "filter_to_clinvar_pathogenic"], [26, 1, 1, "", "filter_to_gencode_cds"], [26, 1, 1, "", "filter_x_nonpar"], [26, 1, 1, "", "filter_y_nonpar"], [26, 1, 1, "", "low_conf_regions_expr"], [26, 1, 1, "", "remove_fields_from_constant"], [26, 1, 1, "", "split_vds_by_strata"], [26, 1, 1, "", "subset_samples_and_variants"]], "gnomad.utils.gen_stats": [[27, 1, 1, "", "from_phred"], [27, 1, 1, "", "get_median_and_mad_expr"], [27, 1, 1, "", "merge_stats_counters_expr"], [27, 1, 1, "", "to_phred"]], "gnomad.utils.intervals": [[29, 1, 1, "", "interval_length"], [29, 1, 1, "", "pad_intervals"], [29, 1, 1, "", "sort_intervals"], [29, 1, 1, "", "union_intervals"]], "gnomad.utils.liftover": [[30, 4, 1, "", "GRCH37_to_GRCH38_CHAIN"], [30, 4, 1, "", "GRCH38_TO_GRCH37_CHAIN"], [30, 1, 1, "", "default_lift_data"], [30, 1, 1, "", "get_liftover_genome"], [30, 1, 1, "", "liftover_expr"], [30, 1, 1, "", "liftover_using_gnomad_map"]], "gnomad.utils.parse": [[31, 1, 1, "", "parse_locus_intervals"], [31, 1, 1, "", "parse_variant"]], "gnomad.utils.plotting": [[32, 1, 1, "", "get_rows_data"], [32, 1, 1, "", "linear_and_log_tabs"], [32, 1, 1, "", "new_show"], [32, 1, 1, "", "pair_plot"], [32, 1, 1, "", "plot_hail_file_metadata"], [32, 1, 1, "", "plot_hail_hist"], [32, 1, 1, "", "plot_hail_hist_both"], [32, 1, 1, "", "plot_hail_hist_cumulative"], [32, 1, 1, "", "plot_multi_hail_hist"], [32, 1, 1, "", "scale_file_sizes"], [32, 1, 1, "", "set_font_size"]], "gnomad.utils.reference_genome": [[33, 1, 1, "", "add_reference_sequence"], [33, 1, 1, "", "get_reference_genome"], [33, 1, 1, "", "get_reference_ht"]], "gnomad.utils.release": [[34, 1, 1, "", "make_faf_index_dict"], [34, 1, 1, "", "make_freq_index_dict"], [34, 1, 1, "", "make_freq_index_dict_from_meta"]], "gnomad.utils.slack": [[35, 2, 1, "", "SlackClient"], [35, 1, 1, "", "slack_notifications"]], "gnomad.utils.slack.SlackClient": [[35, 5, 1, "", "send_file"], [35, 5, 1, "", "send_message"]], "gnomad.utils.sparse_mt": [[36, 1, 1, "", "compute_allele_number_per_ref_site"], [36, 1, 1, "", "compute_coverage_stats"], [36, 1, 1, "", "compute_last_ref_block_end"], [36, 1, 1, "", "compute_stats_per_ref_site"], [36, 1, 1, "", "default_compute_info"], [36, 1, 1, "", "densify_all_reference_sites"], [36, 1, 1, "", "densify_sites"], [36, 1, 1, "", "filter_ref_blocks"], [36, 1, 1, "", "get_allele_number_agg_func"], [36, 1, 1, "", "get_as_info_expr"], [36, 1, 1, "", "get_coverage_agg_func"], [36, 1, 1, "", "get_site_info_expr"], [36, 1, 1, "", "impute_sex_ploidy"], [36, 1, 1, "", "split_info_annotation"], [36, 1, 1, "", "split_lowqual_annotation"]], "gnomad.utils.transcript_annotation": [[37, 4, 1, "", "CELL_LINES"], [37, 4, 1, "", "REPRODUCTIVE_TISSUES"], [37, 1, 1, "", "clean_tissue_name_for_browser"], [37, 1, 1, "", "create_tx_annotation_by_region"], [37, 1, 1, "", "filter_expression_ht_by_tissues"], [37, 1, 1, "", "get_expression_proportion"], [37, 1, 1, "", "get_tissues_to_exclude"], [37, 1, 1, "", "perform_tx_annotation_pipeline"], [37, 1, 1, "", "summarize_transcript_expression"], [37, 1, 1, "", "tissue_expression_ht_to_array"], [37, 1, 1, "", "tx_aggregate_variants"], [37, 1, 1, "", "tx_annotate_variants"], [37, 1, 1, "", "tx_filter_variants_by_csqs"]], "gnomad.utils.vcf": [[38, 4, 1, "", "ALLELE_TYPE_FIELDS"], [38, 4, 1, "", "AS_FIELDS"], [38, 4, 1, "", "AS_VQSR_FIELDS"], [38, 4, 1, "", "ENTRIES"], [38, 4, 1, "", "FAF_GEN_ANC_GROUPS"], [38, 4, 1, "", "FORMAT_DICT"], [38, 4, 1, "", "GROUPS"], [38, 4, 1, "", "HISTS"], [38, 4, 1, "", "INFO_DICT"], [38, 4, 1, "", "IN_SILICO_ANNOTATIONS_INFO_DICT"], [38, 4, 1, "", "JOINT_REGION_FLAG_FIELDS"], [38, 4, 1, "", "REGION_FLAG_FIELDS"], [38, 4, 1, "", "RF_FIELDS"], [38, 4, 1, "", "SEXES"], [38, 4, 1, "", "SITE_FIELDS"], [38, 4, 1, "", "SORT_ORDER"], [38, 4, 1, "", "SPARSE_ENTRIES"], [38, 4, 1, "", "VQSR_FIELDS"], [38, 4, 1, "", "VRS_FIELDS_DICT"], [38, 1, 1, "", "add_as_info_dict"], [38, 1, 1, "", "adjust_vcf_incompatible_types"], [38, 1, 1, "", "build_vcf_export_reference"], [38, 1, 1, "", "create_label_groups"], [38, 1, 1, "", "index_globals"], [38, 1, 1, "", "make_combo_header_text"], [38, 1, 1, "", "make_hist_bin_edges_expr"], [38, 1, 1, "", "make_hist_dict"], [38, 1, 1, "", "make_info_dict"], [38, 1, 1, "", "make_label_combos"], [38, 1, 1, "", "make_vcf_filter_dict"], [38, 1, 1, "", "rekey_new_reference"], [38, 1, 1, "", "set_xx_y_metrics_to_na"]], "gnomad.utils.vep": [[39, 4, 1, "", "CSQ_CODING"], [39, 4, 1, "", "CSQ_SPLICE"], [39, 4, 1, "", "CURRENT_VEP_VERSION"], [39, 4, 1, "", "LOFTEE_LABELS"], [39, 4, 1, "", "LOF_CSQ_SET"], [39, 4, 1, "", "POSSIBLE_REFS"], [39, 4, 1, "", "VEP_CONFIG_PATH"], [39, 4, 1, "", "VEP_CSQ_FIELDS"], [39, 4, 1, "", "VEP_CSQ_HEADER"], [39, 1, 1, "", "add_most_severe_consequence_to_consequence"], [39, 1, 1, "", "add_most_severe_csq_to_tc_within_vep_root"], [39, 1, 1, "", "explode_by_vep_annotation"], [39, 1, 1, "", "filter_vep_to_canonical_transcripts"], [39, 1, 1, "", "filter_vep_to_gene_list"], [39, 1, 1, "", "filter_vep_to_mane_select_transcripts"], [39, 1, 1, "", "filter_vep_to_synonymous_variants"], [39, 1, 1, "", "filter_vep_transcript_csqs"], [39, 1, 1, "", "filter_vep_transcript_csqs_expr"], [39, 1, 1, "", "get_loftee_end_trunc_filter_expr"], [39, 1, 1, "", "get_most_severe_consequence_expr"], [39, 1, 1, "", "get_most_severe_consequence_for_summary"], [39, 1, 1, "", "get_vep_context"], [39, 1, 1, "", "get_vep_csq_header"], [39, 1, 1, "", "get_vep_help"], [39, 1, 1, "", "process_consequences"], [39, 1, 1, "", "update_loftee_end_trunc_filter"], [39, 1, 1, "", "vep_or_lookup_vep"], [39, 1, 1, "", "vep_struct_to_csq"]], "gnomad.variant_qc": [[40, 0, 0, "-", "evaluation"], [42, 0, 0, "-", "ld"], [43, 0, 0, "-", "pipeline"], [44, 0, 0, "-", "random_forest"], [45, 0, 0, "-", "training"]], "gnomad.variant_qc.evaluation": [[40, 1, 1, "", "add_rank"], [40, 1, 1, "", "compute_binned_truth_sample_concordance"], [40, 1, 1, "", "compute_grouped_binned_ht"], [40, 1, 1, "", "compute_ranked_bin"], [40, 1, 1, "", "create_truth_sample_ht"]], "gnomad.variant_qc.ld": [[42, 1, 1, "", "get_r_for_pair_of_variants"], [42, 1, 1, "", "get_r_human_readable"], [42, 1, 1, "", "get_r_within_gene"], [42, 1, 1, "", "get_r_within_gene_in_gen_anc"]], "gnomad.variant_qc.pipeline": [[43, 1, 1, "", "create_binned_ht"], [43, 1, 1, "", "generate_sib_stats"], [43, 1, 1, "", "generate_trio_stats"], [43, 1, 1, "", "score_bin_agg"], [43, 1, 1, "", "train_rf_model"]], "gnomad.variant_qc.random_forest": [[44, 1, 1, "", "apply_rf_model"], [44, 1, 1, "", "check_ht_fields_for_spark"], [44, 1, 1, "", "get_columns_quantiles"], [44, 1, 1, "", "get_features_importance"], [44, 1, 1, "", "get_labels"], [44, 1, 1, "", "get_rf_runs"], [44, 1, 1, "", "get_run_data"], [44, 1, 1, "", "ht_to_rf_df"], [44, 1, 1, "", "load_model"], [44, 1, 1, "", "median_impute_features"], [44, 1, 1, "", "pretty_print_runs"], [44, 1, 1, "", "run_rf_test"], [44, 1, 1, "", "save_model"], [44, 1, 1, "", "test_model"], [44, 1, 1, "", "train_rf"]], "gnomad.variant_qc.training": [[45, 1, 1, "", "sample_training_examples"]]}, "objtypes": {"0": "py:module", "1": "py:function", "2": "py:class", "3": "py:attribute", "4": "py:data", "5": "py:method", "6": "py:property", "7": "py:exception"}, "objnames": {"0": ["py", "module", "Python module"], "1": ["py", "function", "Python function"], "2": ["py", "class", "Python class"], "3": ["py", "attribute", "Python attribute"], "4": ["py", "data", "Python data"], "5": ["py", "method", "Python method"], "6": ["py", "property", "Python property"], "7": ["py", "exception", "Python exception"]}, "titleterms": {"gnomad": [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 49], "assess": [0, 1, 2, 3], "parse_validity_log": 1, "summary_stat": 2, "validity_check": 3, "resourc": [5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 50], "config": 5, "grch37": [6, 7, 8, 9], "gnomad_ld": 7, "reference_data": [9, 12], "grch38": [10, 11, 12], "import_resourc": 13, "resource_util": 15, "sample_qc": [16, 17, 18, 19, 20, 21, 22], "ancestri": 16, "filter": [17, 26], "pipelin": [19, 43], "platform": 20, "related": 21, "sex": 22, "util": [23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39], "annot": 23, "constraint": 24, "file_util": 25, "gen_stat": 27, "interv": 29, "liftov": 30, "pars": 31, "plot": 32, "reference_genom": 33, "releas": 34, "slack": 35, "sparse_mt": 36, "transcript_annot": 37, "input": 37, "hail": 37, "tabl": 37, "output": 37, "vcf": 38, "vep": [39, 47], "variant_qc": [40, 41, 42, 43, 44, 45], "evalu": 40, "ld": 42, "random_forest": 44, "train": 45, "exampl": 46, "variant": 47, "effect": 47, "predictor": 47, "get": 48, "start": 48, "content": 49, "sourc": 50, "custom": 50, "environ": 50, "configur": 50}, "envversion": {"sphinx.domains.c": 2, "sphinx.domains.changeset": 1, "sphinx.domains.citation": 1, "sphinx.domains.cpp": 8, "sphinx.domains.index": 1, "sphinx.domains.javascript": 2, "sphinx.domains.math": 2, "sphinx.domains.python": 3, "sphinx.domains.rst": 2, "sphinx.domains.std": 2, "sphinx.ext.intersphinx": 1, "sphinx.ext.viewcode": 1, "sphinx": 57}, "alltitles": {"gnomad.assessment": [[0, "gnomad-assessment"]], "gnomad.assessment.parse_validity_logs": [[1, "gnomad-assessment-parse-validity-logs"]], "gnomad.assessment.summary_stats": [[2, "gnomad-assessment-summary-stats"]], "gnomad.assessment.validity_checks": [[3, "gnomad-assessment-validity-checks"]], "gnomad": [[4, "gnomad"], [49, "gnomad"]], "gnomad.resources.config": [[5, "gnomad-resources-config"]], "gnomad.resources.grch37.gnomad": [[6, "gnomad-resources-grch37-gnomad"]], "gnomad.resources.grch37.gnomad_ld": [[7, "gnomad-resources-grch37-gnomad-ld"]], "gnomad.resources.grch37": [[8, "gnomad-resources-grch37"]], "gnomad.resources.grch37.reference_data": [[9, "gnomad-resources-grch37-reference-data"]], "gnomad.resources.grch38.gnomad": [[10, "gnomad-resources-grch38-gnomad"]], "gnomad.resources.grch38": [[11, "gnomad-resources-grch38"]], "gnomad.resources.grch38.reference_data": [[12, "gnomad-resources-grch38-reference-data"]], "gnomad.resources.import_resources": [[13, "gnomad-resources-import-resources"]], "gnomad.resources": [[14, "gnomad-resources"]], "gnomad.resources.resource_utils": [[15, "gnomad-resources-resource-utils"]], "gnomad.sample_qc.ancestry": [[16, "gnomad-sample-qc-ancestry"]], "gnomad.sample_qc.filtering": [[17, "gnomad-sample-qc-filtering"]], "gnomad.sample_qc": [[18, "gnomad-sample-qc"]], "gnomad.sample_qc.pipeline": [[19, "gnomad-sample-qc-pipeline"]], "gnomad.sample_qc.platform": [[20, "gnomad-sample-qc-platform"]], "gnomad.sample_qc.relatedness": [[21, "gnomad-sample-qc-relatedness"]], "gnomad.sample_qc.sex": [[22, "gnomad-sample-qc-sex"]], "gnomad.utils.annotations": [[23, "gnomad-utils-annotations"]], "gnomad.utils.constraint": [[24, "gnomad-utils-constraint"]], "gnomad.utils.file_utils": [[25, "gnomad-utils-file-utils"]], "gnomad.utils.filtering": [[26, "gnomad-utils-filtering"]], "gnomad.utils.gen_stats": [[27, "gnomad-utils-gen-stats"]], "gnomad.utils": [[28, "gnomad-utils"]], "gnomad.utils.intervals": [[29, "gnomad-utils-intervals"]], "gnomad.utils.liftover": [[30, "gnomad-utils-liftover"]], "gnomad.utils.parse": [[31, "gnomad-utils-parse"]], "gnomad.utils.plotting": [[32, "gnomad-utils-plotting"]], "gnomad.utils.reference_genome": [[33, "gnomad-utils-reference-genome"]], "gnomad.utils.release": [[34, "gnomad-utils-release"]], "gnomad.utils.slack": [[35, "gnomad-utils-slack"]], "gnomad.utils.sparse_mt": [[36, "gnomad-utils-sparse-mt"]], "gnomad.utils.transcript_annotation": [[37, "gnomad-utils-transcript-annotation"]], "Input Hail Table": [[37, "id1"]], "Output Hail Table": [[37, "id2"]], "gnomad.utils.vcf": [[38, "gnomad-utils-vcf"]], "gnomad.utils.vep": [[39, "gnomad-utils-vep"]], "gnomad.variant_qc.evaluation": [[40, "gnomad-variant-qc-evaluation"]], "gnomad.variant_qc": [[41, "gnomad-variant-qc"]], "gnomad.variant_qc.ld": [[42, "gnomad-variant-qc-ld"]], "gnomad.variant_qc.pipeline": [[43, "gnomad-variant-qc-pipeline"]], "gnomad.variant_qc.random_forest": [[44, "gnomad-variant-qc-random-forest"]], "gnomad.variant_qc.training": [[45, "gnomad-variant-qc-training"]], "Examples": [[46, "examples"]], "Variant Effect Predictor (VEP)": [[47, "variant-effect-predictor-vep"]], "Getting Started": [[48, "getting-started"]], "Contents": [[49, "contents"]], "Resource Sources": [[50, "resource-sources"]], "Custom Sources": [[50, "custom-sources"]], "Environment Configuration": [[50, "environment-configuration"]]}, "indexentries": {"generate_html_report() (in module gnomad.assessment.parse_validity_logs)": [[1, "gnomad.assessment.parse_validity_logs.generate_html_report"]], "gnomad.assessment.parse_validity_logs": [[1, "module-gnomad.assessment.parse_validity_logs"]], "module": [[1, "module-gnomad.assessment.parse_validity_logs"], [2, "module-gnomad.assessment.summary_stats"], [3, "module-gnomad.assessment.validity_checks"], [5, "module-gnomad.resources.config"], [6, "module-gnomad.resources.grch37.gnomad"], [7, "module-gnomad.resources.grch37.gnomad_ld"], [9, "module-gnomad.resources.grch37.reference_data"], [10, "module-gnomad.resources.grch38.gnomad"], [12, "module-gnomad.resources.grch38.reference_data"], [13, "module-gnomad.resources.import_resources"], [15, "module-gnomad.resources.resource_utils"], [16, "module-gnomad.sample_qc.ancestry"], [17, "module-gnomad.sample_qc.filtering"], [19, "module-gnomad.sample_qc.pipeline"], [20, "module-gnomad.sample_qc.platform"], [21, "module-gnomad.sample_qc.relatedness"], [22, "module-gnomad.sample_qc.sex"], [23, "module-gnomad.utils.annotations"], [24, "module-gnomad.utils.constraint"], [25, "module-gnomad.utils.file_utils"], [26, "module-gnomad.utils.filtering"], [27, "module-gnomad.utils.gen_stats"], [29, "module-gnomad.utils.intervals"], [30, "module-gnomad.utils.liftover"], [31, "module-gnomad.utils.parse"], [32, "module-gnomad.utils.plotting"], [33, "module-gnomad.utils.reference_genome"], [34, "module-gnomad.utils.release"], [35, "module-gnomad.utils.slack"], [36, "module-gnomad.utils.sparse_mt"], [37, "module-gnomad.utils.transcript_annotation"], [38, "module-gnomad.utils.vcf"], [39, "module-gnomad.utils.vep"], [40, "module-gnomad.variant_qc.evaluation"], [42, "module-gnomad.variant_qc.ld"], [43, "module-gnomad.variant_qc.pipeline"], [44, "module-gnomad.variant_qc.random_forest"], [45, "module-gnomad.variant_qc.training"]], "parse_log_file() (in module gnomad.assessment.parse_validity_logs)": [[1, "gnomad.assessment.parse_validity_logs.parse_log_file"]], "default_generate_gene_lof_matrix() (in module gnomad.assessment.summary_stats)": [[2, "gnomad.assessment.summary_stats.default_generate_gene_lof_matrix"]], "default_generate_gene_lof_summary() (in module gnomad.assessment.summary_stats)": [[2, "gnomad.assessment.summary_stats.default_generate_gene_lof_summary"]], "freq_bin_expr() (in module gnomad.assessment.summary_stats)": [[2, "gnomad.assessment.summary_stats.freq_bin_expr"]], "generate_filter_combinations() (in module gnomad.assessment.summary_stats)": [[2, "gnomad.assessment.summary_stats.generate_filter_combinations"]], "get_an_criteria() (in module gnomad.assessment.summary_stats)": [[2, "gnomad.assessment.summary_stats.get_an_criteria"]], "get_het_hom_summary_dict() (in module gnomad.assessment.summary_stats)": [[2, "gnomad.assessment.summary_stats.get_het_hom_summary_dict"]], "get_summary_ac_dict() (in module gnomad.assessment.summary_stats)": [[2, "gnomad.assessment.summary_stats.get_summary_ac_dict"]], "get_summary_counts() (in module gnomad.assessment.summary_stats)": [[2, "gnomad.assessment.summary_stats.get_summary_counts"]], "get_summary_counts_dict() (in module gnomad.assessment.summary_stats)": [[2, "gnomad.assessment.summary_stats.get_summary_counts_dict"]], "get_summary_stats_csq_filter_expr() (in module gnomad.assessment.summary_stats)": [[2, "gnomad.assessment.summary_stats.get_summary_stats_csq_filter_expr"]], "get_summary_stats_filter_group_meta() (in module gnomad.assessment.summary_stats)": [[2, "gnomad.assessment.summary_stats.get_summary_stats_filter_group_meta"]], "get_summary_stats_variant_filter_expr() (in module gnomad.assessment.summary_stats)": [[2, "gnomad.assessment.summary_stats.get_summary_stats_variant_filter_expr"]], "get_tx_expression_expr() (in module gnomad.assessment.summary_stats)": [[2, "gnomad.assessment.summary_stats.get_tx_expression_expr"]], "gnomad.assessment.summary_stats": [[2, "module-gnomad.assessment.summary_stats"]], "check_global_and_row_annot_lengths() (in module gnomad.assessment.validity_checks)": [[3, "gnomad.assessment.validity_checks.check_global_and_row_annot_lengths"]], "check_globals_for_retired_terms() (in module gnomad.assessment.validity_checks)": [[3, "gnomad.assessment.validity_checks.check_globals_for_retired_terms"]], "check_missingness_of_struct() (in module gnomad.assessment.validity_checks)": [[3, "gnomad.assessment.validity_checks.check_missingness_of_struct"]], "check_raw_and_adj_callstats() (in module gnomad.assessment.validity_checks)": [[3, "gnomad.assessment.validity_checks.check_raw_and_adj_callstats"]], "check_sex_chr_metrics() (in module gnomad.assessment.validity_checks)": [[3, "gnomad.assessment.validity_checks.check_sex_chr_metrics"]], "compare_row_counts() (in module gnomad.assessment.validity_checks)": [[3, "gnomad.assessment.validity_checks.compare_row_counts"]], "compare_subset_freqs() (in module gnomad.assessment.validity_checks)": [[3, "gnomad.assessment.validity_checks.compare_subset_freqs"]], "compute_missingness() (in module gnomad.assessment.validity_checks)": [[3, "gnomad.assessment.validity_checks.compute_missingness"]], "count_vep_annotated_variants_per_interval() (in module gnomad.assessment.validity_checks)": [[3, "gnomad.assessment.validity_checks.count_vep_annotated_variants_per_interval"]], "custom_record_factory() (in module gnomad.assessment.validity_checks)": [[3, "gnomad.assessment.validity_checks.custom_record_factory"]], "flatten_missingness_struct() (in module gnomad.assessment.validity_checks)": [[3, "gnomad.assessment.validity_checks.flatten_missingness_struct"]], "generate_field_check_expr() (in module gnomad.assessment.validity_checks)": [[3, "gnomad.assessment.validity_checks.generate_field_check_expr"]], "generic_field_check() (in module gnomad.assessment.validity_checks)": [[3, "gnomad.assessment.validity_checks.generic_field_check"]], "generic_field_check_loop() (in module gnomad.assessment.validity_checks)": [[3, "gnomad.assessment.validity_checks.generic_field_check_loop"]], "gnomad.assessment.validity_checks": [[3, "module-gnomad.assessment.validity_checks"]], "make_filters_expr_dict() (in module gnomad.assessment.validity_checks)": [[3, "gnomad.assessment.validity_checks.make_filters_expr_dict"]], "make_group_sum_expr_dict() (in module gnomad.assessment.validity_checks)": [[3, "gnomad.assessment.validity_checks.make_group_sum_expr_dict"]], "pprint_global_anns() (in module gnomad.assessment.validity_checks)": [[3, "gnomad.assessment.validity_checks.pprint_global_anns"]], "sum_group_callstats() (in module gnomad.assessment.validity_checks)": [[3, "gnomad.assessment.validity_checks.sum_group_callstats"]], "summarize_variant_filters() (in module gnomad.assessment.validity_checks)": [[3, "gnomad.assessment.validity_checks.summarize_variant_filters"]], "summarize_variants() (in module gnomad.assessment.validity_checks)": [[3, "gnomad.assessment.validity_checks.summarize_variants"]], "temp_logger() (in module gnomad.assessment.validity_checks)": [[3, "gnomad.assessment.validity_checks.temp_logger"]], "unfurl_array_annotations() (in module gnomad.assessment.validity_checks)": [[3, "gnomad.assessment.validity_checks.unfurl_array_annotations"]], "validate_release_t() (in module gnomad.assessment.validity_checks)": [[3, "gnomad.assessment.validity_checks.validate_release_t"]], "vcf_field_check() (in module gnomad.assessment.validity_checks)": [[3, "gnomad.assessment.validity_checks.vcf_field_check"]], "gnomad (gnomad.resources.config.gnomadpublicresourcesource attribute)": [[5, "gnomad.resources.config.GnomadPublicResourceSource.GNOMAD"]], "google_cloud_public_datasets (gnomad.resources.config.gnomadpublicresourcesource attribute)": [[5, "gnomad.resources.config.GnomadPublicResourceSource.GOOGLE_CLOUD_PUBLIC_DATASETS"]], "gnomadpublicresourcesource (class in gnomad.resources.config)": [[5, "gnomad.resources.config.GnomadPublicResourceSource"]], "registry_of_open_data_on_aws (gnomad.resources.config.gnomadpublicresourcesource attribute)": [[5, "gnomad.resources.config.GnomadPublicResourceSource.REGISTRY_OF_OPEN_DATA_ON_AWS"]], "get_default_public_resource_source() (in module gnomad.resources.config)": [[5, "gnomad.resources.config.get_default_public_resource_source"]], "gnomad.resources.config": [[5, "module-gnomad.resources.config"]], "browser_gene() (in module gnomad.resources.grch37.gnomad)": [[6, "gnomad.resources.grch37.gnomad.browser_gene"]], "constraint() (in module gnomad.resources.grch37.gnomad)": [[6, "gnomad.resources.grch37.gnomad.constraint"]], "coverage() (in module gnomad.resources.grch37.gnomad)": [[6, "gnomad.resources.grch37.gnomad.coverage"]], "gnomad.resources.grch37.gnomad": [[6, "module-gnomad.resources.grch37.gnomad"]], "liftover() (in module gnomad.resources.grch37.gnomad)": [[6, "gnomad.resources.grch37.gnomad.liftover"]], "mnv() (in module gnomad.resources.grch37.gnomad)": [[6, "gnomad.resources.grch37.gnomad.mnv"]], "pext() (in module gnomad.resources.grch37.gnomad)": [[6, "gnomad.resources.grch37.gnomad.pext"]], "public_pca_loadings() (in module gnomad.resources.grch37.gnomad)": [[6, "gnomad.resources.grch37.gnomad.public_pca_loadings"]], "public_release() (in module gnomad.resources.grch37.gnomad)": [[6, "gnomad.resources.grch37.gnomad.public_release"]], "release_vcf_path() (in module gnomad.resources.grch37.gnomad)": [[6, "gnomad.resources.grch37.gnomad.release_vcf_path"]], "gnomad.resources.grch37.gnomad_ld": [[7, "module-gnomad.resources.grch37.gnomad_ld"]], "ld_index() (in module gnomad.resources.grch37.gnomad_ld)": [[7, "gnomad.resources.grch37.gnomad_ld.ld_index"]], "ld_matrix() (in module gnomad.resources.grch37.gnomad_ld)": [[7, "gnomad.resources.grch37.gnomad_ld.ld_matrix"]], "ld_scores() (in module gnomad.resources.grch37.gnomad_ld)": [[7, "gnomad.resources.grch37.gnomad_ld.ld_scores"]], "get_truth_ht() (in module gnomad.resources.grch37.reference_data)": [[9, "gnomad.resources.grch37.reference_data.get_truth_ht"]], "gnomad.resources.grch37.reference_data": [[9, "module-gnomad.resources.grch37.reference_data"]], "transform_grch37_methylation() (in module gnomad.resources.grch37.reference_data)": [[9, "gnomad.resources.grch37.reference_data.transform_grch37_methylation"]], "cohorts_with_gen_anc_stored_as_subgrp (in module gnomad.resources.grch38.gnomad)": [[10, "gnomad.resources.grch38.gnomad.COHORTS_WITH_GEN_ANC_STORED_AS_SUBGRP"]], "downsamplings (in module gnomad.resources.grch38.gnomad)": [[10, "gnomad.resources.grch38.gnomad.DOWNSAMPLINGS"]], "gen_anc_groups (in module gnomad.resources.grch38.gnomad)": [[10, "gnomad.resources.grch38.gnomad.GEN_ANC_GROUPS"]], "gen_anc_groups_to_remove_for_grpmax (in module gnomad.resources.grch38.gnomad)": [[10, "gnomad.resources.grch38.gnomad.GEN_ANC_GROUPS_TO_REMOVE_FOR_GRPMAX"]], "groups (in module gnomad.resources.grch38.gnomad)": [[10, "gnomad.resources.grch38.gnomad.GROUPS"]], "hgdp_gen_anc_groups (in module gnomad.resources.grch38.gnomad)": [[10, "gnomad.resources.grch38.gnomad.HGDP_GEN_ANC_GROUPS"]], "sexes (in module gnomad.resources.grch38.gnomad)": [[10, "gnomad.resources.grch38.gnomad.SEXES"]], "subsets (in module gnomad.resources.grch38.gnomad)": [[10, "gnomad.resources.grch38.gnomad.SUBSETS"]], "tgp_gen_anc_groups (in module gnomad.resources.grch38.gnomad)": [[10, "gnomad.resources.grch38.gnomad.TGP_GEN_ANC_GROUPS"]], "tgp_gen_anc_group_names (in module gnomad.resources.grch38.gnomad)": [[10, "gnomad.resources.grch38.gnomad.TGP_GEN_ANC_GROUP_NAMES"]], "add_grpmaxfaf95_v4() (in module gnomad.resources.grch38.gnomad)": [[10, "gnomad.resources.grch38.gnomad.add_grpMaxFAF95_v4"]], "all_sites_an() (in module gnomad.resources.grch38.gnomad)": [[10, "gnomad.resources.grch38.gnomad.all_sites_an"]], "browser_gene() (in module gnomad.resources.grch38.gnomad)": [[10, "gnomad.resources.grch38.gnomad.browser_gene"]], "browser_variant() (in module gnomad.resources.grch38.gnomad)": [[10, "gnomad.resources.grch38.gnomad.browser_variant"]], "constraint() (in module gnomad.resources.grch38.gnomad)": [[10, "gnomad.resources.grch38.gnomad.constraint"]], "coverage() (in module gnomad.resources.grch38.gnomad)": [[10, "gnomad.resources.grch38.gnomad.coverage"]], "coverage_tsv_path() (in module gnomad.resources.grch38.gnomad)": [[10, "gnomad.resources.grch38.gnomad.coverage_tsv_path"]], "gnomad.resources.grch38.gnomad": [[10, "module-gnomad.resources.grch38.gnomad"]], "gnomad_gks() (in module gnomad.resources.grch38.gnomad)": [[10, "gnomad.resources.grch38.gnomad.gnomad_gks"]], "pext() (in module gnomad.resources.grch38.gnomad)": [[10, "gnomad.resources.grch38.gnomad.pext"]], "public_release() (in module gnomad.resources.grch38.gnomad)": [[10, "gnomad.resources.grch38.gnomad.public_release"]], "release_vcf_path() (in module gnomad.resources.grch38.gnomad)": [[10, "gnomad.resources.grch38.gnomad.release_vcf_path"]], "get_truth_ht() (in module gnomad.resources.grch38.reference_data)": [[12, "gnomad.resources.grch38.reference_data.get_truth_ht"]], "gnomad.resources.grch38.reference_data": [[12, "module-gnomad.resources.grch38.reference_data"]], "transform_grch38_methylation() (in module gnomad.resources.grch38.reference_data)": [[12, "gnomad.resources.grch38.reference_data.transform_grch38_methylation"]], "get_module_importable_resources() (in module gnomad.resources.import_resources)": [[13, "gnomad.resources.import_resources.get_module_importable_resources"]], "get_resources_descriptions() (in module gnomad.resources.import_resources)": [[13, "gnomad.resources.import_resources.get_resources_descriptions"]], "gnomad.resources.import_resources": [[13, "module-gnomad.resources.import_resources"]], "main() (in module gnomad.resources.import_resources)": [[13, "gnomad.resources.import_resources.main"]], "baseresource (class in gnomad.resources.resource_utils)": [[15, "gnomad.resources.resource_utils.BaseResource"]], "baseversionedresource (class in gnomad.resources.resource_utils)": [[15, "gnomad.resources.resource_utils.BaseVersionedResource"]], "blockmatrixresource (class in gnomad.resources.resource_utils)": [[15, "gnomad.resources.resource_utils.BlockMatrixResource"]], "dataexception": [[15, "gnomad.resources.resource_utils.DataException"]], "expressionresource (class in gnomad.resources.resource_utils)": [[15, "gnomad.resources.resource_utils.ExpressionResource"]], "gnomad_public_buckets (in module gnomad.resources.resource_utils)": [[15, "gnomad.resources.resource_utils.GNOMAD_PUBLIC_BUCKETS"]], "gnomadpublicblockmatrixresource (class in gnomad.resources.resource_utils)": [[15, "gnomad.resources.resource_utils.GnomadPublicBlockMatrixResource"]], "gnomadpublicmatrixtableresource (class in gnomad.resources.resource_utils)": [[15, "gnomad.resources.resource_utils.GnomadPublicMatrixTableResource"]], "gnomadpublicpedigreeresource (class in gnomad.resources.resource_utils)": [[15, "gnomad.resources.resource_utils.GnomadPublicPedigreeResource"]], "gnomadpublicresource (class in gnomad.resources.resource_utils)": [[15, "gnomad.resources.resource_utils.GnomadPublicResource"]], "gnomadpublictableresource (class in gnomad.resources.resource_utils)": [[15, "gnomad.resources.resource_utils.GnomadPublicTableResource"]], "matrixtableresource (class in gnomad.resources.resource_utils)": [[15, "gnomad.resources.resource_utils.MatrixTableResource"]], "pedigreeresource (class in gnomad.resources.resource_utils)": [[15, "gnomad.resources.resource_utils.PedigreeResource"]], "resourcenotavailable": [[15, "gnomad.resources.resource_utils.ResourceNotAvailable"]], "tableresource (class in gnomad.resources.resource_utils)": [[15, "gnomad.resources.resource_utils.TableResource"]], "variantdatasetresource (class in gnomad.resources.resource_utils)": [[15, "gnomad.resources.resource_utils.VariantDatasetResource"]], "versionedblockmatrixresource (class in gnomad.resources.resource_utils)": [[15, "gnomad.resources.resource_utils.VersionedBlockMatrixResource"]], "versionedmatrixtableresource (class in gnomad.resources.resource_utils)": [[15, "gnomad.resources.resource_utils.VersionedMatrixTableResource"]], "versionedpedigreeresource (class in gnomad.resources.resource_utils)": [[15, "gnomad.resources.resource_utils.VersionedPedigreeResource"]], "versionedtableresource (class in gnomad.resources.resource_utils)": [[15, "gnomad.resources.resource_utils.VersionedTableResource"]], "versionedvariantdatasetresource (class in gnomad.resources.resource_utils)": [[15, "gnomad.resources.resource_utils.VersionedVariantDatasetResource"]], "bm() (gnomad.resources.resource_utils.blockmatrixresource method)": [[15, "gnomad.resources.resource_utils.BlockMatrixResource.bm"]], "bm() (gnomad.resources.resource_utils.gnomadpublicblockmatrixresource method)": [[15, "gnomad.resources.resource_utils.GnomadPublicBlockMatrixResource.bm"]], "default_version (gnomad.resources.resource_utils.baseversionedresource attribute)": [[15, "gnomad.resources.resource_utils.BaseVersionedResource.default_version"]], "expected_file_extensions (gnomad.resources.resource_utils.baseresource attribute)": [[15, "gnomad.resources.resource_utils.BaseResource.expected_file_extensions"]], "expected_file_extensions (gnomad.resources.resource_utils.blockmatrixresource attribute)": [[15, "gnomad.resources.resource_utils.BlockMatrixResource.expected_file_extensions"]], "expected_file_extensions (gnomad.resources.resource_utils.expressionresource attribute)": [[15, "gnomad.resources.resource_utils.ExpressionResource.expected_file_extensions"]], "expected_file_extensions (gnomad.resources.resource_utils.matrixtableresource attribute)": [[15, "gnomad.resources.resource_utils.MatrixTableResource.expected_file_extensions"]], "expected_file_extensions (gnomad.resources.resource_utils.pedigreeresource attribute)": [[15, "gnomad.resources.resource_utils.PedigreeResource.expected_file_extensions"]], "expected_file_extensions (gnomad.resources.resource_utils.tableresource attribute)": [[15, "gnomad.resources.resource_utils.TableResource.expected_file_extensions"]], "expected_file_extensions (gnomad.resources.resource_utils.variantdatasetresource attribute)": [[15, "gnomad.resources.resource_utils.VariantDatasetResource.expected_file_extensions"]], "gnomad.resources.resource_utils": [[15, "module-gnomad.resources.resource_utils"]], "he() (gnomad.resources.resource_utils.expressionresource method)": [[15, "gnomad.resources.resource_utils.ExpressionResource.he"]], "ht() (gnomad.resources.resource_utils.gnomadpublicpedigreeresource method)": [[15, "gnomad.resources.resource_utils.GnomadPublicPedigreeResource.ht"]], "ht() (gnomad.resources.resource_utils.gnomadpublictableresource method)": [[15, "gnomad.resources.resource_utils.GnomadPublicTableResource.ht"]], "ht() (gnomad.resources.resource_utils.pedigreeresource method)": [[15, "gnomad.resources.resource_utils.PedigreeResource.ht"]], "ht() (gnomad.resources.resource_utils.tableresource method)": [[15, "gnomad.resources.resource_utils.TableResource.ht"]], "import_gencode() (in module gnomad.resources.resource_utils)": [[15, "gnomad.resources.resource_utils.import_gencode"]], "import_resource() (gnomad.resources.resource_utils.baseresource method)": [[15, "gnomad.resources.resource_utils.BaseResource.import_resource"]], "import_resource() (gnomad.resources.resource_utils.blockmatrixresource method)": [[15, "gnomad.resources.resource_utils.BlockMatrixResource.import_resource"]], "import_resource() (gnomad.resources.resource_utils.expressionresource method)": [[15, "gnomad.resources.resource_utils.ExpressionResource.import_resource"]], "import_resource() (gnomad.resources.resource_utils.matrixtableresource method)": [[15, "gnomad.resources.resource_utils.MatrixTableResource.import_resource"]], "import_resource() (gnomad.resources.resource_utils.pedigreeresource method)": [[15, "gnomad.resources.resource_utils.PedigreeResource.import_resource"]], "import_resource() (gnomad.resources.resource_utils.tableresource method)": [[15, "gnomad.resources.resource_utils.TableResource.import_resource"]], "import_resource() (gnomad.resources.resource_utils.variantdatasetresource method)": [[15, "gnomad.resources.resource_utils.VariantDatasetResource.import_resource"]], "import_sites_vcf() (in module gnomad.resources.resource_utils)": [[15, "gnomad.resources.resource_utils.import_sites_vcf"]], "is_resource_available() (gnomad.resources.resource_utils.gnomadpublicresource method)": [[15, "gnomad.resources.resource_utils.GnomadPublicResource.is_resource_available"]], "mt() (gnomad.resources.resource_utils.gnomadpublicmatrixtableresource method)": [[15, "gnomad.resources.resource_utils.GnomadPublicMatrixTableResource.mt"]], "mt() (gnomad.resources.resource_utils.matrixtableresource method)": [[15, "gnomad.resources.resource_utils.MatrixTableResource.mt"]], "path (gnomad.resources.resource_utils.baseresource property)": [[15, "gnomad.resources.resource_utils.BaseResource.path"]], "pedigree() (gnomad.resources.resource_utils.gnomadpublicpedigreeresource method)": [[15, "gnomad.resources.resource_utils.GnomadPublicPedigreeResource.pedigree"]], "pedigree() (gnomad.resources.resource_utils.pedigreeresource method)": [[15, "gnomad.resources.resource_utils.PedigreeResource.pedigree"]], "resource_class (gnomad.resources.resource_utils.baseversionedresource attribute)": [[15, "gnomad.resources.resource_utils.BaseVersionedResource.resource_class"]], "resource_class (gnomad.resources.resource_utils.versionedblockmatrixresource attribute)": [[15, "gnomad.resources.resource_utils.VersionedBlockMatrixResource.resource_class"]], "resource_class (gnomad.resources.resource_utils.versionedmatrixtableresource attribute)": [[15, "gnomad.resources.resource_utils.VersionedMatrixTableResource.resource_class"]], "resource_class (gnomad.resources.resource_utils.versionedpedigreeresource attribute)": [[15, "gnomad.resources.resource_utils.VersionedPedigreeResource.resource_class"]], "resource_class (gnomad.resources.resource_utils.versionedtableresource attribute)": [[15, "gnomad.resources.resource_utils.VersionedTableResource.resource_class"]], "resource_class (gnomad.resources.resource_utils.versionedvariantdatasetresource attribute)": [[15, "gnomad.resources.resource_utils.VersionedVariantDatasetResource.resource_class"]], "vds() (gnomad.resources.resource_utils.variantdatasetresource method)": [[15, "gnomad.resources.resource_utils.VariantDatasetResource.vds"]], "versions (gnomad.resources.resource_utils.baseversionedresource attribute)": [[15, "gnomad.resources.resource_utils.BaseVersionedResource.versions"]], "apply_onnx_classification_model() (in module gnomad.sample_qc.ancestry)": [[16, "gnomad.sample_qc.ancestry.apply_onnx_classification_model"]], "apply_sklearn_classification_model() (in module gnomad.sample_qc.ancestry)": [[16, "gnomad.sample_qc.ancestry.apply_sklearn_classification_model"]], "assign_genetic_ancestry_pcs() (in module gnomad.sample_qc.ancestry)": [[16, "gnomad.sample_qc.ancestry.assign_genetic_ancestry_pcs"]], "convert_sklearn_rf_to_onnx() (in module gnomad.sample_qc.ancestry)": [[16, "gnomad.sample_qc.ancestry.convert_sklearn_rf_to_onnx"]], "gnomad.sample_qc.ancestry": [[16, "module-gnomad.sample_qc.ancestry"]], "pc_project() (in module gnomad.sample_qc.ancestry)": [[16, "gnomad.sample_qc.ancestry.pc_project"]], "run_pca_with_relateds() (in module gnomad.sample_qc.ancestry)": [[16, "gnomad.sample_qc.ancestry.run_pca_with_relateds"]], "compute_qc_metrics_residuals() (in module gnomad.sample_qc.filtering)": [[17, "gnomad.sample_qc.filtering.compute_qc_metrics_residuals"]], "compute_stratified_metrics_filter() (in module gnomad.sample_qc.filtering)": [[17, "gnomad.sample_qc.filtering.compute_stratified_metrics_filter"]], "compute_stratified_sample_qc() (in module gnomad.sample_qc.filtering)": [[17, "gnomad.sample_qc.filtering.compute_stratified_sample_qc"]], "determine_nearest_neighbors() (in module gnomad.sample_qc.filtering)": [[17, "gnomad.sample_qc.filtering.determine_nearest_neighbors"]], "gnomad.sample_qc.filtering": [[17, "module-gnomad.sample_qc.filtering"]], "merge_sample_qc_expr() (in module gnomad.sample_qc.filtering)": [[17, "gnomad.sample_qc.filtering.merge_sample_qc_expr"]], "annotate_sex() (in module gnomad.sample_qc.pipeline)": [[19, "gnomad.sample_qc.pipeline.annotate_sex"]], "filter_rows_for_qc() (in module gnomad.sample_qc.pipeline)": [[19, "gnomad.sample_qc.pipeline.filter_rows_for_qc"]], "get_qc_mt() (in module gnomad.sample_qc.pipeline)": [[19, "gnomad.sample_qc.pipeline.get_qc_mt"]], "gnomad.sample_qc.pipeline": [[19, "module-gnomad.sample_qc.pipeline"]], "infer_sex_karyotype() (in module gnomad.sample_qc.pipeline)": [[19, "gnomad.sample_qc.pipeline.infer_sex_karyotype"]], "assign_platform_from_pcs() (in module gnomad.sample_qc.platform)": [[20, "gnomad.sample_qc.platform.assign_platform_from_pcs"]], "compute_callrate_mt() (in module gnomad.sample_qc.platform)": [[20, "gnomad.sample_qc.platform.compute_callrate_mt"]], "gnomad.sample_qc.platform": [[20, "module-gnomad.sample_qc.platform"]], "run_platform_pca() (in module gnomad.sample_qc.platform)": [[20, "gnomad.sample_qc.platform.run_platform_pca"]], "ambiguous_relationship (in module gnomad.sample_qc.relatedness)": [[21, "gnomad.sample_qc.relatedness.AMBIGUOUS_RELATIONSHIP"]], "duplicate_or_twins (in module gnomad.sample_qc.relatedness)": [[21, "gnomad.sample_qc.relatedness.DUPLICATE_OR_TWINS"]], "parent_child (in module gnomad.sample_qc.relatedness)": [[21, "gnomad.sample_qc.relatedness.PARENT_CHILD"]], "second_degree_relatives (in module gnomad.sample_qc.relatedness)": [[21, "gnomad.sample_qc.relatedness.SECOND_DEGREE_RELATIVES"]], "siblings (in module gnomad.sample_qc.relatedness)": [[21, "gnomad.sample_qc.relatedness.SIBLINGS"]], "unrelated (in module gnomad.sample_qc.relatedness)": [[21, "gnomad.sample_qc.relatedness.UNRELATED"]], "calculate_de_novo_post_prob() (in module gnomad.sample_qc.relatedness)": [[21, "gnomad.sample_qc.relatedness.calculate_de_novo_post_prob"]], "compute_related_samples_to_drop() (in module gnomad.sample_qc.relatedness)": [[21, "gnomad.sample_qc.relatedness.compute_related_samples_to_drop"]], "create_fake_pedigree() (in module gnomad.sample_qc.relatedness)": [[21, "gnomad.sample_qc.relatedness.create_fake_pedigree"]], "default_get_de_novo_expr() (in module gnomad.sample_qc.relatedness)": [[21, "gnomad.sample_qc.relatedness.default_get_de_novo_expr"]], "explode_duplicate_samples_ht() (in module gnomad.sample_qc.relatedness)": [[21, "gnomad.sample_qc.relatedness.explode_duplicate_samples_ht"]], "filter_to_trios() (in module gnomad.sample_qc.relatedness)": [[21, "gnomad.sample_qc.relatedness.filter_to_trios"]], "generate_sib_stats_expr() (in module gnomad.sample_qc.relatedness)": [[21, "gnomad.sample_qc.relatedness.generate_sib_stats_expr"]], "generate_trio_stats_expr() (in module gnomad.sample_qc.relatedness)": [[21, "gnomad.sample_qc.relatedness.generate_trio_stats_expr"]], "get_duplicated_samples() (in module gnomad.sample_qc.relatedness)": [[21, "gnomad.sample_qc.relatedness.get_duplicated_samples"]], "get_duplicated_samples_ht() (in module gnomad.sample_qc.relatedness)": [[21, "gnomad.sample_qc.relatedness.get_duplicated_samples_ht"]], "get_relationship_expr() (in module gnomad.sample_qc.relatedness)": [[21, "gnomad.sample_qc.relatedness.get_relationship_expr"]], "get_slope_int_relationship_expr() (in module gnomad.sample_qc.relatedness)": [[21, "gnomad.sample_qc.relatedness.get_slope_int_relationship_expr"]], "gnomad.sample_qc.relatedness": [[21, "module-gnomad.sample_qc.relatedness"]], "infer_families() (in module gnomad.sample_qc.relatedness)": [[21, "gnomad.sample_qc.relatedness.infer_families"]], "adjust_sex_ploidy() (in module gnomad.sample_qc.sex)": [[22, "gnomad.sample_qc.sex.adjust_sex_ploidy"]], "adjusted_sex_ploidy_expr() (in module gnomad.sample_qc.sex)": [[22, "gnomad.sample_qc.sex.adjusted_sex_ploidy_expr"]], "gaussian_mixture_model_karyotype_assignment() (in module gnomad.sample_qc.sex)": [[22, "gnomad.sample_qc.sex.gaussian_mixture_model_karyotype_assignment"]], "get_chr_x_hom_alt_cutoffs() (in module gnomad.sample_qc.sex)": [[22, "gnomad.sample_qc.sex.get_chr_x_hom_alt_cutoffs"]], "get_ploidy_cutoffs() (in module gnomad.sample_qc.sex)": [[22, "gnomad.sample_qc.sex.get_ploidy_cutoffs"]], "get_sex_expr() (in module gnomad.sample_qc.sex)": [[22, "gnomad.sample_qc.sex.get_sex_expr"]], "gnomad.sample_qc.sex": [[22, "module-gnomad.sample_qc.sex"]], "add_gks_va() (in module gnomad.utils.annotations)": [[23, "gnomad.utils.annotations.add_gks_va"]], "add_gks_vrs() (in module gnomad.utils.annotations)": [[23, "gnomad.utils.annotations.add_gks_vrs"]], "add_variant_type() (in module gnomad.utils.annotations)": [[23, "gnomad.utils.annotations.add_variant_type"]], "age_hists_expr() (in module gnomad.utils.annotations)": [[23, "gnomad.utils.annotations.age_hists_expr"]], "agg_by_strata() (in module gnomad.utils.annotations)": [[23, "gnomad.utils.annotations.agg_by_strata"]], "annotate_adj() (in module gnomad.utils.annotations)": [[23, "gnomad.utils.annotations.annotate_adj"]], "annotate_allele_info() (in module gnomad.utils.annotations)": [[23, "gnomad.utils.annotations.annotate_allele_info"]], "annotate_and_index_source_mt_for_sex_ploidy() (in module gnomad.utils.annotations)": [[23, "gnomad.utils.annotations.annotate_and_index_source_mt_for_sex_ploidy"]], "annotate_downsamplings() (in module gnomad.utils.annotations)": [[23, "gnomad.utils.annotations.annotate_downsamplings"]], "annotate_freq() (in module gnomad.utils.annotations)": [[23, "gnomad.utils.annotations.annotate_freq"]], "annotate_with_ht() (in module gnomad.utils.annotations)": [[23, "gnomad.utils.annotations.annotate_with_ht"]], "annotation_type_in_vcf_info() (in module gnomad.utils.annotations)": [[23, "gnomad.utils.annotations.annotation_type_in_vcf_info"]], "annotation_type_is_numeric() (in module gnomad.utils.annotations)": [[23, "gnomad.utils.annotations.annotation_type_is_numeric"]], "bi_allelic_expr() (in module gnomad.utils.annotations)": [[23, "gnomad.utils.annotations.bi_allelic_expr"]], "bi_allelic_site_inbreeding_expr() (in module gnomad.utils.annotations)": [[23, "gnomad.utils.annotations.bi_allelic_site_inbreeding_expr"]], "build_freq_stratification_list() (in module gnomad.utils.annotations)": [[23, "gnomad.utils.annotations.build_freq_stratification_list"]], "check_annotation_missingness() (in module gnomad.utils.annotations)": [[23, "gnomad.utils.annotations.check_annotation_missingness"]], "compute_freq_by_strata() (in module gnomad.utils.annotations)": [[23, "gnomad.utils.annotations.compute_freq_by_strata"]], "create_frequency_bins_expr() (in module gnomad.utils.annotations)": [[23, "gnomad.utils.annotations.create_frequency_bins_expr"]], "faf_expr() (in module gnomad.utils.annotations)": [[23, "gnomad.utils.annotations.faf_expr"]], "fill_missing_key_combinations() (in module gnomad.utils.annotations)": [[23, "gnomad.utils.annotations.fill_missing_key_combinations"]], "fs_from_sb() (in module gnomad.utils.annotations)": [[23, "gnomad.utils.annotations.fs_from_sb"]], "gen_anc_faf_max_expr() (in module gnomad.utils.annotations)": [[23, "gnomad.utils.annotations.gen_anc_faf_max_expr"]], "generate_freq_group_membership_array() (in module gnomad.utils.annotations)": [[23, "gnomad.utils.annotations.generate_freq_group_membership_array"]], "get_adj_expr() (in module gnomad.utils.annotations)": [[23, "gnomad.utils.annotations.get_adj_expr"]], "get_annotations_hists() (in module gnomad.utils.annotations)": [[23, "gnomad.utils.annotations.get_annotations_hists"]], "get_copy_state_by_sex() (in module gnomad.utils.annotations)": [[23, "gnomad.utils.annotations.get_copy_state_by_sex"]], "get_gq_dp_adj_expr() (in module gnomad.utils.annotations)": [[23, "gnomad.utils.annotations.get_gq_dp_adj_expr"]], "get_het_ab_adj_expr() (in module gnomad.utils.annotations)": [[23, "gnomad.utils.annotations.get_het_ab_adj_expr"]], "get_is_haploid_expr() (in module gnomad.utils.annotations)": [[23, "gnomad.utils.annotations.get_is_haploid_expr"]], "get_lowqual_expr() (in module gnomad.utils.annotations)": [[23, "gnomad.utils.annotations.get_lowqual_expr"]], "gnomad.utils.annotations": [[23, "module-gnomad.utils.annotations"]], "grpmax_expr() (in module gnomad.utils.annotations)": [[23, "gnomad.utils.annotations.grpmax_expr"]], "hemi_expr() (in module gnomad.utils.annotations)": [[23, "gnomad.utils.annotations.hemi_expr"]], "merge_array_expressions() (in module gnomad.utils.annotations)": [[23, "gnomad.utils.annotations.merge_array_expressions"]], "merge_freq_arrays() (in module gnomad.utils.annotations)": [[23, "gnomad.utils.annotations.merge_freq_arrays"]], "merge_histograms() (in module gnomad.utils.annotations)": [[23, "gnomad.utils.annotations.merge_histograms"]], "missing_callstats_expr() (in module gnomad.utils.annotations)": [[23, "gnomad.utils.annotations.missing_callstats_expr"]], "missing_struct_expr() (in module gnomad.utils.annotations)": [[23, "gnomad.utils.annotations.missing_struct_expr"]], "pab_max_expr() (in module gnomad.utils.annotations)": [[23, "gnomad.utils.annotations.pab_max_expr"]], "project_max_expr() (in module gnomad.utils.annotations)": [[23, "gnomad.utils.annotations.project_max_expr"]], "qual_hist_expr() (in module gnomad.utils.annotations)": [[23, "gnomad.utils.annotations.qual_hist_expr"]], "region_flag_expr() (in module gnomad.utils.annotations)": [[23, "gnomad.utils.annotations.region_flag_expr"]], "set_xx_y_metrics_to_na_expr() (in module gnomad.utils.annotations)": [[23, "gnomad.utils.annotations.set_xx_y_metrics_to_na_expr"]], "sor_from_sb() (in module gnomad.utils.annotations)": [[23, "gnomad.utils.annotations.sor_from_sb"]], "unphase_call_expr() (in module gnomad.utils.annotations)": [[23, "gnomad.utils.annotations.unphase_call_expr"]], "update_structured_annotations() (in module gnomad.utils.annotations)": [[23, "gnomad.utils.annotations.update_structured_annotations"]], "coverage_cutoff (in module gnomad.utils.constraint)": [[24, "gnomad.utils.constraint.COVERAGE_CUTOFF"]], "add_gencode_transcript_annotations() (in module gnomad.utils.constraint)": [[24, "gnomad.utils.constraint.add_gencode_transcript_annotations"]], "annotate_exploded_vep_for_constraint_groupings() (in module gnomad.utils.constraint)": [[24, "gnomad.utils.constraint.annotate_exploded_vep_for_constraint_groupings"]], "annotate_mutation_type() (in module gnomad.utils.constraint)": [[24, "gnomad.utils.constraint.annotate_mutation_type"]], "annotate_with_mu() (in module gnomad.utils.constraint)": [[24, "gnomad.utils.constraint.annotate_with_mu"]], "assemble_constraint_context_ht() (in module gnomad.utils.constraint)": [[24, "gnomad.utils.constraint.assemble_constraint_context_ht"]], "build_coverage_model() (in module gnomad.utils.constraint)": [[24, "gnomad.utils.constraint.build_coverage_model"]], "build_models() (in module gnomad.utils.constraint)": [[24, "gnomad.utils.constraint.build_models"]], "build_plateau_models() (in module gnomad.utils.constraint)": [[24, "gnomad.utils.constraint.build_plateau_models"]], "calculate_raw_z_score() (in module gnomad.utils.constraint)": [[24, "gnomad.utils.constraint.calculate_raw_z_score"]], "calculate_raw_z_score_sd() (in module gnomad.utils.constraint)": [[24, "gnomad.utils.constraint.calculate_raw_z_score_sd"]], "collapse_strand() (in module gnomad.utils.constraint)": [[24, "gnomad.utils.constraint.collapse_strand"]], "compute_expected_variants() (in module gnomad.utils.constraint)": [[24, "gnomad.utils.constraint.compute_expected_variants"]], "compute_pli() (in module gnomad.utils.constraint)": [[24, "gnomad.utils.constraint.compute_pli"]], "count_variants_by_group() (in module gnomad.utils.constraint)": [[24, "gnomad.utils.constraint.count_variants_by_group"]], "downsampling_counts_expr() (in module gnomad.utils.constraint)": [[24, "gnomad.utils.constraint.downsampling_counts_expr"]], "explode_downsamplings_oe() (in module gnomad.utils.constraint)": [[24, "gnomad.utils.constraint.explode_downsamplings_oe"]], "get_all_gen_anc_lengths() (in module gnomad.utils.constraint)": [[24, "gnomad.utils.constraint.get_all_gen_anc_lengths"]], "get_constraint_flags() (in module gnomad.utils.constraint)": [[24, "gnomad.utils.constraint.get_constraint_flags"]], "get_constraint_grouping_expr() (in module gnomad.utils.constraint)": [[24, "gnomad.utils.constraint.get_constraint_grouping_expr"]], "get_downsampling_freq_indices() (in module gnomad.utils.constraint)": [[24, "gnomad.utils.constraint.get_downsampling_freq_indices"]], "get_mu_annotation_expr() (in module gnomad.utils.constraint)": [[24, "gnomad.utils.constraint.get_mu_annotation_expr"]], "gnomad.utils.constraint": [[24, "module-gnomad.utils.constraint"]], "oe_aggregation_expr() (in module gnomad.utils.constraint)": [[24, "gnomad.utils.constraint.oe_aggregation_expr"]], "oe_confidence_interval() (in module gnomad.utils.constraint)": [[24, "gnomad.utils.constraint.oe_confidence_interval"]], "transform_methylation_level() (in module gnomad.utils.constraint)": [[24, "gnomad.utils.constraint.transform_methylation_level"]], "trimer_from_heptamer() (in module gnomad.utils.constraint)": [[24, "gnomad.utils.constraint.trimer_from_heptamer"]], "check_file_exists_raise_error() (in module gnomad.utils.file_utils)": [[25, "gnomad.utils.file_utils.check_file_exists_raise_error"]], "create_vds() (in module gnomad.utils.file_utils)": [[25, "gnomad.utils.file_utils.create_vds"]], "file_exists() (in module gnomad.utils.file_utils)": [[25, "gnomad.utils.file_utils.file_exists"]], "get_file_stats() (in module gnomad.utils.file_utils)": [[25, "gnomad.utils.file_utils.get_file_stats"]], "gnomad.utils.file_utils": [[25, "module-gnomad.utils.file_utils"]], "read_list_data() (in module gnomad.utils.file_utils)": [[25, "gnomad.utils.file_utils.read_list_data"]], "repartition_for_join() (in module gnomad.utils.file_utils)": [[25, "gnomad.utils.file_utils.repartition_for_join"]], "select_primitives_from_ht() (in module gnomad.utils.file_utils)": [[25, "gnomad.utils.file_utils.select_primitives_from_ht"]], "write_temp_gcs() (in module gnomad.utils.file_utils)": [[25, "gnomad.utils.file_utils.write_temp_gcs"]], "add_filters_expr() (in module gnomad.utils.filtering)": [[26, "gnomad.utils.filtering.add_filters_expr"]], "combine_functions() (in module gnomad.utils.filtering)": [[26, "gnomad.utils.filtering.combine_functions"]], "filter_arrays_by_meta() (in module gnomad.utils.filtering)": [[26, "gnomad.utils.filtering.filter_arrays_by_meta"]], "filter_by_frequency() (in module gnomad.utils.filtering)": [[26, "gnomad.utils.filtering.filter_by_frequency"]], "filter_by_gencode_intervals() (in module gnomad.utils.filtering)": [[26, "gnomad.utils.filtering.filter_by_gencode_intervals"]], "filter_by_intervals() (in module gnomad.utils.filtering)": [[26, "gnomad.utils.filtering.filter_by_intervals"]], "filter_by_numeric_expr_range() (in module gnomad.utils.filtering)": [[26, "gnomad.utils.filtering.filter_by_numeric_expr_range"]], "filter_for_mu() (in module gnomad.utils.filtering)": [[26, "gnomad.utils.filtering.filter_for_mu"]], "filter_gencode_ht() (in module gnomad.utils.filtering)": [[26, "gnomad.utils.filtering.filter_gencode_ht"]], "filter_low_conf_regions() (in module gnomad.utils.filtering)": [[26, "gnomad.utils.filtering.filter_low_conf_regions"]], "filter_to_adj() (in module gnomad.utils.filtering)": [[26, "gnomad.utils.filtering.filter_to_adj"]], "filter_to_autosomes() (in module gnomad.utils.filtering)": [[26, "gnomad.utils.filtering.filter_to_autosomes"]], "filter_to_clinvar_pathogenic() (in module gnomad.utils.filtering)": [[26, "gnomad.utils.filtering.filter_to_clinvar_pathogenic"]], "filter_to_gencode_cds() (in module gnomad.utils.filtering)": [[26, "gnomad.utils.filtering.filter_to_gencode_cds"]], "filter_x_nonpar() (in module gnomad.utils.filtering)": [[26, "gnomad.utils.filtering.filter_x_nonpar"]], "filter_y_nonpar() (in module gnomad.utils.filtering)": [[26, "gnomad.utils.filtering.filter_y_nonpar"]], "gnomad.utils.filtering": [[26, "module-gnomad.utils.filtering"]], "low_conf_regions_expr() (in module gnomad.utils.filtering)": [[26, "gnomad.utils.filtering.low_conf_regions_expr"]], "remove_fields_from_constant() (in module gnomad.utils.filtering)": [[26, "gnomad.utils.filtering.remove_fields_from_constant"]], "split_vds_by_strata() (in module gnomad.utils.filtering)": [[26, "gnomad.utils.filtering.split_vds_by_strata"]], "subset_samples_and_variants() (in module gnomad.utils.filtering)": [[26, "gnomad.utils.filtering.subset_samples_and_variants"]], "from_phred() (in module gnomad.utils.gen_stats)": [[27, "gnomad.utils.gen_stats.from_phred"]], "get_median_and_mad_expr() (in module gnomad.utils.gen_stats)": [[27, "gnomad.utils.gen_stats.get_median_and_mad_expr"]], "gnomad.utils.gen_stats": [[27, "module-gnomad.utils.gen_stats"]], "merge_stats_counters_expr() (in module gnomad.utils.gen_stats)": [[27, "gnomad.utils.gen_stats.merge_stats_counters_expr"]], "to_phred() (in module gnomad.utils.gen_stats)": [[27, "gnomad.utils.gen_stats.to_phred"]], "gnomad.utils.intervals": [[29, "module-gnomad.utils.intervals"]], "interval_length() (in module gnomad.utils.intervals)": [[29, "gnomad.utils.intervals.interval_length"]], "pad_intervals() (in module gnomad.utils.intervals)": [[29, "gnomad.utils.intervals.pad_intervals"]], "sort_intervals() (in module gnomad.utils.intervals)": [[29, "gnomad.utils.intervals.sort_intervals"]], "union_intervals() (in module gnomad.utils.intervals)": [[29, "gnomad.utils.intervals.union_intervals"]], "grch37_to_grch38_chain (in module gnomad.utils.liftover)": [[30, "gnomad.utils.liftover.GRCH37_to_GRCH38_CHAIN"]], "grch38_to_grch37_chain (in module gnomad.utils.liftover)": [[30, "gnomad.utils.liftover.GRCH38_TO_GRCH37_CHAIN"]], "default_lift_data() (in module gnomad.utils.liftover)": [[30, "gnomad.utils.liftover.default_lift_data"]], "get_liftover_genome() (in module gnomad.utils.liftover)": [[30, "gnomad.utils.liftover.get_liftover_genome"]], "gnomad.utils.liftover": [[30, "module-gnomad.utils.liftover"]], "liftover_expr() (in module gnomad.utils.liftover)": [[30, "gnomad.utils.liftover.liftover_expr"]], "liftover_using_gnomad_map() (in module gnomad.utils.liftover)": [[30, "gnomad.utils.liftover.liftover_using_gnomad_map"]], "gnomad.utils.parse": [[31, "module-gnomad.utils.parse"]], "parse_locus_intervals() (in module gnomad.utils.parse)": [[31, "gnomad.utils.parse.parse_locus_intervals"]], "parse_variant() (in module gnomad.utils.parse)": [[31, "gnomad.utils.parse.parse_variant"]], "get_rows_data() (in module gnomad.utils.plotting)": [[32, "gnomad.utils.plotting.get_rows_data"]], "gnomad.utils.plotting": [[32, "module-gnomad.utils.plotting"]], "linear_and_log_tabs() (in module gnomad.utils.plotting)": [[32, "gnomad.utils.plotting.linear_and_log_tabs"]], "new_show() (in module gnomad.utils.plotting)": [[32, "gnomad.utils.plotting.new_show"]], "pair_plot() (in module gnomad.utils.plotting)": [[32, "gnomad.utils.plotting.pair_plot"]], "plot_hail_file_metadata() (in module gnomad.utils.plotting)": [[32, "gnomad.utils.plotting.plot_hail_file_metadata"]], "plot_hail_hist() (in module gnomad.utils.plotting)": [[32, "gnomad.utils.plotting.plot_hail_hist"]], "plot_hail_hist_both() (in module gnomad.utils.plotting)": [[32, "gnomad.utils.plotting.plot_hail_hist_both"]], "plot_hail_hist_cumulative() (in module gnomad.utils.plotting)": [[32, "gnomad.utils.plotting.plot_hail_hist_cumulative"]], "plot_multi_hail_hist() (in module gnomad.utils.plotting)": [[32, "gnomad.utils.plotting.plot_multi_hail_hist"]], "scale_file_sizes() (in module gnomad.utils.plotting)": [[32, "gnomad.utils.plotting.scale_file_sizes"]], "set_font_size() (in module gnomad.utils.plotting)": [[32, "gnomad.utils.plotting.set_font_size"]], "add_reference_sequence() (in module gnomad.utils.reference_genome)": [[33, "gnomad.utils.reference_genome.add_reference_sequence"]], "get_reference_genome() (in module gnomad.utils.reference_genome)": [[33, "gnomad.utils.reference_genome.get_reference_genome"]], "get_reference_ht() (in module gnomad.utils.reference_genome)": [[33, "gnomad.utils.reference_genome.get_reference_ht"]], "gnomad.utils.reference_genome": [[33, "module-gnomad.utils.reference_genome"]], "gnomad.utils.release": [[34, "module-gnomad.utils.release"]], "make_faf_index_dict() (in module gnomad.utils.release)": [[34, "gnomad.utils.release.make_faf_index_dict"]], "make_freq_index_dict() (in module gnomad.utils.release)": [[34, "gnomad.utils.release.make_freq_index_dict"]], "make_freq_index_dict_from_meta() (in module gnomad.utils.release)": [[34, "gnomad.utils.release.make_freq_index_dict_from_meta"]], "slackclient (class in gnomad.utils.slack)": [[35, "gnomad.utils.slack.SlackClient"]], "gnomad.utils.slack": [[35, "module-gnomad.utils.slack"]], "send_file() (gnomad.utils.slack.slackclient method)": [[35, "gnomad.utils.slack.SlackClient.send_file"]], "send_message() (gnomad.utils.slack.slackclient method)": [[35, "gnomad.utils.slack.SlackClient.send_message"]], "slack_notifications() (in module gnomad.utils.slack)": [[35, "gnomad.utils.slack.slack_notifications"]], "compute_allele_number_per_ref_site() (in module gnomad.utils.sparse_mt)": [[36, "gnomad.utils.sparse_mt.compute_allele_number_per_ref_site"]], "compute_coverage_stats() (in module gnomad.utils.sparse_mt)": [[36, "gnomad.utils.sparse_mt.compute_coverage_stats"]], "compute_last_ref_block_end() (in module gnomad.utils.sparse_mt)": [[36, "gnomad.utils.sparse_mt.compute_last_ref_block_end"]], "compute_stats_per_ref_site() (in module gnomad.utils.sparse_mt)": [[36, "gnomad.utils.sparse_mt.compute_stats_per_ref_site"]], "default_compute_info() (in module gnomad.utils.sparse_mt)": [[36, "gnomad.utils.sparse_mt.default_compute_info"]], "densify_all_reference_sites() (in module gnomad.utils.sparse_mt)": [[36, "gnomad.utils.sparse_mt.densify_all_reference_sites"]], "densify_sites() (in module gnomad.utils.sparse_mt)": [[36, "gnomad.utils.sparse_mt.densify_sites"]], "filter_ref_blocks() (in module gnomad.utils.sparse_mt)": [[36, "gnomad.utils.sparse_mt.filter_ref_blocks"]], "get_allele_number_agg_func() (in module gnomad.utils.sparse_mt)": [[36, "gnomad.utils.sparse_mt.get_allele_number_agg_func"]], "get_as_info_expr() (in module gnomad.utils.sparse_mt)": [[36, "gnomad.utils.sparse_mt.get_as_info_expr"]], "get_coverage_agg_func() (in module gnomad.utils.sparse_mt)": [[36, "gnomad.utils.sparse_mt.get_coverage_agg_func"]], "get_site_info_expr() (in module gnomad.utils.sparse_mt)": [[36, "gnomad.utils.sparse_mt.get_site_info_expr"]], "gnomad.utils.sparse_mt": [[36, "module-gnomad.utils.sparse_mt"]], "impute_sex_ploidy() (in module gnomad.utils.sparse_mt)": [[36, "gnomad.utils.sparse_mt.impute_sex_ploidy"]], "split_info_annotation() (in module gnomad.utils.sparse_mt)": [[36, "gnomad.utils.sparse_mt.split_info_annotation"]], "split_lowqual_annotation() (in module gnomad.utils.sparse_mt)": [[36, "gnomad.utils.sparse_mt.split_lowqual_annotation"]], "cell_lines (in module gnomad.utils.transcript_annotation)": [[37, "gnomad.utils.transcript_annotation.CELL_LINES"]], "reproductive_tissues (in module gnomad.utils.transcript_annotation)": [[37, "gnomad.utils.transcript_annotation.REPRODUCTIVE_TISSUES"]], "clean_tissue_name_for_browser() (in module gnomad.utils.transcript_annotation)": [[37, "gnomad.utils.transcript_annotation.clean_tissue_name_for_browser"]], "create_tx_annotation_by_region() (in module gnomad.utils.transcript_annotation)": [[37, "gnomad.utils.transcript_annotation.create_tx_annotation_by_region"]], "filter_expression_ht_by_tissues() (in module gnomad.utils.transcript_annotation)": [[37, "gnomad.utils.transcript_annotation.filter_expression_ht_by_tissues"]], "get_expression_proportion() (in module gnomad.utils.transcript_annotation)": [[37, "gnomad.utils.transcript_annotation.get_expression_proportion"]], "get_tissues_to_exclude() (in module gnomad.utils.transcript_annotation)": [[37, "gnomad.utils.transcript_annotation.get_tissues_to_exclude"]], "gnomad.utils.transcript_annotation": [[37, "module-gnomad.utils.transcript_annotation"]], "perform_tx_annotation_pipeline() (in module gnomad.utils.transcript_annotation)": [[37, "gnomad.utils.transcript_annotation.perform_tx_annotation_pipeline"]], "summarize_transcript_expression() (in module gnomad.utils.transcript_annotation)": [[37, "gnomad.utils.transcript_annotation.summarize_transcript_expression"]], "tissue_expression_ht_to_array() (in module gnomad.utils.transcript_annotation)": [[37, "gnomad.utils.transcript_annotation.tissue_expression_ht_to_array"]], "tx_aggregate_variants() (in module gnomad.utils.transcript_annotation)": [[37, "gnomad.utils.transcript_annotation.tx_aggregate_variants"]], "tx_annotate_variants() (in module gnomad.utils.transcript_annotation)": [[37, "gnomad.utils.transcript_annotation.tx_annotate_variants"]], "tx_filter_variants_by_csqs() (in module gnomad.utils.transcript_annotation)": [[37, "gnomad.utils.transcript_annotation.tx_filter_variants_by_csqs"]], "allele_type_fields (in module gnomad.utils.vcf)": [[38, "gnomad.utils.vcf.ALLELE_TYPE_FIELDS"]], "as_fields (in module gnomad.utils.vcf)": [[38, "gnomad.utils.vcf.AS_FIELDS"]], "as_vqsr_fields (in module gnomad.utils.vcf)": [[38, "gnomad.utils.vcf.AS_VQSR_FIELDS"]], "entries (in module gnomad.utils.vcf)": [[38, "gnomad.utils.vcf.ENTRIES"]], "faf_gen_anc_groups (in module gnomad.utils.vcf)": [[38, "gnomad.utils.vcf.FAF_GEN_ANC_GROUPS"]], "format_dict (in module gnomad.utils.vcf)": [[38, "gnomad.utils.vcf.FORMAT_DICT"]], "groups (in module gnomad.utils.vcf)": [[38, "gnomad.utils.vcf.GROUPS"]], "hists (in module gnomad.utils.vcf)": [[38, "gnomad.utils.vcf.HISTS"]], "info_dict (in module gnomad.utils.vcf)": [[38, "gnomad.utils.vcf.INFO_DICT"]], "in_silico_annotations_info_dict (in module gnomad.utils.vcf)": [[38, "gnomad.utils.vcf.IN_SILICO_ANNOTATIONS_INFO_DICT"]], "joint_region_flag_fields (in module gnomad.utils.vcf)": [[38, "gnomad.utils.vcf.JOINT_REGION_FLAG_FIELDS"]], "region_flag_fields (in module gnomad.utils.vcf)": [[38, "gnomad.utils.vcf.REGION_FLAG_FIELDS"]], "rf_fields (in module gnomad.utils.vcf)": [[38, "gnomad.utils.vcf.RF_FIELDS"]], "sexes (in module gnomad.utils.vcf)": [[38, "gnomad.utils.vcf.SEXES"]], "site_fields (in module gnomad.utils.vcf)": [[38, "gnomad.utils.vcf.SITE_FIELDS"]], "sort_order (in module gnomad.utils.vcf)": [[38, "gnomad.utils.vcf.SORT_ORDER"]], "sparse_entries (in module gnomad.utils.vcf)": [[38, "gnomad.utils.vcf.SPARSE_ENTRIES"]], "vqsr_fields (in module gnomad.utils.vcf)": [[38, "gnomad.utils.vcf.VQSR_FIELDS"]], "vrs_fields_dict (in module gnomad.utils.vcf)": [[38, "gnomad.utils.vcf.VRS_FIELDS_DICT"]], "add_as_info_dict() (in module gnomad.utils.vcf)": [[38, "gnomad.utils.vcf.add_as_info_dict"]], "adjust_vcf_incompatible_types() (in module gnomad.utils.vcf)": [[38, "gnomad.utils.vcf.adjust_vcf_incompatible_types"]], "build_vcf_export_reference() (in module gnomad.utils.vcf)": [[38, "gnomad.utils.vcf.build_vcf_export_reference"]], "create_label_groups() (in module gnomad.utils.vcf)": [[38, "gnomad.utils.vcf.create_label_groups"]], "gnomad.utils.vcf": [[38, "module-gnomad.utils.vcf"]], "index_globals() (in module gnomad.utils.vcf)": [[38, "gnomad.utils.vcf.index_globals"]], "make_combo_header_text() (in module gnomad.utils.vcf)": [[38, "gnomad.utils.vcf.make_combo_header_text"]], "make_hist_bin_edges_expr() (in module gnomad.utils.vcf)": [[38, "gnomad.utils.vcf.make_hist_bin_edges_expr"]], "make_hist_dict() (in module gnomad.utils.vcf)": [[38, "gnomad.utils.vcf.make_hist_dict"]], "make_info_dict() (in module gnomad.utils.vcf)": [[38, "gnomad.utils.vcf.make_info_dict"]], "make_label_combos() (in module gnomad.utils.vcf)": [[38, "gnomad.utils.vcf.make_label_combos"]], "make_vcf_filter_dict() (in module gnomad.utils.vcf)": [[38, "gnomad.utils.vcf.make_vcf_filter_dict"]], "rekey_new_reference() (in module gnomad.utils.vcf)": [[38, "gnomad.utils.vcf.rekey_new_reference"]], "set_xx_y_metrics_to_na() (in module gnomad.utils.vcf)": [[38, "gnomad.utils.vcf.set_xx_y_metrics_to_na"]], "csq_coding (in module gnomad.utils.vep)": [[39, "gnomad.utils.vep.CSQ_CODING"]], "csq_splice (in module gnomad.utils.vep)": [[39, "gnomad.utils.vep.CSQ_SPLICE"]], "current_vep_version (in module gnomad.utils.vep)": [[39, "gnomad.utils.vep.CURRENT_VEP_VERSION"]], "loftee_labels (in module gnomad.utils.vep)": [[39, "gnomad.utils.vep.LOFTEE_LABELS"]], "lof_csq_set (in module gnomad.utils.vep)": [[39, "gnomad.utils.vep.LOF_CSQ_SET"]], "possible_refs (in module gnomad.utils.vep)": [[39, "gnomad.utils.vep.POSSIBLE_REFS"]], "vep_config_path (in module gnomad.utils.vep)": [[39, "gnomad.utils.vep.VEP_CONFIG_PATH"]], "vep_csq_fields (in module gnomad.utils.vep)": [[39, "gnomad.utils.vep.VEP_CSQ_FIELDS"]], "vep_csq_header (in module gnomad.utils.vep)": [[39, "gnomad.utils.vep.VEP_CSQ_HEADER"]], "add_most_severe_consequence_to_consequence() (in module gnomad.utils.vep)": [[39, "gnomad.utils.vep.add_most_severe_consequence_to_consequence"]], "add_most_severe_csq_to_tc_within_vep_root() (in module gnomad.utils.vep)": [[39, "gnomad.utils.vep.add_most_severe_csq_to_tc_within_vep_root"]], "explode_by_vep_annotation() (in module gnomad.utils.vep)": [[39, "gnomad.utils.vep.explode_by_vep_annotation"]], "filter_vep_to_canonical_transcripts() (in module gnomad.utils.vep)": [[39, "gnomad.utils.vep.filter_vep_to_canonical_transcripts"]], "filter_vep_to_gene_list() (in module gnomad.utils.vep)": [[39, "gnomad.utils.vep.filter_vep_to_gene_list"]], "filter_vep_to_mane_select_transcripts() (in module gnomad.utils.vep)": [[39, "gnomad.utils.vep.filter_vep_to_mane_select_transcripts"]], "filter_vep_to_synonymous_variants() (in module gnomad.utils.vep)": [[39, "gnomad.utils.vep.filter_vep_to_synonymous_variants"]], "filter_vep_transcript_csqs() (in module gnomad.utils.vep)": [[39, "gnomad.utils.vep.filter_vep_transcript_csqs"]], "filter_vep_transcript_csqs_expr() (in module gnomad.utils.vep)": [[39, "gnomad.utils.vep.filter_vep_transcript_csqs_expr"]], "get_loftee_end_trunc_filter_expr() (in module gnomad.utils.vep)": [[39, "gnomad.utils.vep.get_loftee_end_trunc_filter_expr"]], "get_most_severe_consequence_expr() (in module gnomad.utils.vep)": [[39, "gnomad.utils.vep.get_most_severe_consequence_expr"]], "get_most_severe_consequence_for_summary() (in module gnomad.utils.vep)": [[39, "gnomad.utils.vep.get_most_severe_consequence_for_summary"]], "get_vep_context() (in module gnomad.utils.vep)": [[39, "gnomad.utils.vep.get_vep_context"]], "get_vep_csq_header() (in module gnomad.utils.vep)": [[39, "gnomad.utils.vep.get_vep_csq_header"]], "get_vep_help() (in module gnomad.utils.vep)": [[39, "gnomad.utils.vep.get_vep_help"]], "gnomad.utils.vep": [[39, "module-gnomad.utils.vep"]], "process_consequences() (in module gnomad.utils.vep)": [[39, "gnomad.utils.vep.process_consequences"]], "update_loftee_end_trunc_filter() (in module gnomad.utils.vep)": [[39, "gnomad.utils.vep.update_loftee_end_trunc_filter"]], "vep_or_lookup_vep() (in module gnomad.utils.vep)": [[39, "gnomad.utils.vep.vep_or_lookup_vep"]], "vep_struct_to_csq() (in module gnomad.utils.vep)": [[39, "gnomad.utils.vep.vep_struct_to_csq"]], "add_rank() (in module gnomad.variant_qc.evaluation)": [[40, "gnomad.variant_qc.evaluation.add_rank"]], "compute_binned_truth_sample_concordance() (in module gnomad.variant_qc.evaluation)": [[40, "gnomad.variant_qc.evaluation.compute_binned_truth_sample_concordance"]], "compute_grouped_binned_ht() (in module gnomad.variant_qc.evaluation)": [[40, "gnomad.variant_qc.evaluation.compute_grouped_binned_ht"]], "compute_ranked_bin() (in module gnomad.variant_qc.evaluation)": [[40, "gnomad.variant_qc.evaluation.compute_ranked_bin"]], "create_truth_sample_ht() (in module gnomad.variant_qc.evaluation)": [[40, "gnomad.variant_qc.evaluation.create_truth_sample_ht"]], "gnomad.variant_qc.evaluation": [[40, "module-gnomad.variant_qc.evaluation"]], "get_r_for_pair_of_variants() (in module gnomad.variant_qc.ld)": [[42, "gnomad.variant_qc.ld.get_r_for_pair_of_variants"]], "get_r_human_readable() (in module gnomad.variant_qc.ld)": [[42, "gnomad.variant_qc.ld.get_r_human_readable"]], "get_r_within_gene() (in module gnomad.variant_qc.ld)": [[42, "gnomad.variant_qc.ld.get_r_within_gene"]], "get_r_within_gene_in_gen_anc() (in module gnomad.variant_qc.ld)": [[42, "gnomad.variant_qc.ld.get_r_within_gene_in_gen_anc"]], "gnomad.variant_qc.ld": [[42, "module-gnomad.variant_qc.ld"]], "create_binned_ht() (in module gnomad.variant_qc.pipeline)": [[43, "gnomad.variant_qc.pipeline.create_binned_ht"]], "generate_sib_stats() (in module gnomad.variant_qc.pipeline)": [[43, "gnomad.variant_qc.pipeline.generate_sib_stats"]], "generate_trio_stats() (in module gnomad.variant_qc.pipeline)": [[43, "gnomad.variant_qc.pipeline.generate_trio_stats"]], "gnomad.variant_qc.pipeline": [[43, "module-gnomad.variant_qc.pipeline"]], "score_bin_agg() (in module gnomad.variant_qc.pipeline)": [[43, "gnomad.variant_qc.pipeline.score_bin_agg"]], "train_rf_model() (in module gnomad.variant_qc.pipeline)": [[43, "gnomad.variant_qc.pipeline.train_rf_model"]], "apply_rf_model() (in module gnomad.variant_qc.random_forest)": [[44, "gnomad.variant_qc.random_forest.apply_rf_model"]], "check_ht_fields_for_spark() (in module gnomad.variant_qc.random_forest)": [[44, "gnomad.variant_qc.random_forest.check_ht_fields_for_spark"]], "get_columns_quantiles() (in module gnomad.variant_qc.random_forest)": [[44, "gnomad.variant_qc.random_forest.get_columns_quantiles"]], "get_features_importance() (in module gnomad.variant_qc.random_forest)": [[44, "gnomad.variant_qc.random_forest.get_features_importance"]], "get_labels() (in module gnomad.variant_qc.random_forest)": [[44, "gnomad.variant_qc.random_forest.get_labels"]], "get_rf_runs() (in module gnomad.variant_qc.random_forest)": [[44, "gnomad.variant_qc.random_forest.get_rf_runs"]], "get_run_data() (in module gnomad.variant_qc.random_forest)": [[44, "gnomad.variant_qc.random_forest.get_run_data"]], "gnomad.variant_qc.random_forest": [[44, "module-gnomad.variant_qc.random_forest"]], "ht_to_rf_df() (in module gnomad.variant_qc.random_forest)": [[44, "gnomad.variant_qc.random_forest.ht_to_rf_df"]], "load_model() (in module gnomad.variant_qc.random_forest)": [[44, "gnomad.variant_qc.random_forest.load_model"]], "median_impute_features() (in module gnomad.variant_qc.random_forest)": [[44, "gnomad.variant_qc.random_forest.median_impute_features"]], "pretty_print_runs() (in module gnomad.variant_qc.random_forest)": [[44, "gnomad.variant_qc.random_forest.pretty_print_runs"]], "run_rf_test() (in module gnomad.variant_qc.random_forest)": [[44, "gnomad.variant_qc.random_forest.run_rf_test"]], "save_model() (in module gnomad.variant_qc.random_forest)": [[44, "gnomad.variant_qc.random_forest.save_model"]], "test_model() (in module gnomad.variant_qc.random_forest)": [[44, "gnomad.variant_qc.random_forest.test_model"]], "train_rf() (in module gnomad.variant_qc.random_forest)": [[44, "gnomad.variant_qc.random_forest.train_rf"]], "gnomad.variant_qc.training": [[45, "module-gnomad.variant_qc.training"]], "sample_training_examples() (in module gnomad.variant_qc.training)": [[45, "gnomad.variant_qc.training.sample_training_examples"]]}})