<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>gnomad.sample_qc.relatedness &mdash; gnomad master documentation</title>
      <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../../../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../../../_static/theme_overrides.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../../../_static/jquery.js"></script>
        <script src="../../../_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script data-url_root="../../../" id="documentation_options" src="../../../_static/documentation_options.js"></script>
        <script src="../../../_static/doctools.js"></script>
        <script src="../../../_static/sphinx_highlight.js"></script>
    <script src="../../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../../index.html" class="icon icon-home">
            gnomad
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../../../getting_started.html">Getting Started</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../examples/index.html">Examples</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../api_reference/index.html">API Reference</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../resource_sources.html">Resource Sources</a></li>
<li class="toctree-l1"><a class="reference external" href="https://github.com/broadinstitute/gnomad_methods/releases">Change Log</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">gnomad</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../../index.html">Module code</a></li>
      <li class="breadcrumb-item active">gnomad.sample_qc.relatedness</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for gnomad.sample_qc.relatedness</h1><div class="highlight"><pre>
<span></span><span class="c1"># noqa: D100</span>

<span class="kn">import</span> <span class="nn">logging</span>
<span class="kn">import</span> <span class="nn">random</span>
<span class="kn">from</span> <span class="nn">collections</span> <span class="kn">import</span> <span class="n">defaultdict</span>
<span class="kn">from</span> <span class="nn">typing</span> <span class="kn">import</span> <span class="n">Dict</span><span class="p">,</span> <span class="n">Iterable</span><span class="p">,</span> <span class="n">List</span><span class="p">,</span> <span class="n">Optional</span><span class="p">,</span> <span class="n">Set</span><span class="p">,</span> <span class="n">Tuple</span><span class="p">,</span> <span class="n">Union</span>

<span class="kn">import</span> <span class="nn">hail</span> <span class="k">as</span> <span class="nn">hl</span>
<span class="kn">import</span> <span class="nn">networkx</span> <span class="k">as</span> <span class="nn">nx</span>

<span class="kn">from</span> <span class="nn">gnomad.utils.annotations</span> <span class="kn">import</span> <span class="n">get_copy_state_by_sex</span>
<span class="kn">from</span> <span class="nn">gnomad.utils.filtering</span> <span class="kn">import</span> <span class="n">add_filters_expr</span>

<span class="n">logging</span><span class="o">.</span><span class="n">basicConfig</span><span class="p">(</span><span class="nb">format</span><span class="o">=</span><span class="s2">&quot;</span><span class="si">%(levelname)s</span><span class="s2"> (</span><span class="si">%(name)s</span><span class="s2"> </span><span class="si">%(lineno)s</span><span class="s2">): </span><span class="si">%(message)s</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="n">logger</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">(</span><span class="vm">__name__</span><span class="p">)</span>
<span class="n">logger</span><span class="o">.</span><span class="n">setLevel</span><span class="p">(</span><span class="n">logging</span><span class="o">.</span><span class="n">INFO</span><span class="p">)</span>

<span class="n">UNRELATED</span> <span class="o">=</span> <span class="s2">&quot;unrelated&quot;</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">String representation for a pair of unrelated individuals in this module.</span>
<span class="sd">Typically &gt;2nd degree relatives, but the threshold is user-dependant.</span>
<span class="sd">&quot;&quot;&quot;</span>

<span class="n">SECOND_DEGREE_RELATIVES</span> <span class="o">=</span> <span class="s2">&quot;second degree relatives&quot;</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">String representation for a pair of 2nd degree relatives in this module.</span>
<span class="sd">&quot;&quot;&quot;</span>

<span class="n">PARENT_CHILD</span> <span class="o">=</span> <span class="s2">&quot;parent-child&quot;</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">String representation for a parent-child pair in this module.</span>
<span class="sd">&quot;&quot;&quot;</span>

<span class="n">SIBLINGS</span> <span class="o">=</span> <span class="s2">&quot;siblings&quot;</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">String representation for a sibling pair in this module.</span>
<span class="sd">&quot;&quot;&quot;</span>

<span class="n">DUPLICATE_OR_TWINS</span> <span class="o">=</span> <span class="s2">&quot;duplicate/twins&quot;</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">String representation for a pair of samples who are identical (either MZ twins of duplicate) in this module.</span>
<span class="sd">&quot;&quot;&quot;</span>

<span class="n">AMBIGUOUS_RELATIONSHIP</span> <span class="o">=</span> <span class="s2">&quot;ambiguous&quot;</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">String representation for a pair of samples whose relationship is ambiguous.</span>
<span class="sd">This is used in the case of a pair of samples which kinship/IBD values do not correspond to any biological relationship between two individuals.</span>
<span class="sd">&quot;&quot;&quot;</span>


<div class="viewcode-block" id="get_duplicated_samples"><a class="viewcode-back" href="../../../api_reference/sample_qc/relatedness.html#gnomad.sample_qc.relatedness.get_duplicated_samples">[docs]</a><span class="k">def</span> <span class="nf">get_duplicated_samples</span><span class="p">(</span>
    <span class="n">relationship_ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span>
    <span class="n">i_col</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;i&quot;</span><span class="p">,</span>
    <span class="n">j_col</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;j&quot;</span><span class="p">,</span>
    <span class="n">rel_col</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;relationship&quot;</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">Set</span><span class="p">[</span><span class="nb">str</span><span class="p">]]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Extract the list of duplicate samples using a Table ouput from pc_relate.</span>

<span class="sd">    :param relationship_ht: Table with relationships between pairs of samples</span>
<span class="sd">    :param i_col: Column containing the 1st sample</span>
<span class="sd">    :param j_col: Column containing the 2nd sample</span>
<span class="sd">    :param rel_col: Column containing the sample pair relationship annotated with get_relationship_expr</span>
<span class="sd">    :return: List of sets of samples that are duplicates</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="nf">get_all_dups</span><span class="p">(</span>
        <span class="n">s</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">dups</span><span class="p">:</span> <span class="n">Set</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span> <span class="n">samples_duplicates</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Set</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Tuple</span><span class="p">[</span><span class="n">Set</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Set</span><span class="p">[</span><span class="nb">str</span><span class="p">]]]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Create the set of all duplicated samples corresponding to `s` that are found in `sample_duplicates`.</span>

<span class="sd">        Also return the remaining sample duplicates after removing all duplicated samples corresponding to `s`.</span>

<span class="sd">        Works by recursively adding duplicated samples to the set.</span>

<span class="sd">        :param s: sample to identify duplicates for</span>
<span class="sd">        :param dups: set of corresponding samples already identified</span>
<span class="sd">        :param samples_duplicates: dict of sample -&gt; duplicate-pair left to assign</span>
<span class="sd">        :return: (set of duplicates corresponding to s found in samples_duplicates, remaining samples_duplicates)</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">samples_duplicates</span><span class="p">:</span>
            <span class="n">dups</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">s</span><span class="p">)</span>
            <span class="n">s_dups</span> <span class="o">=</span> <span class="n">samples_duplicates</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="n">s</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">s_dup</span> <span class="ow">in</span> <span class="n">s_dups</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">s_dup</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">dups</span><span class="p">:</span>
                    <span class="n">dups</span><span class="p">,</span> <span class="n">samples_duplicates</span> <span class="o">=</span> <span class="n">get_all_dups</span><span class="p">(</span>
                        <span class="n">s_dup</span><span class="p">,</span> <span class="n">dups</span><span class="p">,</span> <span class="n">samples_duplicates</span>
                    <span class="p">)</span>
        <span class="k">return</span> <span class="n">dups</span><span class="p">,</span> <span class="n">samples_duplicates</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Computing duplicate sets&quot;</span><span class="p">)</span>
    <span class="n">dup_pairs</span> <span class="o">=</span> <span class="n">relationship_ht</span><span class="o">.</span><span class="n">aggregate</span><span class="p">(</span>
        <span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
            <span class="n">relationship_ht</span><span class="p">[</span><span class="n">rel_col</span><span class="p">]</span> <span class="o">==</span> <span class="n">DUPLICATE_OR_TWINS</span><span class="p">,</span>
            <span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">collect</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">tuple</span><span class="p">([</span><span class="n">relationship_ht</span><span class="p">[</span><span class="n">i_col</span><span class="p">],</span> <span class="n">relationship_ht</span><span class="p">[</span><span class="n">j_col</span><span class="p">]])),</span>
        <span class="p">)</span>
    <span class="p">)</span>

    <span class="n">samples_duplicates</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">set</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">j</span> <span class="ow">in</span> <span class="n">dup_pairs</span><span class="p">:</span>
        <span class="n">samples_duplicates</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">j</span><span class="p">)</span>
        <span class="n">samples_duplicates</span><span class="p">[</span><span class="n">j</span><span class="p">]</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>

    <span class="n">duplicated_samples</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">while</span> <span class="nb">len</span><span class="p">(</span><span class="n">samples_duplicates</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">dup_set</span><span class="p">,</span> <span class="n">samples_duplicates</span> <span class="o">=</span> <span class="n">get_all_dups</span><span class="p">(</span>
            <span class="nb">list</span><span class="p">(</span><span class="n">samples_duplicates</span><span class="p">)[</span><span class="mi">0</span><span class="p">],</span> <span class="nb">set</span><span class="p">(),</span> <span class="n">samples_duplicates</span>
        <span class="p">)</span>
        <span class="n">duplicated_samples</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">dup_set</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">duplicated_samples</span></div>


<div class="viewcode-block" id="get_duplicated_samples_ht"><a class="viewcode-back" href="../../../api_reference/sample_qc/relatedness.html#gnomad.sample_qc.relatedness.get_duplicated_samples_ht">[docs]</a><span class="k">def</span> <span class="nf">get_duplicated_samples_ht</span><span class="p">(</span>
    <span class="n">duplicated_samples</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">Set</span><span class="p">[</span><span class="nb">str</span><span class="p">]],</span>
    <span class="n">samples_rankings_ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span>
    <span class="n">rank_ann</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;rank&quot;</span><span class="p">,</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Create a HT with duplicated samples sets.</span>

<span class="sd">    Each row is indexed by the sample that is kept and also contains the set of duplicate samples that should be filtered.</span>

<span class="sd">    `samples_rankings_ht` is a HT containing a global rank for each of the samples (smaller is better).</span>

<span class="sd">    :param duplicated_samples: List of sets of duplicated samples</span>
<span class="sd">    :param samples_rankings_ht: HT with global rank for each sample</span>
<span class="sd">    :param rank_ann: Annotation in `samples_ranking_ht` containing each sample global rank (smaller is better).</span>
<span class="sd">    :return: HT with duplicate sample sets, including which to keep/filter</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">dups_ht</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="o">.</span><span class="n">parallelize</span><span class="p">(</span>
        <span class="p">[</span>
            <span class="n">hl</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span><span class="n">dup_set</span><span class="o">=</span><span class="n">i</span><span class="p">,</span> <span class="n">dups</span><span class="o">=</span><span class="n">duplicated_samples</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>
            <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">duplicated_samples</span><span class="p">))</span>
        <span class="p">]</span>
    <span class="p">)</span>
    <span class="n">dups_ht</span> <span class="o">=</span> <span class="n">dups_ht</span><span class="o">.</span><span class="n">explode</span><span class="p">(</span><span class="n">dups_ht</span><span class="o">.</span><span class="n">dups</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s2">&quot;_dup&quot;</span><span class="p">)</span>
    <span class="n">dups_ht</span> <span class="o">=</span> <span class="n">dups_ht</span><span class="o">.</span><span class="n">key_by</span><span class="p">(</span><span class="s2">&quot;_dup&quot;</span><span class="p">)</span>
    <span class="n">dups_ht</span> <span class="o">=</span> <span class="n">dups_ht</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span><span class="n">rank</span><span class="o">=</span><span class="n">samples_rankings_ht</span><span class="p">[</span><span class="n">dups_ht</span><span class="o">.</span><span class="n">key</span><span class="p">][</span><span class="n">rank_ann</span><span class="p">])</span>
    <span class="n">dups_cols</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">bind</span><span class="p">(</span>
        <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span><span class="n">kept</span><span class="o">=</span><span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">filtered</span><span class="o">=</span><span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">:]),</span>
        <span class="n">hl</span><span class="o">.</span><span class="n">sorted</span><span class="p">(</span>
            <span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">collect</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">tuple</span><span class="p">([</span><span class="n">dups_ht</span><span class="o">.</span><span class="n">_dup</span><span class="p">,</span> <span class="n">dups_ht</span><span class="o">.</span><span class="n">rank</span><span class="p">])),</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
        <span class="p">)</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">]),</span>
    <span class="p">)</span>
    <span class="n">dups_ht</span> <span class="o">=</span> <span class="n">dups_ht</span><span class="o">.</span><span class="n">group_by</span><span class="p">(</span><span class="n">dups_ht</span><span class="o">.</span><span class="n">dup_set</span><span class="p">)</span><span class="o">.</span><span class="n">aggregate</span><span class="p">(</span><span class="o">**</span><span class="n">dups_cols</span><span class="p">)</span>

    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">dups_ht</span><span class="o">.</span><span class="n">kept</span><span class="p">,</span> <span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">StructExpression</span><span class="p">):</span>
        <span class="n">dups_ht</span> <span class="o">=</span> <span class="n">dups_ht</span><span class="o">.</span><span class="n">key_by</span><span class="p">(</span><span class="o">**</span><span class="n">dups_ht</span><span class="o">.</span><span class="n">kept</span><span class="p">)</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s2">&quot;kept&quot;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">dups_ht</span> <span class="o">=</span> <span class="n">dups_ht</span><span class="o">.</span><span class="n">key_by</span><span class="p">(</span>
            <span class="n">s</span><span class="o">=</span><span class="n">dups_ht</span><span class="o">.</span><span class="n">kept</span>
        <span class="p">)</span>  <span class="c1"># Since there is no defined name in the case of a non-struct type, use `s`</span>
    <span class="k">return</span> <span class="n">dups_ht</span></div>


<div class="viewcode-block" id="explode_duplicate_samples_ht"><a class="viewcode-back" href="../../../api_reference/sample_qc/relatedness.html#gnomad.sample_qc.relatedness.explode_duplicate_samples_ht">[docs]</a><span class="k">def</span> <span class="nf">explode_duplicate_samples_ht</span><span class="p">(</span><span class="n">dups_ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Explode the result of `get_duplicated_samples_ht`, so that each line contains a single sample.</span>

<span class="sd">    An additional annotation is added: `dup_filtered` indicating which of the duplicated samples was kept.</span>
<span class="sd">    Requires a field `filtered` which type should be the same as the input duplicated samples Table key.</span>

<span class="sd">    :param dups_ht: Input HT</span>
<span class="sd">    :return: Flattened HT</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="nf">get_dups_to_keep_expr</span><span class="p">():</span>
        <span class="k">if</span> <span class="n">dups_ht</span><span class="o">.</span><span class="n">filtered</span><span class="o">.</span><span class="n">dtype</span><span class="o">.</span><span class="n">element_type</span> <span class="o">==</span> <span class="n">dups_ht</span><span class="o">.</span><span class="n">key</span><span class="o">.</span><span class="n">dtype</span><span class="p">:</span>
            <span class="k">return</span> <span class="p">(</span><span class="n">dups_ht</span><span class="o">.</span><span class="n">key</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>
        <span class="k">elif</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">dups_ht</span><span class="o">.</span><span class="n">key</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span>
            <span class="n">dups_ht</span><span class="o">.</span><span class="n">filtered</span><span class="o">.</span><span class="n">dtype</span><span class="o">.</span><span class="n">element_type</span> <span class="o">==</span> <span class="n">dups_ht</span><span class="o">.</span><span class="n">key</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">dtype</span>
        <span class="p">):</span>
            <span class="k">return</span> <span class="p">(</span><span class="n">dups_ht</span><span class="o">.</span><span class="n">key</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="kc">False</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span>
                <span class="s2">&quot;Cannot explode table as types of the filtered field&quot;</span>
                <span class="sa">f</span><span class="s2">&quot; (</span><span class="si">{</span><span class="n">dups_ht</span><span class="o">.</span><span class="n">filtered</span><span class="o">.</span><span class="n">dtype</span><span class="si">}</span><span class="s2">) and the key (</span><span class="si">{</span><span class="n">dups_ht</span><span class="o">.</span><span class="n">key</span><span class="o">.</span><span class="n">dtype</span><span class="si">}</span><span class="s2">) are&quot;</span>
                <span class="s2">&quot; incompatible.&quot;</span>
            <span class="p">)</span>

    <span class="n">dups_ht</span> <span class="o">=</span> <span class="n">dups_ht</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span>
        <span class="n">dups</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="n">get_dups_to_keep_expr</span><span class="p">()])</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span>
            <span class="n">dups_ht</span><span class="o">.</span><span class="n">filtered</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="kc">True</span><span class="p">))</span>
        <span class="p">)</span>
    <span class="p">)</span>
    <span class="n">dups_ht</span> <span class="o">=</span> <span class="n">dups_ht</span><span class="o">.</span><span class="n">explode</span><span class="p">(</span><span class="s2">&quot;dups&quot;</span><span class="p">)</span>
    <span class="n">dups_ht</span> <span class="o">=</span> <span class="n">dups_ht</span><span class="o">.</span><span class="n">key_by</span><span class="p">()</span>
    <span class="k">return</span> <span class="n">dups_ht</span><span class="o">.</span><span class="n">select</span><span class="p">(</span><span class="n">s</span><span class="o">=</span><span class="n">dups_ht</span><span class="o">.</span><span class="n">dups</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">dup_filtered</span><span class="o">=</span><span class="n">dups_ht</span><span class="o">.</span><span class="n">dups</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span><span class="o">.</span><span class="n">key_by</span><span class="p">(</span><span class="s2">&quot;s&quot;</span><span class="p">)</span></div>


<div class="viewcode-block" id="get_relationship_expr"><a class="viewcode-back" href="../../../api_reference/sample_qc/relatedness.html#gnomad.sample_qc.relatedness.get_relationship_expr">[docs]</a><span class="k">def</span> <span class="nf">get_relationship_expr</span><span class="p">(</span>  <span class="c1"># TODO: The threshold detection could be easily automated by fitting distributions over the data.</span>
    <span class="n">kin_expr</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">NumericExpression</span><span class="p">,</span>
    <span class="n">ibd0_expr</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">NumericExpression</span><span class="p">,</span>
    <span class="n">ibd1_expr</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">NumericExpression</span><span class="p">,</span>
    <span class="n">ibd2_expr</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">NumericExpression</span><span class="p">,</span>
    <span class="n">first_degree_kin_thresholds</span><span class="p">:</span> <span class="n">Tuple</span><span class="p">[</span><span class="nb">float</span><span class="p">,</span> <span class="nb">float</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="mf">0.19</span><span class="p">,</span> <span class="mf">0.4</span><span class="p">),</span>
    <span class="n">second_degree_min_kin</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.1</span><span class="p">,</span>
    <span class="n">ibd0_0_max</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.025</span><span class="p">,</span>
    <span class="n">ibd0_25_thresholds</span><span class="p">:</span> <span class="n">Tuple</span><span class="p">[</span><span class="nb">float</span><span class="p">,</span> <span class="nb">float</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="mf">0.1</span><span class="p">,</span> <span class="mf">0.425</span><span class="p">),</span>
    <span class="c1"># ibd0_50_thresholds = [0.37, 0.625], Not useful for relationship inference</span>
    <span class="c1"># ibd0_100_threshold = 0.625  , Not useful for relationship inference</span>
    <span class="n">ibd1_0_thresholds</span><span class="p">:</span> <span class="n">Tuple</span><span class="p">[</span><span class="nb">float</span><span class="p">,</span> <span class="nb">float</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="o">-</span><span class="mf">0.15</span><span class="p">,</span> <span class="mf">0.1</span><span class="p">),</span>
    <span class="c1"># ibd1_25_thresholds: Tuple[float, float] = (0.1, 0.37), Not useful for</span>
    <span class="c1"># relationship inference</span>
    <span class="n">ibd1_50_thresholds</span><span class="p">:</span> <span class="n">Tuple</span><span class="p">[</span><span class="nb">float</span><span class="p">,</span> <span class="nb">float</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="mf">0.275</span><span class="p">,</span> <span class="mf">0.75</span><span class="p">),</span>
    <span class="n">ibd1_100_min</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.75</span><span class="p">,</span>
    <span class="n">ibd2_0_max</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.125</span><span class="p">,</span>
    <span class="n">ibd2_25_thresholds</span><span class="p">:</span> <span class="n">Tuple</span><span class="p">[</span><span class="nb">float</span><span class="p">,</span> <span class="nb">float</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="mf">0.1</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">),</span>
    <span class="n">ibd2_100_thresholds</span><span class="p">:</span> <span class="n">Tuple</span><span class="p">[</span><span class="nb">float</span><span class="p">,</span> <span class="nb">float</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="mf">0.75</span><span class="p">,</span> <span class="mf">1.25</span><span class="p">),</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">StringExpression</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Return an expression indicating the relationship between a pair of samples given their kin coefficient and IBDO, IBD1, IBD2 values.</span>

<span class="sd">    The kinship coefficient values in the defaults are in line with those output from</span>
<span class="sd">    `hail.methods.pc_relate &lt;https://hail.is/docs/0.2/methods/genetics.html?highlight=pc_relate#hail.methods.pc_relate&gt;`.</span>

<span class="sd">    :param kin_expr: Kin coefficient expression</span>
<span class="sd">    :param ibd0_expr: IBDO expression</span>
<span class="sd">    :param ibd1_expr: IBD1 expression</span>
<span class="sd">    :param ibd2_expr: IDB2 expression</span>
<span class="sd">    :param first_degree_kin_thresholds: (min, max) kinship threshold for 1st degree relatives</span>
<span class="sd">    :param second_degree_min_kin: min kinship threshold for 2nd degree relatives</span>
<span class="sd">    :param ibd0_0_max: max IBD0 threshold for 0 IBD0 sharing</span>
<span class="sd">    :param ibd0_25_thresholds: (min, max) thresholds for 0.25 IBD0 sharing</span>
<span class="sd">    :param ibd1_0_thresholds: (min, max) thresholds for 0 IBD1 sharing. Note that the min is there because pc_relate can output large negative values in some corner cases.</span>
<span class="sd">    :param ibd1_50_thresholds: (min, max) thresholds for 0.5 IBD1 sharing</span>
<span class="sd">    :param ibd1_100_min: min IBD1 threshold for 1.0 IBD1 sharing</span>
<span class="sd">    :param ibd2_0_max: max IBD2 threshold for 0 IBD2 sharing</span>
<span class="sd">    :param ibd2_25_thresholds: (min, max) thresholds for 0.25 IBD2 sharing</span>
<span class="sd">    :param ibd2_100_thresholds: (min, max) thresholds for 1.00 IBD2 sharing. Note that the min is there because pc_relate can output much larger IBD2 values in some corner cases.</span>
<span class="sd">    :return: The relationship annotation using the constants defined in this module.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="p">(</span>
        <span class="n">hl</span><span class="o">.</span><span class="n">case</span><span class="p">()</span>
        <span class="o">.</span><span class="n">when</span><span class="p">(</span><span class="n">kin_expr</span> <span class="o">&lt;</span> <span class="n">second_degree_min_kin</span><span class="p">,</span> <span class="n">UNRELATED</span><span class="p">)</span>
        <span class="o">.</span><span class="n">when</span><span class="p">((</span><span class="n">kin_expr</span> <span class="o">&lt;</span> <span class="n">first_degree_kin_thresholds</span><span class="p">[</span><span class="mi">0</span><span class="p">]),</span> <span class="n">SECOND_DEGREE_RELATIVES</span><span class="p">)</span>
        <span class="o">.</span><span class="n">when</span><span class="p">(</span>
            <span class="p">(</span><span class="n">kin_expr</span> <span class="o">&lt;</span> <span class="n">first_degree_kin_thresholds</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
            <span class="o">&amp;</span> <span class="p">(</span><span class="n">ibd0_expr</span> <span class="o">&lt;=</span> <span class="n">ibd0_0_max</span><span class="p">)</span>
            <span class="o">&amp;</span> <span class="p">(</span><span class="n">ibd1_expr</span> <span class="o">&gt;=</span> <span class="n">ibd1_100_min</span><span class="p">)</span>
            <span class="o">&amp;</span> <span class="p">(</span><span class="n">ibd2_expr</span> <span class="o">&lt;=</span> <span class="n">ibd2_0_max</span><span class="p">),</span>
            <span class="n">PARENT_CHILD</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="o">.</span><span class="n">when</span><span class="p">(</span>
            <span class="p">(</span><span class="n">kin_expr</span> <span class="o">&lt;</span> <span class="n">first_degree_kin_thresholds</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
            <span class="o">&amp;</span> <span class="p">(</span><span class="n">ibd0_expr</span> <span class="o">&gt;=</span> <span class="n">ibd0_25_thresholds</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
            <span class="o">&amp;</span> <span class="p">(</span><span class="n">ibd0_expr</span> <span class="o">&lt;=</span> <span class="n">ibd0_25_thresholds</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
            <span class="o">&amp;</span> <span class="p">(</span><span class="n">ibd1_expr</span> <span class="o">&gt;=</span> <span class="n">ibd1_50_thresholds</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
            <span class="o">&amp;</span> <span class="p">(</span><span class="n">ibd1_expr</span> <span class="o">&lt;=</span> <span class="n">ibd1_50_thresholds</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
            <span class="o">&amp;</span> <span class="p">(</span><span class="n">ibd2_expr</span> <span class="o">&gt;=</span> <span class="n">ibd2_25_thresholds</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
            <span class="o">&amp;</span> <span class="p">(</span><span class="n">ibd2_expr</span> <span class="o">&lt;=</span> <span class="n">ibd2_25_thresholds</span><span class="p">[</span><span class="mi">1</span><span class="p">]),</span>
            <span class="n">SIBLINGS</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="o">.</span><span class="n">when</span><span class="p">(</span>
            <span class="p">(</span><span class="n">kin_expr</span> <span class="o">&gt;</span> <span class="n">first_degree_kin_thresholds</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
            <span class="o">&amp;</span> <span class="p">(</span><span class="n">ibd0_expr</span> <span class="o">&lt;</span> <span class="n">ibd0_0_max</span><span class="p">)</span>
            <span class="o">&amp;</span> <span class="p">(</span><span class="n">ibd1_expr</span> <span class="o">&gt;=</span> <span class="n">ibd1_0_thresholds</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
            <span class="o">&amp;</span> <span class="p">(</span><span class="n">ibd1_expr</span> <span class="o">&lt;=</span> <span class="n">ibd1_0_thresholds</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
            <span class="o">&amp;</span> <span class="p">(</span><span class="n">ibd2_expr</span> <span class="o">&gt;=</span> <span class="n">ibd2_100_thresholds</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
            <span class="o">&amp;</span> <span class="p">(</span><span class="n">ibd2_expr</span> <span class="o">&lt;=</span> <span class="n">ibd2_100_thresholds</span><span class="p">[</span><span class="mi">1</span><span class="p">]),</span>
            <span class="n">DUPLICATE_OR_TWINS</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="o">.</span><span class="n">default</span><span class="p">(</span><span class="n">AMBIGUOUS_RELATIONSHIP</span><span class="p">)</span>
    <span class="p">)</span></div>


<div class="viewcode-block" id="get_slope_int_relationship_expr"><a class="viewcode-back" href="../../../api_reference/sample_qc/relatedness.html#gnomad.sample_qc.relatedness.get_slope_int_relationship_expr">[docs]</a><span class="k">def</span> <span class="nf">get_slope_int_relationship_expr</span><span class="p">(</span>
    <span class="n">kin_expr</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">NumericExpression</span><span class="p">,</span>
    <span class="n">y_expr</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">NumericExpression</span><span class="p">,</span>
    <span class="n">parent_child_max_y</span><span class="p">:</span> <span class="nb">float</span><span class="p">,</span>
    <span class="n">second_degree_sibling_lower_cutoff_slope</span><span class="p">:</span> <span class="nb">float</span><span class="p">,</span>
    <span class="n">second_degree_sibling_lower_cutoff_intercept</span><span class="p">:</span> <span class="nb">float</span><span class="p">,</span>
    <span class="n">second_degree_upper_sibling_lower_cutoff_slope</span><span class="p">:</span> <span class="nb">float</span><span class="p">,</span>
    <span class="n">second_degree_upper_sibling_lower_cutoff_intercept</span><span class="p">:</span> <span class="nb">float</span><span class="p">,</span>
    <span class="n">duplicate_twin_min_kin</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.42</span><span class="p">,</span>
    <span class="n">second_degree_min_kin</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.1</span><span class="p">,</span>
    <span class="n">duplicate_twin_ibd1_min</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="o">-</span><span class="mf">0.15</span><span class="p">,</span>
    <span class="n">duplicate_twin_ibd1_max</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.1</span><span class="p">,</span>
    <span class="n">ibd1_expr</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">NumericExpression</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Return an expression indicating the relationship between a pair of samples given slope and intercept cutoffs.</span>

<span class="sd">    The kinship coefficient (`kin_expr`) and an additional metric (`y_expr`) are used</span>
<span class="sd">    to define the relationship between a pair of samples. For this function the</span>
<span class="sd">    slope and intercepts should refer to cutoff lines where the x-axis, or independent</span>
<span class="sd">    variable is the kinship coefficient and the y-axis, or dependent variable, is</span>
<span class="sd">    the metric defined by `y_expr`. Typically, the y-axis metric IBS0, IBS0/IBS2, or</span>
<span class="sd">    IBD0.</span>

<span class="sd">    .. note::</span>

<span class="sd">        No defaults are provided for the slope and intercept cutoffs because they are</span>
<span class="sd">        highly dependent on the dataset and the metric used in `y_expr`.</span>

<span class="sd">    The relationship expression is determined as follows:</span>
<span class="sd">        - If `kin_expr` &lt; `second_degree_min_kin` -&gt;  UNRELATED</span>
<span class="sd">        - If `kin_expr` &gt; `duplicate_twin_min_kin`:</span>
<span class="sd">            - If `y_expr` &lt; `parent_child_max_y`:</span>
<span class="sd">                - If `ibd1_expr` is defined:</span>
<span class="sd">                    - If `duplicate_twin_ibd1_min` &lt;= `ibd1_expr` &lt;= `</span>
<span class="sd">                      duplicate_twin_ibd1_max` -&gt; DUPLICATE_OR_TWINS</span>
<span class="sd">                    - Else -&gt; AMBIGUOUS_RELATIONSHIP</span>
<span class="sd">                - Else -&gt; DUPLICATE_OR_TWINS</span>
<span class="sd">        - If `y_expr` &lt; `parent_child_max_y` -&gt; PARENT_CHILD</span>
<span class="sd">        - If pair is over second_degree_sibling_lower_cutoff line:</span>
<span class="sd">            - If pair is over second_degree_upper_sibling_lower_cutoff line -&gt; SIBLINGS</span>
<span class="sd">            - Else -&gt; SECOND_DEGREE_RELATIVES</span>
<span class="sd">        - If none of the above conditions are met -&gt; AMBIGUOUS_RELATIONSHIP</span>

<span class="sd">    :param kin_expr: Kin coefficient expression. Used as the x-axis, or independent</span>
<span class="sd">        variable, for the slope and intercept cutoffs.</span>
<span class="sd">    :param y_expr: Expression for the metric to use as the y-axis, or dependent</span>
<span class="sd">        variable, for the slope and intercept cutoffs. This is typically an expression</span>
<span class="sd">        for IBS0, IBS0/IBS2, or IBD0.</span>
<span class="sd">    :param parent_child_max_y: Maximum value of the metric defined by `y_expr` for a</span>
<span class="sd">        parent-child pair.</span>
<span class="sd">    :param second_degree_sibling_lower_cutoff_slope: Slope of the line to use as a</span>
<span class="sd">        lower cutoff for second degree relatives and siblings from parent-child pairs.</span>
<span class="sd">    :param second_degree_sibling_lower_cutoff_intercept: Intercept of the line to use</span>
<span class="sd">        as a lower cutoff for second degree relatives and siblings from parent-child</span>
<span class="sd">        pairs.</span>
<span class="sd">    :param second_degree_upper_sibling_lower_cutoff_slope: Slope of the line to use as</span>
<span class="sd">        an upper cutoff for second degree relatives and a lower cutoff for siblings.</span>
<span class="sd">    :param second_degree_upper_sibling_lower_cutoff_intercept: Intercept of the line to</span>
<span class="sd">        use as an upper cutoff for second degree relatives and a lower cutoff for</span>
<span class="sd">        siblings.</span>
<span class="sd">    :param duplicate_twin_min_kin: Minimum kinship for duplicate or twin pairs.</span>
<span class="sd">        Default is 0.42.</span>
<span class="sd">    :param second_degree_min_kin: Minimum kinship threshold for 2nd degree relatives.</span>
<span class="sd">        Default is 0.08838835. Bycroft et al. (2018) calculates a theoretical kinship</span>
<span class="sd">        of 0.08838835 for a second degree relationship cutoff, but this cutoff should be</span>
<span class="sd">        determined by evaluation of the kinship distribution.</span>
<span class="sd">    :param ibd1_expr: Optional IBD1 expression. If this expression is provided,</span>
<span class="sd">        `duplicate_twin_ibd1_min` and `duplicate_twin_ibd1_max` will be used as an</span>
<span class="sd">        additional cutoff for duplicate or twin pairs.</span>
<span class="sd">    :param duplicate_twin_ibd1_min: Minimum IBD1 cutoff for duplicate or twin pairs.</span>
<span class="sd">        Note: the min is because pc_relate can output large negative values in some</span>
<span class="sd">        corner cases.</span>
<span class="sd">    :param duplicate_twin_ibd1_max: Maximum IBD1 cutoff for duplicate or twin pairs.</span>
<span class="sd">    :return: The relationship annotation using the constants defined in this module.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Only use a duplicate/twin IBD1 cutoff if an ibd1_expr is supplied.</span>
    <span class="k">if</span> <span class="n">ibd1_expr</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">dup_twin_ibd1_expr</span> <span class="o">=</span> <span class="p">(</span><span class="n">ibd1_expr</span> <span class="o">&gt;=</span> <span class="n">duplicate_twin_ibd1_min</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span>
            <span class="n">ibd1_expr</span> <span class="o">&lt;=</span> <span class="n">duplicate_twin_ibd1_max</span>
        <span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">dup_twin_ibd1_expr</span> <span class="o">=</span> <span class="kc">True</span>

    <span class="k">return</span> <span class="p">(</span>
        <span class="n">hl</span><span class="o">.</span><span class="n">case</span><span class="p">()</span>
        <span class="o">.</span><span class="n">when</span><span class="p">(</span><span class="n">kin_expr</span> <span class="o">&lt;</span> <span class="n">second_degree_min_kin</span><span class="p">,</span> <span class="n">UNRELATED</span><span class="p">)</span>
        <span class="o">.</span><span class="n">when</span><span class="p">(</span>
            <span class="n">kin_expr</span> <span class="o">&gt;</span> <span class="n">duplicate_twin_min_kin</span><span class="p">,</span>
            <span class="n">hl</span><span class="o">.</span><span class="n">if_else</span><span class="p">(</span>
                <span class="n">dup_twin_ibd1_expr</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">y_expr</span> <span class="o">&lt;</span> <span class="n">parent_child_max_y</span><span class="p">),</span>
                <span class="n">DUPLICATE_OR_TWINS</span><span class="p">,</span>
                <span class="n">AMBIGUOUS_RELATIONSHIP</span><span class="p">,</span>
            <span class="p">),</span>
        <span class="p">)</span>
        <span class="o">.</span><span class="n">when</span><span class="p">(</span><span class="n">y_expr</span> <span class="o">&lt;</span> <span class="n">parent_child_max_y</span><span class="p">,</span> <span class="n">PARENT_CHILD</span><span class="p">)</span>
        <span class="o">.</span><span class="n">when</span><span class="p">(</span>
            <span class="n">y_expr</span>
            <span class="o">&gt;</span> <span class="n">second_degree_sibling_lower_cutoff_slope</span> <span class="o">*</span> <span class="n">kin_expr</span>
            <span class="o">+</span> <span class="n">second_degree_sibling_lower_cutoff_intercept</span><span class="p">,</span>
            <span class="n">hl</span><span class="o">.</span><span class="n">if_else</span><span class="p">(</span>
                <span class="n">y_expr</span>
                <span class="o">&gt;</span> <span class="n">second_degree_upper_sibling_lower_cutoff_slope</span> <span class="o">*</span> <span class="n">kin_expr</span>
                <span class="o">+</span> <span class="n">second_degree_upper_sibling_lower_cutoff_intercept</span><span class="p">,</span>
                <span class="n">SIBLINGS</span><span class="p">,</span>
                <span class="n">SECOND_DEGREE_RELATIVES</span><span class="p">,</span>
            <span class="p">),</span>
        <span class="p">)</span>
        <span class="o">.</span><span class="n">default</span><span class="p">(</span><span class="n">AMBIGUOUS_RELATIONSHIP</span><span class="p">)</span>
    <span class="p">)</span></div>


<div class="viewcode-block" id="infer_families"><a class="viewcode-back" href="../../../api_reference/sample_qc/relatedness.html#gnomad.sample_qc.relatedness.infer_families">[docs]</a><span class="k">def</span> <span class="nf">infer_families</span><span class="p">(</span>
    <span class="n">relationship_ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span>
    <span class="n">sex</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">bool</span><span class="p">]],</span>
    <span class="n">duplicate_samples_ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span>
    <span class="n">i_col</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;i&quot;</span><span class="p">,</span>
    <span class="n">j_col</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;j&quot;</span><span class="p">,</span>
    <span class="n">relationship_col</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;relationship&quot;</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">hl</span><span class="o">.</span><span class="n">Pedigree</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Generate a pedigree containing trios inferred from the `relationship_ht`.</span>

<span class="sd">    This function takes a hail Table with a row for each pair of related individuals i, j in the data (it&#39;s OK to have</span>
<span class="sd">    unrelated samples too).</span>

<span class="sd">    The `relationship_col` should be a column specifying the relationship between each two samples as defined in this</span>
<span class="sd">     module&#39;s constants.</span>

<span class="sd">    This function returns a pedigree containing trios inferred from the data. Family ID can be the same for multiple</span>
<span class="sd">    trios if one or more members of the trios are related (e.g. sibs, multi-generational family). Trios are ordered by family ID.</span>

<span class="sd">    .. note::</span>

<span class="sd">        This function only returns complete trios defined as: one child, one father, and one mother (sex is required for both parents).</span>

<span class="sd">    :param relationship_ht: Input relationship table</span>
<span class="sd">    :param sex: A Table or dict giving the sex for each sample (`TRUE`=XX, `FALSE`=XY). If a Table is given, it should have a field `is_xx`.</span>
<span class="sd">    :param duplicated_samples: All duplicated samples TO REMOVE (If not provided, this function won&#39;t work as it assumes that each child has exactly two parents)</span>
<span class="sd">    :param i_col: Column containing the 1st sample of the pair in the relationship table</span>
<span class="sd">    :param j_col: Column containing the 2nd sample of the pair in the relationship table</span>
<span class="sd">    :param relationship_col: Column contatining the relationship for the sample pair as defined in this module constants.</span>
<span class="sd">    :return: Pedigree of complete trios</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="nf">group_parent_child_pairs_by_fam</span><span class="p">(</span>
        <span class="n">parent_child_pairs</span><span class="p">:</span> <span class="n">Iterable</span><span class="p">[</span><span class="n">Tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]],</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="n">Tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]]]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Group parent-child pairs into a list of families.</span>

<span class="sd">        A family here is defined as a list of sample-pairs which all share at least one sample with at least one other</span>
<span class="sd">        sample-pair in the list.</span>

<span class="sd">        :param parent_child_pairs: All the parent-children pairs</span>
<span class="sd">        :return: A list of families, where each element of the list is a list of the parent-children pairs</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">fam_id</span> <span class="o">=</span> <span class="mi">1</span>  <span class="c1"># stores the current family id</span>
        <span class="n">s_fam</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">()</span>  <span class="c1"># stores the family id for each sample</span>
        <span class="n">fams</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">list</span><span class="p">)</span>  <span class="c1"># stores fam_id -&gt; sample-pairs</span>
        <span class="k">for</span> <span class="n">pair</span> <span class="ow">in</span> <span class="n">parent_child_pairs</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">pair</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="ow">in</span> <span class="n">s_fam</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">pair</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="ow">in</span> <span class="n">s_fam</span><span class="p">:</span>
                    <span class="k">if</span> <span class="p">(</span>
                        <span class="n">s_fam</span><span class="p">[</span><span class="n">pair</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span> <span class="o">!=</span> <span class="n">s_fam</span><span class="p">[</span><span class="n">pair</span><span class="p">[</span><span class="mi">1</span><span class="p">]]</span>
                    <span class="p">):</span>  <span class="c1"># If both samples are in different families, merge the families</span>
                        <span class="n">new_fam_id</span> <span class="o">=</span> <span class="n">s_fam</span><span class="p">[</span><span class="n">pair</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span>
                        <span class="n">fam_id_to_merge</span> <span class="o">=</span> <span class="n">s_fam</span><span class="p">[</span><span class="n">pair</span><span class="p">[</span><span class="mi">1</span><span class="p">]]</span>
                        <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">s_fam</span><span class="p">:</span>
                            <span class="k">if</span> <span class="n">s_fam</span><span class="p">[</span><span class="n">s</span><span class="p">]</span> <span class="o">==</span> <span class="n">fam_id_to_merge</span><span class="p">:</span>
                                <span class="n">s_fam</span><span class="p">[</span><span class="n">s</span><span class="p">]</span> <span class="o">=</span> <span class="n">new_fam_id</span>
                        <span class="n">fams</span><span class="p">[</span><span class="n">new_fam_id</span><span class="p">]</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">fams</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="n">fam_id_to_merge</span><span class="p">))</span>
                <span class="k">else</span><span class="p">:</span>  <span class="c1"># If only the 1st sample in the pair is already in a family, assign the 2nd sample in the pair to the same family</span>
                    <span class="n">s_fam</span><span class="p">[</span><span class="n">pair</span><span class="p">[</span><span class="mi">1</span><span class="p">]]</span> <span class="o">=</span> <span class="n">s_fam</span><span class="p">[</span><span class="n">pair</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span>
                <span class="n">fams</span><span class="p">[</span><span class="n">s_fam</span><span class="p">[</span><span class="n">pair</span><span class="p">[</span><span class="mi">0</span><span class="p">]]]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">pair</span><span class="p">)</span>
            <span class="k">elif</span> <span class="p">(</span>
                <span class="n">pair</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="ow">in</span> <span class="n">s_fam</span>
            <span class="p">):</span>  <span class="c1"># If only the 2nd sample in the pair is already in a family, assign the 1st sample in the pair to the same family</span>
                <span class="n">s_fam</span><span class="p">[</span><span class="n">pair</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span> <span class="o">=</span> <span class="n">s_fam</span><span class="p">[</span><span class="n">pair</span><span class="p">[</span><span class="mi">1</span><span class="p">]]</span>
                <span class="n">fams</span><span class="p">[</span><span class="n">s_fam</span><span class="p">[</span><span class="n">pair</span><span class="p">[</span><span class="mi">1</span><span class="p">]]]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">pair</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>  <span class="c1"># If none of the samples in the pair is already in a family, create a new family</span>
                <span class="n">s_fam</span><span class="p">[</span><span class="n">pair</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span> <span class="o">=</span> <span class="n">fam_id</span>
                <span class="n">s_fam</span><span class="p">[</span><span class="n">pair</span><span class="p">[</span><span class="mi">1</span><span class="p">]]</span> <span class="o">=</span> <span class="n">fam_id</span>
                <span class="n">fams</span><span class="p">[</span><span class="n">fam_id</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">pair</span><span class="p">)</span>
                <span class="n">fam_id</span> <span class="o">+=</span> <span class="mi">1</span>

        <span class="k">return</span> <span class="nb">list</span><span class="p">(</span><span class="n">fams</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>

    <span class="k">def</span> <span class="nf">get_trios</span><span class="p">(</span>
        <span class="n">fam_id</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
        <span class="n">parent_child_pairs</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">Tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]],</span>
        <span class="n">related_pairs</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="n">Tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">],</span> <span class="nb">str</span><span class="p">],</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">hl</span><span class="o">.</span><span class="n">Trio</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Generate trios based on the list of parent-child pairs in the family and all related pairs in the data.</span>

<span class="sd">        Only complete parent/offspring trios are included in the results.</span>

<span class="sd">        The trios are assembled as follows:</span>
<span class="sd">        1. All pairs of unrelated samples with different sexes within the family are extracted as possible parent pairs</span>
<span class="sd">        2. For each possible parent pair, a list of all children is constructed (each child in the list has a parent-offspring pair with each parent)</span>
<span class="sd">        3. If there are multiple children for a given parent pair, all children should be siblings with each other</span>
<span class="sd">        4. Check that each child was only assigned a single pair of parents. If a child is found to have multiple parent pairs, they are ALL discarded.</span>

<span class="sd">        :param fam_id: The family ID</span>
<span class="sd">        :param parent_child_pairs: The parent-child pairs for this family</span>
<span class="sd">        :param related_pairs: All related sample pairs in the data</span>
<span class="sd">        :return: List of trios in the family</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">def</span> <span class="nf">get_possible_parents</span><span class="p">(</span><span class="n">samples</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">Tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]]:</span>
<span class="w">            </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">            Return all pairs of unrelated samples with different sexes within the family are extracted as possible parent pairs.</span>

<span class="sd">            :param samples: All samples in the family</span>
<span class="sd">            :return: Possible parent pairs</span>
<span class="sd">            &quot;&quot;&quot;</span>
            <span class="n">possible_parents</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">samples</span><span class="p">)):</span>
                <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">samples</span><span class="p">)):</span>
                    <span class="k">if</span> <span class="p">(</span>
                        <span class="n">related_pairs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="nb">tuple</span><span class="p">(</span><span class="nb">sorted</span><span class="p">([</span><span class="n">samples</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">samples</span><span class="p">[</span><span class="n">j</span><span class="p">]])))</span>
                        <span class="ow">is</span> <span class="kc">None</span>
                    <span class="p">):</span>
                        <span class="k">if</span> <span class="n">sex</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">samples</span><span class="p">[</span><span class="n">i</span><span class="p">])</span> <span class="ow">is</span> <span class="kc">False</span> <span class="ow">and</span> <span class="n">sex</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">samples</span><span class="p">[</span><span class="n">j</span><span class="p">])</span> <span class="ow">is</span> <span class="kc">True</span><span class="p">:</span>
                            <span class="n">possible_parents</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">samples</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">samples</span><span class="p">[</span><span class="n">j</span><span class="p">]))</span>
                        <span class="k">elif</span> <span class="p">(</span>
                            <span class="n">sex</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">samples</span><span class="p">[</span><span class="n">i</span><span class="p">])</span> <span class="ow">is</span> <span class="kc">True</span> <span class="ow">and</span> <span class="n">sex</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">samples</span><span class="p">[</span><span class="n">j</span><span class="p">])</span> <span class="ow">is</span> <span class="kc">False</span>
                        <span class="p">):</span>
                            <span class="n">possible_parents</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">samples</span><span class="p">[</span><span class="n">j</span><span class="p">],</span> <span class="n">samples</span><span class="p">[</span><span class="n">i</span><span class="p">]))</span>
            <span class="k">return</span> <span class="n">possible_parents</span>

        <span class="k">def</span> <span class="nf">get_children</span><span class="p">(</span><span class="n">possible_parents</span><span class="p">:</span> <span class="n">Tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">            </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">            Construct a list of all children for a given possible parent pair.</span>

<span class="sd">            Each child in the list has a parent-offspring pair with each parent.</span>

<span class="sd">            :param possible_parents: A pair of possible parents</span>
<span class="sd">            :return: The list of all children (if any) corresponding to the possible parents</span>
<span class="sd">            &quot;&quot;&quot;</span>
            <span class="n">possible_offsprings</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span>
                <span class="nb">set</span>
            <span class="p">)</span>  <span class="c1"># stores sample -&gt; set of parents in the possible_parents where (sample, parent) is found in possible_child_pairs</span>
            <span class="k">for</span> <span class="n">pair</span> <span class="ow">in</span> <span class="n">parent_child_pairs</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">possible_parents</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="n">pair</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span>
                    <span class="n">possible_offsprings</span><span class="p">[</span><span class="n">pair</span><span class="p">[</span><span class="mi">1</span><span class="p">]]</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">possible_parents</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
                <span class="k">elif</span> <span class="n">possible_parents</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="n">pair</span><span class="p">[</span><span class="mi">1</span><span class="p">]:</span>
                    <span class="n">possible_offsprings</span><span class="p">[</span><span class="n">pair</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">possible_parents</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
                <span class="k">elif</span> <span class="n">possible_parents</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="n">pair</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span>
                    <span class="n">possible_offsprings</span><span class="p">[</span><span class="n">pair</span><span class="p">[</span><span class="mi">1</span><span class="p">]]</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">possible_parents</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
                <span class="k">elif</span> <span class="n">possible_parents</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="n">pair</span><span class="p">[</span><span class="mi">1</span><span class="p">]:</span>
                    <span class="n">possible_offsprings</span><span class="p">[</span><span class="n">pair</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">possible_parents</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>

            <span class="k">return</span> <span class="p">[</span>
                <span class="n">s</span> <span class="k">for</span> <span class="n">s</span><span class="p">,</span> <span class="n">parents</span> <span class="ow">in</span> <span class="n">possible_offsprings</span><span class="o">.</span><span class="n">items</span><span class="p">()</span> <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">parents</span><span class="p">)</span> <span class="o">==</span> <span class="mi">2</span>
            <span class="p">]</span>

        <span class="k">def</span> <span class="nf">check_sibs</span><span class="p">(</span><span class="n">children</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="nb">bool</span><span class="p">:</span>
<span class="w">            </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">            Confirm that all children of a parent pair are siblings with each other.</span>

<span class="sd">            If there are multiple children for a given parent pair, all children should be siblings with each other.</span>

<span class="sd">            :param children: List of all children for a given parent pair</span>
<span class="sd">            :return: Whether all children in the list are siblings</span>
<span class="sd">            &quot;&quot;&quot;</span>
            <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">children</span><span class="p">)):</span>
                <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">children</span><span class="p">)):</span>
                    <span class="k">if</span> <span class="p">(</span>
                        <span class="n">related_pairs</span><span class="p">[</span><span class="nb">tuple</span><span class="p">(</span><span class="nb">sorted</span><span class="p">([</span><span class="n">children</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">children</span><span class="p">[</span><span class="n">j</span><span class="p">]]))]</span>
                        <span class="o">!=</span> <span class="n">SIBLINGS</span>
                    <span class="p">):</span>
                        <span class="k">return</span> <span class="kc">False</span>
            <span class="k">return</span> <span class="kc">True</span>

        <span class="k">def</span> <span class="nf">discard_multi_parents_children</span><span class="p">(</span><span class="n">trios</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">hl</span><span class="o">.</span><span class="n">Trio</span><span class="p">]):</span>
<span class="w">            </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">            Check that each child was only assigned a single pair of parents.</span>

<span class="sd">            If a child is found to have multiple parent pairs, they are ALL discarded.</span>

<span class="sd">            :param trios: All trios formed for this family</span>
<span class="sd">            :return: The list of trios for which each child has a single parents pair.</span>
<span class="sd">            &quot;&quot;&quot;</span>
            <span class="n">children_trios</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">list</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">trio</span> <span class="ow">in</span> <span class="n">trios</span><span class="p">:</span>
                <span class="n">children_trios</span><span class="p">[</span><span class="n">trio</span><span class="o">.</span><span class="n">s</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">trio</span><span class="p">)</span>

            <span class="k">for</span> <span class="n">s</span><span class="p">,</span> <span class="n">s_trios</span> <span class="ow">in</span> <span class="n">children_trios</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">s_trios</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
                    <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
                        <span class="s2">&quot;Discarded duplicated child </span><span class="si">%s</span><span class="s2"> found multiple in trios: </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span>
                        <span class="n">s</span><span class="p">,</span>
                        <span class="s2">&quot;, &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="nb">str</span><span class="p">(</span><span class="n">trio</span><span class="p">)</span> <span class="k">for</span> <span class="n">trio</span> <span class="ow">in</span> <span class="n">s_trios</span><span class="p">]),</span>
                    <span class="p">)</span>

            <span class="k">return</span> <span class="p">[</span><span class="n">trios</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">trios</span> <span class="ow">in</span> <span class="n">children_trios</span><span class="o">.</span><span class="n">values</span><span class="p">()</span> <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">trios</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">]</span>

        <span class="c1"># Get all possible pairs of parents in (father, mother) order</span>
        <span class="n">all_possible_parents</span> <span class="o">=</span> <span class="n">get_possible_parents</span><span class="p">(</span>
            <span class="nb">list</span><span class="p">({</span><span class="n">s</span> <span class="k">for</span> <span class="n">pair</span> <span class="ow">in</span> <span class="n">parent_child_pairs</span> <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">pair</span><span class="p">})</span>
        <span class="p">)</span>

        <span class="n">trios</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">possible_parents</span> <span class="ow">in</span> <span class="n">all_possible_parents</span><span class="p">:</span>
            <span class="n">children</span> <span class="o">=</span> <span class="n">get_children</span><span class="p">(</span><span class="n">possible_parents</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">check_sibs</span><span class="p">(</span><span class="n">children</span><span class="p">):</span>
                <span class="n">trios</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span>
                    <span class="p">[</span>
                        <span class="n">hl</span><span class="o">.</span><span class="n">Trio</span><span class="p">(</span>
                            <span class="n">s</span><span class="o">=</span><span class="n">s</span><span class="p">,</span>
                            <span class="n">fam_id</span><span class="o">=</span><span class="n">fam_id</span><span class="p">,</span>
                            <span class="n">pat_id</span><span class="o">=</span><span class="n">possible_parents</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
                            <span class="n">mat_id</span><span class="o">=</span><span class="n">possible_parents</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span>
                            <span class="n">is_female</span><span class="o">=</span><span class="n">sex</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">s</span><span class="p">),</span>
                        <span class="p">)</span>
                        <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">children</span>
                    <span class="p">]</span>
                <span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
                    <span class="s2">&quot;Discarded family with same parents, and multiple offspring that&quot;</span>
                    <span class="s2">&quot; weren&#39;t siblings:</span><span class="se">\n</span><span class="s2">Mother: </span><span class="si">%s</span><span class="se">\n</span><span class="s2">Father:</span><span class="si">%s</span><span class="se">\n</span><span class="s2">Children:</span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span>
                    <span class="n">possible_parents</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
                    <span class="n">possible_parents</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span>
                    <span class="s2">&quot;, &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">children</span><span class="p">),</span>
                <span class="p">)</span>

        <span class="k">return</span> <span class="n">discard_multi_parents_children</span><span class="p">(</span><span class="n">trios</span><span class="p">)</span>

    <span class="c1"># Get all the relations we care about:</span>
    <span class="c1"># =&gt; Remove unrelateds and duplicates</span>
    <span class="n">dups</span> <span class="o">=</span> <span class="n">duplicate_samples_ht</span><span class="o">.</span><span class="n">aggregate</span><span class="p">(</span>
        <span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">explode</span><span class="p">(</span>
            <span class="k">lambda</span> <span class="n">dup</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">collect_as_set</span><span class="p">(</span><span class="n">dup</span><span class="p">),</span> <span class="n">duplicate_samples_ht</span><span class="o">.</span><span class="n">filtered</span>
        <span class="p">),</span>
        <span class="n">_localize</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">relationship_ht</span> <span class="o">=</span> <span class="n">relationship_ht</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
        <span class="o">~</span><span class="n">dups</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="n">relationship_ht</span><span class="p">[</span><span class="n">i_col</span><span class="p">])</span>
        <span class="o">&amp;</span> <span class="o">~</span><span class="n">dups</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="n">relationship_ht</span><span class="p">[</span><span class="n">j_col</span><span class="p">])</span>
        <span class="o">&amp;</span> <span class="p">(</span><span class="n">relationship_ht</span><span class="p">[</span><span class="n">relationship_col</span><span class="p">]</span> <span class="o">!=</span> <span class="n">UNRELATED</span><span class="p">)</span>
    <span class="p">)</span>

    <span class="c1"># Check relatedness table format</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">relationship_ht</span><span class="p">[</span><span class="n">i_col</span><span class="p">]</span><span class="o">.</span><span class="n">dtype</span> <span class="o">==</span> <span class="n">relationship_ht</span><span class="p">[</span><span class="n">j_col</span><span class="p">]</span><span class="o">.</span><span class="n">dtype</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span>
            <span class="s2">&quot;i_col and j_col of the relatedness table need to be of the same type.&quot;</span>
        <span class="p">)</span>

    <span class="c1"># If i_col and j_col aren&#39;t str, then convert them</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">relationship_ht</span><span class="p">[</span><span class="n">i_col</span><span class="p">],</span> <span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">StringExpression</span><span class="p">):</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
            <span class="s2">&quot;Pedigrees can only be constructed from string IDs, but your relatedness_ht&quot;</span>
            <span class="s2">&quot; ID column is of type: </span><span class="si">%s</span><span class="s2">. Expression will be converted to string in&quot;</span>
            <span class="s2">&quot; Pedigrees.&quot;</span><span class="p">,</span>
            <span class="n">relationship_ht</span><span class="p">[</span><span class="n">i_col</span><span class="p">]</span><span class="o">.</span><span class="n">dtype</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">relationship_ht</span><span class="p">[</span><span class="n">i_col</span><span class="p">],</span> <span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">StructExpression</span><span class="p">):</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
                <span class="s2">&quot;Struct fields </span><span class="si">%s</span><span class="s2"> will be joined by underscores to use as sample names&quot;</span>
                <span class="s2">&quot; in Pedigree.&quot;</span><span class="p">,</span>
                <span class="nb">list</span><span class="p">(</span><span class="n">relationship_ht</span><span class="p">[</span><span class="n">i_col</span><span class="p">]),</span>
            <span class="p">)</span>
            <span class="n">relationship_ht</span> <span class="o">=</span> <span class="n">relationship_ht</span><span class="o">.</span><span class="n">key_by</span><span class="p">(</span>
                <span class="o">**</span><span class="p">{</span>
                    <span class="n">i_col</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">delimit</span><span class="p">(</span>
                        <span class="n">hl</span><span class="o">.</span><span class="n">array</span><span class="p">(</span>
                            <span class="p">[</span>
                                <span class="n">hl</span><span class="o">.</span><span class="n">str</span><span class="p">(</span><span class="n">relationship_ht</span><span class="p">[</span><span class="n">i_col</span><span class="p">][</span><span class="n">x</span><span class="p">])</span>
                                <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">relationship_ht</span><span class="p">[</span><span class="n">i_col</span><span class="p">]</span>
                            <span class="p">]</span>
                        <span class="p">),</span>
                        <span class="s2">&quot;_&quot;</span><span class="p">,</span>
                    <span class="p">),</span>
                    <span class="n">j_col</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">delimit</span><span class="p">(</span>
                        <span class="n">hl</span><span class="o">.</span><span class="n">array</span><span class="p">(</span>
                            <span class="p">[</span>
                                <span class="n">hl</span><span class="o">.</span><span class="n">str</span><span class="p">(</span><span class="n">relationship_ht</span><span class="p">[</span><span class="n">j_col</span><span class="p">][</span><span class="n">x</span><span class="p">])</span>
                                <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">relationship_ht</span><span class="p">[</span><span class="n">j_col</span><span class="p">]</span>
                            <span class="p">]</span>
                        <span class="p">),</span>
                        <span class="s2">&quot;_&quot;</span><span class="p">,</span>
                    <span class="p">),</span>
                <span class="p">}</span>
            <span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span>
                <span class="s2">&quot;The `i_col` and `j_col` columns of the `relationship_ht` argument&quot;</span>
                <span class="s2">&quot; passed to infer_families are not of type StringExpression or Struct.&quot;</span>
            <span class="p">)</span>

    <span class="c1"># If sex is a Table, extract sex information as a Dict</span>
    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">sex</span><span class="p">,</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">):</span>
        <span class="n">sex</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">tuple</span><span class="p">([</span><span class="n">sex</span><span class="o">.</span><span class="n">s</span><span class="p">,</span> <span class="n">sex</span><span class="o">.</span><span class="n">is_xx</span><span class="p">])</span><span class="o">.</span><span class="n">collect</span><span class="p">())</span>

    <span class="c1"># Collect all related sample pairs and</span>
    <span class="c1"># create a dictionnary with pairs as keys and relationships as values</span>
    <span class="c1"># Sample-pairs are tuples ordered by sample name</span>
    <span class="n">related_pairs</span> <span class="o">=</span> <span class="p">{</span>
        <span class="nb">tuple</span><span class="p">(</span><span class="nb">sorted</span><span class="p">([</span><span class="n">i</span><span class="p">,</span> <span class="n">j</span><span class="p">])):</span> <span class="n">rel</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">j</span><span class="p">,</span> <span class="n">rel</span> <span class="ow">in</span> <span class="n">hl</span><span class="o">.</span><span class="n">tuple</span><span class="p">(</span>
            <span class="p">[</span><span class="n">relationship_ht</span><span class="o">.</span><span class="n">i</span><span class="p">,</span> <span class="n">relationship_ht</span><span class="o">.</span><span class="n">j</span><span class="p">,</span> <span class="n">relationship_ht</span><span class="o">.</span><span class="n">relationship</span><span class="p">]</span>
        <span class="p">)</span><span class="o">.</span><span class="n">collect</span><span class="p">()</span>
    <span class="p">}</span>

    <span class="n">parent_child_pairs_by_fam</span> <span class="o">=</span> <span class="n">group_parent_child_pairs_by_fam</span><span class="p">(</span>
        <span class="p">[</span><span class="n">pair</span> <span class="k">for</span> <span class="n">pair</span><span class="p">,</span> <span class="n">rel</span> <span class="ow">in</span> <span class="n">related_pairs</span><span class="o">.</span><span class="n">items</span><span class="p">()</span> <span class="k">if</span> <span class="n">rel</span> <span class="o">==</span> <span class="n">PARENT_CHILD</span><span class="p">]</span>
    <span class="p">)</span>
    <span class="k">return</span> <span class="n">hl</span><span class="o">.</span><span class="n">Pedigree</span><span class="p">(</span>
        <span class="p">[</span>
            <span class="n">trio</span>
            <span class="k">for</span> <span class="n">fam_index</span><span class="p">,</span> <span class="n">parent_child_pairs</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">parent_child_pairs_by_fam</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">trio</span> <span class="ow">in</span> <span class="n">get_trios</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">fam_index</span><span class="p">),</span> <span class="n">parent_child_pairs</span><span class="p">,</span> <span class="n">related_pairs</span><span class="p">)</span>
        <span class="p">]</span>
    <span class="p">)</span></div>


<div class="viewcode-block" id="create_fake_pedigree"><a class="viewcode-back" href="../../../api_reference/sample_qc/relatedness.html#gnomad.sample_qc.relatedness.create_fake_pedigree">[docs]</a><span class="k">def</span> <span class="nf">create_fake_pedigree</span><span class="p">(</span>
    <span class="n">n</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
    <span class="n">sample_list</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span>
    <span class="n">exclude_real_probands</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="n">max_tries</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">10</span><span class="p">,</span>
    <span class="n">real_pedigree</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">hl</span><span class="o">.</span><span class="n">Pedigree</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">sample_list_stratification</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">hl</span><span class="o">.</span><span class="n">Pedigree</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Generate a pedigree made of trios created by sampling 3 random samples in the sample list.</span>

<span class="sd">    - If `real_pedigree` is given, then children in the resulting fake trios will not</span>
<span class="sd">      include any trio with proband - parents that are in the real ones.</span>
<span class="sd">    - Each sample can be used only once as a proband in the resulting trios.</span>
<span class="sd">    - Sex of probands in fake trios is random.</span>

<span class="sd">    :param n: Number of fake trios desired in the pedigree.</span>
<span class="sd">    :param sample_list: List of samples.</span>
<span class="sd">    :param exclude_real_probands: If set, then fake trios probands cannot be in the</span>
<span class="sd">        real trios probands.</span>
<span class="sd">    :param max_tries: Maximum number of sampling to try before bailing out (preventing</span>
<span class="sd">        infinite loop if `n` is too large w.r.t. the number of samples).</span>
<span class="sd">    :param real_pedigree: Optional pedigree to exclude children from.</span>
<span class="sd">    :param sample_list_stratification: Optional dictionary with samples as keys and</span>
<span class="sd">        a value that should be used to stratify samples in `sample_list` into groups</span>
<span class="sd">        that the trio should be picked from. This ensures that each fake trio will</span>
<span class="sd">        contain samples from only the same stratification. For example, if all samples</span>
<span class="sd">        within a fake trio should be chosen from the same platform, this can be a</span>
<span class="sd">        dictionary of sample: platform.</span>
<span class="sd">    :return: Fake pedigree.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">real_trios</span> <span class="o">=</span> <span class="p">(</span>
        <span class="p">{</span><span class="n">trio</span><span class="o">.</span><span class="n">s</span><span class="p">:</span> <span class="n">trio</span> <span class="k">for</span> <span class="n">trio</span> <span class="ow">in</span> <span class="n">real_pedigree</span><span class="o">.</span><span class="n">trios</span><span class="p">}</span>
        <span class="k">if</span> <span class="n">real_pedigree</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span>
        <span class="k">else</span> <span class="nb">dict</span><span class="p">()</span>
    <span class="p">)</span>

    <span class="k">if</span> <span class="n">sample_list_stratification</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">sample_list_stratified</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">list</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">sample_list</span><span class="p">:</span>
            <span class="n">s_strata</span> <span class="o">=</span> <span class="n">sample_list_stratification</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">s</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">s_strata</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;Sample </span><span class="si">{</span><span class="n">s</span><span class="si">}</span><span class="s2"> not found in &#39;sample_list_stratification&#39; dict!&quot;</span>
                <span class="p">)</span>
            <span class="n">sample_list_stratified</span><span class="p">[</span><span class="n">s_strata</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">s</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">sample_list_stratified</span> <span class="o">=</span> <span class="kc">None</span>

    <span class="k">if</span> <span class="n">exclude_real_probands</span> <span class="ow">and</span> <span class="nb">len</span><span class="p">(</span><span class="n">real_trios</span><span class="p">)</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">sample_list</span><span class="p">)):</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
            <span class="s2">&quot;All samples are in the real probands list; cannot create any fake&quot;</span>
            <span class="s2">&quot; pedigrees with exclude_real_probands=True. Returning an empty Pedigree.&quot;</span>
        <span class="p">)</span>
        <span class="k">return</span> <span class="n">hl</span><span class="o">.</span><span class="n">Pedigree</span><span class="p">([])</span>

    <span class="n">fake_trios</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">tries</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">while</span> <span class="nb">len</span><span class="p">(</span><span class="n">fake_trios</span><span class="p">)</span> <span class="o">&lt;</span> <span class="n">n</span> <span class="ow">and</span> <span class="n">tries</span> <span class="o">&lt;</span> <span class="n">max_tries</span><span class="p">:</span>
        <span class="n">s</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">sample_list</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">sample_list_stratified</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">curr_sample_list</span> <span class="o">=</span> <span class="n">sample_list</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">curr_sample_list</span> <span class="o">=</span> <span class="n">sample_list_stratified</span><span class="p">[</span><span class="n">sample_list_stratification</span><span class="p">[</span><span class="n">s</span><span class="p">]]</span>

        <span class="n">mat_id</span><span class="p">,</span> <span class="n">pat_id</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">sample</span><span class="p">(</span><span class="n">curr_sample_list</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span>
        <span class="k">if</span> <span class="p">(</span>
            <span class="n">s</span> <span class="ow">in</span> <span class="n">real_trios</span>
            <span class="ow">and</span> <span class="p">(</span>
                <span class="n">exclude_real_probands</span>
                <span class="ow">or</span> <span class="p">{</span><span class="n">mat_id</span><span class="p">,</span> <span class="n">pat_id</span><span class="p">}</span> <span class="o">==</span> <span class="p">{</span><span class="n">real_trios</span><span class="p">[</span><span class="n">s</span><span class="p">]</span><span class="o">.</span><span class="n">mat_id</span><span class="p">,</span> <span class="n">real_trios</span><span class="p">[</span><span class="n">s</span><span class="p">]</span><span class="o">.</span><span class="n">pat_id</span><span class="p">}</span>
            <span class="p">)</span>
        <span class="p">)</span> <span class="ow">or</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">fake_trios</span><span class="p">:</span>
            <span class="n">tries</span> <span class="o">+=</span> <span class="mi">1</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">tries</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="n">fake_trios</span><span class="p">[</span><span class="n">s</span><span class="p">]</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">Trio</span><span class="p">(</span>
                <span class="n">s</span><span class="o">=</span><span class="n">s</span><span class="p">,</span>
                <span class="n">pat_id</span><span class="o">=</span><span class="n">pat_id</span><span class="p">,</span>
                <span class="n">mat_id</span><span class="o">=</span><span class="n">mat_id</span><span class="p">,</span>
                <span class="n">fam_id</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;fake_</span><span class="si">{</span><span class="nb">str</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">fake_trios</span><span class="p">))</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span>
                <span class="n">is_female</span><span class="o">=</span><span class="nb">bool</span><span class="p">(</span><span class="n">random</span><span class="o">.</span><span class="n">getrandbits</span><span class="p">(</span><span class="mi">1</span><span class="p">)),</span>
            <span class="p">)</span>

    <span class="k">if</span> <span class="n">tries</span> <span class="o">==</span> <span class="n">max_tries</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
            <span class="s2">&quot;Only returning </span><span class="si">%d</span><span class="s2"> fake trios; random trio sampling stopped after reaching&quot;</span>
            <span class="s2">&quot; the maximum </span><span class="si">%d</span><span class="s2"> iterations&quot;</span><span class="p">,</span>
            <span class="nb">len</span><span class="p">(</span><span class="n">fake_trios</span><span class="p">),</span>
            <span class="n">max_tries</span><span class="p">,</span>
        <span class="p">)</span>

    <span class="k">return</span> <span class="n">hl</span><span class="o">.</span><span class="n">Pedigree</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">fake_trios</span><span class="o">.</span><span class="n">values</span><span class="p">()))</span></div>


<div class="viewcode-block" id="compute_related_samples_to_drop"><a class="viewcode-back" href="../../../api_reference/sample_qc/relatedness.html#gnomad.sample_qc.relatedness.compute_related_samples_to_drop">[docs]</a><span class="k">def</span> <span class="nf">compute_related_samples_to_drop</span><span class="p">(</span>
    <span class="n">relatedness_ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span>
    <span class="n">rank_ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span>
    <span class="n">kin_threshold</span><span class="p">:</span> <span class="nb">float</span><span class="p">,</span>
    <span class="n">filtered_samples</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">SetExpression</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">min_related_hard_filter</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">keep_samples</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">SetExpression</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">keep_samples_when_related</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Compute a Table with the list of samples to drop (and their global rank) to get the maximal independent set of unrelated samples.</span>

<span class="sd">    .. note::</span>

<span class="sd">        - `relatedness_ht` should be keyed by exactly two fields of the same type, identifying the pair of samples for each row.</span>
<span class="sd">        - `rank_ht` should be keyed by a single key of the same type as a single sample identifier in `relatedness_ht`.</span>

<span class="sd">    :param relatedness_ht: relatedness HT, as produced by e.g. pc-relate</span>
<span class="sd">    :param kin_threshold: Kinship threshold to consider two samples as related</span>
<span class="sd">    :param rank_ht: Table with a global rank for each sample (smaller is preferred)</span>
<span class="sd">    :param filtered_samples: An optional set of samples to exclude (e.g. these samples were hard-filtered)  These samples will then appear in the resulting samples to drop.</span>
<span class="sd">    :param min_related_hard_filter: If provided, any sample that is related to more samples than this parameter will be filtered prior to computing the maximal independent set and appear in the results.</span>
<span class="sd">    :param keep_samples: An optional set of samples that must be kept. An error is raised (when `keep_samples_when_related` is False) if any two samples in the list are among the related pairs.</span>
<span class="sd">    :param keep_samples_when_related: Don&#39;t raise an error if `keep_samples` contains related samples, and keep related samples. Default is False.</span>
<span class="sd">    :return: A Table with the list of the samples to drop along with their rank.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Make sure that the key types are valid</span>
    <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">relatedness_ht</span><span class="o">.</span><span class="n">key</span><span class="p">))</span> <span class="o">==</span> <span class="mi">2</span>
    <span class="k">assert</span> <span class="n">relatedness_ht</span><span class="o">.</span><span class="n">key</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">dtype</span> <span class="o">==</span> <span class="n">relatedness_ht</span><span class="o">.</span><span class="n">key</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">dtype</span>
    <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">rank_ht</span><span class="o">.</span><span class="n">key</span><span class="p">))</span> <span class="o">==</span> <span class="mi">1</span>
    <span class="k">assert</span> <span class="n">relatedness_ht</span><span class="o">.</span><span class="n">key</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">dtype</span> <span class="o">==</span> <span class="n">rank_ht</span><span class="o">.</span><span class="n">key</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">dtype</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Filtering related samples using a kin threshold of </span><span class="si">%f</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">kin_threshold</span><span class="p">)</span>
    <span class="n">relatedness_ht</span> <span class="o">=</span> <span class="n">relatedness_ht</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="n">relatedness_ht</span><span class="o">.</span><span class="n">kin</span> <span class="o">&gt;</span> <span class="n">kin_threshold</span><span class="p">)</span>

    <span class="n">filtered_samples_rel</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
    <span class="k">if</span> <span class="n">min_related_hard_filter</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
            <span class="s2">&quot;Computing samples related to too many individuals (&gt;</span><span class="si">%d</span><span class="s2">) for exclusion&quot;</span><span class="p">,</span>
            <span class="n">min_related_hard_filter</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="n">gbi</span> <span class="o">=</span> <span class="n">relatedness_ht</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span><span class="n">s</span><span class="o">=</span><span class="p">[</span><span class="n">relatedness_ht</span><span class="o">.</span><span class="n">key</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">relatedness_ht</span><span class="o">.</span><span class="n">key</span><span class="p">[</span><span class="mi">1</span><span class="p">]])</span>
        <span class="n">gbi</span> <span class="o">=</span> <span class="n">gbi</span><span class="o">.</span><span class="n">explode</span><span class="p">(</span><span class="n">gbi</span><span class="o">.</span><span class="n">s</span><span class="p">)</span>
        <span class="n">gbi</span> <span class="o">=</span> <span class="n">gbi</span><span class="o">.</span><span class="n">group_by</span><span class="p">(</span><span class="n">gbi</span><span class="o">.</span><span class="n">s</span><span class="p">)</span><span class="o">.</span><span class="n">aggregate</span><span class="p">(</span><span class="n">n</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">count</span><span class="p">())</span>
        <span class="n">filtered_samples_rel</span> <span class="o">=</span> <span class="n">gbi</span><span class="o">.</span><span class="n">aggregate</span><span class="p">(</span>
            <span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="n">gbi</span><span class="o">.</span><span class="n">n</span> <span class="o">&gt;</span> <span class="n">min_related_hard_filter</span><span class="p">,</span> <span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">collect_as_set</span><span class="p">(</span><span class="n">gbi</span><span class="o">.</span><span class="n">s</span><span class="p">))</span>
        <span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
            <span class="s2">&quot;Found </span><span class="si">%d</span><span class="s2"> samples with too many 1st/2nd degree relatives. These samples&quot;</span>
            <span class="s2">&quot; will be excluded.&quot;</span><span class="p">,</span>
            <span class="nb">len</span><span class="p">(</span><span class="n">filtered_samples_rel</span><span class="p">),</span>
        <span class="p">)</span>

    <span class="k">if</span> <span class="n">filtered_samples</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">filtered_samples_rel</span> <span class="o">=</span> <span class="n">filtered_samples_rel</span><span class="o">.</span><span class="n">union</span><span class="p">(</span>
            <span class="n">relatedness_ht</span><span class="o">.</span><span class="n">aggregate</span><span class="p">(</span>
                <span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">explode</span><span class="p">(</span>
                    <span class="k">lambda</span> <span class="n">s</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">collect_as_set</span><span class="p">(</span><span class="n">s</span><span class="p">),</span>
                    <span class="n">hl</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="n">relatedness_ht</span><span class="o">.</span><span class="n">key</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">relatedness_ht</span><span class="o">.</span><span class="n">key</span><span class="p">[</span><span class="mi">1</span><span class="p">]])</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
                        <span class="k">lambda</span> <span class="n">s</span><span class="p">:</span> <span class="n">filtered_samples</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="n">s</span><span class="p">)</span>
                    <span class="p">),</span>
                <span class="p">)</span>
            <span class="p">)</span>
        <span class="p">)</span>

    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">filtered_samples_rel</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">filtered_samples_lit</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">literal</span><span class="p">(</span><span class="n">filtered_samples_rel</span><span class="p">)</span>
        <span class="n">relatedness_ht</span> <span class="o">=</span> <span class="n">relatedness_ht</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
            <span class="n">filtered_samples_lit</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="n">relatedness_ht</span><span class="o">.</span><span class="n">key</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
            <span class="o">|</span> <span class="n">filtered_samples_lit</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="n">relatedness_ht</span><span class="o">.</span><span class="n">key</span><span class="p">[</span><span class="mi">1</span><span class="p">]),</span>
            <span class="n">keep</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="p">)</span>

    <span class="k">if</span> <span class="n">keep_samples</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
            <span class="s2">&quot;Number of samples in the provided list of samples to keep: </span><span class="si">%d</span><span class="s2">&quot;</span><span class="p">,</span>
            <span class="n">hl</span><span class="o">.</span><span class="n">eval</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">len</span><span class="p">(</span><span class="n">keep_samples</span><span class="p">)),</span>
        <span class="p">)</span>
        <span class="n">i</span> <span class="o">=</span> <span class="n">relatedness_ht</span><span class="o">.</span><span class="n">key</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">j</span> <span class="o">=</span> <span class="n">relatedness_ht</span><span class="o">.</span><span class="n">key</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
        <span class="n">keep_samples_rel</span> <span class="o">=</span> <span class="n">relatedness_ht</span><span class="o">.</span><span class="n">aggregate</span><span class="p">(</span>
            <span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
                <span class="n">keep_samples</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="o">&amp;</span> <span class="n">keep_samples</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="n">j</span><span class="p">),</span>
                <span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">collect_as_set</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span><span class="n">i</span><span class="o">=</span><span class="n">i</span><span class="p">,</span> <span class="n">j</span><span class="o">=</span><span class="n">j</span><span class="p">,</span> <span class="n">kin</span><span class="o">=</span><span class="n">relatedness_ht</span><span class="o">.</span><span class="n">kin</span><span class="p">)),</span>
            <span class="p">)</span>
        <span class="p">)</span>
        <span class="n">num_keep_samples_rel</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">keep_samples_rel</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">num_keep_samples_rel</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
                <span class="s2">&quot;The following pairs are in the list of samples to keep, but are &quot;</span>
                <span class="s2">&quot;related:</span><span class="se">\n</span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span>
                <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">map</span><span class="p">(</span><span class="nb">str</span><span class="p">,</span> <span class="n">keep_samples_rel</span><span class="p">)),</span>
            <span class="p">)</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">keep_samples_when_related</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                    <span class="s2">&quot;Both samples can not be retained when &#39;keep_samples_when_related&#39; &quot;</span>
                    <span class="s2">&quot;is False!&quot;</span>
                <span class="p">)</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Annotating related sample pairs with rank.&quot;</span><span class="p">)</span>
    <span class="n">i</span><span class="p">,</span> <span class="n">j</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">relatedness_ht</span><span class="o">.</span><span class="n">key</span><span class="p">)</span>
    <span class="n">relatedness_ht</span> <span class="o">=</span> <span class="n">relatedness_ht</span><span class="o">.</span><span class="n">key_by</span><span class="p">(</span><span class="n">s</span><span class="o">=</span><span class="n">relatedness_ht</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>
    <span class="n">relatedness_ht</span> <span class="o">=</span> <span class="n">relatedness_ht</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span>
        <span class="o">**</span><span class="p">{</span><span class="n">i</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span><span class="n">s</span><span class="o">=</span><span class="n">relatedness_ht</span><span class="o">.</span><span class="n">s</span><span class="p">,</span> <span class="n">rank</span><span class="o">=</span><span class="n">rank_ht</span><span class="p">[</span><span class="n">relatedness_ht</span><span class="o">.</span><span class="n">key</span><span class="p">]</span><span class="o">.</span><span class="n">rank</span><span class="p">)}</span>
    <span class="p">)</span>
    <span class="n">relatedness_ht</span> <span class="o">=</span> <span class="n">relatedness_ht</span><span class="o">.</span><span class="n">key_by</span><span class="p">(</span><span class="n">s</span><span class="o">=</span><span class="n">relatedness_ht</span><span class="p">[</span><span class="n">j</span><span class="p">])</span>
    <span class="n">relatedness_ht</span> <span class="o">=</span> <span class="n">relatedness_ht</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span>
        <span class="o">**</span><span class="p">{</span><span class="n">j</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span><span class="n">s</span><span class="o">=</span><span class="n">relatedness_ht</span><span class="o">.</span><span class="n">s</span><span class="p">,</span> <span class="n">rank</span><span class="o">=</span><span class="n">rank_ht</span><span class="p">[</span><span class="n">relatedness_ht</span><span class="o">.</span><span class="n">key</span><span class="p">]</span><span class="o">.</span><span class="n">rank</span><span class="p">)}</span>
    <span class="p">)</span>
    <span class="n">relatedness_ht</span> <span class="o">=</span> <span class="n">relatedness_ht</span><span class="o">.</span><span class="n">key_by</span><span class="p">(</span><span class="n">i</span><span class="p">,</span> <span class="n">j</span><span class="p">)</span>
    <span class="n">relatedness_ht</span> <span class="o">=</span> <span class="n">relatedness_ht</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s2">&quot;s&quot;</span><span class="p">)</span>
    <span class="n">relatedness_ht</span> <span class="o">=</span> <span class="n">relatedness_ht</span><span class="o">.</span><span class="n">persist</span><span class="p">()</span>

    <span class="k">def</span> <span class="nf">maximal_independent_set_keep_samples</span><span class="p">(</span>
        <span class="n">pair_graph</span><span class="p">:</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">,</span>
        <span class="n">keep</span><span class="p">:</span> <span class="nb">set</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(),</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Find maximal independent set with the ability to retain specific samples.</span>

<span class="sd">        Results will be the same as Hail&#39;s maximal_independent_set when `keep`</span>
<span class="sd">        is empty. Otherwise, `keep` will always be kept, while still being</span>
<span class="sd">        included in the graph to compute the degree (number of samples each sample is</span>
<span class="sd">        related to) of each node (sample).</span>

<span class="sd">        :param pair_graph: A networkx graph of related sample pairs.</span>
<span class="sd">        :param keep: Set of samples that must be kept even when the degree</span>
<span class="sd">            (number of total samples this sample is related to) is highest and would</span>
<span class="sd">            typically be removed to get the maximal independent set.</span>
<span class="sd">        :return: List of related samples to drop.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">drop_samples</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="c1"># Get a list of all connected components in pair_graph.</span>
        <span class="c1"># This is a list of sample sets where each set contains samples that are</span>
        <span class="c1"># connected in the graph and share no connections to samples in any other set.</span>
        <span class="n">connected_samples</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">nx</span><span class="o">.</span><span class="n">connected_components</span><span class="p">(</span><span class="n">pair_graph</span><span class="p">))</span>
        <span class="c1"># For each of the connected components, determine the sample with the largest</span>
        <span class="c1"># degree (number of samples it is related to), followed by highest rank, that</span>
        <span class="c1"># is not in keep.</span>
        <span class="c1"># Add this sample to drop_samples and then compute the subgraph with this</span>
        <span class="c1"># sample excluded. Repeat the process on this subgraph. Continue until all</span>
        <span class="c1"># connected components contain only a single sample, unless connected samples</span>
        <span class="c1"># are in `keep`.</span>
        <span class="k">for</span> <span class="n">con</span> <span class="ow">in</span> <span class="n">connected_samples</span><span class="p">:</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">con</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
                <span class="c1"># Build a list of (degree, rank, sample) sorted by highest degree and</span>
                <span class="c1"># then highest rank.</span>
                <span class="n">degree_rank_list</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span>
                    <span class="p">[(</span><span class="n">pair_graph</span><span class="o">.</span><span class="n">degree</span><span class="p">[</span><span class="n">s</span><span class="p">],</span> <span class="n">s</span><span class="o">.</span><span class="n">rank</span><span class="p">,</span> <span class="n">s</span><span class="p">)</span> <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">con</span><span class="p">],</span>
                    <span class="n">reverse</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                <span class="p">)</span>
                <span class="c1"># Get the sample with the highest degree and highest rank not in keep.</span>
                <span class="n">highest_degree_s</span> <span class="o">=</span> <span class="n">degree_rank_list</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="mi">0</span><span class="p">)[</span><span class="mi">2</span><span class="p">]</span>
                <span class="k">while</span> <span class="n">highest_degree_s</span><span class="o">.</span><span class="n">s</span> <span class="ow">in</span> <span class="n">keep</span> <span class="ow">and</span> <span class="nb">len</span><span class="p">(</span><span class="n">degree_rank_list</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                    <span class="n">highest_degree_s</span> <span class="o">=</span> <span class="n">degree_rank_list</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="mi">0</span><span class="p">)[</span><span class="mi">2</span><span class="p">]</span>
                <span class="k">if</span> <span class="n">highest_degree_s</span><span class="o">.</span><span class="n">s</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">keep</span><span class="p">:</span>
                    <span class="c1"># Add sample with the highest degree and highest rank not in keep</span>
                    <span class="c1"># to the list of samples to drop.</span>
                    <span class="n">drop_samples</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">highest_degree_s</span><span class="p">)</span>
                    <span class="c1"># Remove that sample from the samples in the current connected</span>
                    <span class="c1"># component and if there is more than one sample left, get the</span>
                    <span class="c1"># subgraph containing those samples and repeat this process.</span>
                    <span class="n">new_con</span> <span class="o">=</span> <span class="n">con</span> <span class="o">-</span> <span class="p">{</span><span class="n">highest_degree_s</span><span class="p">}</span>
                    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">new_con</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
                        <span class="n">drop_samples</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span>
                            <span class="n">maximal_independent_set_keep_samples</span><span class="p">(</span>
                                <span class="n">pair_graph</span><span class="o">.</span><span class="n">subgraph</span><span class="p">(</span><span class="n">new_con</span><span class="p">),</span>
                                <span class="n">keep</span><span class="o">=</span><span class="n">keep</span><span class="p">,</span>
                            <span class="p">)</span>
                        <span class="p">)</span>

        <span class="k">return</span> <span class="n">drop_samples</span>

    <span class="k">if</span> <span class="n">keep_samples</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">related_samples_to_drop_ht</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">maximal_independent_set</span><span class="p">(</span>
            <span class="n">relatedness_ht</span><span class="p">[</span><span class="n">i</span><span class="p">],</span>
            <span class="n">relatedness_ht</span><span class="p">[</span><span class="n">j</span><span class="p">],</span>
            <span class="n">keep</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
            <span class="n">tie_breaker</span><span class="o">=</span><span class="k">lambda</span> <span class="n">l</span><span class="p">,</span> <span class="n">r</span><span class="p">:</span> <span class="n">l</span><span class="o">.</span><span class="n">rank</span> <span class="o">-</span> <span class="n">r</span><span class="o">.</span><span class="n">rank</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="n">related_samples_to_drop_ht</span> <span class="o">=</span> <span class="n">related_samples_to_drop_ht</span><span class="o">.</span><span class="n">key_by</span><span class="p">()</span>
        <span class="n">related_samples_to_drop_ht</span> <span class="o">=</span> <span class="n">related_samples_to_drop_ht</span><span class="o">.</span><span class="n">select</span><span class="p">(</span>
            <span class="o">**</span><span class="n">related_samples_to_drop_ht</span><span class="o">.</span><span class="n">node</span>
        <span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">related_pair_graph</span> <span class="o">=</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">()</span>
        <span class="n">related_pair_graph</span><span class="o">.</span><span class="n">add_edges_from</span><span class="p">(</span>
            <span class="nb">list</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">relatedness_ht</span><span class="o">.</span><span class="n">i</span><span class="o">.</span><span class="n">collect</span><span class="p">(),</span> <span class="n">relatedness_ht</span><span class="o">.</span><span class="n">j</span><span class="o">.</span><span class="n">collect</span><span class="p">()))</span>
        <span class="p">)</span>
        <span class="n">related_samples_to_drop_ht</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="o">.</span><span class="n">parallelize</span><span class="p">(</span>
            <span class="n">maximal_independent_set_keep_samples</span><span class="p">(</span>
                <span class="n">related_pair_graph</span><span class="p">,</span> <span class="n">keep</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">eval</span><span class="p">(</span><span class="n">keep_samples</span><span class="p">)</span>
            <span class="p">),</span>
            <span class="n">hl</span><span class="o">.</span><span class="n">tstruct</span><span class="p">(</span><span class="n">s</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">tstr</span><span class="p">,</span> <span class="n">rank</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">tint64</span><span class="p">),</span>
        <span class="p">)</span>
    <span class="n">related_samples_to_drop_ht</span> <span class="o">=</span> <span class="n">related_samples_to_drop_ht</span><span class="o">.</span><span class="n">key_by</span><span class="p">(</span><span class="s2">&quot;s&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">filtered_samples_rel</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">related_samples_to_drop_ht</span> <span class="o">=</span> <span class="n">related_samples_to_drop_ht</span><span class="o">.</span><span class="n">union</span><span class="p">(</span>
            <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="o">.</span><span class="n">parallelize</span><span class="p">(</span>
                <span class="p">[</span><span class="n">hl</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span><span class="n">s</span><span class="o">=</span><span class="n">s</span><span class="p">,</span> <span class="n">rank</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">null</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">tint64</span><span class="p">))</span> <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">filtered_samples_rel</span><span class="p">],</span>
                <span class="n">key</span><span class="o">=</span><span class="s2">&quot;s&quot;</span><span class="p">,</span>
            <span class="p">)</span>
        <span class="p">)</span>

    <span class="k">return</span> <span class="n">related_samples_to_drop_ht</span></div>


<div class="viewcode-block" id="filter_to_trios"><a class="viewcode-back" href="../../../api_reference/sample_qc/relatedness.html#gnomad.sample_qc.relatedness.filter_to_trios">[docs]</a><span class="k">def</span> <span class="nf">filter_to_trios</span><span class="p">(</span>
    <span class="n">mtds</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span> <span class="n">hl</span><span class="o">.</span><span class="n">MatrixTable</span><span class="p">,</span> <span class="n">hl</span><span class="o">.</span><span class="n">vds</span><span class="o">.</span><span class="n">VariantDataset</span><span class="p">],</span> <span class="n">fam_ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Union</span><span class="p">[</span><span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span> <span class="n">hl</span><span class="o">.</span><span class="n">MatrixTable</span><span class="p">,</span> <span class="n">hl</span><span class="o">.</span><span class="n">vds</span><span class="o">.</span><span class="n">VariantDataset</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Filter a Table, MatrixTable or VariantDataset to a set of trios in `fam_ht`.</span>

<span class="sd">    .. note::</span>

<span class="sd">        Using `filter_cols` in MatrixTable will not affect the number of rows</span>
<span class="sd">        (variants), however, using `filter_samples` in VariantDataset will remove the</span>
<span class="sd">        variants that are not present in any of the trios.</span>

<span class="sd">    :param mtds: A Variant Dataset or a Matrix Table or a Table to filter to only trios.</span>
<span class="sd">    :param fam_ht: A Table of trios to filter to, loaded using `hl.import_fam`.</span>
<span class="sd">    :return: A Table, MatrixTable or VariantDataset with only the trios in `fam_ht`.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Filter to samples present in any of the trios.</span>
    <span class="n">fam_ht</span> <span class="o">=</span> <span class="n">fam_ht</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span><span class="n">fam_members</span><span class="o">=</span><span class="p">[</span><span class="n">fam_ht</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="n">fam_ht</span><span class="o">.</span><span class="n">pat_id</span><span class="p">,</span> <span class="n">fam_ht</span><span class="o">.</span><span class="n">mat_id</span><span class="p">])</span>
    <span class="n">fam_ht</span> <span class="o">=</span> <span class="n">fam_ht</span><span class="o">.</span><span class="n">explode</span><span class="p">(</span><span class="s2">&quot;fam_members&quot;</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s2">&quot;s&quot;</span><span class="p">)</span>
    <span class="n">fam_ht</span> <span class="o">=</span> <span class="n">fam_ht</span><span class="o">.</span><span class="n">key_by</span><span class="p">(</span><span class="s2">&quot;s&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">select</span><span class="p">()</span><span class="o">.</span><span class="n">distinct</span><span class="p">()</span>

    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">mtds</span><span class="p">,</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">):</span>
        <span class="n">mtds</span> <span class="o">=</span> <span class="n">mtds</span><span class="o">.</span><span class="n">semi_join</span><span class="p">(</span><span class="n">fam_ht</span><span class="p">)</span>
    <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">mtds</span><span class="p">,</span> <span class="n">hl</span><span class="o">.</span><span class="n">MatrixTable</span><span class="p">):</span>
        <span class="n">mtds</span> <span class="o">=</span> <span class="n">mtds</span><span class="o">.</span><span class="n">filter_cols</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">is_defined</span><span class="p">(</span><span class="n">fam_ht</span><span class="p">[</span><span class="n">mtds</span><span class="o">.</span><span class="n">col_key</span><span class="p">]))</span>
    <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">mtds</span><span class="p">,</span> <span class="n">hl</span><span class="o">.</span><span class="n">vds</span><span class="o">.</span><span class="n">VariantDataset</span><span class="p">):</span>
        <span class="n">mtds</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">vds</span><span class="o">.</span><span class="n">filter_samples</span><span class="p">(</span><span class="n">mtds</span><span class="p">,</span> <span class="n">fam_ht</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;mtds must be a Table, MatrixTable, or VariantDataset&quot;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">mtds</span></div>


<div class="viewcode-block" id="generate_trio_stats_expr"><a class="viewcode-back" href="../../../api_reference/sample_qc/relatedness.html#gnomad.sample_qc.relatedness.generate_trio_stats_expr">[docs]</a><span class="k">def</span> <span class="nf">generate_trio_stats_expr</span><span class="p">(</span>
    <span class="n">trio_mt</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">MatrixTable</span><span class="p">,</span>
    <span class="n">transmitted_strata</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">BooleanExpression</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;raw&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">},</span>
    <span class="n">de_novo_strata</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">BooleanExpression</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;raw&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">},</span>
    <span class="n">ac_strata</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">BooleanExpression</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;raw&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">},</span>
    <span class="n">proband_is_female_expr</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">BooleanExpression</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">StructExpression</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Generate a row-wise expression containing trio transmission stats.</span>

<span class="sd">    The expression will generate the following counts:</span>
<span class="sd">        - Number of alleles in het parents transmitted to the proband</span>
<span class="sd">        - Number of alleles in het parents not transmitted to the proband</span>
<span class="sd">        - Number of de novo mutations</span>
<span class="sd">        - Parent allele count</span>
<span class="sd">        - Proband allele count</span>

<span class="sd">    Transmission and de novo mutation metrics and allele counts can be stratified using additional filters.</span>
<span class="sd">    `transmitted_strata`, `de_novo_strata`, and `ac_strata` all expect a dictionary of filtering expressions keyed</span>
<span class="sd">    by their desired suffix to append for labeling. The default will perform counts using all genotypes and append</span>
<span class="sd">    &#39;raw&#39; to the label.</span>

<span class="sd">    .. note::</span>

<span class="sd">        Expects that `mt` is dense if dealing with a sparse MT `hl.experimental.densify` must be run first.</span>

<span class="sd">    :param trio_mt: A trio standard trio MT (with the format as produced by hail.methods.trio_matrix)</span>
<span class="sd">    :param transmitted_strata: Strata for the transmission counts</span>
<span class="sd">    :param de_novo_strata: Strata for the de novo counts</span>
<span class="sd">    :param ac_strata: Strata for the parent and child allele counts</span>
<span class="sd">    :param proband_is_xx_expr: An optional expression giving the karyotype of the proband (XX=True, XY=False). If not given, DNMs are only computed for autosomes.</span>
<span class="sd">    :return: An expression with the counts</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Create map for transmitted, untransmitted and DNM</span>
    <span class="n">hom_ref</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">het</span> <span class="o">=</span> <span class="mi">1</span>
    <span class="n">hom_var</span> <span class="o">=</span> <span class="mi">2</span>

    <span class="n">auto_or_par</span> <span class="o">=</span> <span class="mi">2</span>
    <span class="n">hemi_x</span> <span class="o">=</span> <span class="mi">1</span>
    <span class="n">hemi_y</span> <span class="o">=</span> <span class="mi">0</span>

    <span class="n">trans_config_counts</span> <span class="o">=</span> <span class="p">{</span>
        <span class="c1"># kid, dad, mom, copy -&gt; t, u</span>
        <span class="p">(</span><span class="n">hom_ref</span><span class="p">,</span> <span class="n">het</span><span class="p">,</span> <span class="n">het</span><span class="p">,</span> <span class="n">auto_or_par</span><span class="p">):</span> <span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">),</span>
        <span class="p">(</span><span class="n">hom_ref</span><span class="p">,</span> <span class="n">hom_ref</span><span class="p">,</span> <span class="n">het</span><span class="p">,</span> <span class="n">auto_or_par</span><span class="p">):</span> <span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">),</span>
        <span class="p">(</span><span class="n">hom_ref</span><span class="p">,</span> <span class="n">het</span><span class="p">,</span> <span class="n">hom_ref</span><span class="p">,</span> <span class="n">auto_or_par</span><span class="p">):</span> <span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">),</span>
        <span class="p">(</span><span class="n">het</span><span class="p">,</span> <span class="n">het</span><span class="p">,</span> <span class="n">het</span><span class="p">,</span> <span class="n">auto_or_par</span><span class="p">):</span> <span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">),</span>
        <span class="p">(</span><span class="n">het</span><span class="p">,</span> <span class="n">hom_ref</span><span class="p">,</span> <span class="n">het</span><span class="p">,</span> <span class="n">auto_or_par</span><span class="p">):</span> <span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">),</span>
        <span class="p">(</span><span class="n">het</span><span class="p">,</span> <span class="n">het</span><span class="p">,</span> <span class="n">hom_ref</span><span class="p">,</span> <span class="n">auto_or_par</span><span class="p">):</span> <span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">),</span>
        <span class="p">(</span><span class="n">het</span><span class="p">,</span> <span class="n">hom_var</span><span class="p">,</span> <span class="n">het</span><span class="p">,</span> <span class="n">auto_or_par</span><span class="p">):</span> <span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">),</span>
        <span class="p">(</span><span class="n">het</span><span class="p">,</span> <span class="n">het</span><span class="p">,</span> <span class="n">hom_var</span><span class="p">,</span> <span class="n">auto_or_par</span><span class="p">):</span> <span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">),</span>
        <span class="p">(</span><span class="n">hom_var</span><span class="p">,</span> <span class="n">het</span><span class="p">,</span> <span class="n">het</span><span class="p">,</span> <span class="n">auto_or_par</span><span class="p">):</span> <span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">0</span><span class="p">),</span>
        <span class="p">(</span><span class="n">hom_var</span><span class="p">,</span> <span class="n">het</span><span class="p">,</span> <span class="n">hom_var</span><span class="p">,</span> <span class="n">auto_or_par</span><span class="p">):</span> <span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">),</span>
        <span class="p">(</span><span class="n">hom_var</span><span class="p">,</span> <span class="n">hom_var</span><span class="p">,</span> <span class="n">het</span><span class="p">,</span> <span class="n">auto_or_par</span><span class="p">):</span> <span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">),</span>
        <span class="p">(</span><span class="n">hom_ref</span><span class="p">,</span> <span class="n">hom_ref</span><span class="p">,</span> <span class="n">het</span><span class="p">,</span> <span class="n">hemi_x</span><span class="p">):</span> <span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">),</span>
        <span class="p">(</span><span class="n">hom_ref</span><span class="p">,</span> <span class="n">hom_var</span><span class="p">,</span> <span class="n">het</span><span class="p">,</span> <span class="n">hemi_x</span><span class="p">):</span> <span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">),</span>
        <span class="p">(</span><span class="n">hom_var</span><span class="p">,</span> <span class="n">hom_ref</span><span class="p">,</span> <span class="n">het</span><span class="p">,</span> <span class="n">hemi_x</span><span class="p">):</span> <span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">),</span>
        <span class="p">(</span><span class="n">hom_var</span><span class="p">,</span> <span class="n">hom_var</span><span class="p">,</span> <span class="n">het</span><span class="p">,</span> <span class="n">hemi_x</span><span class="p">):</span> <span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">),</span>
    <span class="p">}</span>

    <span class="n">trans_count_map</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">literal</span><span class="p">(</span><span class="n">trans_config_counts</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">_get_copy_state</span><span class="p">(</span><span class="n">locus</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">LocusExpression</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">Int32Expression</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Get copy-state int from LocusExpression for indexing into trans_count_map.&quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="p">(</span>
            <span class="n">hl</span><span class="o">.</span><span class="n">case</span><span class="p">()</span>
            <span class="o">.</span><span class="n">when</span><span class="p">(</span><span class="n">locus</span><span class="o">.</span><span class="n">in_autosome_or_par</span><span class="p">(),</span> <span class="n">auto_or_par</span><span class="p">)</span>
            <span class="o">.</span><span class="n">when</span><span class="p">(</span><span class="n">locus</span><span class="o">.</span><span class="n">in_x_nonpar</span><span class="p">(),</span> <span class="n">hemi_x</span><span class="p">)</span>
            <span class="o">.</span><span class="n">when</span><span class="p">(</span><span class="n">locus</span><span class="o">.</span><span class="n">in_y_nonpar</span><span class="p">(),</span> <span class="n">hemi_y</span><span class="p">)</span>
            <span class="o">.</span><span class="n">or_missing</span><span class="p">()</span>
        <span class="p">)</span>

    <span class="k">def</span> <span class="nf">_is_dnm</span><span class="p">(</span>
        <span class="n">proband_gt</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">CallExpression</span><span class="p">,</span>
        <span class="n">father_gt</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">CallExpression</span><span class="p">,</span>
        <span class="n">mother_gt</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">CallExpression</span><span class="p">,</span>
        <span class="n">locus</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">LocusExpression</span><span class="p">,</span>
        <span class="n">proband_is_xx</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">BooleanExpression</span><span class="p">],</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">BooleanExpression</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Determine whether a trio genotype combination is a DNM.&quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">proband_is_xx</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
                <span class="s2">&quot;Since no proband sex expression was given to generate_trio_stats_expr,&quot;</span>
                <span class="s2">&quot; only DNMs in autosomes will be counted.&quot;</span>
            <span class="p">)</span>
            <span class="k">return</span> <span class="n">hl</span><span class="o">.</span><span class="n">or_missing</span><span class="p">(</span>
                <span class="n">locus</span><span class="o">.</span><span class="n">in_autosome</span><span class="p">(),</span>
                <span class="n">proband_gt</span><span class="o">.</span><span class="n">is_het</span><span class="p">()</span> <span class="o">&amp;</span> <span class="n">father_gt</span><span class="o">.</span><span class="n">is_hom_ref</span><span class="p">()</span> <span class="o">&amp;</span> <span class="n">mother_gt</span><span class="o">.</span><span class="n">is_hom_ref</span><span class="p">(),</span>
            <span class="p">)</span>
        <span class="k">return</span> <span class="n">hl</span><span class="o">.</span><span class="n">if_else</span><span class="p">(</span>
            <span class="n">locus</span><span class="o">.</span><span class="n">in_autosome_or_par</span><span class="p">()</span> <span class="o">|</span> <span class="p">(</span><span class="n">proband_is_xx</span> <span class="o">&amp;</span> <span class="n">locus</span><span class="o">.</span><span class="n">in_x_nonpar</span><span class="p">()),</span>
            <span class="n">proband_gt</span><span class="o">.</span><span class="n">is_het</span><span class="p">()</span> <span class="o">&amp;</span> <span class="n">father_gt</span><span class="o">.</span><span class="n">is_hom_ref</span><span class="p">()</span> <span class="o">&amp;</span> <span class="n">mother_gt</span><span class="o">.</span><span class="n">is_hom_ref</span><span class="p">(),</span>
            <span class="n">hl</span><span class="o">.</span><span class="n">or_missing</span><span class="p">(</span>
                <span class="o">~</span><span class="n">proband_is_xx</span><span class="p">,</span> <span class="n">proband_gt</span><span class="o">.</span><span class="n">is_hom_var</span><span class="p">()</span> <span class="o">&amp;</span> <span class="n">father_gt</span><span class="o">.</span><span class="n">is_hom_ref</span><span class="p">()</span>
            <span class="p">),</span>
        <span class="p">)</span>

    <span class="k">def</span> <span class="nf">_ac_an_parent_child_count</span><span class="p">(</span>
        <span class="n">proband_gt</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">CallExpression</span><span class="p">,</span>
        <span class="n">father_gt</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">CallExpression</span><span class="p">,</span>
        <span class="n">mother_gt</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">CallExpression</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">Int64Expression</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Get AC and AN for parents and children.&quot;&quot;&quot;</span>
        <span class="n">ac_parent_expr</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span>
            <span class="n">father_gt</span><span class="o">.</span><span class="n">n_alt_alleles</span><span class="p">()</span> <span class="o">+</span> <span class="n">mother_gt</span><span class="o">.</span><span class="n">n_alt_alleles</span><span class="p">()</span>
        <span class="p">)</span>
        <span class="n">an_parent_expr</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span>
            <span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">is_defined</span><span class="p">(</span><span class="n">father_gt</span><span class="p">)</span> <span class="o">+</span> <span class="n">hl</span><span class="o">.</span><span class="n">is_defined</span><span class="p">(</span><span class="n">mother_gt</span><span class="p">))</span> <span class="o">*</span> <span class="mi">2</span>
        <span class="p">)</span>
        <span class="n">ac_child_expr</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">proband_gt</span><span class="o">.</span><span class="n">n_alt_alleles</span><span class="p">())</span>
        <span class="n">an_child_expr</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">is_defined</span><span class="p">(</span><span class="n">proband_gt</span><span class="p">)</span> <span class="o">*</span> <span class="mi">2</span><span class="p">)</span>

        <span class="k">return</span> <span class="p">{</span>
            <span class="s2">&quot;ac_parents&quot;</span><span class="p">:</span> <span class="n">ac_parent_expr</span><span class="p">,</span>
            <span class="s2">&quot;an_parents&quot;</span><span class="p">:</span> <span class="n">an_parent_expr</span><span class="p">,</span>
            <span class="s2">&quot;ac_children&quot;</span><span class="p">:</span> <span class="n">ac_child_expr</span><span class="p">,</span>
            <span class="s2">&quot;an_children&quot;</span><span class="p">:</span> <span class="n">an_child_expr</span><span class="p">,</span>
        <span class="p">}</span>

    <span class="c1"># Create transmission counters</span>
    <span class="n">trio_stats</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span>
        <span class="o">**</span><span class="p">{</span>
            <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">name2</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
                <span class="p">(</span>
                    <span class="n">trio_mt</span><span class="o">.</span><span class="n">proband_entry</span><span class="o">.</span><span class="n">GT</span><span class="o">.</span><span class="n">is_non_ref</span><span class="p">()</span>
                    <span class="o">|</span> <span class="n">trio_mt</span><span class="o">.</span><span class="n">father_entry</span><span class="o">.</span><span class="n">GT</span><span class="o">.</span><span class="n">is_non_ref</span><span class="p">()</span>
                    <span class="o">|</span> <span class="n">trio_mt</span><span class="o">.</span><span class="n">mother_entry</span><span class="o">.</span><span class="n">GT</span><span class="o">.</span><span class="n">is_non_ref</span><span class="p">()</span>
                <span class="p">)</span>
                <span class="o">&amp;</span> <span class="n">expr</span><span class="p">,</span>
                <span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span>
                    <span class="n">trans_count_map</span><span class="o">.</span><span class="n">get</span><span class="p">(</span>
                        <span class="p">(</span>
                            <span class="n">trio_mt</span><span class="o">.</span><span class="n">proband_entry</span><span class="o">.</span><span class="n">GT</span><span class="o">.</span><span class="n">n_alt_alleles</span><span class="p">(),</span>
                            <span class="n">trio_mt</span><span class="o">.</span><span class="n">father_entry</span><span class="o">.</span><span class="n">GT</span><span class="o">.</span><span class="n">n_alt_alleles</span><span class="p">(),</span>
                            <span class="n">trio_mt</span><span class="o">.</span><span class="n">mother_entry</span><span class="o">.</span><span class="n">GT</span><span class="o">.</span><span class="n">n_alt_alleles</span><span class="p">(),</span>
                            <span class="n">_get_copy_state</span><span class="p">(</span><span class="n">trio_mt</span><span class="o">.</span><span class="n">locus</span><span class="p">),</span>
                        <span class="p">),</span>
                        <span class="n">default</span><span class="o">=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">),</span>
                    <span class="p">)[</span><span class="n">i</span><span class="p">]</span>
                <span class="p">),</span>
            <span class="p">)</span>
            <span class="k">for</span> <span class="n">name</span><span class="p">,</span> <span class="n">expr</span> <span class="ow">in</span> <span class="n">transmitted_strata</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
            <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">name2</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">([</span><span class="s2">&quot;n_transmitted&quot;</span><span class="p">,</span> <span class="s2">&quot;n_untransmitted&quot;</span><span class="p">])</span>
        <span class="p">}</span>
    <span class="p">)</span>

    <span class="c1"># Create de novo counters</span>
    <span class="n">trio_stats</span> <span class="o">=</span> <span class="n">trio_stats</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span>
        <span class="o">**</span><span class="p">{</span>
            <span class="sa">f</span><span class="s2">&quot;n_de_novos_</span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
                <span class="n">_is_dnm</span><span class="p">(</span>
                    <span class="n">trio_mt</span><span class="o">.</span><span class="n">proband_entry</span><span class="o">.</span><span class="n">GT</span><span class="p">,</span>
                    <span class="n">trio_mt</span><span class="o">.</span><span class="n">father_entry</span><span class="o">.</span><span class="n">GT</span><span class="p">,</span>
                    <span class="n">trio_mt</span><span class="o">.</span><span class="n">mother_entry</span><span class="o">.</span><span class="n">GT</span><span class="p">,</span>
                    <span class="n">trio_mt</span><span class="o">.</span><span class="n">locus</span><span class="p">,</span>
                    <span class="n">proband_is_female_expr</span><span class="p">,</span>
                <span class="p">)</span>
                <span class="o">&amp;</span> <span class="n">expr</span><span class="p">,</span>
                <span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">count</span><span class="p">(),</span>
            <span class="p">)</span>
            <span class="k">for</span> <span class="n">name</span><span class="p">,</span> <span class="n">expr</span> <span class="ow">in</span> <span class="n">de_novo_strata</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
        <span class="p">}</span>
    <span class="p">)</span>

    <span class="n">trio_stats</span> <span class="o">=</span> <span class="n">trio_stats</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span>
        <span class="o">**</span><span class="p">{</span>
            <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">name2</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
                <span class="n">expr</span><span class="p">,</span>
                <span class="n">_ac_an_parent_child_count</span><span class="p">(</span>
                    <span class="n">trio_mt</span><span class="o">.</span><span class="n">proband_entry</span><span class="o">.</span><span class="n">GT</span><span class="p">,</span>
                    <span class="n">trio_mt</span><span class="o">.</span><span class="n">father_entry</span><span class="o">.</span><span class="n">GT</span><span class="p">,</span>
                    <span class="n">trio_mt</span><span class="o">.</span><span class="n">mother_entry</span><span class="o">.</span><span class="n">GT</span><span class="p">,</span>
                <span class="p">)[</span><span class="n">name2</span><span class="p">],</span>
            <span class="p">)</span>
            <span class="k">for</span> <span class="n">name</span><span class="p">,</span> <span class="n">expr</span> <span class="ow">in</span> <span class="n">ac_strata</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
            <span class="k">for</span> <span class="n">name2</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;ac_parents&quot;</span><span class="p">,</span> <span class="s2">&quot;an_parents&quot;</span><span class="p">,</span> <span class="s2">&quot;ac_children&quot;</span><span class="p">,</span> <span class="s2">&quot;an_children&quot;</span><span class="p">]</span>
        <span class="p">}</span>
    <span class="p">)</span>

    <span class="k">return</span> <span class="n">trio_stats</span></div>


<div class="viewcode-block" id="generate_sib_stats_expr"><a class="viewcode-back" href="../../../api_reference/sample_qc/relatedness.html#gnomad.sample_qc.relatedness.generate_sib_stats_expr">[docs]</a><span class="k">def</span> <span class="nf">generate_sib_stats_expr</span><span class="p">(</span>
    <span class="n">mt</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">MatrixTable</span><span class="p">,</span>
    <span class="n">sib_ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span>
    <span class="n">i_col</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;i&quot;</span><span class="p">,</span>
    <span class="n">j_col</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;j&quot;</span><span class="p">,</span>
    <span class="n">strata</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">BooleanExpression</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;raw&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">},</span>
    <span class="n">is_female</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">BooleanExpression</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">StructExpression</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Generate a row-wise expression containing the number of alternate alleles in common between sibling pairs.</span>

<span class="sd">    The sibling sharing counts can be stratified using additional filters using `stata`.</span>

<span class="sd">    .. note::</span>

<span class="sd">        This function expects that the `mt` has either been split or filtered to only bi-allelics</span>
<span class="sd">        If a sample has multiple sibling pairs, only one pair will be counted</span>

<span class="sd">    :param mt: Input matrix table</span>
<span class="sd">    :param sib_ht: Table defining sibling pairs with one sample in a col (`i_col`) and the second in another col (`j_col`)</span>
<span class="sd">    :param i_col: Column containing the 1st sample of the pair in the relationship table</span>
<span class="sd">    :param j_col: Column containing the 2nd sample of the pair in the relationship table</span>
<span class="sd">    :param strata: Dict with additional strata to use when computing shared sibling variant counts</span>
<span class="sd">    :param is_female: An optional column in mt giving the sample sex. If not given, counts are only computed for autosomes.</span>
<span class="sd">    :return: A Table with the sibling shared variant counts</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="nf">_get_alt_count</span><span class="p">(</span><span class="n">locus</span><span class="p">,</span> <span class="n">gt</span><span class="p">,</span> <span class="n">is_female</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Calculate alt allele count with sex info if present.&quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">is_female</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">hl</span><span class="o">.</span><span class="n">or_missing</span><span class="p">(</span><span class="n">locus</span><span class="o">.</span><span class="n">in_autosome</span><span class="p">(),</span> <span class="n">gt</span><span class="o">.</span><span class="n">n_alt_alleles</span><span class="p">())</span>
        <span class="k">return</span> <span class="p">(</span>
            <span class="n">hl</span><span class="o">.</span><span class="n">case</span><span class="p">()</span>
            <span class="o">.</span><span class="n">when</span><span class="p">(</span><span class="n">locus</span><span class="o">.</span><span class="n">in_autosome_or_par</span><span class="p">(),</span> <span class="n">gt</span><span class="o">.</span><span class="n">n_alt_alleles</span><span class="p">())</span>
            <span class="o">.</span><span class="n">when</span><span class="p">(</span>
                <span class="o">~</span><span class="n">is_female</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">locus</span><span class="o">.</span><span class="n">in_x_nonpar</span><span class="p">()</span> <span class="o">|</span> <span class="n">locus</span><span class="o">.</span><span class="n">in_y_nonpar</span><span class="p">()),</span>
                <span class="n">hl</span><span class="o">.</span><span class="n">min</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">gt</span><span class="o">.</span><span class="n">n_alt_alleles</span><span class="p">()),</span>
            <span class="p">)</span>
            <span class="o">.</span><span class="n">when</span><span class="p">(</span><span class="n">is_female</span> <span class="o">&amp;</span> <span class="n">locus</span><span class="o">.</span><span class="n">in_y_nonpar</span><span class="p">(),</span> <span class="mi">0</span><span class="p">)</span>
            <span class="o">.</span><span class="n">default</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
        <span class="p">)</span>

    <span class="k">if</span> <span class="n">is_female</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
            <span class="s2">&quot;Since no sex expression was given to generate_sib_stats_expr, only&quot;</span>
            <span class="s2">&quot; variants in autosomes will be counted.&quot;</span>
        <span class="p">)</span>

    <span class="c1"># If a sample is in sib_ht more than one time, keep only one of the sibling pairs</span>
    <span class="c1"># First filter to only samples found in mt to keep as many pairs as possible</span>
    <span class="n">s_to_keep</span> <span class="o">=</span> <span class="n">mt</span><span class="o">.</span><span class="n">aggregate_cols</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">collect_as_set</span><span class="p">(</span><span class="n">mt</span><span class="o">.</span><span class="n">s</span><span class="p">),</span> <span class="n">_localize</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="n">sib_ht</span> <span class="o">=</span> <span class="n">sib_ht</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
        <span class="n">s_to_keep</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="n">sib_ht</span><span class="p">[</span><span class="n">i_col</span><span class="p">]</span><span class="o">.</span><span class="n">s</span><span class="p">)</span> <span class="o">&amp;</span> <span class="n">s_to_keep</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="n">sib_ht</span><span class="p">[</span><span class="n">j_col</span><span class="p">]</span><span class="o">.</span><span class="n">s</span><span class="p">)</span>
    <span class="p">)</span>
    <span class="n">sib_ht</span> <span class="o">=</span> <span class="n">sib_ht</span><span class="o">.</span><span class="n">add_index</span><span class="p">(</span><span class="s2">&quot;sib_idx&quot;</span><span class="p">)</span>
    <span class="n">sib_ht</span> <span class="o">=</span> <span class="n">sib_ht</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span><span class="n">sibs</span><span class="o">=</span><span class="p">[</span><span class="n">sib_ht</span><span class="p">[</span><span class="n">i_col</span><span class="p">]</span><span class="o">.</span><span class="n">s</span><span class="p">,</span> <span class="n">sib_ht</span><span class="p">[</span><span class="n">j_col</span><span class="p">]</span><span class="o">.</span><span class="n">s</span><span class="p">])</span>
    <span class="n">sib_ht</span> <span class="o">=</span> <span class="n">sib_ht</span><span class="o">.</span><span class="n">explode</span><span class="p">(</span><span class="s2">&quot;sibs&quot;</span><span class="p">)</span>
    <span class="n">sib_ht</span> <span class="o">=</span> <span class="n">sib_ht</span><span class="o">.</span><span class="n">group_by</span><span class="p">(</span><span class="s2">&quot;sibs&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">aggregate</span><span class="p">(</span>
        <span class="n">sib_idx</span><span class="o">=</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">take</span><span class="p">(</span><span class="n">sib_ht</span><span class="o">.</span><span class="n">sib_idx</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">ordering</span><span class="o">=</span><span class="n">sib_ht</span><span class="o">.</span><span class="n">sib_idx</span><span class="p">)[</span><span class="mi">0</span><span class="p">])</span>
    <span class="p">)</span>
    <span class="n">sib_ht</span> <span class="o">=</span> <span class="n">sib_ht</span><span class="o">.</span><span class="n">group_by</span><span class="p">(</span><span class="n">sib_ht</span><span class="o">.</span><span class="n">sib_idx</span><span class="p">)</span><span class="o">.</span><span class="n">aggregate</span><span class="p">(</span><span class="n">sibs</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">collect</span><span class="p">(</span><span class="n">sib_ht</span><span class="o">.</span><span class="n">sibs</span><span class="p">))</span>
    <span class="n">sib_ht</span> <span class="o">=</span> <span class="n">sib_ht</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">len</span><span class="p">(</span><span class="n">sib_ht</span><span class="o">.</span><span class="n">sibs</span><span class="p">)</span> <span class="o">==</span> <span class="mi">2</span><span class="p">)</span><span class="o">.</span><span class="n">persist</span><span class="p">()</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
        <span class="s2">&quot;Generating sibling variant sharing counts using </span><span class="si">%d</span><span class="s2"> pairs.&quot;</span><span class="p">,</span> <span class="n">sib_ht</span><span class="o">.</span><span class="n">count</span><span class="p">()</span>
    <span class="p">)</span>
    <span class="n">sib_ht</span> <span class="o">=</span> <span class="n">sib_ht</span><span class="o">.</span><span class="n">explode</span><span class="p">(</span><span class="s2">&quot;sibs&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">key_by</span><span class="p">(</span><span class="s2">&quot;sibs&quot;</span><span class="p">)[</span><span class="n">mt</span><span class="o">.</span><span class="n">s</span><span class="p">]</span>

    <span class="c1"># Create sibling sharing counters</span>
    <span class="n">sib_stats</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span>
        <span class="o">**</span><span class="p">{</span>
            <span class="sa">f</span><span class="s2">&quot;n_sib_shared_variants_</span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span>
                <span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
                    <span class="n">expr</span><span class="p">,</span>
                    <span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">group_by</span><span class="p">(</span>
                        <span class="n">sib_ht</span><span class="o">.</span><span class="n">sib_idx</span><span class="p">,</span>
                        <span class="n">hl</span><span class="o">.</span><span class="n">or_missing</span><span class="p">(</span>
                            <span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">is_defined</span><span class="p">(</span><span class="n">mt</span><span class="o">.</span><span class="n">GT</span><span class="p">))</span> <span class="o">==</span> <span class="mi">2</span><span class="p">,</span>
                            <span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">min</span><span class="p">(</span><span class="n">_get_alt_count</span><span class="p">(</span><span class="n">mt</span><span class="o">.</span><span class="n">locus</span><span class="p">,</span> <span class="n">mt</span><span class="o">.</span><span class="n">GT</span><span class="p">,</span> <span class="n">is_female</span><span class="p">)),</span>
                        <span class="p">),</span>
                    <span class="p">),</span>
                <span class="p">)</span><span class="o">.</span><span class="n">values</span><span class="p">()</span>
            <span class="p">)</span>
            <span class="k">for</span> <span class="n">name</span><span class="p">,</span> <span class="n">expr</span> <span class="ow">in</span> <span class="n">strata</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
        <span class="p">}</span>
    <span class="p">)</span>

    <span class="n">sib_stats</span> <span class="o">=</span> <span class="n">sib_stats</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span>
        <span class="o">**</span><span class="p">{</span>
            <span class="sa">f</span><span class="s2">&quot;ac_sibs_</span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
                <span class="n">expr</span> <span class="o">&amp;</span> <span class="n">hl</span><span class="o">.</span><span class="n">is_defined</span><span class="p">(</span><span class="n">sib_ht</span><span class="o">.</span><span class="n">sib_idx</span><span class="p">),</span> <span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">mt</span><span class="o">.</span><span class="n">GT</span><span class="o">.</span><span class="n">n_alt_alleles</span><span class="p">())</span>
            <span class="p">)</span>
            <span class="k">for</span> <span class="n">name</span><span class="p">,</span> <span class="n">expr</span> <span class="ow">in</span> <span class="n">strata</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
        <span class="p">}</span>
    <span class="p">)</span>

    <span class="k">return</span> <span class="n">sib_stats</span></div>


<div class="viewcode-block" id="calculate_de_novo_post_prob"><a class="viewcode-back" href="../../../api_reference/sample_qc/relatedness.html#gnomad.sample_qc.relatedness.calculate_de_novo_post_prob">[docs]</a><span class="k">def</span> <span class="nf">calculate_de_novo_post_prob</span><span class="p">(</span>
    <span class="n">proband_pl_expr</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">ArrayExpression</span><span class="p">,</span>
    <span class="n">father_pl_expr</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">ArrayExpression</span><span class="p">,</span>
    <span class="n">mother_pl_expr</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">ArrayExpression</span><span class="p">,</span>
    <span class="n">diploid_expr</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">BooleanExpression</span><span class="p">,</span>
    <span class="n">hemi_x_expr</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">BooleanExpression</span><span class="p">,</span>
    <span class="n">hemi_y_expr</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">BooleanExpression</span><span class="p">,</span>
    <span class="n">freq_prior_expr</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">Float64Expression</span><span class="p">,</span>
    <span class="n">min_pop_prior</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">float</span><span class="p">]</span> <span class="o">=</span> <span class="mi">100</span> <span class="o">/</span> <span class="mf">3e7</span><span class="p">,</span>
    <span class="n">de_novo_prior</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">float</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1</span> <span class="o">/</span> <span class="mf">3e7</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">Float64Expression</span><span class="p">:</span>
<span class="w">    </span><span class="sa">r</span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Calculate the posterior probability of a *de novo* mutation.</span>

<span class="sd">    This function computes the posterior probability of a *de novo* mutation (`P_dn`)</span>
<span class="sd">    based on the genotype likelihoods of the proband and parents, along with the</span>
<span class="sd">    population frequency prior for the variant. The method is adapted from Kaitlin</span>
<span class="sd">    Samocha&#39;s `de novo caller &lt;https://github.com/ksamocha/de_novo_scripts&gt;`_</span>
<span class="sd">    and Hail&#39;s `de_novo &lt;https://hail.is/docs/0.2/methods/genetics.html#hail.methods.de_novo&gt;`_ function.</span>
<span class="sd">    However, neither approach explicitly documented how to compute *de novo*</span>
<span class="sd">    probabilities for hemizygous genotypes in XY individuals. To address this,</span>
<span class="sd">    we provide the full set of equations in this docstring.</span>

<span class="sd">    The posterior probability of an event being truly *de novo* vs. the probability it was a missed heterozygote call in one of the two parents is:</span>

<span class="sd">    .. math::</span>

<span class="sd">        P_{dn} = \frac{P(DN \mid \text{data})}{P(DN \mid \text{data}) + P(\text{missed het in parent(s)} \mid \text{data})}</span>

<span class="sd">    The terms are defined as follows:</span>

<span class="sd">    - :math:`P(DN \mid \text{data})` is the probability that the variant is *de novo*, given the observed genotype data.</span>

<span class="sd">    - :math:`P(\text{missed het in parent(s)} \mid \text{data})` is the probability that the heterozygous variant was **missed in at least one parent**.</span>

<span class="sd">    Applying Bayesian Theorem to the numerator and denominator yields:</span>

<span class="sd">    .. math::</span>

<span class="sd">        P_{dn} = \frac{P(\text{data} \mid DN) \cdot P(DN)}{P(\text{data} \mid DN) \cdot P(DN) + P(\text{data} \mid \text{missed het in parent(s)}) \cdot P(\text{missed het in parent(s)})}</span>

<span class="sd">    where:</span>

<span class="sd">    - :math:`P(\text{data} \mid DN)`: Probability of observing the data under the assumption of a *de novo* mutation.</span>

<span class="sd">      - **Autosomes and PAR regions**:</span>

<span class="sd">        .. math::</span>

<span class="sd">            P(\text{data} \mid DN) = P(\text{hom_ref in father}) \cdot P(\text{hom_ref in mother}) \cdot P(\text{het in proband})</span>

<span class="sd">      Probability of a observing a *de novo* mutation given the data specifically for hemizygous calls in XY individuals</span>

<span class="sd">          Note that hemizygous calls in XY individuals will be reported as homozygous alternate without any sex ploidy adjustments, which is why the formulas below use `P(hom_alt in proband)`</span>

<span class="sd">      - **X non-PAR regions (XY only)**:</span>

<span class="sd">        .. math::</span>

<span class="sd">            P(\text{data} \mid DN) = P(\text{hom_ref in mother}) \cdot P(\text{hom_alt in proband})</span>

<span class="sd">      - **Y non-PAR regions (XY only)**:</span>

<span class="sd">        .. math::</span>

<span class="sd">            P(\text{data} \mid DN) = P(\text{hom_ref in father}) \cdot P(\text{hom_alt in proband})</span>

<span class="sd">    - :math:`P(DN)`: The prior probability of a *de novo* mutation from literature, defined as:</span>

<span class="sd">      .. math::</span>

<span class="sd">          P(DN) = \frac{1}{3 \times 10^7}</span>

<span class="sd">    - :math:`P(\text{data} \mid \text{missed het in parent(s)})`: Probability of observing the data under the assumption of a missed het in a parent.</span>

<span class="sd">      - **Autosomes and PAR regions**:</span>

<span class="sd">        .. math::</span>

<span class="sd">            P(\text{data} \mid \text{missed het in parents}) = ( P(\text{het in father}) \cdot P(\text{hom_ref in mother}) + P(\text{hom_ref in father}) \cdot P(\text{het in mother})) \cdot P(\text{het in proband})</span>

<span class="sd">      - **X non-PAR regions (XY only)**:</span>

<span class="sd">        .. math::</span>

<span class="sd">            P(\text{data} \mid \text{missed het in mother}) = (P(\text{het in mother}) + P(\text{hom_alt in mother})) \cdot P(\text{hom_alt in proband})</span>

<span class="sd">      - **Y non-PAR regions (XY only)**:</span>

<span class="sd">        .. math::</span>

<span class="sd">            P(\text{data} \mid \text{missed het in father}) = (P(\text{het in father}) + P(\text{hom_alt in father})) \cdot P(\text{hom_alt in proband})</span>

<span class="sd">    - :math:`P(\text{missed het in parent(s)}`: Prior that at least one parent is heterozygous. Depends on alternate allele frequency:</span>

<span class="sd">    .. math::</span>

<span class="sd">        P(\text{het in one parent}) = 1 - (1 - \text{freq_prior})^4</span>

<span class="sd">    where :math:`\text{freq_prior}` is the population frequency prior for the variant.</span>

<span class="sd">    :param proband_pl_expr: Phred-scaled genotype likelihoods for the proband.</span>
<span class="sd">    :param father_pl_expr: Phred-scaled genotype likelihoods for the father.</span>
<span class="sd">    :param mother_pl_expr: Phred-scaled genotype likelihoods for the mother.</span>
<span class="sd">    :param diploid_expr: Boolean expression indicating a diploid genotype.</span>
<span class="sd">    :param hemi_x_expr: Boolean expression indicating a hemizygous genotype on the X chromosome.</span>
<span class="sd">    :param hemi_y_expr: Boolean expression indicating a hemizygous genotype on the Y chromosome.</span>
<span class="sd">    :param freq_prior_expr: Population frequency prior for the variant.</span>
<span class="sd">    :param min_pop_prior: Minimum population frequency prior (default: :math:`\text{100/3e7}`).</span>
<span class="sd">    :param de_novo_prior: Prior probability of a *de novo* mutation (default: :math:`\text{1/3e7}`).</span>
<span class="sd">    :return: Posterior probability of a *de novo* mutation (`P_dn`).</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="nf">_get_freq_prior</span><span class="p">(</span><span class="n">freq_prior</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">Float64Expression</span><span class="p">,</span> <span class="n">min_prior</span><span class="o">=</span><span class="mi">100</span> <span class="o">/</span> <span class="mf">3e7</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Get the population frequency prior for a *de novo* mutation.</span>

<span class="sd">        :param freq_prior: The population frequency prior for the variant.</span>
<span class="sd">        :param min_prior: The minimum population frequency prior. Default is</span>
<span class="sd">           100/3e7, taken from Kaitlin Samocha&#39;s [*de novo* caller](https://github.com/ksamocha/de_novo_scripts).</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="n">hl</span><span class="o">.</span><span class="n">max</span><span class="p">(</span>
            <span class="n">hl</span><span class="o">.</span><span class="n">or_else</span><span class="p">(</span>
                <span class="n">hl</span><span class="o">.</span><span class="n">case</span><span class="p">()</span>
                <span class="o">.</span><span class="n">when</span><span class="p">((</span><span class="n">freq_prior</span> <span class="o">&gt;=</span> <span class="mi">0</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">freq_prior</span> <span class="o">&lt;=</span> <span class="mi">1</span><span class="p">),</span> <span class="n">freq_prior</span><span class="p">)</span>
                <span class="o">.</span><span class="n">or_error</span><span class="p">(</span>
                    <span class="n">hl</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                        <span class="s2">&quot;de_novo: expect 0 &lt;= freq_prior_expr &lt;= 1, found </span><span class="si">%.3e</span><span class="s2">&quot;</span><span class="p">,</span>
                        <span class="n">freq_prior</span><span class="p">,</span>
                    <span class="p">)</span>
                <span class="p">),</span>
                <span class="mf">0.0</span><span class="p">,</span>
            <span class="p">),</span>
            <span class="n">min_prior</span><span class="p">,</span>
        <span class="p">)</span>

    <span class="k">def</span> <span class="nf">_transform_pl_to_pp</span><span class="p">(</span>
        <span class="n">pl_expr</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">ArrayExpression</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">ArrayExpression</span><span class="p">:</span>
<span class="w">        </span><span class="sa">r</span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Transform the Phred-scaled likelihoods (PL) into the probability of observing each genotype (PP).</span>

<span class="sd">        .. note::</span>
<span class="sd">           The Phred-scaled likelihoods (PL) are converted back into conditional genotype</span>
<span class="sd">           probabilities (PP) given the data, as computed by HaplotypeCaller, using the</span>
<span class="sd">           following relationship:</span>

<span class="sd">           .. math::</span>

<span class="sd">             {PL} = -10 \cdot \log_{10}{P(\text{Genotype} \mid \text{Data})}</span>

<span class="sd">        :param pl_expr: ArrayExpression of PL values.</span>
<span class="sd">        :return: ArrayExpression of the probability of observing each genotype (PP).</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="n">hl</span><span class="o">.</span><span class="n">bind</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span> <span class="o">/</span> <span class="n">hl</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">x</span><span class="p">),</span> <span class="mi">10</span> <span class="o">**</span> <span class="p">(</span><span class="o">-</span><span class="n">pl_expr</span> <span class="o">/</span> <span class="mi">10</span><span class="p">))</span>

    <span class="c1"># Adjust frequency prior</span>
    <span class="n">freq_prior_expr</span> <span class="o">=</span> <span class="n">_get_freq_prior</span><span class="p">(</span><span class="n">freq_prior_expr</span><span class="p">,</span> <span class="n">min_pop_prior</span><span class="p">)</span>
    <span class="n">prior_one_parent_het</span> <span class="o">=</span> <span class="mi">1</span> <span class="o">-</span> <span class="p">(</span><span class="mi">1</span> <span class="o">-</span> <span class="n">freq_prior_expr</span><span class="p">)</span> <span class="o">**</span> <span class="mi">4</span>

    <span class="c1"># Convert PL to probabilities</span>
    <span class="n">proband_pp_expr</span><span class="p">,</span> <span class="n">father_pp_expr</span><span class="p">,</span> <span class="n">mother_pp_expr</span> <span class="o">=</span> <span class="p">[</span>
        <span class="n">_transform_pl_to_pp</span><span class="p">(</span><span class="n">pl</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">pl</span> <span class="ow">in</span> <span class="p">[</span><span class="n">proband_pl_expr</span><span class="p">,</span> <span class="n">father_pl_expr</span><span class="p">,</span> <span class="n">mother_pl_expr</span><span class="p">]</span>
    <span class="p">]</span>

    <span class="c1"># Compute `P(data | DN)`</span>
    <span class="n">prob_data_given_dn_expr</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">hl</span><span class="o">.</span><span class="n">case</span><span class="p">()</span>
        <span class="o">.</span><span class="n">when</span><span class="p">(</span><span class="n">hemi_x_expr</span><span class="p">,</span> <span class="n">mother_pp_expr</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">*</span> <span class="n">proband_pp_expr</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span>
        <span class="o">.</span><span class="n">when</span><span class="p">(</span><span class="n">hemi_y_expr</span><span class="p">,</span> <span class="n">father_pp_expr</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">*</span> <span class="n">proband_pp_expr</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span>
        <span class="o">.</span><span class="n">when</span><span class="p">(</span><span class="n">diploid_expr</span><span class="p">,</span> <span class="n">father_pp_expr</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">*</span> <span class="n">mother_pp_expr</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">*</span> <span class="n">proband_pp_expr</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
        <span class="o">.</span><span class="n">or_missing</span><span class="p">()</span>
    <span class="p">)</span>

    <span class="c1"># Compute `P(data | missed het in parent(s))`</span>
    <span class="n">prob_data_missed_het_expr</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">hl</span><span class="o">.</span><span class="n">case</span><span class="p">()</span>
        <span class="o">.</span><span class="n">when</span><span class="p">(</span>
            <span class="n">hemi_x_expr</span><span class="p">,</span>
            <span class="p">(</span><span class="n">mother_pp_expr</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">+</span> <span class="n">mother_pp_expr</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span>
            <span class="o">*</span> <span class="n">proband_pp_expr</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span>
            <span class="o">*</span> <span class="n">prior_one_parent_het</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="o">.</span><span class="n">when</span><span class="p">(</span>
            <span class="n">hemi_y_expr</span><span class="p">,</span>
            <span class="p">(</span><span class="n">father_pp_expr</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">+</span> <span class="n">father_pp_expr</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span>
            <span class="o">*</span> <span class="n">proband_pp_expr</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span>
            <span class="o">*</span> <span class="n">prior_one_parent_het</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="o">.</span><span class="n">when</span><span class="p">(</span>
            <span class="n">diploid_expr</span><span class="p">,</span>
            <span class="p">(</span>
                <span class="n">father_pp_expr</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">*</span> <span class="n">mother_pp_expr</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
                <span class="o">+</span> <span class="n">father_pp_expr</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">*</span> <span class="n">mother_pp_expr</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
            <span class="p">)</span>
            <span class="o">*</span> <span class="n">proband_pp_expr</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
            <span class="o">*</span> <span class="n">prior_one_parent_het</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="o">.</span><span class="n">or_missing</span><span class="p">()</span>
    <span class="p">)</span>

    <span class="c1"># Calculate posterior probability of *de novo* mutation</span>
    <span class="n">prob_dn_given_data_expr</span> <span class="o">=</span> <span class="n">prob_data_given_dn_expr</span> <span class="o">*</span> <span class="n">de_novo_prior</span>
    <span class="n">p_dn_expr</span> <span class="o">=</span> <span class="n">prob_dn_given_data_expr</span> <span class="o">/</span> <span class="p">(</span>
        <span class="n">prob_dn_given_data_expr</span> <span class="o">+</span> <span class="n">prob_data_missed_het_expr</span>
    <span class="p">)</span>
    <span class="k">return</span> <span class="n">p_dn_expr</span></div>


<div class="viewcode-block" id="default_get_de_novo_expr"><a class="viewcode-back" href="../../../api_reference/sample_qc/relatedness.html#gnomad.sample_qc.relatedness.default_get_de_novo_expr">[docs]</a><span class="k">def</span> <span class="nf">default_get_de_novo_expr</span><span class="p">(</span>
    <span class="n">locus_expr</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">LocusExpression</span><span class="p">,</span>
    <span class="n">alleles_expr</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">ArrayExpression</span><span class="p">,</span>
    <span class="n">proband_expr</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">StructExpression</span><span class="p">,</span>
    <span class="n">father_expr</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">StructExpression</span><span class="p">,</span>
    <span class="n">mother_expr</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">StructExpression</span><span class="p">,</span>
    <span class="n">is_xx_expr</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">BooleanExpression</span><span class="p">,</span>
    <span class="n">freq_prior_expr</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">Float64Expression</span><span class="p">,</span>
    <span class="n">min_pop_prior</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mi">100</span> <span class="o">/</span> <span class="mf">3e7</span><span class="p">,</span>
    <span class="n">de_novo_prior</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mi">1</span> <span class="o">/</span> <span class="mf">3e7</span><span class="p">,</span>
    <span class="n">min_dp_ratio</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.1</span><span class="p">,</span>
    <span class="n">min_gq</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">20</span><span class="p">,</span>
    <span class="n">min_proband_ab</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.2</span><span class="p">,</span>
    <span class="n">max_parent_ab</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.05</span><span class="p">,</span>
    <span class="n">min_de_novo_p</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.05</span><span class="p">,</span>
    <span class="n">high_conf_dp_ratio</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.2</span><span class="p">,</span>
    <span class="n">dp_threshold_snp</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">10</span><span class="p">,</span>
    <span class="n">high_med_conf_ab</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.3</span><span class="p">,</span>
    <span class="n">low_conf_ab</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.2</span><span class="p">,</span>
    <span class="n">high_conf_p</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.99</span><span class="p">,</span>
    <span class="n">med_conf_p</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.5</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">StructExpression</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Get the *de novo* status of a variant based on the proband and parent genotypes.</span>

<span class="sd">    Confidence thresholds (from Kaitlin Samocha&#39;s `de novo caller &lt;https://github.com/ksamocha/de_novo_scripts&gt;`_):</span>

<span class="sd">    +----------------+--------------+-----------------------+------+------+-------+------+</span>
<span class="sd">    |   Category     | P(*de novo*) | AB                    | AD   | DP   | DR    | GQ   |</span>
<span class="sd">    +================+==============+=======================+======+======+=======+======+</span>
<span class="sd">    | FAIL           | &lt; 0.05       | AB(parents) &gt; 0.05 OR |  0   |      | &lt; 0.1 | &lt; 20 |</span>
<span class="sd">    |                |              | AB(proband) &lt; 0.2     |      |      |       |      |</span>
<span class="sd">    +----------------+--------------+-----------------------+------+------+-------+------+</span>
<span class="sd">    | HIGH (Indel)   | &gt; 0.99       | &gt; 0.3                 |      |      | &gt; 0.2 |      |</span>
<span class="sd">    +----------------+--------------+-----------------------+------+------+-------+------+</span>
<span class="sd">    | HIGH (SNV) 1   | &gt; 0.99       | &gt; 0.3                 |      |      | &gt; 0.2 |      |</span>
<span class="sd">    +----------------+--------------+-----------------------+------+------+-------+------+</span>
<span class="sd">    | HIGH (SNV) 2   | &gt; 0.5        | &gt; 0.3                 |      | &gt; 10 |       |      |</span>
<span class="sd">    +----------------+--------------+-----------------------+------+------+-------+------+</span>
<span class="sd">    | MEDIUM         | &gt; 0.5        | &gt; 0.3                 |      |      |       |      |</span>
<span class="sd">    +----------------+--------------+-----------------------+------+------+-------+------+</span>
<span class="sd">    | LOW            | &gt;= 0.05      | &gt;= 0.2                |      |      |       |      |</span>
<span class="sd">    +----------------+--------------+-----------------------+------+------+-------+------+</span>

<span class="sd">    * AB: Proband AB. FAIL criteria also includes threshold for parent(s).</span>

<span class="sd">    * AD: Sum of parent(s) AD.</span>

<span class="sd">    * DP: Proband DP.</span>

<span class="sd">    * DR: Defined as DP(proband) / DP(parent(s)).</span>

<span class="sd">    * GQ: Proband GQ.</span>

<span class="sd">    .. note::</span>

<span class="sd">        The “LOW” confidence category differs slightly from the criteria in the</span>
<span class="sd">        original code (P(*de novo*) &gt; 0.05  and  AB(proband &gt; 0.2), as it is</span>
<span class="sd">        designed to fill the gap for variants that do not meet the FAIL criteria but</span>
<span class="sd">        would otherwise remain unclassified.</span>

<span class="sd">        The *de novo* confidence is calculated as a simplified version of the one</span>
<span class="sd">        previously described in Kaitlin Samocha&#39;s [*de novo* caller](https://github.com/ksamocha/de_novo_scripts)</span>
<span class="sd">        and Hail&#39;s [*de_novo*](https://hail.is/docs/0.2/methods/genetics.html#hail.methods.de_novo)</span>
<span class="sd">        method. This simplified version is the same as Hail&#39;s methods when using the</span>
<span class="sd">        `ignore_in_sample_allele_frequency` parameter. The main difference is that</span>
<span class="sd">        this mode should be used when families larger than a single trio are in the</span>
<span class="sd">        dataset, in which an allele might be *de novo* in a parent and transmitted to a</span>
<span class="sd">        child in the dataset. This mode will not consider the allele count (AC) in</span>
<span class="sd">        the dataset, and will only consider the Phred-scaled likelihoods (PL) of the</span>
<span class="sd">        child and parents, allele balance (AB) of the child and parents,</span>
<span class="sd">        the genotype quality (GQ) of the child, the depth (DP) of the child and</span>
<span class="sd">        parents, and the population frequency prior.</span>

<span class="sd">    .. warning::</span>

<span class="sd">        This method assumes that the PL and AD fields are present in the genotype fields</span>
<span class="sd">        of the child and parents. If they are missing, this method will not work.</span>
<span class="sd">        Many of our larger datasets have the PL and AD fields intentionally removed to</span>
<span class="sd">        save storage space. If this is the reason that the PL and AD fields are</span>
<span class="sd">        missing, the only way to use this method is to set them to their approximate</span>
<span class="sd">        values:</span>

<span class="sd">        .. code-block:: python</span>

<span class="sd">                PL=hl.or_else(PL, [0, GQ, 2 * GQ])</span>
<span class="sd">                AD=hl.or_else(AD, [DP, 0])</span>

<span class="sd">    :param locus_expr: Variant locus.</span>
<span class="sd">    :param alleles_expr: Variant alleles. Function assumes all variants are</span>
<span class="sd">       biallelic, meaning that multiallelic variants in the input dataset should be</span>
<span class="sd">       split prior to running this function.</span>
<span class="sd">    :param proband_expr: Proband genotype info; required fields: GT, DP, GQ, AD, PL.</span>
<span class="sd">    :param father_expr: Father genotype info; required fields: GT, DP, GQ, AD, PL.</span>
<span class="sd">    :param mother_expr: Mother genotype info; required fields: GT, DP, GQ, AD, PL.</span>
<span class="sd">    :param is_xx_expr: Whether the proband has XX sex karyotype.</span>
<span class="sd">    :param freq_prior_expr: Population frequency prior for the variant.</span>
<span class="sd">    :param min_pop_prior: Minimum population frequency prior. Default is 100 / 3e7.</span>
<span class="sd">    :param de_novo_prior: Prior probability of a *de novo* mutation. Default is 1 / 3e7.</span>
<span class="sd">    :param min_dp_ratio: Minimum depth ratio for proband to parents. Default is 0.1.</span>
<span class="sd">    :param min_gq: Minimum genotype quality for the proband. Default is 20.</span>
<span class="sd">    :param min_proband_ab: Minimum allele balance for the proband. Default is 0.2.</span>
<span class="sd">    :param max_parent_ab: Maximum allele balance for parents. Default is 0.05.</span>
<span class="sd">    :param min_de_novo_p: Minimum probability for variant to be called *de novo*. Default is 0.05.</span>
<span class="sd">    :param high_conf_dp_ratio: DP ratio threshold of proband DP to combined DP in parents for high confidence. Default is 0.2.</span>
<span class="sd">    :param dp_threshold_snp: Minimum depth for high-confidence SNPs. Default is 10.</span>
<span class="sd">    :param high_med_conf_ab: AB threshold for high/medium confidence. Default is 0.3.</span>
<span class="sd">    :param low_conf_ab: AB threshold for low confidence. Default is 0.2.</span>
<span class="sd">    :param high_conf_p: P(*de novo*) threshold for high confidence. Default is 0.99.</span>
<span class="sd">    :param med_conf_p: P(*de novo*) threshold for medium confidence. Default is 0.5.</span>
<span class="sd">    :return: StructExpression with variant *de novo* status and confidence of *de novo* call.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Check whether multiallelics have been split</span>
    <span class="n">alleles_expr</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">hl</span><span class="o">.</span><span class="n">case</span><span class="p">()</span>
        <span class="o">.</span><span class="n">when</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">len</span><span class="p">(</span><span class="n">alleles_expr</span><span class="p">)</span> <span class="o">==</span> <span class="mi">2</span><span class="p">,</span> <span class="n">alleles_expr</span><span class="p">)</span>
        <span class="o">.</span><span class="n">or_error</span><span class="p">(</span><span class="s2">&quot;Must split multiallelic variants prior to running this function.&quot;</span><span class="p">)</span>
    <span class="p">)</span>

    <span class="c1"># Determine genomic context</span>
    <span class="n">diploid_expr</span><span class="p">,</span> <span class="n">hemi_x_expr</span><span class="p">,</span> <span class="n">hemi_y_expr</span> <span class="o">=</span> <span class="n">get_copy_state_by_sex</span><span class="p">(</span>
        <span class="n">locus_expr</span><span class="p">,</span> <span class="n">is_xx_expr</span>
    <span class="p">)</span>

    <span class="n">p_de_novo</span> <span class="o">=</span> <span class="n">calculate_de_novo_post_prob</span><span class="p">(</span>
        <span class="n">proband_expr</span><span class="o">.</span><span class="n">PL</span><span class="p">,</span>
        <span class="n">father_expr</span><span class="o">.</span><span class="n">PL</span><span class="p">,</span>
        <span class="n">mother_expr</span><span class="o">.</span><span class="n">PL</span><span class="p">,</span>
        <span class="n">diploid_expr</span><span class="p">,</span>
        <span class="n">hemi_x_expr</span><span class="p">,</span>
        <span class="n">hemi_y_expr</span><span class="p">,</span>
        <span class="n">freq_prior_expr</span><span class="p">,</span>
        <span class="n">min_pop_prior</span><span class="o">=</span><span class="n">min_pop_prior</span><span class="p">,</span>
        <span class="n">de_novo_prior</span><span class="o">=</span><span class="n">de_novo_prior</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="c1"># Calculate DP ratio</span>
    <span class="n">parent_dp</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">hl</span><span class="o">.</span><span class="n">case</span><span class="p">()</span>
        <span class="o">.</span><span class="n">when</span><span class="p">(</span><span class="n">diploid_expr</span><span class="p">,</span> <span class="n">father_expr</span><span class="o">.</span><span class="n">DP</span> <span class="o">+</span> <span class="n">mother_expr</span><span class="o">.</span><span class="n">DP</span><span class="p">)</span>
        <span class="o">.</span><span class="n">when</span><span class="p">(</span><span class="n">hemi_x_expr</span><span class="p">,</span> <span class="n">mother_expr</span><span class="o">.</span><span class="n">DP</span><span class="p">)</span>
        <span class="o">.</span><span class="n">when</span><span class="p">(</span><span class="n">hemi_y_expr</span><span class="p">,</span> <span class="n">father_expr</span><span class="o">.</span><span class="n">DP</span><span class="p">)</span>
        <span class="o">.</span><span class="n">or_missing</span><span class="p">()</span>
    <span class="p">)</span>
    <span class="n">dp_ratio</span> <span class="o">=</span> <span class="n">proband_expr</span><span class="o">.</span><span class="n">DP</span> <span class="o">/</span> <span class="n">parent_dp</span>

    <span class="c1"># Calculate proband AB</span>
    <span class="n">proband_ab</span> <span class="o">=</span> <span class="n">proband_expr</span><span class="o">.</span><span class="n">AD</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">/</span> <span class="n">hl</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">proband_expr</span><span class="o">.</span><span class="n">AD</span><span class="p">)</span>

    <span class="n">is_de_novo</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">diploid_expr</span>
        <span class="o">&amp;</span> <span class="p">(</span>
            <span class="n">proband_expr</span><span class="o">.</span><span class="n">GT</span><span class="o">.</span><span class="n">is_het</span><span class="p">()</span>
            <span class="o">&amp;</span> <span class="n">father_expr</span><span class="o">.</span><span class="n">GT</span><span class="o">.</span><span class="n">is_hom_ref</span><span class="p">()</span>
            <span class="o">&amp;</span> <span class="n">mother_expr</span><span class="o">.</span><span class="n">GT</span><span class="o">.</span><span class="n">is_hom_ref</span><span class="p">()</span>
        <span class="p">)</span>
        <span class="o">|</span> <span class="n">hemi_x_expr</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">proband_expr</span><span class="o">.</span><span class="n">GT</span><span class="o">.</span><span class="n">is_hom_var</span><span class="p">()</span> <span class="o">&amp;</span> <span class="n">mother_expr</span><span class="o">.</span><span class="n">GT</span><span class="o">.</span><span class="n">is_hom_ref</span><span class="p">())</span>
        <span class="o">|</span> <span class="n">hemi_y_expr</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">proband_expr</span><span class="o">.</span><span class="n">GT</span><span class="o">.</span><span class="n">is_hom_var</span><span class="p">()</span> <span class="o">&amp;</span> <span class="n">father_expr</span><span class="o">.</span><span class="n">GT</span><span class="o">.</span><span class="n">is_hom_ref</span><span class="p">())</span>
    <span class="p">)</span>

    <span class="c1"># Confidence assignment</span>
    <span class="n">confidence_expr</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">hl</span><span class="o">.</span><span class="n">case</span><span class="p">()</span>
        <span class="o">.</span><span class="n">when</span><span class="p">(</span>
            <span class="p">(</span>
                <span class="p">(</span>
                    <span class="p">(</span><span class="n">p_de_novo</span> <span class="o">&gt;</span> <span class="n">high_conf_p</span><span class="p">)</span>
                    <span class="o">&amp;</span> <span class="p">(</span><span class="n">proband_ab</span> <span class="o">&gt;</span> <span class="n">high_med_conf_ab</span><span class="p">)</span>
                    <span class="o">&amp;</span> <span class="p">(</span><span class="n">dp_ratio</span> <span class="o">&gt;</span> <span class="n">high_conf_dp_ratio</span><span class="p">)</span>
                <span class="p">)</span>
                <span class="o">|</span> <span class="p">(</span>
                    <span class="n">hl</span><span class="o">.</span><span class="n">is_snp</span><span class="p">(</span><span class="n">alleles_expr</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">alleles_expr</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
                    <span class="o">&amp;</span> <span class="p">(</span><span class="n">p_de_novo</span> <span class="o">&gt;</span> <span class="n">med_conf_p</span><span class="p">)</span>
                    <span class="o">&amp;</span> <span class="p">(</span><span class="n">proband_ab</span> <span class="o">&gt;</span> <span class="n">high_med_conf_ab</span><span class="p">)</span>
                    <span class="o">&amp;</span> <span class="p">(</span><span class="n">proband_expr</span><span class="o">.</span><span class="n">DP</span> <span class="o">&gt;</span> <span class="n">dp_threshold_snp</span><span class="p">)</span>
                <span class="p">)</span>
            <span class="p">),</span>
            <span class="s2">&quot;HIGH&quot;</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="o">.</span><span class="n">when</span><span class="p">((</span><span class="n">p_de_novo</span> <span class="o">&gt;</span> <span class="n">med_conf_p</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">proband_ab</span> <span class="o">&gt;</span> <span class="n">high_med_conf_ab</span><span class="p">),</span> <span class="s2">&quot;MEDIUM&quot;</span><span class="p">)</span>
        <span class="o">.</span><span class="n">when</span><span class="p">((</span><span class="n">p_de_novo</span> <span class="o">&gt;=</span> <span class="n">min_de_novo_p</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">proband_ab</span> <span class="o">&gt;=</span> <span class="n">low_conf_ab</span><span class="p">),</span> <span class="s2">&quot;LOW&quot;</span><span class="p">)</span>
        <span class="o">.</span><span class="n">or_missing</span><span class="p">()</span>
    <span class="p">)</span>

    <span class="n">fail_parent_sum_ad_0_expr</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">hl</span><span class="o">.</span><span class="n">case</span><span class="p">()</span>
        <span class="o">.</span><span class="n">when</span><span class="p">(</span>
            <span class="n">diploid_expr</span><span class="p">,</span> <span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">father_expr</span><span class="o">.</span><span class="n">AD</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">)</span> <span class="o">|</span> <span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">mother_expr</span><span class="o">.</span><span class="n">AD</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">)</span>
        <span class="p">)</span>
        <span class="o">.</span><span class="n">when</span><span class="p">(</span><span class="n">hemi_x_expr</span><span class="p">,</span> <span class="n">hl</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">mother_expr</span><span class="o">.</span><span class="n">AD</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">)</span>
        <span class="o">.</span><span class="n">when</span><span class="p">(</span><span class="n">hemi_y_expr</span><span class="p">,</span> <span class="n">hl</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">father_expr</span><span class="o">.</span><span class="n">AD</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">)</span>
        <span class="o">.</span><span class="n">or_missing</span><span class="p">()</span>
    <span class="p">)</span>

    <span class="n">fail_max_parent_ab_expr</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">hl</span><span class="o">.</span><span class="n">case</span><span class="p">()</span>
        <span class="o">.</span><span class="n">when</span><span class="p">(</span>
            <span class="n">diploid_expr</span><span class="p">,</span>
            <span class="p">(</span><span class="n">father_expr</span><span class="o">.</span><span class="n">AD</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">/</span> <span class="n">hl</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">father_expr</span><span class="o">.</span><span class="n">AD</span><span class="p">)</span> <span class="o">&gt;</span> <span class="n">max_parent_ab</span><span class="p">)</span>
            <span class="o">|</span> <span class="p">(</span><span class="n">mother_expr</span><span class="o">.</span><span class="n">AD</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">/</span> <span class="n">hl</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">mother_expr</span><span class="o">.</span><span class="n">AD</span><span class="p">)</span> <span class="o">&gt;</span> <span class="n">max_parent_ab</span><span class="p">),</span>
        <span class="p">)</span>
        <span class="o">.</span><span class="n">when</span><span class="p">(</span><span class="n">hemi_x_expr</span><span class="p">,</span> <span class="n">mother_expr</span><span class="o">.</span><span class="n">AD</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">/</span> <span class="n">hl</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">mother_expr</span><span class="o">.</span><span class="n">AD</span><span class="p">)</span> <span class="o">&gt;</span> <span class="n">max_parent_ab</span><span class="p">)</span>
        <span class="o">.</span><span class="n">when</span><span class="p">(</span><span class="n">hemi_y_expr</span><span class="p">,</span> <span class="n">father_expr</span><span class="o">.</span><span class="n">AD</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">/</span> <span class="n">hl</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">father_expr</span><span class="o">.</span><span class="n">AD</span><span class="p">)</span> <span class="o">&gt;</span> <span class="n">max_parent_ab</span><span class="p">)</span>
        <span class="o">.</span><span class="n">or_missing</span><span class="p">()</span>
    <span class="p">)</span>

    <span class="c1"># Fail checks</span>
    <span class="n">fail_checks_expr</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;min_dp_ratio&quot;</span><span class="p">:</span> <span class="n">dp_ratio</span> <span class="o">&lt;</span> <span class="n">min_dp_ratio</span><span class="p">,</span>
        <span class="s2">&quot;parent_sum_ad_0&quot;</span><span class="p">:</span> <span class="n">fail_parent_sum_ad_0_expr</span><span class="p">,</span>
        <span class="s2">&quot;max_parent_ab&quot;</span><span class="p">:</span> <span class="n">fail_max_parent_ab_expr</span><span class="p">,</span>
        <span class="s2">&quot;min_proband_ab&quot;</span><span class="p">:</span> <span class="n">proband_ab</span> <span class="o">&lt;</span> <span class="n">min_proband_ab</span><span class="p">,</span>
        <span class="s2">&quot;min_proband_gq&quot;</span><span class="p">:</span> <span class="n">proband_expr</span><span class="o">.</span><span class="n">GQ</span> <span class="o">&lt;</span> <span class="n">min_gq</span><span class="p">,</span>
        <span class="s2">&quot;min_de_novo_p&quot;</span><span class="p">:</span> <span class="n">p_de_novo</span> <span class="o">&lt;=</span> <span class="n">min_de_novo_p</span><span class="p">,</span>
    <span class="p">}</span>

    <span class="n">fail</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">any</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">fail_checks_expr</span><span class="o">.</span><span class="n">values</span><span class="p">()))</span>
    <span class="n">result_expr</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span>
        <span class="n">is_de_novo</span><span class="o">=</span><span class="n">is_de_novo</span><span class="p">,</span>
        <span class="n">p_de_novo</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">or_missing</span><span class="p">(</span><span class="n">is_de_novo</span> <span class="o">&amp;</span> <span class="o">~</span><span class="n">fail</span><span class="p">,</span> <span class="n">p_de_novo</span><span class="p">),</span>
        <span class="n">confidence</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">or_missing</span><span class="p">(</span><span class="n">is_de_novo</span> <span class="o">&amp;</span> <span class="o">~</span><span class="n">fail</span><span class="p">,</span> <span class="n">confidence_expr</span><span class="p">),</span>
        <span class="n">fail_reason</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">or_missing</span><span class="p">(</span>
            <span class="n">is_de_novo</span> <span class="o">&amp;</span> <span class="n">fail</span><span class="p">,</span>
            <span class="n">add_filters_expr</span><span class="p">(</span><span class="n">filters</span><span class="o">=</span><span class="n">fail_checks_expr</span><span class="p">),</span>
        <span class="p">),</span>
    <span class="p">)</span>
    <span class="k">return</span> <span class="n">result_expr</span></div>
</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p></p>
  </div>

   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>