<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>gnomad.utils.vcf &mdash; gnomad master documentation</title>
      <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../../../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../../../_static/theme_overrides.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../../../_static/jquery.js"></script>
        <script src="../../../_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script data-url_root="../../../" id="documentation_options" src="../../../_static/documentation_options.js"></script>
        <script src="../../../_static/doctools.js"></script>
        <script src="../../../_static/sphinx_highlight.js"></script>
    <script src="../../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../../index.html" class="icon icon-home">
            gnomad
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../../../getting_started.html">Getting Started</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../examples/index.html">Examples</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../api_reference/index.html">API Reference</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../resource_sources.html">Resource Sources</a></li>
<li class="toctree-l1"><a class="reference external" href="https://github.com/broadinstitute/gnomad_methods/releases">Change Log</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">gnomad</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../../index.html">Module code</a></li>
      <li class="breadcrumb-item active">gnomad.utils.vcf</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for gnomad.utils.vcf</h1><div class="highlight"><pre>
<span></span><span class="c1"># noqa: D100</span>

<span class="kn">import</span> <span class="nn">copy</span>
<span class="kn">import</span> <span class="nn">itertools</span>
<span class="kn">import</span> <span class="nn">logging</span>
<span class="kn">from</span> <span class="nn">typing</span> <span class="kn">import</span> <span class="n">Dict</span><span class="p">,</span> <span class="n">List</span><span class="p">,</span> <span class="n">Optional</span><span class="p">,</span> <span class="n">Union</span>

<span class="kn">import</span> <span class="nn">hail</span> <span class="k">as</span> <span class="nn">hl</span>

<span class="kn">from</span> <span class="nn">gnomad.sample_qc.ancestry</span> <span class="kn">import</span> <span class="n">GEN_ANC_NAMES</span>

<span class="n">logging</span><span class="o">.</span><span class="n">basicConfig</span><span class="p">(</span><span class="nb">format</span><span class="o">=</span><span class="s2">&quot;</span><span class="si">%(levelname)s</span><span class="s2"> (</span><span class="si">%(name)s</span><span class="s2"> </span><span class="si">%(lineno)s</span><span class="s2">): </span><span class="si">%(message)s</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="n">logger</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">(</span><span class="vm">__name__</span><span class="p">)</span>
<span class="n">logger</span><span class="o">.</span><span class="n">setLevel</span><span class="p">(</span><span class="n">logging</span><span class="o">.</span><span class="n">INFO</span><span class="p">)</span>

<span class="n">SORT_ORDER</span> <span class="o">=</span> <span class="p">[</span>
    <span class="s2">&quot;subset&quot;</span><span class="p">,</span>
    <span class="s2">&quot;downsampling&quot;</span><span class="p">,</span>
    <span class="s2">&quot;grpmax&quot;</span><span class="p">,</span>
    <span class="s2">&quot;popmax&quot;</span><span class="p">,</span>
    <span class="s2">&quot;gen_anc&quot;</span><span class="p">,</span>
    <span class="s2">&quot;pop&quot;</span><span class="p">,</span>
    <span class="s2">&quot;subgrp&quot;</span><span class="p">,</span>
    <span class="s2">&quot;subpop&quot;</span><span class="p">,</span>
    <span class="s2">&quot;sex&quot;</span><span class="p">,</span>
    <span class="s2">&quot;group&quot;</span><span class="p">,</span>
<span class="p">]</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">Order to sort subgroupings during VCF export.</span>
<span class="sd">Ensures that INFO labels in VCF are in desired order (e.g., raw_AC_afr_XX).</span>
<span class="sd">&quot;&quot;&quot;</span>

<span class="n">GROUPS</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;adj&quot;</span><span class="p">,</span> <span class="s2">&quot;raw&quot;</span><span class="p">]</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">Group names used to generate labels for high quality genotypes and all raw genotypes. Used in VCF export.</span>
<span class="sd">&quot;&quot;&quot;</span>

<span class="n">HISTS</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;gq_hist_alt&quot;</span><span class="p">,</span> <span class="s2">&quot;gq_hist_all&quot;</span><span class="p">,</span> <span class="s2">&quot;dp_hist_alt&quot;</span><span class="p">,</span> <span class="s2">&quot;dp_hist_all&quot;</span><span class="p">,</span> <span class="s2">&quot;ab_hist_alt&quot;</span><span class="p">]</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">Quality histograms used in VCF export.</span>
<span class="sd">&quot;&quot;&quot;</span>

<span class="n">FAF_GEN_ANC_GROUPS</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;v3&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;afr&quot;</span><span class="p">,</span> <span class="s2">&quot;amr&quot;</span><span class="p">,</span> <span class="s2">&quot;eas&quot;</span><span class="p">,</span> <span class="s2">&quot;nfe&quot;</span><span class="p">,</span> <span class="s2">&quot;sas&quot;</span><span class="p">],</span>
    <span class="s2">&quot;v4&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;afr&quot;</span><span class="p">,</span> <span class="s2">&quot;amr&quot;</span><span class="p">,</span> <span class="s2">&quot;eas&quot;</span><span class="p">,</span> <span class="s2">&quot;mid&quot;</span><span class="p">,</span> <span class="s2">&quot;nfe&quot;</span><span class="p">,</span> <span class="s2">&quot;sas&quot;</span><span class="p">],</span>
<span class="p">}</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">Genetic ancestry groups that are included in filtering allele frequency (faf) calculations. Used in VCF export.</span>
<span class="sd">&quot;&quot;&quot;</span>

<span class="n">SEXES</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;XX&quot;</span><span class="p">,</span> <span class="s2">&quot;XY&quot;</span><span class="p">]</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">Sample sexes used in VCF export.</span>

<span class="sd">Used to stratify frequency annotations (AC, AN, AF) for each sex.</span>
<span class="sd">Note that sample sexes in gnomAD v3 and earlier were &#39;male&#39; and &#39;female&#39;.</span>
<span class="sd">&quot;&quot;&quot;</span>

<span class="n">AS_FIELDS</span> <span class="o">=</span> <span class="p">[</span>
    <span class="s2">&quot;AS_FS&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AS_MQ&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AS_MQRankSum&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AS_pab_max&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AS_QUALapprox&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AS_QD&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AS_ReadPosRankSum&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AS_SB_TABLE&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AS_SOR&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AS_VarDP&quot;</span><span class="p">,</span>
    <span class="s2">&quot;InbreedingCoeff&quot;</span><span class="p">,</span>
<span class="p">]</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">Allele-specific variant annotations.</span>
<span class="sd">&quot;&quot;&quot;</span>

<span class="n">SITE_FIELDS</span> <span class="o">=</span> <span class="p">[</span>
    <span class="s2">&quot;FS&quot;</span><span class="p">,</span>
    <span class="s2">&quot;MQ&quot;</span><span class="p">,</span>
    <span class="s2">&quot;MQRankSum&quot;</span><span class="p">,</span>
    <span class="s2">&quot;QUALapprox&quot;</span><span class="p">,</span>
    <span class="s2">&quot;QD&quot;</span><span class="p">,</span>
    <span class="s2">&quot;ReadPosRankSum&quot;</span><span class="p">,</span>
    <span class="s2">&quot;SB&quot;</span><span class="p">,</span>
    <span class="s2">&quot;SOR&quot;</span><span class="p">,</span>
    <span class="s2">&quot;VarDP&quot;</span><span class="p">,</span>
<span class="p">]</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">Site level variant annotations.</span>
<span class="sd">&quot;&quot;&quot;</span>

<span class="n">ALLELE_TYPE_FIELDS</span> <span class="o">=</span> <span class="p">[</span>
    <span class="s2">&quot;allele_type&quot;</span><span class="p">,</span>
    <span class="s2">&quot;has_star&quot;</span><span class="p">,</span>
    <span class="s2">&quot;n_alt_alleles&quot;</span><span class="p">,</span>
    <span class="s2">&quot;original_alleles&quot;</span><span class="p">,</span>
    <span class="s2">&quot;variant_type&quot;</span><span class="p">,</span>
    <span class="s2">&quot;was_mixed&quot;</span><span class="p">,</span>
<span class="p">]</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">Allele-type annotations.</span>
<span class="sd">&quot;&quot;&quot;</span>

<span class="n">REGION_FLAG_FIELDS</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;decoy&quot;</span><span class="p">,</span> <span class="s2">&quot;lcr&quot;</span><span class="p">,</span> <span class="s2">&quot;nonpar&quot;</span><span class="p">,</span> <span class="s2">&quot;non_par&quot;</span><span class="p">,</span> <span class="s2">&quot;segdup&quot;</span><span class="p">]</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">Annotations about variant region type.</span>

<span class="sd">.. note::</span>
<span class="sd">    decoy resource files do not currently exist for GRCh38/hg38.</span>
<span class="sd">&quot;&quot;&quot;</span>

<span class="n">JOINT_REGION_FLAG_FIELDS</span> <span class="o">=</span> <span class="p">[</span>
    <span class="s2">&quot;fail_interval_qc&quot;</span><span class="p">,</span>
    <span class="s2">&quot;outside_broad_capture_region&quot;</span><span class="p">,</span>
    <span class="s2">&quot;outside_ukb_capture_region&quot;</span><span class="p">,</span>
    <span class="s2">&quot;outside_broad_calling_region&quot;</span><span class="p">,</span>
    <span class="s2">&quot;outside_ukb_calling_region&quot;</span><span class="p">,</span>
    <span class="s2">&quot;not_called_in_exomes&quot;</span><span class="p">,</span>
    <span class="s2">&quot;not_called_in_genomes&quot;</span><span class="p">,</span>
<span class="p">]</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">Annotations about variant region type that are specifically created for joint dataset of exomes and genomes from gnomAD v4.1.</span>
<span class="sd">&quot;&quot;&quot;</span>

<span class="n">RF_FIELDS</span> <span class="o">=</span> <span class="p">[</span>
    <span class="s2">&quot;rf_positive_label&quot;</span><span class="p">,</span>
    <span class="s2">&quot;rf_negative_label&quot;</span><span class="p">,</span>
    <span class="s2">&quot;rf_label&quot;</span><span class="p">,</span>
    <span class="s2">&quot;rf_train&quot;</span><span class="p">,</span>
    <span class="s2">&quot;rf_tp_probability&quot;</span><span class="p">,</span>
<span class="p">]</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">Annotations specific to the variant QC using a random forest model.</span>
<span class="sd">&quot;&quot;&quot;</span>

<span class="n">AS_VQSR_FIELDS</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;AS_culprit&quot;</span><span class="p">,</span> <span class="s2">&quot;AS_VQSLOD&quot;</span><span class="p">]</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">Allele-specific VQSR annotations.</span>
<span class="sd">&quot;&quot;&quot;</span>

<span class="n">VQSR_FIELDS</span> <span class="o">=</span> <span class="n">AS_VQSR_FIELDS</span> <span class="o">+</span> <span class="p">[</span><span class="s2">&quot;NEGATIVE_TRAIN_SITE&quot;</span><span class="p">,</span> <span class="s2">&quot;POSITIVE_TRAIN_SITE&quot;</span><span class="p">]</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">Annotations specific to VQSR.</span>
<span class="sd">&quot;&quot;&quot;</span>

<span class="n">INFO_VCF_AS_PIPE_DELIMITED_FIELDS</span> <span class="o">=</span> <span class="p">[</span>
    <span class="s2">&quot;AS_QUALapprox&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AS_VarDP&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AS_MQ_DP&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AS_RAW_MQ&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AS_SB_TABLE&quot;</span><span class="p">,</span>
<span class="p">]</span>

<span class="n">INFO_DICT</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;FS&quot;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="s2">&quot;Phred-scaled p-value of Fisher&#39;s exact test for strand bias&quot;</span>
    <span class="p">},</span>
    <span class="s2">&quot;InbreedingCoeff&quot;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s2">&quot;Number&quot;</span><span class="p">:</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="p">(</span>
            <span class="s2">&quot;Inbreeding coefficient, the excess heterozygosity at a variant site,&quot;</span>
            <span class="s2">&quot; computed as 1 - (the number of heterozygous genotypes)/(the number of&quot;</span>
            <span class="s2">&quot; heterozygous genotypes expected under Hardy-Weinberg equilibrium)&quot;</span>
        <span class="p">),</span>
    <span class="p">},</span>
    <span class="s2">&quot;inbreeding_coeff&quot;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s2">&quot;Number&quot;</span><span class="p">:</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="p">(</span>
            <span class="s2">&quot;Inbreeding coefficient, the excess heterozygosity at a variant site,&quot;</span>
            <span class="s2">&quot; computed as 1 - (the number of heterozygous genotypes)/(the number of&quot;</span>
            <span class="s2">&quot; heterozygous genotypes expected under Hardy-Weinberg equilibrium)&quot;</span>
        <span class="p">),</span>
    <span class="p">},</span>
    <span class="s2">&quot;MQ&quot;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="p">(</span>
            <span class="s2">&quot;Root mean square of the mapping quality of reads across all samples&quot;</span>
        <span class="p">)</span>
    <span class="p">},</span>
    <span class="s2">&quot;MQRankSum&quot;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="p">(</span>
            <span class="s2">&quot;Z-score from Wilcoxon rank sum test of alternate vs. reference read&quot;</span>
            <span class="s2">&quot; mapping qualities&quot;</span>
        <span class="p">)</span>
    <span class="p">},</span>
    <span class="s2">&quot;QD&quot;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="p">(</span>
            <span class="s2">&quot;Variant call confidence normalized by depth of sample reads supporting a&quot;</span>
            <span class="s2">&quot; variant&quot;</span>
        <span class="p">)</span>
    <span class="p">},</span>
    <span class="s2">&quot;ReadPosRankSum&quot;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="p">(</span>
            <span class="s2">&quot;Z-score from Wilcoxon rank sum test of alternate vs. reference read&quot;</span>
            <span class="s2">&quot; position bias&quot;</span>
        <span class="p">)</span>
    <span class="p">},</span>
    <span class="s2">&quot;SOR&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="s2">&quot;Strand bias estimated by the symmetric odds ratio test&quot;</span><span class="p">},</span>
    <span class="s2">&quot;POSITIVE_TRAIN_SITE&quot;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="p">(</span>
            <span class="s2">&quot;Variant was used to build the positive training set of high-quality&quot;</span>
            <span class="s2">&quot; variants for VQSR&quot;</span>
        <span class="p">)</span>
    <span class="p">},</span>
    <span class="s2">&quot;NEGATIVE_TRAIN_SITE&quot;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="p">(</span>
            <span class="s2">&quot;Variant was used to build the negative training set of low-quality&quot;</span>
            <span class="s2">&quot; variants for VQSR&quot;</span>
        <span class="p">)</span>
    <span class="p">},</span>
    <span class="s2">&quot;positive_train_site&quot;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="p">(</span>
            <span class="s2">&quot;Variant was used to build the positive training set of high-quality&quot;</span>
            <span class="s2">&quot; variants for VQSR&quot;</span>
        <span class="p">)</span>
    <span class="p">},</span>
    <span class="s2">&quot;negative_train_site&quot;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="p">(</span>
            <span class="s2">&quot;Variant was used to build the negative training set of low-quality&quot;</span>
            <span class="s2">&quot; variants for VQSR&quot;</span>
        <span class="p">)</span>
    <span class="p">},</span>
    <span class="s2">&quot;BaseQRankSum&quot;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="p">(</span>
            <span class="s2">&quot;Z-score from Wilcoxon rank sum test of alternate vs. reference base&quot;</span>
            <span class="s2">&quot; qualities&quot;</span>
        <span class="p">),</span>
    <span class="p">},</span>
    <span class="s2">&quot;VarDP&quot;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="p">(</span>
            <span class="s2">&quot;Depth over variant genotypes (does not include depth of reference samples)&quot;</span>
        <span class="p">)</span>
    <span class="p">},</span>
    <span class="s2">&quot;VQSLOD&quot;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="p">(</span>
            <span class="s2">&quot;Log-odds ratio of being a true variant versus being a false positive under&quot;</span>
            <span class="s2">&quot; the trained VQSR Gaussian mixture model&quot;</span>
        <span class="p">),</span>
    <span class="p">},</span>
    <span class="s2">&quot;culprit&quot;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="s2">&quot;Worst-performing annotation in the VQSR Gaussian mixture model&quot;</span><span class="p">,</span>
    <span class="p">},</span>
    <span class="s2">&quot;decoy&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="s2">&quot;Variant falls within a reference decoy region&quot;</span><span class="p">},</span>
    <span class="s2">&quot;lcr&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="s2">&quot;Variant falls within a low complexity region&quot;</span><span class="p">},</span>
    <span class="s2">&quot;nonpar&quot;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="p">(</span>
            <span class="s2">&quot;Variant (on sex chromosome) falls outside a pseudoautosomal region&quot;</span>
        <span class="p">)</span>
    <span class="p">},</span>
    <span class="s2">&quot;non_par&quot;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="p">(</span>
            <span class="s2">&quot;Variant (on sex chromosome) falls outside a pseudoautosomal region&quot;</span>
        <span class="p">)</span>
    <span class="p">},</span>
    <span class="s2">&quot;segdup&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="s2">&quot;Variant falls within a segmental duplication region&quot;</span><span class="p">},</span>
    <span class="s2">&quot;fail_interval_qc&quot;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="p">(</span>
            <span class="s2">&quot;Less than 85 percent of samples meet 20X coverage if variant is in&quot;</span>
            <span class="s2">&quot; autosomal or PAR regions or 10X coverage for non-PAR regions of&quot;</span>
            <span class="s2">&quot; chromosomes X and Y.&quot;</span>
        <span class="p">)</span>
    <span class="p">},</span>
    <span class="s2">&quot;outside_ukb_capture_region&quot;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="s2">&quot;Variant falls outside of UK Biobank exome capture regions.&quot;</span>
    <span class="p">},</span>
    <span class="s2">&quot;outside_broad_capture_region&quot;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="s2">&quot;Variant falls outside of Broad exome capture regions.&quot;</span>
    <span class="p">},</span>
    <span class="s2">&quot;rf_positive_label&quot;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="p">(</span>
            <span class="s2">&quot;Variant was labelled as a positive example for training of random forest&quot;</span>
            <span class="s2">&quot; model&quot;</span>
        <span class="p">)</span>
    <span class="p">},</span>
    <span class="s2">&quot;rf_negative_label&quot;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="p">(</span>
            <span class="s2">&quot;Variant was labelled as a negative example for training of random forest&quot;</span>
            <span class="s2">&quot; model&quot;</span>
        <span class="p">)</span>
    <span class="p">},</span>
    <span class="s2">&quot;rf_label&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="s2">&quot;Random forest training label&quot;</span><span class="p">},</span>
    <span class="s2">&quot;rf_train&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="s2">&quot;Variant was used in training random forest model&quot;</span><span class="p">},</span>
    <span class="s2">&quot;rf_tp_probability&quot;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="p">(</span>
            <span class="s2">&quot;Probability of a called variant being a true variant as determined by&quot;</span>
            <span class="s2">&quot; random forest model&quot;</span>
        <span class="p">)</span>
    <span class="p">},</span>
    <span class="s2">&quot;transmitted_singleton&quot;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="p">(</span>
            <span class="s2">&quot;Variant was a callset-wide doubleton that was transmitted within a family&quot;</span>
            <span class="s2">&quot; from a parent to a child (i.e., a singleton amongst unrelated samples in&quot;</span>
            <span class="s2">&quot; cohort)&quot;</span>
        <span class="p">)</span>
    <span class="p">},</span>
    <span class="s2">&quot;sibling_singleton&quot;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="p">(</span>
            <span class="s2">&quot;Variant was a callset-wide doubleton that was present only in two siblings&quot;</span>
            <span class="s2">&quot; (i.e., a singleton amongst unrelated samples in cohort).&quot;</span>
        <span class="p">)</span>
    <span class="p">},</span>
    <span class="s2">&quot;original_alleles&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="s2">&quot;Alleles before splitting multiallelics&quot;</span><span class="p">},</span>
    <span class="s2">&quot;variant_type&quot;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="s2">&quot;Variant type (snv, indel, multi-snv, multi-indel, or mixed)&quot;</span>
    <span class="p">},</span>
    <span class="s2">&quot;allele_type&quot;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="s2">&quot;Allele type (snv, insertion, deletion, or mixed)&quot;</span><span class="p">,</span>
    <span class="p">},</span>
    <span class="s2">&quot;n_alt_alleles&quot;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s2">&quot;Number&quot;</span><span class="p">:</span> <span class="s2">&quot;1&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="s2">&quot;Total number of alternate alleles observed at variant locus&quot;</span><span class="p">,</span>
    <span class="p">},</span>
    <span class="s2">&quot;was_mixed&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="s2">&quot;Variant type was mixed&quot;</span><span class="p">},</span>
    <span class="s2">&quot;has_star&quot;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="p">(</span>
            <span class="s2">&quot;Variant locus coincides with a spanning deletion (represented by a star)&quot;</span>
            <span class="s2">&quot; observed elsewhere in the callset&quot;</span>
        <span class="p">)</span>
    <span class="p">},</span>
    <span class="s2">&quot;AS_pab_max&quot;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s2">&quot;Number&quot;</span><span class="p">:</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="p">(</span>
            <span class="s2">&quot;Maximum p-value over callset for binomial test of observed allele balance&quot;</span>
            <span class="s2">&quot; for a heterozygous genotype, given expectation of 0.5&quot;</span>
        <span class="p">),</span>
    <span class="p">},</span>
    <span class="s2">&quot;monoallelic&quot;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="s2">&quot;All samples are homozygous alternate for the variant&quot;</span>
    <span class="p">},</span>
    <span class="s2">&quot;only_het&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="s2">&quot;All samples are heterozygous for the variant&quot;</span><span class="p">},</span>
    <span class="s2">&quot;QUALapprox&quot;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s2">&quot;Number&quot;</span><span class="p">:</span> <span class="s2">&quot;1&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="s2">&quot;Sum of PL[0] values; used to approximate the QUAL score&quot;</span><span class="p">,</span>
    <span class="p">},</span>
    <span class="s2">&quot;AS_SB_TABLE&quot;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s2">&quot;Number&quot;</span><span class="p">:</span> <span class="s2">&quot;.&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="p">(</span>
            <span class="s2">&quot;Allele-specific forward/reverse read counts for strand bias tests&quot;</span>
        <span class="p">),</span>
    <span class="p">},</span>
<span class="p">}</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">Dictionary used during VCF export to export row (variant) annotations.</span>
<span class="sd">&quot;&quot;&quot;</span>

<span class="n">JOINT_REGION_FLAGS_INFO_DICT</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;fail_interval_qc&quot;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="p">(</span>
            <span class="s2">&quot;Less than 85 percent of samples meet 20X coverage if variant is in&quot;</span>
            <span class="s2">&quot; autosomal or PAR regions or 10X coverage for non-PAR regions of&quot;</span>
            <span class="s2">&quot; chromosomes X and Y.&quot;</span>
        <span class="p">)</span>
    <span class="p">},</span>
    <span class="s2">&quot;outside_ukb_capture_region&quot;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="s2">&quot;Variant falls outside of the UK Biobank exome capture regions.&quot;</span>
    <span class="p">},</span>
    <span class="s2">&quot;outside_broad_capture_region&quot;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="s2">&quot;Variant falls outside of the Broad exome capture regions.&quot;</span>
    <span class="p">},</span>
    <span class="s2">&quot;outside_ukb_calling_region&quot;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="p">(</span>
            <span class="s2">&quot;Variant falls outside of the UK Biobank exome capture regions plus 150 bp&quot;</span>
            <span class="s2">&quot; padding.&quot;</span>
        <span class="p">)</span>
    <span class="p">},</span>
    <span class="s2">&quot;outside_broad_calling_region&quot;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="p">(</span>
            <span class="s2">&quot;Variant falls outside of the Broad exome capture regions plus 150 bp&quot;</span>
            <span class="s2">&quot; padding.&quot;</span>
        <span class="p">)</span>
    <span class="p">},</span>
    <span class="s2">&quot;not_called_in_exomes&quot;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="s2">&quot;Variant was not called in the gnomAD exomes.&quot;</span>
    <span class="p">},</span>
    <span class="s2">&quot;not_called_in_genomes&quot;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="s2">&quot;Variant was not called in the gnomAD genomes.&quot;</span>
    <span class="p">},</span>
<span class="p">}</span>


<span class="n">IN_SILICO_ANNOTATIONS_INFO_DICT</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;cadd_raw_score&quot;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s2">&quot;Number&quot;</span><span class="p">:</span> <span class="s2">&quot;1&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="p">(</span>
            <span class="s2">&quot;Raw CADD scores are interpretable as the extent to which the annotation&quot;</span>
            <span class="s2">&quot; profile for a given variant suggests that the variant is likely to be&quot;</span>
            <span class="s2">&quot; &#39;observed&#39; (negative values) vs &#39;simulated&#39; (positive values). Larger&quot;</span>
            <span class="s2">&quot; values are more deleterious.&quot;</span>
        <span class="p">),</span>
    <span class="p">},</span>
    <span class="s2">&quot;cadd_phred&quot;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s2">&quot;Number&quot;</span><span class="p">:</span> <span class="s2">&quot;1&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="p">(</span>
            <span class="s2">&quot;Cadd Phred-like scores (&#39;scaled C-scores&#39;) ranging from 1 to 99, based on&quot;</span>
            <span class="s2">&quot; the rank of each variant relative to all possible 8.6 billion&quot;</span>
            <span class="s2">&quot; substitutions in the human reference genome. Larger values are more&quot;</span>
            <span class="s2">&quot; deleterious.&quot;</span>
        <span class="p">),</span>
    <span class="p">},</span>
    <span class="s2">&quot;revel_max&quot;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s2">&quot;Number&quot;</span><span class="p">:</span> <span class="s2">&quot;1&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="p">(</span>
            <span class="s2">&quot;The maximum REVEL score at a site&#39;s MANE Select or canonical&quot;</span>
            <span class="s2">&quot; transcript. It&#39;s an ensemble score for predicting the pathogenicity of&quot;</span>
            <span class="s2">&quot; missense variants (based on 13 other variant predictors). Scores ranges&quot;</span>
            <span class="s2">&quot; from 0 to 1. Variants with higher scores are predicted to be more likely&quot;</span>
            <span class="s2">&quot; to be deleterious.&quot;</span>
        <span class="p">),</span>
    <span class="p">},</span>
    <span class="s2">&quot;spliceai_ds_max&quot;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s2">&quot;Number&quot;</span><span class="p">:</span> <span class="s2">&quot;1&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="p">(</span>
            <span class="s2">&quot;Illumina&#39;s SpliceAI max delta score; interpreted as the probability of the&quot;</span>
            <span class="s2">&quot; variant being splice-altering.&quot;</span>
        <span class="p">),</span>
    <span class="p">},</span>
    <span class="s2">&quot;pangolin_largest_ds&quot;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s2">&quot;Number&quot;</span><span class="p">:</span> <span class="s2">&quot;1&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="p">(</span>
            <span class="s2">&quot;Pangolin&#39;s largest delta score across 2 splicing consequences, which&quot;</span>
            <span class="s2">&quot; reflects the probability of the variant being splice-altering&quot;</span>
        <span class="p">),</span>
    <span class="p">},</span>
    <span class="s2">&quot;phylop&quot;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s2">&quot;Number&quot;</span><span class="p">:</span> <span class="s2">&quot;1&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="p">(</span>
            <span class="s2">&quot;Base-wise conservation score across the 241 placental mammals in the&quot;</span>
            <span class="s2">&quot; Zoonomia project. Score ranges from -20 to 9.28, and reflects&quot;</span>
            <span class="s2">&quot; acceleration (faster evolution than expected under neutral drift,&quot;</span>
            <span class="s2">&quot; assigned negative scores) as well as conservation (slower than expected&quot;</span>
            <span class="s2">&quot; evolution, assigned positive scores).&quot;</span>
        <span class="p">),</span>
    <span class="p">},</span>
    <span class="s2">&quot;sift_max&quot;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s2">&quot;Number&quot;</span><span class="p">:</span> <span class="s2">&quot;1&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="p">(</span>
            <span class="s2">&quot;Score reflecting the scaled probability of the amino acid substitution&quot;</span>
            <span class="s2">&quot; being tolerated, ranging from 0 to 1. Scores below 0.05 are predicted to&quot;</span>
            <span class="s2">&quot; impact protein function. We prioritize max scores for MANE Select&quot;</span>
            <span class="s2">&quot; transcripts where possible and otherwise report a score for the canonical&quot;</span>
            <span class="s2">&quot; transcript.&quot;</span>
        <span class="p">),</span>
    <span class="p">},</span>
    <span class="s2">&quot;polyphen_max&quot;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s2">&quot;Number&quot;</span><span class="p">:</span> <span class="s2">&quot;1&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="p">(</span>
            <span class="s2">&quot;Score that predicts the possible impact of an amino acid substitution on&quot;</span>
            <span class="s2">&quot; the structure and function of a human protein, ranging from 0.0&quot;</span>
            <span class="s2">&quot; (tolerated) to 1.0 (deleterious).  We prioritize max scores for MANE&quot;</span>
            <span class="s2">&quot; Select transcripts where possible and otherwise report a score for the&quot;</span>
            <span class="s2">&quot; canonical transcript.&quot;</span>
        <span class="p">),</span>
    <span class="p">},</span>
<span class="p">}</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">Dictionary with in silico score descriptions to include in the VCF INFO header.</span>
<span class="sd">&quot;&quot;&quot;</span>


<span class="n">VRS_FIELDS_DICT</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;VRS_Allele_IDs&quot;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s2">&quot;Number&quot;</span><span class="p">:</span> <span class="s2">&quot;R&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="p">(</span>
            <span class="s2">&quot;The computed identifiers for the GA4GH VRS Alleles corresponding to the&quot;</span>
            <span class="s2">&quot; values in the REF and ALT fields&quot;</span>
        <span class="p">),</span>
    <span class="p">},</span>
    <span class="s2">&quot;VRS_Starts&quot;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s2">&quot;Number&quot;</span><span class="p">:</span> <span class="s2">&quot;R&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="p">(</span>
            <span class="s2">&quot;Interresidue coordinates used as the location starts for the GA4GH VRS&quot;</span>
            <span class="s2">&quot; Alleles corresponding to the values in the REF and ALT fields&quot;</span>
        <span class="p">),</span>
    <span class="p">},</span>
    <span class="s2">&quot;VRS_Ends&quot;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s2">&quot;Number&quot;</span><span class="p">:</span> <span class="s2">&quot;R&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="p">(</span>
            <span class="s2">&quot;Interresidue coordinates used as the location ends for the GA4GH VRS&quot;</span>
            <span class="s2">&quot; Alleles corresponding to the values in the REF and ALT fields&quot;</span>
        <span class="p">),</span>
    <span class="p">},</span>
    <span class="s2">&quot;VRS_States&quot;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s2">&quot;Number&quot;</span><span class="p">:</span> <span class="s2">&quot;.&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="p">(</span>
            <span class="s2">&quot;The literal sequence states used for the GA4GH VRS Alleles corresponding&quot;</span>
            <span class="s2">&quot; to the values in the REF and ALT fields&quot;</span>
        <span class="p">),</span>
    <span class="p">},</span>
<span class="p">}</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">Dictionary with VRS annotations to include in the VCF INFO field and VCF header.</span>
<span class="sd">&quot;&quot;&quot;</span>


<span class="n">ENTRIES</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;GT&quot;</span><span class="p">,</span> <span class="s2">&quot;GQ&quot;</span><span class="p">,</span> <span class="s2">&quot;DP&quot;</span><span class="p">,</span> <span class="s2">&quot;AD&quot;</span><span class="p">,</span> <span class="s2">&quot;MIN_DP&quot;</span><span class="p">,</span> <span class="s2">&quot;PGT&quot;</span><span class="p">,</span> <span class="s2">&quot;PID&quot;</span><span class="p">,</span> <span class="s2">&quot;PL&quot;</span><span class="p">,</span> <span class="s2">&quot;SB&quot;</span><span class="p">]</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">Densified entries to be selected during VCF export.</span>
<span class="sd">&quot;&quot;&quot;</span>

<span class="n">SPARSE_ENTRIES</span> <span class="o">=</span> <span class="p">[</span>
    <span class="s2">&quot;END&quot;</span><span class="p">,</span>
    <span class="s2">&quot;DP&quot;</span><span class="p">,</span>
    <span class="s2">&quot;GQ&quot;</span><span class="p">,</span>
    <span class="s2">&quot;LA&quot;</span><span class="p">,</span>
    <span class="s2">&quot;LAD&quot;</span><span class="p">,</span>
    <span class="s2">&quot;LGT&quot;</span><span class="p">,</span>
    <span class="s2">&quot;LPGT&quot;</span><span class="p">,</span>
    <span class="s2">&quot;LPL&quot;</span><span class="p">,</span>
    <span class="s2">&quot;MIN_DP&quot;</span><span class="p">,</span>
    <span class="s2">&quot;PID&quot;</span><span class="p">,</span>
    <span class="s2">&quot;RGQ&quot;</span><span class="p">,</span>
    <span class="s2">&quot;SB&quot;</span><span class="p">,</span>
<span class="p">]</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">Sparse entries to be selected and densified during VCF export.</span>
<span class="sd">&quot;&quot;&quot;</span>

<span class="n">FORMAT_DICT</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;GT&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="s2">&quot;Genotype&quot;</span><span class="p">,</span> <span class="s2">&quot;Number&quot;</span><span class="p">:</span> <span class="s2">&quot;1&quot;</span><span class="p">,</span> <span class="s2">&quot;Type&quot;</span><span class="p">:</span> <span class="s2">&quot;String&quot;</span><span class="p">},</span>
    <span class="s2">&quot;AD&quot;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="s2">&quot;Allelic depths for the ref and alt alleles in the order listed&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Number&quot;</span><span class="p">:</span> <span class="s2">&quot;R&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Type&quot;</span><span class="p">:</span> <span class="s2">&quot;Integer&quot;</span><span class="p">,</span>
    <span class="p">},</span>
    <span class="s2">&quot;DP&quot;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="p">(</span>
            <span class="s2">&quot;Approximate read depth (reads with MQ=255 or with bad mates are filtered)&quot;</span>
        <span class="p">),</span>
        <span class="s2">&quot;Number&quot;</span><span class="p">:</span> <span class="s2">&quot;1&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Type&quot;</span><span class="p">:</span> <span class="s2">&quot;Integer&quot;</span><span class="p">,</span>
    <span class="p">},</span>
    <span class="s2">&quot;GQ&quot;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="p">(</span>
            <span class="s2">&quot;Phred-scaled confidence that the genotype assignment is correct. Value is&quot;</span>
            <span class="s2">&quot; the difference between the second lowest PL and the lowest PL (always&quot;</span>
            <span class="s2">&quot; normalized to 0).&quot;</span>
        <span class="p">),</span>
        <span class="s2">&quot;Number&quot;</span><span class="p">:</span> <span class="s2">&quot;1&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Type&quot;</span><span class="p">:</span> <span class="s2">&quot;Integer&quot;</span><span class="p">,</span>
    <span class="p">},</span>
    <span class="s2">&quot;MIN_DP&quot;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="s2">&quot;Minimum DP observed within the GVCF block&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Number&quot;</span><span class="p">:</span> <span class="s2">&quot;1&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Type&quot;</span><span class="p">:</span> <span class="s2">&quot;Integer&quot;</span><span class="p">,</span>
    <span class="p">},</span>
    <span class="s2">&quot;PGT&quot;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="p">(</span>
            <span class="s2">&quot;Physical phasing haplotype information, describing how the alternate&quot;</span>
            <span class="s2">&quot; alleles are phased in relation to one another&quot;</span>
        <span class="p">),</span>
        <span class="s2">&quot;Number&quot;</span><span class="p">:</span> <span class="s2">&quot;1&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Type&quot;</span><span class="p">:</span> <span class="s2">&quot;String&quot;</span><span class="p">,</span>
    <span class="p">},</span>
    <span class="s2">&quot;PID&quot;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="p">(</span>
            <span class="s2">&quot;Physical phasing ID information, where each unique ID within a given&quot;</span>
            <span class="s2">&quot; sample (but not across samples) connects records within a phasing group&quot;</span>
        <span class="p">),</span>
        <span class="s2">&quot;Number&quot;</span><span class="p">:</span> <span class="s2">&quot;1&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Type&quot;</span><span class="p">:</span> <span class="s2">&quot;String&quot;</span><span class="p">,</span>
    <span class="p">},</span>
    <span class="s2">&quot;PL&quot;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="p">(</span>
            <span class="s2">&quot;Normalized, phred-scaled likelihoods for genotypes as defined in the VCF&quot;</span>
            <span class="s2">&quot; specification&quot;</span>
        <span class="p">),</span>
        <span class="s2">&quot;Number&quot;</span><span class="p">:</span> <span class="s2">&quot;G&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Type&quot;</span><span class="p">:</span> <span class="s2">&quot;Integer&quot;</span><span class="p">,</span>
    <span class="p">},</span>
    <span class="s2">&quot;SB&quot;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="p">(</span>
            <span class="s2">&quot;Per-sample component statistics which comprise the Fisher&#39;s exact test to&quot;</span>
            <span class="s2">&quot; detect strand bias. Values are: depth of reference allele on forward&quot;</span>
            <span class="s2">&quot; strand, depth of reference allele on reverse strand, depth of alternate&quot;</span>
            <span class="s2">&quot; allele on forward strand, depth of alternate allele on reverse strand.&quot;</span>
        <span class="p">),</span>
        <span class="s2">&quot;Number&quot;</span><span class="p">:</span> <span class="s2">&quot;4&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Type&quot;</span><span class="p">:</span> <span class="s2">&quot;Integer&quot;</span><span class="p">,</span>
    <span class="p">},</span>
    <span class="s2">&quot;RGQ&quot;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="p">(</span>
            <span class="s2">&quot;Unconditional reference genotype confidence, encoded as a phred&quot;</span>
            <span class="s2">&quot; quality -10*log10 p(genotype call is wrong)&quot;</span>
        <span class="p">),</span>
        <span class="s2">&quot;Number&quot;</span><span class="p">:</span> <span class="s2">&quot;1&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Type&quot;</span><span class="p">:</span> <span class="s2">&quot;Integer&quot;</span><span class="p">,</span>
    <span class="p">},</span>
<span class="p">}</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">Dictionary used during VCF export to export MatrixTable entries.</span>
<span class="sd">&quot;&quot;&quot;</span>

<span class="n">JOINT_FILTERS_INFO_DICT</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;exomes_filters&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="s2">&quot;Filters&#39; values from the exomes dataset.&quot;</span><span class="p">},</span>
    <span class="s2">&quot;genomes_filters&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="s2">&quot;Filters&#39; values from the genomes dataset.&quot;</span><span class="p">},</span>
<span class="p">}</span>


<div class="viewcode-block" id="adjust_vcf_incompatible_types"><a class="viewcode-back" href="../../../api_reference/utils/vcf.html#gnomad.utils.vcf.adjust_vcf_incompatible_types">[docs]</a><span class="k">def</span> <span class="nf">adjust_vcf_incompatible_types</span><span class="p">(</span>
    <span class="n">ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span>
    <span class="n">pipe_delimited_annotations</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="n">INFO_VCF_AS_PIPE_DELIMITED_FIELDS</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Create a Table ready for vcf export.</span>

<span class="sd">    In particular, the following conversions are done:</span>
<span class="sd">        - All int64 are coerced to int32</span>
<span class="sd">        - Fields specified by `pipe_delimited_annotations` are converted from arrays to pipe-delimited strings</span>

<span class="sd">    :param ht: Input Table.</span>
<span class="sd">    :param pipe_delimited_annotations: List of info fields (they must be fields of the ht.info Struct).</span>
<span class="sd">    :return: Table ready for VCF export.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="nf">get_pipe_expr</span><span class="p">(</span><span class="n">array_expr</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">ArrayExpression</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">StringExpression</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">hl</span><span class="o">.</span><span class="n">delimit</span><span class="p">(</span><span class="n">array_expr</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">or_else</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">str</span><span class="p">(</span><span class="n">x</span><span class="p">),</span> <span class="s2">&quot;&quot;</span><span class="p">)),</span> <span class="s2">&quot;|&quot;</span><span class="p">)</span>

    <span class="c1"># Make sure the HT is keyed by locus, alleles</span>
    <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">key_by</span><span class="p">(</span><span class="s2">&quot;locus&quot;</span><span class="p">,</span> <span class="s2">&quot;alleles&quot;</span><span class="p">)</span>

    <span class="n">info_type_convert_expr</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="c1"># Convert int64 fields to int32 (int64 isn&#39;t supported by VCF)</span>
    <span class="k">for</span> <span class="n">f</span><span class="p">,</span> <span class="n">ft</span> <span class="ow">in</span> <span class="n">ht</span><span class="o">.</span><span class="n">info</span><span class="o">.</span><span class="n">dtype</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="k">if</span> <span class="n">ft</span> <span class="o">==</span> <span class="n">hl</span><span class="o">.</span><span class="n">dtype</span><span class="p">(</span><span class="s2">&quot;int64&quot;</span><span class="p">):</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
                <span class="s2">&quot;Coercing field info.</span><span class="si">%s</span><span class="s2"> from int64 to int32 for VCF output. Value&quot;</span>
                <span class="s2">&quot; will be capped at int32 max value.&quot;</span><span class="p">,</span>
                <span class="n">f</span><span class="p">,</span>
            <span class="p">)</span>
            <span class="n">info_type_convert_expr</span><span class="o">.</span><span class="n">update</span><span class="p">(</span>
                <span class="p">{</span>
                    <span class="n">f</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">or_missing</span><span class="p">(</span>
                        <span class="n">hl</span><span class="o">.</span><span class="n">is_defined</span><span class="p">(</span><span class="n">ht</span><span class="o">.</span><span class="n">info</span><span class="p">[</span><span class="n">f</span><span class="p">]),</span>
                        <span class="n">hl</span><span class="o">.</span><span class="n">int32</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">min</span><span class="p">(</span><span class="mi">2</span><span class="o">**</span><span class="mi">31</span> <span class="o">-</span> <span class="mi">1</span><span class="p">,</span> <span class="n">ht</span><span class="o">.</span><span class="n">info</span><span class="p">[</span><span class="n">f</span><span class="p">])),</span>
                    <span class="p">)</span>
                <span class="p">}</span>
            <span class="p">)</span>
        <span class="k">elif</span> <span class="n">ft</span> <span class="o">==</span> <span class="n">hl</span><span class="o">.</span><span class="n">dtype</span><span class="p">(</span><span class="s2">&quot;array&lt;int64&gt;&quot;</span><span class="p">):</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
                <span class="s2">&quot;Coercing field info.</span><span class="si">%s</span><span class="s2"> from array&lt;int64&gt; to array&lt;int32&gt; for VCF&quot;</span>
                <span class="s2">&quot; output. Array values will be capped at int32 max value.&quot;</span><span class="p">,</span>
                <span class="n">f</span><span class="p">,</span>
            <span class="p">)</span>
            <span class="n">info_type_convert_expr</span><span class="o">.</span><span class="n">update</span><span class="p">(</span>
                <span class="p">{</span>
                    <span class="n">f</span><span class="p">:</span> <span class="n">ht</span><span class="o">.</span><span class="n">info</span><span class="p">[</span><span class="n">f</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">(</span>
                        <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">or_missing</span><span class="p">(</span>
                            <span class="n">hl</span><span class="o">.</span><span class="n">is_defined</span><span class="p">(</span><span class="n">x</span><span class="p">),</span> <span class="n">hl</span><span class="o">.</span><span class="n">int32</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">min</span><span class="p">(</span><span class="mi">2</span><span class="o">**</span><span class="mi">31</span> <span class="o">-</span> <span class="mi">1</span><span class="p">,</span> <span class="n">x</span><span class="p">))</span>
                        <span class="p">)</span>
                    <span class="p">)</span>
                <span class="p">}</span>
            <span class="p">)</span>

    <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span><span class="n">info</span><span class="o">=</span><span class="n">ht</span><span class="o">.</span><span class="n">info</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span><span class="o">**</span><span class="n">info_type_convert_expr</span><span class="p">))</span>

    <span class="n">info_expr</span> <span class="o">=</span> <span class="p">{}</span>

    <span class="c1"># Make sure to pipe-delimit fields that need to.</span>
    <span class="c1"># Note: the expr needs to be prefixed by &quot;|&quot; because GATK expect one value for the ref (always empty)</span>
    <span class="c1"># Note2: this doesn&#39;t produce the correct annotation for AS_SB_TABLE, it</span>
    <span class="c1"># is handled below</span>
    <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">pipe_delimited_annotations</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">ht</span><span class="o">.</span><span class="n">info</span> <span class="ow">and</span> <span class="n">f</span> <span class="o">!=</span> <span class="s2">&quot;AS_SB_TABLE&quot;</span><span class="p">:</span>
            <span class="n">info_expr</span><span class="p">[</span><span class="n">f</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;|&quot;</span> <span class="o">+</span> <span class="n">get_pipe_expr</span><span class="p">(</span><span class="n">ht</span><span class="o">.</span><span class="n">info</span><span class="p">[</span><span class="n">f</span><span class="p">])</span>

    <span class="c1"># Flatten SB if it is an array of arrays</span>
    <span class="k">if</span> <span class="s2">&quot;SB&quot;</span> <span class="ow">in</span> <span class="n">ht</span><span class="o">.</span><span class="n">info</span> <span class="ow">and</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">ht</span><span class="o">.</span><span class="n">info</span><span class="o">.</span><span class="n">SB</span><span class="p">,</span> <span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">ArrayNumericExpression</span><span class="p">):</span>
        <span class="n">info_expr</span><span class="p">[</span><span class="s2">&quot;SB&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">info</span><span class="o">.</span><span class="n">SB</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">ht</span><span class="o">.</span><span class="n">info</span><span class="o">.</span><span class="n">SB</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>

    <span class="k">if</span> <span class="s2">&quot;AS_SB_TABLE&quot;</span> <span class="ow">in</span> <span class="n">ht</span><span class="o">.</span><span class="n">info</span><span class="p">:</span>
        <span class="n">info_expr</span><span class="p">[</span><span class="s2">&quot;AS_SB_TABLE&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">get_pipe_expr</span><span class="p">(</span>
            <span class="n">ht</span><span class="o">.</span><span class="n">info</span><span class="o">.</span><span class="n">AS_SB_TABLE</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">delimit</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="s2">&quot;,&quot;</span><span class="p">))</span>
        <span class="p">)</span>

    <span class="c1"># Annotate with new expression</span>
    <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span><span class="n">info</span><span class="o">=</span><span class="n">ht</span><span class="o">.</span><span class="n">info</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span><span class="o">**</span><span class="n">info_expr</span><span class="p">))</span>

    <span class="k">return</span> <span class="n">ht</span></div>


<div class="viewcode-block" id="make_label_combos"><a class="viewcode-back" href="../../../api_reference/utils/vcf.html#gnomad.utils.vcf.make_label_combos">[docs]</a><span class="k">def</span> <span class="nf">make_label_combos</span><span class="p">(</span>
    <span class="n">label_groups</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]],</span>
    <span class="n">sort_order</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="n">SORT_ORDER</span><span class="p">,</span>
    <span class="n">label_delimiter</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;_&quot;</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Make combinations of all possible labels for a supplied dictionary of label groups.</span>

<span class="sd">    For example, if label_groups is {&quot;sex&quot;: [&quot;XY&quot;, &quot;XX&quot;], &quot;gen_anc&quot;: [&quot;afr&quot;, &quot;nfe&quot;, &quot;amr&quot;]},</span>
<span class="sd">    this function will return [&quot;afr_XY&quot;, &quot;afr_XX&quot;, &quot;nfe_XY&quot;, &quot;nfe_XX&quot;, &quot;amr_XY&quot;, &quot;amr_XX&quot;]</span>

<span class="sd">    :param label_groups: Dictionary containing an entry for each label group, where key is the name of the grouping,</span>
<span class="sd">        e.g. &quot;sex&quot; or &quot;gen_anc&quot;, and value is a list of all possible values for that grouping (e.g. [&quot;XY&quot;, &quot;XX&quot;] or [&quot;afr&quot;, &quot;nfe&quot;, &quot;amr&quot;]).</span>
<span class="sd">    :param sort_order: List containing order to sort label group combinations. Default is SORT_ORDER.</span>
<span class="sd">    :param label_delimiter: String to use as delimiter when making group label combinations.</span>
<span class="sd">    :return: list of all possible combinations of values for the supplied label groupings.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">copy_label_groups</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="n">label_groups</span><span class="p">)</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">copy_label_groups</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
        <span class="k">return</span> <span class="p">[</span><span class="n">item</span> <span class="k">for</span> <span class="n">sublist</span> <span class="ow">in</span> <span class="n">copy_label_groups</span><span class="o">.</span><span class="n">values</span><span class="p">()</span> <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">sublist</span><span class="p">]</span>
    <span class="n">anchor_group</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">copy_label_groups</span><span class="o">.</span><span class="n">keys</span><span class="p">(),</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">sort_order</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">x</span><span class="p">))[</span>
        <span class="mi">0</span>
    <span class="p">]</span>
    <span class="n">anchor_val</span> <span class="o">=</span> <span class="n">copy_label_groups</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="n">anchor_group</span><span class="p">)</span>
    <span class="n">combos</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">x</span><span class="p">,</span> <span class="n">y</span> <span class="ow">in</span> <span class="n">itertools</span><span class="o">.</span><span class="n">product</span><span class="p">(</span>
        <span class="n">anchor_val</span><span class="p">,</span>
        <span class="n">make_label_combos</span><span class="p">(</span><span class="n">copy_label_groups</span><span class="p">,</span> <span class="n">label_delimiter</span><span class="o">=</span><span class="n">label_delimiter</span><span class="p">),</span>
    <span class="p">):</span>
        <span class="n">combos</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">x</span><span class="si">}{</span><span class="n">label_delimiter</span><span class="si">}{</span><span class="n">y</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">combos</span></div>


<div class="viewcode-block" id="index_globals"><a class="viewcode-back" href="../../../api_reference/utils/vcf.html#gnomad.utils.vcf.index_globals">[docs]</a><span class="k">def</span> <span class="nf">index_globals</span><span class="p">(</span>
    <span class="n">globals_array</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]],</span>
    <span class="n">label_groups</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]],</span>
    <span class="n">label_delimiter</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;_&quot;</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">int</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Create a dictionary keyed by the specified label groupings with values describing the corresponding index of each grouping entry in the meta_array annotation.</span>

<span class="sd">    :param globals_array: Ordered list containing dictionary entries describing all the grouping combinations contained in the globals_array annotation.</span>
<span class="sd">       Keys are the grouping type (e.g., &#39;group&#39;, &#39;gen_anc&#39;, &#39;sex&#39;) and values are the grouping attribute (e.g., &#39;adj&#39;, &#39;eas&#39;, &#39;XY&#39;).</span>
<span class="sd">    :param label_groups: Dictionary containing an entry for each label group, where key is the name of the grouping,</span>
<span class="sd">        e.g. &quot;sex&quot; or &quot;gen_anc&quot;, and value is a list of all possible values for that grouping (e.g. [&quot;XY&quot;, &quot;XX&quot;] or [&quot;afr&quot;, &quot;nfe&quot;, &quot;amr&quot;])</span>
<span class="sd">    :param label_delimiter: String used as delimiter when making group label combinations.</span>
<span class="sd">    :return: Dictionary keyed by specified label grouping combinations, with values describing the corresponding index</span>
<span class="sd">        of each grouping entry in the globals</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">combos</span> <span class="o">=</span> <span class="n">make_label_combos</span><span class="p">(</span><span class="n">label_groups</span><span class="p">,</span> <span class="n">label_delimiter</span><span class="o">=</span><span class="n">label_delimiter</span><span class="p">)</span>
    <span class="n">index_dict</span> <span class="o">=</span> <span class="p">{}</span>

    <span class="k">for</span> <span class="n">combo</span> <span class="ow">in</span> <span class="n">combos</span><span class="p">:</span>
        <span class="n">combo_fields</span> <span class="o">=</span> <span class="n">combo</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">label_delimiter</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">globals_array</span><span class="p">):</span>
            <span class="k">if</span> <span class="nb">set</span><span class="p">(</span><span class="n">v</span><span class="o">.</span><span class="n">values</span><span class="p">())</span> <span class="o">==</span> <span class="nb">set</span><span class="p">(</span><span class="n">combo_fields</span><span class="p">):</span>
                <span class="n">index_dict</span><span class="o">.</span><span class="n">update</span><span class="p">({</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">combo</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">:</span> <span class="n">i</span><span class="p">})</span>
    <span class="k">return</span> <span class="n">index_dict</span></div>


<div class="viewcode-block" id="make_combo_header_text"><a class="viewcode-back" href="../../../api_reference/utils/vcf.html#gnomad.utils.vcf.make_combo_header_text">[docs]</a><span class="k">def</span> <span class="nf">make_combo_header_text</span><span class="p">(</span>
    <span class="n">preposition</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">combo_dict</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">],</span>
    <span class="n">gen_anc_names</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">],</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Programmatically generate text to populate the VCF header description for a given variant annotation with specific groupings and subset.</span>

<span class="sd">    For example, if preposition is &quot;for&quot;, group_types is [&quot;group&quot;, &quot;gen_anc&quot;, &quot;sex&quot;], and combo_fields is [&quot;adj&quot;, &quot;afr&quot;, &quot;XX&quot;],</span>
<span class="sd">    this function will return the string &quot; for XX samples in the African-American/African genetic ancestry group&quot;.</span>

<span class="sd">    :param preposition: Relevant preposition to precede automatically generated text.</span>
<span class="sd">    :param combo_dict: Dict with grouping types as keys and values for grouping type as values. This function generates text for these values.</span>
<span class="sd">        Possible grouping types are: &quot;group&quot;, &quot;gen_anc&quot;, &quot;sex&quot;, and &quot;subgroup&quot;.</span>
<span class="sd">        Example input: {&quot;gen_anc&quot;: &quot;afr&quot;, &quot;sex&quot;: &quot;XX&quot;}</span>
<span class="sd">    :param gen_anc_names: Dict with global genetic ancestry group names (keys) and genetic ancestry group descriptions (values).</span>
<span class="sd">    :return: String with automatically generated description text for a given set of combo fields.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">header_text</span> <span class="o">=</span> <span class="s2">&quot; &quot;</span> <span class="o">+</span> <span class="n">preposition</span>

    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">combo_dict</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">combo_dict</span><span class="p">[</span><span class="s2">&quot;group&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;adj&quot;</span><span class="p">:</span>
            <span class="k">return</span> <span class="s2">&quot;&quot;</span>

    <span class="k">if</span> <span class="s2">&quot;sex&quot;</span> <span class="ow">in</span> <span class="n">combo_dict</span><span class="p">:</span>
        <span class="n">header_text</span> <span class="o">=</span> <span class="n">header_text</span> <span class="o">+</span> <span class="s2">&quot; &quot;</span> <span class="o">+</span> <span class="n">combo_dict</span><span class="p">[</span><span class="s2">&quot;sex&quot;</span><span class="p">]</span>

    <span class="n">header_text</span> <span class="o">=</span> <span class="n">header_text</span> <span class="o">+</span> <span class="s2">&quot; samples&quot;</span>

    <span class="k">if</span> <span class="s2">&quot;subgrp&quot;</span> <span class="ow">in</span> <span class="n">combo_dict</span> <span class="ow">or</span> <span class="s2">&quot;gen_anc&quot;</span> <span class="ow">in</span> <span class="n">combo_dict</span><span class="p">:</span>
        <span class="k">if</span> <span class="s2">&quot;subgrp&quot;</span> <span class="ow">in</span> <span class="n">combo_dict</span><span class="p">:</span>
            <span class="n">header_text</span> <span class="o">=</span> <span class="p">(</span>
                <span class="n">header_text</span>
                <span class="o">+</span> <span class="sa">f</span><span class="s2">&quot; in the </span><span class="si">{</span><span class="n">gen_anc_names</span><span class="p">[</span><span class="n">combo_dict</span><span class="p">[</span><span class="s1">&#39;subgrp&#39;</span><span class="p">]]</span><span class="si">}</span><span class="s2"> genetic ancestry subgroup&quot;</span>
            <span class="p">)</span>

        <span class="k">else</span><span class="p">:</span>
            <span class="n">header_text</span> <span class="o">=</span> <span class="p">(</span>
                <span class="n">header_text</span>
                <span class="o">+</span> <span class="sa">f</span><span class="s2">&quot; in the </span><span class="si">{</span><span class="n">gen_anc_names</span><span class="p">[</span><span class="n">combo_dict</span><span class="p">[</span><span class="s1">&#39;gen_anc&#39;</span><span class="p">]]</span><span class="si">}</span><span class="s2"> genetic ancestry group&quot;</span>
            <span class="p">)</span>

    <span class="k">if</span> <span class="s2">&quot;group&quot;</span> <span class="ow">in</span> <span class="n">combo_dict</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">combo_dict</span><span class="p">[</span><span class="s2">&quot;group&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;raw&quot;</span><span class="p">:</span>
            <span class="n">header_text</span> <span class="o">=</span> <span class="n">header_text</span> <span class="o">+</span> <span class="s2">&quot;, before removing low-confidence genotypes&quot;</span>

    <span class="k">return</span> <span class="n">header_text</span></div>


<div class="viewcode-block" id="create_label_groups"><a class="viewcode-back" href="../../../api_reference/utils/vcf.html#gnomad.utils.vcf.create_label_groups">[docs]</a><span class="k">def</span> <span class="nf">create_label_groups</span><span class="p">(</span>
    <span class="n">gen_ancs</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span>
    <span class="n">sexes</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="n">SEXES</span><span class="p">,</span>
    <span class="n">all_groups</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="n">GROUPS</span><span class="p">,</span>
    <span class="n">gen_anc_sex_groups</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;adj&quot;</span><span class="p">],</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Generate a list of label group dictionaries needed to populate info dictionary.</span>

<span class="sd">    Label dictionaries are passed as input to `make_info_dict`.</span>

<span class="sd">    :param gen_ancs: List of genetic ancestry group names.</span>
<span class="sd">    :param sexes: List of sample sexes.</span>
<span class="sd">    :param all_groups: List of data types (raw, adj). Default is `GROUPS`, which is [&quot;raw&quot;, &quot;adj&quot;].</span>
<span class="sd">    :param gen_anc_sex_groups: List of data types (raw, adj) to populate with gen_ancs and sexes. Default is [&quot;adj&quot;].</span>
<span class="sd">    :return: List of label group dictionaries.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="p">[</span>
        <span class="c1"># This is to capture raw frequency fields, which are</span>
        <span class="c1"># not stratified by sex or genetic ancestry group (e.g., only AC_raw</span>
        <span class="c1"># exists, not AC_XX_raw)</span>
        <span class="nb">dict</span><span class="p">(</span><span class="n">group</span><span class="o">=</span><span class="n">all_groups</span><span class="p">),</span>
        <span class="nb">dict</span><span class="p">(</span><span class="n">group</span><span class="o">=</span><span class="n">gen_anc_sex_groups</span><span class="p">,</span> <span class="n">sex</span><span class="o">=</span><span class="n">sexes</span><span class="p">),</span>
        <span class="nb">dict</span><span class="p">(</span><span class="n">group</span><span class="o">=</span><span class="n">gen_anc_sex_groups</span><span class="p">,</span> <span class="n">gen_anc</span><span class="o">=</span><span class="n">gen_ancs</span><span class="p">),</span>
        <span class="nb">dict</span><span class="p">(</span><span class="n">group</span><span class="o">=</span><span class="n">gen_anc_sex_groups</span><span class="p">,</span> <span class="n">gen_anc</span><span class="o">=</span><span class="n">gen_ancs</span><span class="p">,</span> <span class="n">sex</span><span class="o">=</span><span class="n">sexes</span><span class="p">),</span>
    <span class="p">]</span></div>


<div class="viewcode-block" id="make_info_dict"><a class="viewcode-back" href="../../../api_reference/utils/vcf.html#gnomad.utils.vcf.make_info_dict">[docs]</a><span class="k">def</span> <span class="nf">make_info_dict</span><span class="p">(</span>
    <span class="n">prefix</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="p">,</span>
    <span class="n">suffix</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="p">,</span>
    <span class="n">prefix_before_metric</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
    <span class="n">gen_anc_names</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="n">GEN_ANC_NAMES</span><span class="p">,</span>
    <span class="n">label_groups</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">label_delimiter</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;_&quot;</span><span class="p">,</span>
    <span class="n">bin_edges</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">faf</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="n">grpmax</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="n">fafmax</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="n">callstats</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="n">freq_ctt</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="n">freq_cmh</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="n">freq_stat_union</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="n">description_text</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="p">,</span>
    <span class="n">age_hist_distribution</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">sort_order</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="n">SORT_ORDER</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Generate dictionary of Number and Description attributes of VCF INFO fields.</span>

<span class="sd">    Used to populate the INFO fields of the VCF header during export.</span>

<span class="sd">    Creates:</span>
<span class="sd">        - INFO fields for age histograms (bin freq, n_smaller, and n_larger for heterozygous and homozygous variant carriers)</span>
<span class="sd">        - INFO fields for grpmax AC, AN, AF, nhomalt, and grpmax genetic ancestry group</span>
<span class="sd">        - INFO fields for AC, AN, AF, nhomalt for each combination of sample genetic ancestry group, sex, and subgroup, both for adj and raw data</span>
<span class="sd">        - INFO fields for filtering allele frequency (faf) annotations</span>

<span class="sd">    :param prefix: Prefix string for data, e.g. &quot;gnomAD&quot;. Default is empty string.</span>
<span class="sd">    :param suffix: Suffix string for data, e.g. &quot;gnomAD&quot;. Default is empty string.</span>
<span class="sd">    :param prefix_before_metric: Whether prefix should be added before the metric (AC, AN, AF, nhomalt, faf95, faf99) in INFO field. Default is True.</span>
<span class="sd">    :param gen_anc_names: Dict with global genetic ancestry group names (keys) and genetic ancestry group descriptions (values). Default is GEN_ANC_NAMES.</span>
<span class="sd">    :param label_groups: Dictionary containing an entry for each label group, where key is the name of the grouping,</span>
<span class="sd">        e.g. &quot;sex&quot; or &quot;gen_anc&quot;, and value is a list of all possible values for that grouping (e.g. [&quot;XY&quot;, &quot;XX&quot;] or [&quot;afr&quot;, &quot;nfe&quot;, &quot;amr&quot;]).</span>
<span class="sd">    :param label_delimiter: String to use as delimiter when making group label combinations.</span>
<span class="sd">    :param bin_edges: Dictionary keyed by annotation type, with values that reflect the bin edges corresponding to the annotation.</span>
<span class="sd">    :param faf: If True, use alternate logic to auto-populate dictionary values associated with filter allele frequency annotations.</span>
<span class="sd">    :param grpmax: If True, use alternate logic to auto-populate dictionary values associated with grpmax annotations.</span>
<span class="sd">    :param fafmax: If True, use alternate logic to auto-populate dictionary values associated with fafmax annotations.</span>
<span class="sd">    :param callstats: If True, use alternate logic to auto-populate dictionary values associated with callstats annotations.</span>
<span class="sd">    :param freq_ctt: If True, use alternate logic to auto-populate dictionary values associated with frequency contingency table test (CTT) annotations.</span>
<span class="sd">    :param freq_cmh: If True, use alternate logic to auto-populate dictionary values associated with frequency Cochran-Mantel-Haenszel (CMH) annotations.</span>
<span class="sd">    :param freq_stat_union: If True, use alternate logic to auto-populate dictionary values associated with the union of the contingency table and Cochran-Mantel-Haenszel tests.</span>
<span class="sd">    :param description_text: Optional text to append to the end of descriptions. Needs to start with a space if specified.</span>
<span class="sd">    :param str age_hist_distribution: Pipe-delimited string of overall age distribution.</span>
<span class="sd">    :param sort_order: List containing order to sort label group combinations. Default is SORT_ORDER.</span>
<span class="sd">    :return: Dictionary keyed by VCF INFO annotations, where values are dictionaries of Number and Description attributes.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">prefix</span> <span class="o">!=</span> <span class="s2">&quot;&quot;</span><span class="p">:</span>
        <span class="n">prefix</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">prefix</span><span class="si">}{</span><span class="n">label_delimiter</span><span class="si">}</span><span class="s2">&quot;</span>
    <span class="k">if</span> <span class="n">suffix</span> <span class="o">!=</span> <span class="s2">&quot;&quot;</span><span class="p">:</span>
        <span class="n">suffix</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">label_delimiter</span><span class="si">}{</span><span class="n">suffix</span><span class="si">}</span><span class="s2">&quot;</span>

    <span class="n">info_dict</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">()</span>

    <span class="k">if</span> <span class="n">age_hist_distribution</span><span class="p">:</span>
        <span class="n">age_hist_dict</span> <span class="o">=</span> <span class="p">{</span>
            <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2">age_hist_het_bin_freq</span><span class="si">{</span><span class="n">suffix</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">:</span> <span class="p">{</span>
                <span class="s2">&quot;Number&quot;</span><span class="p">:</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;Histogram of ages of heterozygous individuals</span><span class="si">{</span><span class="n">description_text</span><span class="si">}</span><span class="s2">;&quot;</span>
                    <span class="sa">f</span><span class="s2">&quot; bin edges are: </span><span class="si">{</span><span class="n">bin_edges</span><span class="p">[</span><span class="s1">&#39;het&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">; total number of individuals&quot;</span>
                    <span class="sa">f</span><span class="s2">&quot; of any genotype bin: </span><span class="si">{</span><span class="n">age_hist_distribution</span><span class="si">}</span><span class="s2">&quot;</span>
                <span class="p">),</span>
            <span class="p">},</span>
            <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2">age_hist_het_n_smaller</span><span class="si">{</span><span class="n">suffix</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">:</span> <span class="p">{</span>
                <span class="s2">&quot;Number&quot;</span><span class="p">:</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="p">(</span>
                    <span class="s2">&quot;Count of age values falling below lowest histogram bin edge for&quot;</span>
                    <span class="sa">f</span><span class="s2">&quot; heterozygous individuals</span><span class="si">{</span><span class="n">description_text</span><span class="si">}</span><span class="s2">&quot;</span>
                <span class="p">),</span>
            <span class="p">},</span>
            <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2">age_hist_het_n_larger</span><span class="si">{</span><span class="n">suffix</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">:</span> <span class="p">{</span>
                <span class="s2">&quot;Number&quot;</span><span class="p">:</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="p">(</span>
                    <span class="s2">&quot;Count of age values falling above highest histogram bin edge for&quot;</span>
                    <span class="sa">f</span><span class="s2">&quot; heterozygous individuals</span><span class="si">{</span><span class="n">description_text</span><span class="si">}</span><span class="s2">&quot;</span>
                <span class="p">),</span>
            <span class="p">},</span>
            <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2">age_hist_hom_bin_freq</span><span class="si">{</span><span class="n">suffix</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">:</span> <span class="p">{</span>
                <span class="s2">&quot;Number&quot;</span><span class="p">:</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="p">(</span>
                    <span class="s2">&quot;Histogram of ages of homozygous alternate&quot;</span>
                    <span class="sa">f</span><span class="s2">&quot; individuals</span><span class="si">{</span><span class="n">description_text</span><span class="si">}</span><span class="s2">; bin edges are:&quot;</span>
                    <span class="sa">f</span><span class="s2">&quot; </span><span class="si">{</span><span class="n">bin_edges</span><span class="p">[</span><span class="s1">&#39;hom&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">; total number of individuals of any genotype&quot;</span>
                    <span class="sa">f</span><span class="s2">&quot; bin: </span><span class="si">{</span><span class="n">age_hist_distribution</span><span class="si">}</span><span class="s2">&quot;</span>
                <span class="p">),</span>
            <span class="p">},</span>
            <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2">age_hist_hom_n_smaller</span><span class="si">{</span><span class="n">suffix</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">:</span> <span class="p">{</span>
                <span class="s2">&quot;Number&quot;</span><span class="p">:</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="p">(</span>
                    <span class="s2">&quot;Count of age values falling below lowest histogram bin edge for&quot;</span>
                    <span class="sa">f</span><span class="s2">&quot; homozygous alternate individuals</span><span class="si">{</span><span class="n">description_text</span><span class="si">}</span><span class="s2">&quot;</span>
                <span class="p">),</span>
            <span class="p">},</span>
            <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2">age_hist_hom_n_larger</span><span class="si">{</span><span class="n">suffix</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">:</span> <span class="p">{</span>
                <span class="s2">&quot;Number&quot;</span><span class="p">:</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="p">(</span>
                    <span class="s2">&quot;Count of age values falling above highest histogram bin edge for&quot;</span>
                    <span class="sa">f</span><span class="s2">&quot; homozygous alternate individuals</span><span class="si">{</span><span class="n">description_text</span><span class="si">}</span><span class="s2">&quot;</span>
                <span class="p">),</span>
            <span class="p">},</span>
        <span class="p">}</span>
        <span class="n">info_dict</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">age_hist_dict</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">grpmax</span><span class="p">:</span>
        <span class="n">grpmax_dict</span> <span class="o">=</span> <span class="p">{</span>
            <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2">grpmax</span><span class="si">{</span><span class="n">suffix</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">:</span> <span class="p">{</span>
                <span class="s2">&quot;Number&quot;</span><span class="p">:</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="p">(</span>
                    <span class="s2">&quot;Genetic ancestry group with the maximum allele&quot;</span>
                    <span class="sa">f</span><span class="s2">&quot; frequency</span><span class="si">{</span><span class="n">description_text</span><span class="si">}</span><span class="s2">&quot;</span>
                <span class="p">),</span>
            <span class="p">},</span>
            <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2">AC</span><span class="si">{</span><span class="n">label_delimiter</span><span class="si">}</span><span class="s2">grpmax</span><span class="si">{</span><span class="n">suffix</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">:</span> <span class="p">{</span>
                <span class="s2">&quot;Number&quot;</span><span class="p">:</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="p">(</span>
                    <span class="s2">&quot;Allele count in the genetic ancestry group with the maximum allele&quot;</span>
                    <span class="sa">f</span><span class="s2">&quot; frequency</span><span class="si">{</span><span class="n">description_text</span><span class="si">}</span><span class="s2">&quot;</span>
                <span class="p">),</span>
            <span class="p">},</span>
            <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2">AN</span><span class="si">{</span><span class="n">label_delimiter</span><span class="si">}</span><span class="s2">grpmax</span><span class="si">{</span><span class="n">suffix</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">:</span> <span class="p">{</span>
                <span class="s2">&quot;Number&quot;</span><span class="p">:</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="p">(</span>
                    <span class="s2">&quot;Total number of alleles in the genetic ancestry group with the&quot;</span>
                    <span class="sa">f</span><span class="s2">&quot; maximum allele frequency</span><span class="si">{</span><span class="n">description_text</span><span class="si">}</span><span class="s2">&quot;</span>
                <span class="p">),</span>
            <span class="p">},</span>
            <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2">AF</span><span class="si">{</span><span class="n">label_delimiter</span><span class="si">}</span><span class="s2">grpmax</span><span class="si">{</span><span class="n">suffix</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">:</span> <span class="p">{</span>
                <span class="s2">&quot;Number&quot;</span><span class="p">:</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="p">(</span>
                    <span class="s2">&quot;Maximum allele frequency across genetic ancestry&quot;</span>
                    <span class="sa">f</span><span class="s2">&quot; groups</span><span class="si">{</span><span class="n">description_text</span><span class="si">}</span><span class="s2">&quot;</span>
                <span class="p">),</span>
            <span class="p">},</span>
            <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2">nhomalt</span><span class="si">{</span><span class="n">label_delimiter</span><span class="si">}</span><span class="s2">grpmax</span><span class="si">{</span><span class="n">suffix</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">:</span> <span class="p">{</span>
                <span class="s2">&quot;Number&quot;</span><span class="p">:</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="p">(</span>
                    <span class="s2">&quot;Count of homozygous individuals in the genetic ancestry group&quot;</span>
                    <span class="sa">f</span><span class="s2">&quot; with the maximum allele frequency</span><span class="si">{</span><span class="n">description_text</span><span class="si">}</span><span class="s2">&quot;</span>
                <span class="p">),</span>
            <span class="p">},</span>
        <span class="p">}</span>
        <span class="n">info_dict</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">grpmax_dict</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">fafmax</span><span class="p">:</span>
        <span class="n">fafmax_dict</span> <span class="o">=</span> <span class="p">{</span>
            <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2">fafmax</span><span class="si">{</span><span class="n">label_delimiter</span><span class="si">}</span><span class="s2">faf95</span><span class="si">{</span><span class="n">label_delimiter</span><span class="si">}</span><span class="s2">max</span><span class="si">{</span><span class="n">suffix</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">:</span> <span class="p">{</span>
                <span class="s2">&quot;Number&quot;</span><span class="p">:</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="p">(</span>
                    <span class="s2">&quot;Maximum filtering allele frequency (using Poisson 95% CI)&quot;</span>
                    <span class="sa">f</span><span class="s2">&quot; across genetic ancestry groups</span><span class="si">{</span><span class="n">description_text</span><span class="si">}</span><span class="s2">&quot;</span>
                <span class="p">),</span>
            <span class="p">},</span>
            <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2">fafmax</span><span class="si">{</span><span class="n">label_delimiter</span><span class="si">}</span><span class="s2">faf95</span><span class="si">{</span><span class="n">label_delimiter</span><span class="si">}</span><span class="s2">max</span><span class="si">{</span><span class="n">label_delimiter</span><span class="si">}</span><span class="s2">gen</span><span class="si">{</span><span class="n">label_delimiter</span><span class="si">}</span><span class="s2">anc</span><span class="si">{</span><span class="n">suffix</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">:</span> <span class="p">{</span>
                <span class="s2">&quot;Number&quot;</span><span class="p">:</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="p">(</span>
                    <span class="s2">&quot;Genetic ancestry group with maximum filtering allele&quot;</span>
                    <span class="sa">f</span><span class="s2">&quot; frequency (using Poisson 95% CI)</span><span class="si">{</span><span class="n">description_text</span><span class="si">}</span><span class="s2">&quot;</span>
                <span class="p">),</span>
            <span class="p">},</span>
            <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2">fafmax</span><span class="si">{</span><span class="n">label_delimiter</span><span class="si">}</span><span class="s2">faf99</span><span class="si">{</span><span class="n">label_delimiter</span><span class="si">}</span><span class="s2">max</span><span class="si">{</span><span class="n">suffix</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">:</span> <span class="p">{</span>
                <span class="s2">&quot;Number&quot;</span><span class="p">:</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="p">(</span>
                    <span class="s2">&quot;Maximum filtering allele frequency (using Poisson 99% CI)&quot;</span>
                    <span class="sa">f</span><span class="s2">&quot; across genetic ancestry groups</span><span class="si">{</span><span class="n">description_text</span><span class="si">}</span><span class="s2">&quot;</span>
                <span class="p">),</span>
            <span class="p">},</span>
            <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2">fafmax</span><span class="si">{</span><span class="n">label_delimiter</span><span class="si">}</span><span class="s2">faf99</span><span class="si">{</span><span class="n">label_delimiter</span><span class="si">}</span><span class="s2">max</span><span class="si">{</span><span class="n">label_delimiter</span><span class="si">}</span><span class="s2">gen</span><span class="si">{</span><span class="n">label_delimiter</span><span class="si">}</span><span class="s2">anc</span><span class="si">{</span><span class="n">suffix</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">:</span> <span class="p">{</span>
                <span class="s2">&quot;Number&quot;</span><span class="p">:</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="p">(</span>
                    <span class="s2">&quot;Genetic ancestry group with maximum filtering allele&quot;</span>
                    <span class="sa">f</span><span class="s2">&quot; frequency (using Poisson 99% CI)</span><span class="si">{</span><span class="n">description_text</span><span class="si">}</span><span class="s2">&quot;</span>
                <span class="p">),</span>
            <span class="p">},</span>
        <span class="p">}</span>

        <span class="n">info_dict</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">fafmax_dict</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">callstats</span> <span class="ow">or</span> <span class="n">faf</span> <span class="ow">or</span> <span class="n">freq_ctt</span><span class="p">:</span>
        <span class="n">group_types</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">label_groups</span><span class="o">.</span><span class="n">keys</span><span class="p">(),</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">sort_order</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">x</span><span class="p">))</span>
        <span class="n">combos</span> <span class="o">=</span> <span class="n">make_label_combos</span><span class="p">(</span><span class="n">label_groups</span><span class="p">,</span> <span class="n">label_delimiter</span><span class="o">=</span><span class="n">label_delimiter</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">combo</span> <span class="ow">in</span> <span class="n">combos</span><span class="p">:</span>
            <span class="n">combo_fields</span> <span class="o">=</span> <span class="n">combo</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">label_delimiter</span><span class="p">)</span>
            <span class="n">group_dict</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">group_types</span><span class="p">,</span> <span class="n">combo_fields</span><span class="p">))</span>

            <span class="n">for_combo</span> <span class="o">=</span> <span class="n">make_combo_header_text</span><span class="p">(</span><span class="s2">&quot;for&quot;</span><span class="p">,</span> <span class="n">group_dict</span><span class="p">,</span> <span class="n">gen_anc_names</span><span class="p">)</span>
            <span class="n">in_combo</span> <span class="o">=</span> <span class="n">make_combo_header_text</span><span class="p">(</span><span class="s2">&quot;in&quot;</span><span class="p">,</span> <span class="n">group_dict</span><span class="p">,</span> <span class="n">gen_anc_names</span><span class="p">)</span>

            <span class="n">metrics</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;AC&quot;</span><span class="p">,</span> <span class="s2">&quot;AN&quot;</span><span class="p">,</span> <span class="s2">&quot;AF&quot;</span><span class="p">,</span> <span class="s2">&quot;nhomalt&quot;</span><span class="p">,</span> <span class="s2">&quot;faf95&quot;</span><span class="p">,</span> <span class="s2">&quot;faf99&quot;</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">freq_ctt</span><span class="p">:</span>
                <span class="n">metrics</span> <span class="o">+=</span> <span class="p">[</span><span class="s2">&quot;CTT_odds_ratio&quot;</span><span class="p">,</span> <span class="s2">&quot;CTT_p_value&quot;</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">prefix_before_metric</span><span class="p">:</span>
                <span class="n">metric_label_dict</span> <span class="o">=</span> <span class="p">{</span>
                    <span class="n">metric</span><span class="p">:</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">prefix</span><span class="si">}{</span><span class="n">metric</span><span class="si">}{</span><span class="n">label_delimiter</span><span class="si">}{</span><span class="n">combo</span><span class="si">}{</span><span class="n">suffix</span><span class="si">}</span><span class="s2">&quot;</span>
                    <span class="k">for</span> <span class="n">metric</span> <span class="ow">in</span> <span class="n">metrics</span>
                <span class="p">}</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">metric_label_dict</span> <span class="o">=</span> <span class="p">{</span>
                    <span class="n">metric</span><span class="p">:</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">metric</span><span class="si">}{</span><span class="n">label_delimiter</span><span class="si">}{</span><span class="n">prefix</span><span class="si">}{</span><span class="n">combo</span><span class="si">}{</span><span class="n">suffix</span><span class="si">}</span><span class="s2">&quot;</span>
                    <span class="k">for</span> <span class="n">metric</span> <span class="ow">in</span> <span class="n">metrics</span>
                <span class="p">}</span>

            <span class="k">if</span> <span class="n">callstats</span><span class="p">:</span>
                <span class="n">combo_dict</span> <span class="o">=</span> <span class="p">{</span>
                    <span class="n">metric_label_dict</span><span class="p">[</span><span class="s2">&quot;AC&quot;</span><span class="p">]:</span> <span class="p">{</span>
                        <span class="s2">&quot;Number&quot;</span><span class="p">:</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span>
                        <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="p">(</span>
                            <span class="sa">f</span><span class="s2">&quot;Alternate allele count</span><span class="si">{</span><span class="n">for_combo</span><span class="si">}{</span><span class="n">description_text</span><span class="si">}</span><span class="s2">&quot;</span>
                        <span class="p">),</span>
                    <span class="p">},</span>
                    <span class="n">metric_label_dict</span><span class="p">[</span><span class="s2">&quot;AN&quot;</span><span class="p">]:</span> <span class="p">{</span>
                        <span class="s2">&quot;Number&quot;</span><span class="p">:</span> <span class="s2">&quot;1&quot;</span><span class="p">,</span>
                        <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="p">(</span>
                            <span class="sa">f</span><span class="s2">&quot;Total number of alleles</span><span class="si">{</span><span class="n">in_combo</span><span class="si">}{</span><span class="n">description_text</span><span class="si">}</span><span class="s2">&quot;</span>
                        <span class="p">),</span>
                    <span class="p">},</span>
                    <span class="n">metric_label_dict</span><span class="p">[</span><span class="s2">&quot;AF&quot;</span><span class="p">]:</span> <span class="p">{</span>
                        <span class="s2">&quot;Number&quot;</span><span class="p">:</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span>
                        <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="p">(</span>
                            <span class="sa">f</span><span class="s2">&quot;Alternate allele frequency</span><span class="si">{</span><span class="n">in_combo</span><span class="si">}{</span><span class="n">description_text</span><span class="si">}</span><span class="s2">&quot;</span>
                        <span class="p">),</span>
                    <span class="p">},</span>
                    <span class="n">metric_label_dict</span><span class="p">[</span><span class="s2">&quot;nhomalt&quot;</span><span class="p">]:</span> <span class="p">{</span>
                        <span class="s2">&quot;Number&quot;</span><span class="p">:</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span>
                        <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="p">(</span>
                            <span class="s2">&quot;Count of homozygous&quot;</span>
                            <span class="sa">f</span><span class="s2">&quot; individuals</span><span class="si">{</span><span class="n">in_combo</span><span class="si">}{</span><span class="n">description_text</span><span class="si">}</span><span class="s2">&quot;</span>
                        <span class="p">),</span>
                    <span class="p">},</span>
                <span class="p">}</span>
            <span class="k">elif</span> <span class="n">faf</span><span class="p">:</span>
                <span class="k">if</span> <span class="p">(</span><span class="s2">&quot;XX&quot;</span> <span class="ow">in</span> <span class="n">combo_fields</span><span class="p">)</span> <span class="o">|</span> <span class="p">(</span><span class="s2">&quot;XY&quot;</span> <span class="ow">in</span> <span class="n">combo_fields</span><span class="p">):</span>
                    <span class="n">faf_description_text</span> <span class="o">=</span> <span class="p">(</span>
                        <span class="n">description_text</span> <span class="o">+</span> <span class="s2">&quot; in non-PAR regions of sex chromosomes only&quot;</span>
                    <span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">faf_description_text</span> <span class="o">=</span> <span class="n">description_text</span>
                <span class="n">combo_dict</span> <span class="o">=</span> <span class="p">{</span>
                    <span class="n">metric_label_dict</span><span class="p">[</span><span class="s2">&quot;faf95&quot;</span><span class="p">]:</span> <span class="p">{</span>
                        <span class="s2">&quot;Number&quot;</span><span class="p">:</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span>
                        <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="p">(</span>
                            <span class="s2">&quot;Filtering allele frequency (using Poisson 95%&quot;</span>
                            <span class="sa">f</span><span class="s2">&quot; CI)</span><span class="si">{</span><span class="n">for_combo</span><span class="si">}{</span><span class="n">faf_description_text</span><span class="si">}</span><span class="s2">&quot;</span>
                        <span class="p">),</span>
                    <span class="p">},</span>
                    <span class="n">metric_label_dict</span><span class="p">[</span><span class="s2">&quot;faf99&quot;</span><span class="p">]:</span> <span class="p">{</span>
                        <span class="s2">&quot;Number&quot;</span><span class="p">:</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span>
                        <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="p">(</span>
                            <span class="s2">&quot;Filtering allele frequency (using Poisson 99%&quot;</span>
                            <span class="sa">f</span><span class="s2">&quot; CI)</span><span class="si">{</span><span class="n">for_combo</span><span class="si">}{</span><span class="n">faf_description_text</span><span class="si">}</span><span class="s2">&quot;</span>
                        <span class="p">),</span>
                    <span class="p">},</span>
                <span class="p">}</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">combo_dict</span> <span class="o">=</span> <span class="p">{</span>
                    <span class="n">metric_label_dict</span><span class="p">[</span><span class="s2">&quot;CTT_odds_ratio&quot;</span><span class="p">]:</span> <span class="p">{</span>
                        <span class="s2">&quot;Number&quot;</span><span class="p">:</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span>
                        <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="p">(</span>
                            <span class="s2">&quot;Odds ratio from from Hail&#39;s contingency_table_test with&quot;</span>
                            <span class="s2">&quot; `min_cell_count=100` comparing allele frequencies&quot;</span>
                            <span class="sa">f</span><span class="s2">&quot; between exomes and genomes</span><span class="si">{</span><span class="n">for_combo</span><span class="si">}{</span><span class="n">description_text</span><span class="si">}</span><span class="s2">&quot;</span>
                        <span class="p">),</span>
                    <span class="p">},</span>
                    <span class="n">metric_label_dict</span><span class="p">[</span><span class="s2">&quot;CTT_p_value&quot;</span><span class="p">]:</span> <span class="p">{</span>
                        <span class="s2">&quot;Number&quot;</span><span class="p">:</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span>
                        <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="p">(</span>
                            <span class="s2">&quot;P-value from Hail&#39;s contingency_table_test with&quot;</span>
                            <span class="s2">&quot; `min_cell_count=100` comparing allele frequencies&quot;</span>
                            <span class="sa">f</span><span class="s2">&quot; between exomes and genomes</span><span class="si">{</span><span class="n">for_combo</span><span class="si">}{</span><span class="n">description_text</span><span class="si">}</span><span class="s2">&quot;</span>
                        <span class="p">),</span>
                    <span class="p">},</span>
                <span class="p">}</span>
            <span class="n">info_dict</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">combo_dict</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">freq_cmh</span><span class="p">:</span>
        <span class="n">cmh_dict</span> <span class="o">=</span> <span class="p">{</span>
            <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2">CMH_chisq</span><span class="si">{</span><span class="n">suffix</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">:</span> <span class="p">{</span>
                <span class="s2">&quot;Number&quot;</span><span class="p">:</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="p">(</span>
                    <span class="s2">&quot;Chi-squared test statistic from the Cochran-Mantel-Haenszel test&quot;</span>
                    <span class="s2">&quot; comparing allele frequencies between exomes and genomes&quot;</span>
                    <span class="sa">f</span><span class="s2">&quot; stratified by genetic ancestry group</span><span class="si">{</span><span class="n">description_text</span><span class="si">}</span><span class="s2">&quot;</span>
                <span class="p">),</span>
            <span class="p">},</span>
            <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2">CMH_p_value</span><span class="si">{</span><span class="n">suffix</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">:</span> <span class="p">{</span>
                <span class="s2">&quot;Number&quot;</span><span class="p">:</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="p">(</span>
                    <span class="s2">&quot;Odds ratio from Cochran-Mantel-Haenszel test comparing allele&quot;</span>
                    <span class="s2">&quot; frequencies between exomes and genomes stratified by genetic&quot;</span>
                    <span class="sa">f</span><span class="s2">&quot; ancestry group</span><span class="si">{</span><span class="n">description_text</span><span class="si">}</span><span class="s2">&quot;</span>
                <span class="p">),</span>
            <span class="p">},</span>
        <span class="p">}</span>
        <span class="n">info_dict</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">cmh_dict</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">freq_stat_union</span><span class="p">:</span>
        <span class="n">freq_stat_union_dict</span> <span class="o">=</span> <span class="p">{</span>
            <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2">stat_union_p_value</span><span class="si">{</span><span class="n">suffix</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">:</span> <span class="p">{</span>
                <span class="s2">&quot;Number&quot;</span><span class="p">:</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;p-value from the contingency table or Cochran-Mantel-Haenszel tests</span><span class="si">{</span><span class="n">description_text</span><span class="si">}</span><span class="s2">&quot;</span>
                <span class="p">),</span>
            <span class="p">},</span>
            <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2">stat_union_test_name</span><span class="si">{</span><span class="n">suffix</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">:</span> <span class="p">{</span>
                <span class="s2">&quot;Number&quot;</span><span class="p">:</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;Name of the test, either contingency_table_test or cochran_mantel_haenszel_test, used to compare allele frequencies between exomes and genomes</span><span class="si">{</span><span class="n">description_text</span><span class="si">}</span><span class="s2">&quot;</span>
                <span class="p">),</span>
            <span class="p">},</span>
            <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2">stat_union_gen_ancs</span><span class="si">{</span><span class="n">suffix</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">:</span> <span class="p">{</span>
                <span class="s2">&quot;Number&quot;</span><span class="p">:</span> <span class="s2">&quot;.&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;List of genetic ancestry groups included in the test. If stat_union_test_name is contingency_table_test, the length of gen_ancs is one and if stat_union_test_name is &#39;cochran_mantel_haenszel_test&#39;, the length of &#39;gen_ancs&#39; is greater than one</span><span class="si">{</span><span class="n">description_text</span><span class="si">}</span><span class="s2">&quot;</span>
                <span class="p">),</span>
            <span class="p">},</span>
        <span class="p">}</span>
        <span class="n">info_dict</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">freq_stat_union_dict</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">info_dict</span></div>


<div class="viewcode-block" id="add_as_info_dict"><a class="viewcode-back" href="../../../api_reference/utils/vcf.html#gnomad.utils.vcf.add_as_info_dict">[docs]</a><span class="k">def</span> <span class="nf">add_as_info_dict</span><span class="p">(</span>
    <span class="n">info_dict</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="n">INFO_DICT</span><span class="p">,</span> <span class="n">as_fields</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="n">AS_FIELDS</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Update info dictionary with allele-specific terms and their descriptions.</span>

<span class="sd">    Used in VCF export.</span>

<span class="sd">    :param info_dict: Dictionary containing site-level annotations and their descriptions. Default is INFO_DICT.</span>
<span class="sd">    :param as_fields: List containing allele-specific fields to be added to info_dict. Default is AS_FIELDS.</span>
<span class="sd">    :return: Dictionary with allele specific annotations, their descriptions, and their VCF number field.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">as_dict</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">field</span> <span class="ow">in</span> <span class="n">as_fields</span><span class="p">:</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="c1"># Strip AS_ from field name</span>
            <span class="n">site_field</span> <span class="o">=</span> <span class="n">field</span><span class="p">[</span><span class="mi">3</span><span class="p">:]</span>

            <span class="c1"># Get site description from info dictionary and make first letter lower case</span>
            <span class="n">first_letter</span> <span class="o">=</span> <span class="n">info_dict</span><span class="p">[</span><span class="n">site_field</span><span class="p">][</span><span class="s2">&quot;Description&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span>
            <span class="n">rest_of_description</span> <span class="o">=</span> <span class="n">info_dict</span><span class="p">[</span><span class="n">site_field</span><span class="p">][</span><span class="s2">&quot;Description&quot;</span><span class="p">][</span><span class="mi">1</span><span class="p">:]</span>

            <span class="n">as_dict</span><span class="p">[</span><span class="n">field</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>
            <span class="n">as_dict</span><span class="p">[</span><span class="n">field</span><span class="p">][</span><span class="s2">&quot;Number&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;A&quot;</span>
            <span class="n">as_dict</span><span class="p">[</span><span class="n">field</span><span class="p">][</span>
                <span class="s2">&quot;Description&quot;</span>
            <span class="p">]</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;Allele-specific </span><span class="si">{</span><span class="n">first_letter</span><span class="si">}{</span><span class="n">rest_of_description</span><span class="si">}</span><span class="s2">&quot;</span>

        <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2"> is not present in input info dictionary!&quot;</span><span class="p">,</span> <span class="n">field</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">as_dict</span></div>


<div class="viewcode-block" id="make_vcf_filter_dict"><a class="viewcode-back" href="../../../api_reference/utils/vcf.html#gnomad.utils.vcf.make_vcf_filter_dict">[docs]</a><span class="k">def</span> <span class="nf">make_vcf_filter_dict</span><span class="p">(</span>
    <span class="n">snp_cutoff</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">float</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">indel_cutoff</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">float</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">inbreeding_cutoff</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">float</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">variant_qc_filter</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;RF&quot;</span><span class="p">,</span>
    <span class="n">joint</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Generate dictionary of Number and Description attributes to be used in the VCF header, specifically for FILTER annotations.</span>

<span class="sd">    Generates descriptions for:</span>
<span class="sd">        - AC0 filter</span>
<span class="sd">        - InbreedingCoeff filter</span>
<span class="sd">        - Variant QC filter (RF or AS_VQSR)</span>
<span class="sd">        - PASS (passed all variant filters)</span>

<span class="sd">    :param snp_cutoff: Minimum SNP cutoff score from random forest model.</span>
<span class="sd">    :param indel_cutoff: Minimum indel cutoff score from random forest model.</span>
<span class="sd">    :param inbreeding_cutoff: Inbreeding coefficient hard cutoff.</span>
<span class="sd">    :param variant_qc_filter: Method used for variant QC filter. One of &#39;RF&#39; or &#39;AS_VQSR&#39;. Default is &#39;RF&#39;.</span>
<span class="sd">    :param joint: Whether the filter dictionary is for the joint release. Default is False.</span>
<span class="sd">    :return: Dictionary keyed by VCF FILTER annotations, where values are Dictionaries of Number and Description attributes.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">variant_qc_filter_dict</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;RF&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;Failed random forest filtering thresholds of </span><span class="si">{</span><span class="n">snp_cutoff</span><span class="si">}</span><span class="s2"> for SNPs&quot;</span>
                <span class="sa">f</span><span class="s2">&quot; and </span><span class="si">{</span><span class="n">indel_cutoff</span><span class="si">}</span><span class="s2"> for indels (probabilities of being a true&quot;</span>
                <span class="s2">&quot; positive variant)&quot;</span>
            <span class="p">)</span>
        <span class="p">},</span>
        <span class="s2">&quot;AS_VQSR&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;Failed VQSR filtering thresholds of </span><span class="si">{</span><span class="n">snp_cutoff</span><span class="si">}</span><span class="s2"> for SNPs and&quot;</span>
                <span class="sa">f</span><span class="s2">&quot; </span><span class="si">{</span><span class="n">indel_cutoff</span><span class="si">}</span><span class="s2"> for indels&quot;</span>
            <span class="p">)</span>
        <span class="p">},</span>
    <span class="p">}</span>

    <span class="k">if</span> <span class="n">variant_qc_filter</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">variant_qc_filter_dict</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">variant_qc_filter</span><span class="si">}</span><span class="s2"> is not a valid value for &#39;variant_qc_filter&#39;. It must&quot;</span>
            <span class="s2">&quot; be &#39;RF&#39; or &#39;AS_VQSR&#39;&quot;</span>
        <span class="p">)</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">joint</span> <span class="ow">and</span> <span class="p">(</span>
        <span class="n">snp_cutoff</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">or</span> <span class="n">indel_cutoff</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">or</span> <span class="n">inbreeding_cutoff</span> <span class="ow">is</span> <span class="kc">None</span>
    <span class="p">):</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
            <span class="s2">&quot;snp_cutoff, indel_cutoff, and inbreeding_cutoff must be specified to generate filter descriptions.&quot;</span>
        <span class="p">)</span>

    <span class="k">if</span> <span class="n">joint</span><span class="p">:</span>
        <span class="n">filter_dict</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;PASS&quot;</span><span class="p">:</span> <span class="p">{</span>
                <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="s2">&quot;Either passed all variant filters in both exomes and &quot;</span>
                <span class="s2">&quot;genomes, or passed all variant filters in either &quot;</span>
                <span class="s2">&quot;exomes or genomes while being absent from the other &quot;</span>
                <span class="s2">&quot;dataset&quot;</span>
            <span class="p">},</span>
            <span class="s2">&quot;EXOMES_FILTERED&quot;</span><span class="p">:</span> <span class="p">{</span>
                <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="s2">&quot;Failed variant filters in the exomes dataset and either &quot;</span>
                <span class="s2">&quot;passed all variant filters in the genomes dataset or the variant was &quot;</span>
                <span class="s2">&quot;not present in the genomes dataset. Refer to &#39;exomes_filters&#39; within &quot;</span>
                <span class="s2">&quot;INFO for more information&quot;</span>
            <span class="p">},</span>
            <span class="s2">&quot;GENOMES_FILTERED&quot;</span><span class="p">:</span> <span class="p">{</span>
                <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="s2">&quot;Failed variant filters in the genomes dataset and either &quot;</span>
                <span class="s2">&quot;passed all variant filters in the exomes dataset or the variant was &quot;</span>
                <span class="s2">&quot;not present in the exomes dataset. Refer to &#39;genomes_filters&#39; within &quot;</span>
                <span class="s2">&quot;INFO for more information&quot;</span>
            <span class="p">},</span>
            <span class="s2">&quot;BOTH_FILTERED&quot;</span><span class="p">:</span> <span class="p">{</span>
                <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="s2">&quot;Failed variant filters in both exomes and genomes datasets. &quot;</span>
                <span class="s2">&quot;Refer to &#39;exomes_filters&#39; and &#39;genomes_filters&#39; within INFO for more information&quot;</span>
            <span class="p">},</span>
        <span class="p">}</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">filter_dict</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;AC0&quot;</span><span class="p">:</span> <span class="p">{</span>
                <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="p">(</span>
                    <span class="s2">&quot;Allele count is zero after filtering out low-confidence genotypes (GQ&quot;</span>
                    <span class="s2">&quot; &lt; 20; DP &lt; 10; and AB &lt; 0.2 for het calls)&quot;</span>
                <span class="p">)</span>
            <span class="p">},</span>
            <span class="s2">&quot;InbreedingCoeff&quot;</span><span class="p">:</span> <span class="p">{</span>
                <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="sa">f</span><span class="s2">&quot;Inbreeding coefficient &lt; </span><span class="si">{</span><span class="n">inbreeding_cutoff</span><span class="si">}</span><span class="s2">&quot;</span>
            <span class="p">},</span>
            <span class="s2">&quot;PASS&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="s2">&quot;Passed all variant filters&quot;</span><span class="p">},</span>
            <span class="n">variant_qc_filter</span><span class="p">:</span> <span class="n">variant_qc_filter_dict</span><span class="p">[</span><span class="n">variant_qc_filter</span><span class="p">],</span>
        <span class="p">}</span>

    <span class="k">return</span> <span class="n">filter_dict</span></div>


<div class="viewcode-block" id="make_hist_bin_edges_expr"><a class="viewcode-back" href="../../../api_reference/utils/vcf.html#gnomad.utils.vcf.make_hist_bin_edges_expr">[docs]</a><span class="k">def</span> <span class="nf">make_hist_bin_edges_expr</span><span class="p">(</span>
    <span class="n">ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span>
    <span class="n">hists</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="n">HISTS</span><span class="p">,</span>
    <span class="n">ann_with_hists</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">prefix</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="p">,</span>
    <span class="n">label_delimiter</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;_&quot;</span><span class="p">,</span>
    <span class="n">include_age_hists</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Create dictionaries containing variant histogram annotations and their associated bin edges, formatted into a string separated by pipe delimiters.</span>

<span class="sd">    :param ht: Table containing histogram variant annotations.</span>
<span class="sd">    :param hists: List of variant histogram annotations. Default is HISTS.</span>
<span class="sd">    :param ann_with_hists: Name of row annotation containing histogram data. In exomes or</span>
<span class="sd">        genomes release HT, `histograms` is a row, but in the joint release HT, it&#39;s</span>
<span class="sd">        under the row of `exomes`, `genomes`, or `joint`.</span>
<span class="sd">    :param prefix: Prefix text for age histogram bin edges.  Default is empty string.</span>
<span class="sd">    :param label_delimiter: String used as delimiter between prefix and histogram annotation.</span>
<span class="sd">    :param include_age_hists: Include age histogram annotations.</span>
<span class="sd">    :return: Dictionary keyed by histogram annotation name, with corresponding</span>
<span class="sd">        reformatted bin edges for values.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Add underscore to prefix if it isn&#39;t empty</span>
    <span class="k">if</span> <span class="n">prefix</span><span class="p">:</span>
        <span class="n">prefix</span> <span class="o">+=</span> <span class="n">label_delimiter</span>

    <span class="n">edges_dict</span> <span class="o">=</span> <span class="p">{}</span>

    <span class="k">if</span> <span class="n">include_age_hists</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">call_type</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;het&quot;</span><span class="p">,</span> <span class="s2">&quot;hom&quot;</span><span class="p">]:</span>
            <span class="k">if</span> <span class="n">ann_with_hists</span><span class="p">:</span>
                <span class="n">bin_edges</span> <span class="o">=</span> <span class="p">(</span>
                    <span class="n">ht</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
                        <span class="n">hl</span><span class="o">.</span><span class="n">is_defined</span><span class="p">(</span>
                            <span class="n">ht</span><span class="p">[</span><span class="n">ann_with_hists</span><span class="p">]</span>
                            <span class="o">.</span><span class="n">histograms</span><span class="o">.</span><span class="n">age_hists</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;age_hist_</span><span class="si">{</span><span class="n">call_type</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">]</span>
                            <span class="o">.</span><span class="n">bin_edges</span>
                        <span class="p">)</span>
                    <span class="p">)[</span><span class="n">ann_with_hists</span><span class="p">]</span>
                    <span class="o">.</span><span class="n">histograms</span><span class="o">.</span><span class="n">age_hists</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;age_hist_</span><span class="si">{</span><span class="n">call_type</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">]</span>
                    <span class="o">.</span><span class="n">bin_edges</span><span class="o">.</span><span class="n">take</span><span class="p">(</span><span class="mi">1</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
                <span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">bin_edges</span> <span class="o">=</span> <span class="p">(</span>
                    <span class="n">ht</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
                        <span class="n">hl</span><span class="o">.</span><span class="n">is_defined</span><span class="p">(</span>
                            <span class="n">ht</span><span class="o">.</span><span class="n">histograms</span><span class="o">.</span><span class="n">age_hists</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;age_hist_</span><span class="si">{</span><span class="n">call_type</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">bin_edges</span>
                        <span class="p">)</span>
                    <span class="p">)</span>
                    <span class="o">.</span><span class="n">histograms</span><span class="o">.</span><span class="n">age_hists</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;age_hist_</span><span class="si">{</span><span class="n">call_type</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">]</span>
                    <span class="o">.</span><span class="n">bin_edges</span><span class="o">.</span><span class="n">take</span><span class="p">(</span><span class="mi">1</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
                <span class="p">)</span>

            <span class="k">if</span> <span class="n">bin_edges</span><span class="p">:</span>
                <span class="n">edges_dict</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">prefix</span><span class="si">}{</span><span class="n">call_type</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;|&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
                    <span class="nb">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">x</span><span class="si">:</span><span class="s2">.1f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">bin_edges</span><span class="p">)</span>
                <span class="p">)</span>

    <span class="k">for</span> <span class="n">hist</span> <span class="ow">in</span> <span class="n">hists</span><span class="p">:</span>
        <span class="c1"># Parse hists calculated on both raw and adj-filtered data</span>
        <span class="k">for</span> <span class="n">hist_type</span> <span class="ow">in</span> <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2">raw_qual_hists&quot;</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2">qual_hists&quot;</span><span class="p">]:</span>
            <span class="n">hist_name</span> <span class="o">=</span> <span class="n">hist</span> <span class="k">if</span> <span class="s2">&quot;raw&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">hist_type</span> <span class="k">else</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">prefix</span><span class="si">}{</span><span class="n">hist</span><span class="si">}</span><span class="s2">_raw&quot;</span>

            <span class="k">if</span> <span class="n">ann_with_hists</span><span class="p">:</span>
                <span class="n">bin_edges</span> <span class="o">=</span> <span class="p">(</span>
                    <span class="n">ht</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
                        <span class="n">hl</span><span class="o">.</span><span class="n">is_defined</span><span class="p">(</span>
                            <span class="n">ht</span><span class="p">[</span><span class="n">ann_with_hists</span><span class="p">]</span><span class="o">.</span><span class="n">histograms</span><span class="p">[</span><span class="n">hist_type</span><span class="p">][</span><span class="n">hist</span><span class="p">]</span><span class="o">.</span><span class="n">bin_edges</span>
                        <span class="p">)</span>
                    <span class="p">)[</span><span class="n">ann_with_hists</span><span class="p">]</span>
                    <span class="o">.</span><span class="n">histograms</span><span class="p">[</span><span class="n">hist_type</span><span class="p">][</span><span class="n">hist</span><span class="p">]</span>
                    <span class="o">.</span><span class="n">bin_edges</span><span class="o">.</span><span class="n">take</span><span class="p">(</span><span class="mi">1</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
                <span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">bin_edges</span> <span class="o">=</span> <span class="p">(</span>
                    <span class="n">ht</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">is_defined</span><span class="p">(</span><span class="n">ht</span><span class="o">.</span><span class="n">histograms</span><span class="p">[</span><span class="n">hist_type</span><span class="p">][</span><span class="n">hist</span><span class="p">]</span><span class="o">.</span><span class="n">bin_edges</span><span class="p">))</span>
                    <span class="o">.</span><span class="n">histograms</span><span class="p">[</span><span class="n">hist_type</span><span class="p">][</span><span class="n">hist</span><span class="p">]</span>
                    <span class="o">.</span><span class="n">bin_edges</span><span class="o">.</span><span class="n">take</span><span class="p">(</span><span class="mi">1</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
                <span class="p">)</span>
            <span class="k">if</span> <span class="n">bin_edges</span><span class="p">:</span>
                <span class="n">edges_dict</span><span class="p">[</span><span class="n">hist_name</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;|&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
                    <span class="nb">map</span><span class="p">(</span>
                        <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">x</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">if</span> <span class="s2">&quot;ab&quot;</span> <span class="ow">in</span> <span class="n">hist</span> <span class="k">else</span> <span class="nb">str</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">x</span><span class="p">)),</span> <span class="n">bin_edges</span>
                    <span class="p">)</span>
                <span class="p">)</span>

    <span class="k">return</span> <span class="n">edges_dict</span></div>


<div class="viewcode-block" id="make_hist_dict"><a class="viewcode-back" href="../../../api_reference/utils/vcf.html#gnomad.utils.vcf.make_hist_dict">[docs]</a><span class="k">def</span> <span class="nf">make_hist_dict</span><span class="p">(</span>
    <span class="n">bin_edges</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]],</span>
    <span class="n">adj</span><span class="p">:</span> <span class="nb">bool</span><span class="p">,</span>
    <span class="n">hist_metric_list</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="n">HISTS</span><span class="p">,</span>
    <span class="n">label_delimiter</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;_&quot;</span><span class="p">,</span>
    <span class="n">drop_n_smaller_larger</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="n">prefix</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="p">,</span>
    <span class="n">suffix</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="p">,</span>
    <span class="n">description_text</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Generate dictionary of Number and Description attributes to be used in the VCF header, specifically for histogram annotations.</span>

<span class="sd">    :param bin_edges: Dictionary keyed by histogram annotation name, with corresponding string-reformatted bin edges for values.</span>
<span class="sd">    :param adj: Whether to create a header dict for raw or adj quality histograms.</span>
<span class="sd">    :param hist_metric_list: List of hists for which to build hist info dict</span>
<span class="sd">    :param label_delimiter: String used as delimiter in values stored in hist_metric_list.</span>
<span class="sd">    :param drop_n_smaller_larger: Whether to drop n_smaller and n_larger annotations from header dict. Default is False.</span>
<span class="sd">    :param prefix: Prefix text for histogram annotations. Default is empty string.</span>
<span class="sd">    :param suffix: Suffix text for histogram annotations. Default is empty string.</span>
<span class="sd">    :param description_text: Optional text to append to the end of descriptions. Needs to start with a space if specified.</span>
<span class="sd">    :return: Dictionary keyed by VCF INFO annotations, where values are Dictionaries of Number and Description attributes.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">prefix</span> <span class="o">!=</span> <span class="s2">&quot;&quot;</span><span class="p">:</span>
        <span class="n">prefix</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">prefix</span><span class="si">}{</span><span class="n">label_delimiter</span><span class="si">}</span><span class="s2">&quot;</span>
    <span class="k">if</span> <span class="n">suffix</span> <span class="o">!=</span> <span class="s2">&quot;&quot;</span><span class="p">:</span>
        <span class="n">suffix</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">label_delimiter</span><span class="si">}{</span><span class="n">suffix</span><span class="si">}</span><span class="s2">&quot;</span>

    <span class="n">header_hist_dict</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">hist</span> <span class="ow">in</span> <span class="n">hist_metric_list</span><span class="p">:</span>
        <span class="c1"># Get hists for both raw and adj data</span>
        <span class="c1"># Add &quot;_raw&quot; to quality histograms calculated on raw data</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">adj</span><span class="p">:</span>
            <span class="n">hist</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">hist</span><span class="si">}</span><span class="s2">_raw&quot;</span>

        <span class="n">edges</span> <span class="o">=</span> <span class="n">bin_edges</span><span class="p">[</span><span class="n">hist</span><span class="p">]</span>
        <span class="n">hist_fields</span> <span class="o">=</span> <span class="n">hist</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">label_delimiter</span><span class="p">)</span>
        <span class="n">hist_text</span> <span class="o">=</span> <span class="n">hist_fields</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>

        <span class="k">if</span> <span class="n">hist_fields</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;alt&quot;</span><span class="p">:</span>
            <span class="n">hist_text</span> <span class="o">=</span> <span class="n">hist_text</span> <span class="o">+</span> <span class="s2">&quot; in heterozygous individuals&quot;</span>
        <span class="k">if</span> <span class="n">adj</span><span class="p">:</span>
            <span class="n">hist_text</span> <span class="o">=</span> <span class="n">hist_text</span> <span class="o">+</span> <span class="s2">&quot; calculated on high quality genotypes&quot;</span>

        <span class="n">hist_dict</span> <span class="o">=</span> <span class="p">{</span>
            <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">prefix</span><span class="si">}{</span><span class="n">hist</span><span class="si">}</span><span class="s2">_bin_freq</span><span class="si">{</span><span class="n">suffix</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">:</span> <span class="p">{</span>
                <span class="s2">&quot;Number&quot;</span><span class="p">:</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;Histogram for </span><span class="si">{</span><span class="n">hist_text</span><span class="si">}{</span><span class="n">description_text</span><span class="si">}</span><span class="s2">; bin edges are:&quot;</span>
                    <span class="sa">f</span><span class="s2">&quot; </span><span class="si">{</span><span class="n">edges</span><span class="si">}</span><span class="s2">&quot;</span>
                <span class="p">),</span>
            <span class="p">},</span>
        <span class="p">}</span>
        <span class="c1"># These annotations are frequently zero and are dropped from gnomad</span>
        <span class="c1"># releases for most histograms.</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">drop_n_smaller_larger</span><span class="p">:</span>
            <span class="n">hist_dict</span><span class="o">.</span><span class="n">update</span><span class="p">(</span>
                <span class="p">{</span>
                    <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">prefix</span><span class="si">}{</span><span class="n">hist</span><span class="si">}</span><span class="s2">_n_smaller</span><span class="si">{</span><span class="n">suffix</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">:</span> <span class="p">{</span>
                        <span class="s2">&quot;Number&quot;</span><span class="p">:</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span>
                        <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="p">(</span>
                            <span class="sa">f</span><span class="s2">&quot;Count of </span><span class="si">{</span><span class="n">hist_fields</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span><span class="si">}</span><span class="s2"> values falling below&quot;</span>
                            <span class="sa">f</span><span class="s2">&quot; lowest histogram bin edge </span><span class="si">{</span><span class="n">hist_text</span><span class="si">}{</span><span class="n">description_text</span><span class="si">}</span><span class="s2">&quot;</span>
                        <span class="p">),</span>
                    <span class="p">},</span>
                    <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">prefix</span><span class="si">}{</span><span class="n">hist</span><span class="si">}</span><span class="s2">_n_larger</span><span class="si">{</span><span class="n">suffix</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">:</span> <span class="p">{</span>
                        <span class="s2">&quot;Number&quot;</span><span class="p">:</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span>
                        <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="p">(</span>
                            <span class="sa">f</span><span class="s2">&quot;Count of </span><span class="si">{</span><span class="n">hist_fields</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span><span class="si">}</span><span class="s2"> values falling above&quot;</span>
                            <span class="sa">f</span><span class="s2">&quot; highest histogram bin edge </span><span class="si">{</span><span class="n">hist_text</span><span class="si">}{</span><span class="n">description_text</span><span class="si">}</span><span class="s2">&quot;</span>
                        <span class="p">),</span>
                    <span class="p">},</span>
                <span class="p">}</span>
            <span class="p">)</span>
        <span class="c1"># Only add n_larger for dp qual histograms.</span>
        <span class="k">if</span> <span class="s2">&quot;dp&quot;</span> <span class="ow">in</span> <span class="n">hist</span><span class="p">:</span>
            <span class="n">hist_dict</span><span class="o">.</span><span class="n">update</span><span class="p">(</span>
                <span class="p">{</span>
                    <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">prefix</span><span class="si">}{</span><span class="n">hist</span><span class="si">}</span><span class="s2">_n_larger</span><span class="si">{</span><span class="n">suffix</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">:</span> <span class="p">{</span>
                        <span class="s2">&quot;Number&quot;</span><span class="p">:</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span>
                        <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="p">(</span>
                            <span class="sa">f</span><span class="s2">&quot;Count of </span><span class="si">{</span><span class="n">hist_fields</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span><span class="si">}</span><span class="s2"> values falling above&quot;</span>
                            <span class="sa">f</span><span class="s2">&quot; highest histogram bin edge </span><span class="si">{</span><span class="n">hist_text</span><span class="si">}{</span><span class="n">description_text</span><span class="si">}</span><span class="s2">&quot;</span>
                        <span class="p">),</span>
                    <span class="p">},</span>
                <span class="p">}</span>
            <span class="p">)</span>

        <span class="n">header_hist_dict</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">hist_dict</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">header_hist_dict</span></div>


<div class="viewcode-block" id="set_xx_y_metrics_to_na"><a class="viewcode-back" href="../../../api_reference/utils/vcf.html#gnomad.utils.vcf.set_xx_y_metrics_to_na">[docs]</a><span class="k">def</span> <span class="nf">set_xx_y_metrics_to_na</span><span class="p">(</span>
    <span class="n">t</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span> <span class="n">hl</span><span class="o">.</span><span class="n">MatrixTable</span><span class="p">],</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">Int32Expression</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Set AC, AN, and nhomalt chrY variant annotations for XX samples to NA (instead of 0).</span>

<span class="sd">    :param t: Table/MatrixTable containing XX variant annotations.</span>
<span class="sd">    :return: Dictionary with reset annotations</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">metrics</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">t</span><span class="o">.</span><span class="n">row</span><span class="o">.</span><span class="n">info</span><span class="p">)</span>
    <span class="n">xx_metrics</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">metrics</span> <span class="k">if</span> <span class="s2">&quot;_female&quot;</span> <span class="ow">in</span> <span class="n">x</span> <span class="ow">or</span> <span class="s2">&quot;_XX&quot;</span> <span class="ow">in</span> <span class="n">x</span><span class="p">]</span>

    <span class="n">xx_metrics_dict</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">metric</span> <span class="ow">in</span> <span class="n">xx_metrics</span><span class="p">:</span>
        <span class="n">xx_metrics_dict</span><span class="o">.</span><span class="n">update</span><span class="p">(</span>
            <span class="p">{</span>
                <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">metric</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">or_missing</span><span class="p">(</span>
                    <span class="p">(</span><span class="o">~</span><span class="n">t</span><span class="o">.</span><span class="n">locus</span><span class="o">.</span><span class="n">in_y_nonpar</span><span class="p">()</span> <span class="o">&amp;</span> <span class="o">~</span><span class="n">t</span><span class="o">.</span><span class="n">locus</span><span class="o">.</span><span class="n">in_y_par</span><span class="p">()),</span>
                    <span class="n">t</span><span class="o">.</span><span class="n">info</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">metric</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">],</span>
                <span class="p">)</span>
            <span class="p">}</span>
        <span class="p">)</span>
    <span class="k">return</span> <span class="n">xx_metrics_dict</span></div>


<div class="viewcode-block" id="build_vcf_export_reference"><a class="viewcode-back" href="../../../api_reference/utils/vcf.html#gnomad.utils.vcf.build_vcf_export_reference">[docs]</a><span class="k">def</span> <span class="nf">build_vcf_export_reference</span><span class="p">(</span>
    <span class="n">name</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">build</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;GRCh38&quot;</span><span class="p">,</span>
    <span class="n">keep_contigs</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;chr</span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">23</span><span class="p">)]</span> <span class="o">+</span> <span class="p">[</span><span class="s2">&quot;chrX&quot;</span><span class="p">,</span> <span class="s2">&quot;chrY&quot;</span><span class="p">],</span>
    <span class="n">keep_chrM</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">hl</span><span class="o">.</span><span class="n">ReferenceGenome</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Create export reference based on reference genome defined by `build`.</span>

<span class="sd">    By default this will return a new reference with all non-standard contigs eliminated. Keeps chr 1-22, Y, X, and M.</span>

<span class="sd">    An example of a non-standard contig is: ##contig=&lt;ID=chr3_GL000221v1_random,length=155397,assembly=GRCh38&gt;</span>

<span class="sd">    :param name: Name to use for new reference.</span>
<span class="sd">    :param build: Reference genome build to use as starting reference genome.</span>
<span class="sd">    :param keep_contigs: Contigs to keep from reference genome defined by `build`. Default is autosomes and sex chromosomes.</span>
<span class="sd">    :param keep_chrM: Whether to keep chrM. Default is True.</span>
<span class="sd">    :return: Reference genome for VCF export containing only contigs in `keep_contigs`.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">ref</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">get_reference</span><span class="p">(</span><span class="n">build</span><span class="p">)</span>
    <span class="n">ref_args</span> <span class="o">=</span> <span class="p">{}</span>

    <span class="k">if</span> <span class="n">keep_chrM</span><span class="p">:</span>
        <span class="n">keep_contigs</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">ref</span><span class="o">.</span><span class="n">mt_contigs</span><span class="p">)</span>
        <span class="n">ref_args</span><span class="o">.</span><span class="n">update</span><span class="p">({</span><span class="s2">&quot;mt_contigs&quot;</span><span class="p">:</span> <span class="n">ref</span><span class="o">.</span><span class="n">mt_contigs</span><span class="p">})</span>

    <span class="n">ref_args</span><span class="o">.</span><span class="n">update</span><span class="p">(</span>
        <span class="p">{</span>
            <span class="s2">&quot;name&quot;</span><span class="p">:</span> <span class="n">name</span><span class="p">,</span>
            <span class="s2">&quot;contigs&quot;</span><span class="p">:</span> <span class="n">keep_contigs</span><span class="p">,</span>
            <span class="s2">&quot;lengths&quot;</span><span class="p">:</span> <span class="p">{</span><span class="n">contig</span><span class="p">:</span> <span class="n">ref</span><span class="o">.</span><span class="n">lengths</span><span class="p">[</span><span class="n">contig</span><span class="p">]</span> <span class="k">for</span> <span class="n">contig</span> <span class="ow">in</span> <span class="n">keep_contigs</span><span class="p">},</span>
            <span class="s2">&quot;x_contigs&quot;</span><span class="p">:</span> <span class="n">ref</span><span class="o">.</span><span class="n">x_contigs</span><span class="p">,</span>
            <span class="s2">&quot;y_contigs&quot;</span><span class="p">:</span> <span class="n">ref</span><span class="o">.</span><span class="n">y_contigs</span><span class="p">,</span>
            <span class="s2">&quot;par&quot;</span><span class="p">:</span> <span class="p">[</span>
                <span class="p">(</span><span class="n">interval</span><span class="o">.</span><span class="n">start</span><span class="o">.</span><span class="n">contig</span><span class="p">,</span> <span class="n">interval</span><span class="o">.</span><span class="n">start</span><span class="o">.</span><span class="n">position</span><span class="p">,</span> <span class="n">interval</span><span class="o">.</span><span class="n">end</span><span class="o">.</span><span class="n">position</span><span class="p">)</span>
                <span class="k">for</span> <span class="n">interval</span> <span class="ow">in</span> <span class="n">ref</span><span class="o">.</span><span class="n">par</span>
            <span class="p">],</span>
        <span class="p">}</span>
    <span class="p">)</span>

    <span class="n">export_reference</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">ReferenceGenome</span><span class="p">(</span><span class="o">**</span><span class="n">ref_args</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">export_reference</span></div>


<div class="viewcode-block" id="rekey_new_reference"><a class="viewcode-back" href="../../../api_reference/utils/vcf.html#gnomad.utils.vcf.rekey_new_reference">[docs]</a><span class="k">def</span> <span class="nf">rekey_new_reference</span><span class="p">(</span>
    <span class="n">t</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span> <span class="n">hl</span><span class="o">.</span><span class="n">MatrixTable</span><span class="p">],</span> <span class="n">reference</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">ReferenceGenome</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Union</span><span class="p">[</span><span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span> <span class="n">hl</span><span class="o">.</span><span class="n">MatrixTable</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Re-key Table or MatrixTable with a new reference genome.</span>

<span class="sd">    :param t: Input Table/MatrixTable.</span>
<span class="sd">    :param reference: Reference genome to re-key with.</span>
<span class="sd">    :return: Re-keyed Table/MatrixTable</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">t</span> <span class="o">=</span> <span class="n">t</span><span class="o">.</span><span class="n">rename</span><span class="p">({</span><span class="s2">&quot;locus&quot;</span><span class="p">:</span> <span class="s2">&quot;locus_original&quot;</span><span class="p">})</span>
    <span class="n">locus_expr</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">locus</span><span class="p">(</span>
        <span class="n">t</span><span class="o">.</span><span class="n">locus_original</span><span class="o">.</span><span class="n">contig</span><span class="p">,</span>
        <span class="n">t</span><span class="o">.</span><span class="n">locus_original</span><span class="o">.</span><span class="n">position</span><span class="p">,</span>
        <span class="n">reference_genome</span><span class="o">=</span><span class="n">reference</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">t</span><span class="p">,</span> <span class="n">hl</span><span class="o">.</span><span class="n">MatrixTable</span><span class="p">):</span>
        <span class="n">t</span> <span class="o">=</span> <span class="n">t</span><span class="o">.</span><span class="n">annotate_rows</span><span class="p">(</span><span class="n">locus</span><span class="o">=</span><span class="n">locus_expr</span><span class="p">)</span>
        <span class="n">t</span> <span class="o">=</span> <span class="n">t</span><span class="o">.</span><span class="n">key_rows_by</span><span class="p">(</span><span class="s2">&quot;locus&quot;</span><span class="p">,</span> <span class="s2">&quot;alleles&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s2">&quot;locus_original&quot;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">t</span> <span class="o">=</span> <span class="n">t</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span><span class="n">locus</span><span class="o">=</span><span class="n">locus_expr</span><span class="p">)</span>
        <span class="n">t</span> <span class="o">=</span> <span class="n">t</span><span class="o">.</span><span class="n">key_by</span><span class="p">(</span><span class="s2">&quot;locus&quot;</span><span class="p">,</span> <span class="s2">&quot;alleles&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s2">&quot;locus_original&quot;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">t</span></div>
</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p></p>
  </div>

   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>