<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>gnomad.variant_qc.ld &mdash; gnomad master documentation</title>
      <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../../_static/theme_overrides.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../../_static/jquery.js"></script>
        <script src="../../_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
        <script src="../../_static/doctools.js"></script>
        <script src="../../_static/sphinx_highlight.js"></script>
    <script src="../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="gnomad.variant_qc.pipeline" href="pipeline.html" />
    <link rel="prev" title="gnomad.variant_qc.evaluation" href="evaluation.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../index.html" class="icon icon-home">
            gnomad
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../getting_started.html">Getting Started</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../examples/index.html">Examples</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../index.html">API Reference</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../utils/index.html">utils</a></li>
<li class="toctree-l2"><a class="reference internal" href="../resources/index.html">resources</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="index.html">variant_qc</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="evaluation.html">evaluation</a></li>
<li class="toctree-l3 current"><a class="current reference internal" href="#">ld</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#gnomad.variant_qc.ld.get_r_human_readable"><code class="docutils literal notranslate"><span class="pre">get_r_human_readable()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#gnomad.variant_qc.ld.get_r_for_pair_of_variants"><code class="docutils literal notranslate"><span class="pre">get_r_for_pair_of_variants()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#gnomad.variant_qc.ld.get_r_within_gene_in_gen_anc"><code class="docutils literal notranslate"><span class="pre">get_r_within_gene_in_gen_anc()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#gnomad.variant_qc.ld.get_r_within_gene"><code class="docutils literal notranslate"><span class="pre">get_r_within_gene()</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="pipeline.html">pipeline</a></li>
<li class="toctree-l3"><a class="reference internal" href="random_forest.html">random_forest</a></li>
<li class="toctree-l3"><a class="reference internal" href="training.html">training</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../sample_qc/index.html">sample_qc</a></li>
<li class="toctree-l2"><a class="reference internal" href="../assessment/index.html">assessment</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../resource_sources.html">Resource Sources</a></li>
<li class="toctree-l1"><a class="reference external" href="https://github.com/broadinstitute/gnomad_methods/releases">Change Log</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">gnomad</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../index.html">gnomad</a></li>
          <li class="breadcrumb-item"><a href="index.html">gnomad.variant_qc</a></li>
      <li class="breadcrumb-item active">gnomad.variant_qc.ld</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../../_sources/api_reference/variant_qc/ld.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="gnomad-variant-qc-ld">
<h1>gnomad.variant_qc.ld<a class="headerlink" href="#gnomad-variant-qc-ld" title="Permalink to this heading"></a></h1>
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#gnomad.variant_qc.ld.get_r_human_readable" title="gnomad.variant_qc.ld.get_r_human_readable"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.variant_qc.ld.get_r_human_readable</span></code></a>(...)</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#gnomad.variant_qc.ld.get_r_for_pair_of_variants" title="gnomad.variant_qc.ld.get_r_for_pair_of_variants"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.variant_qc.ld.get_r_for_pair_of_variants</span></code></a>(bm, ...)</p></td>
<td><p>Get <cite>r</cite> value (LD) for pair of variants <cite>var1</cite> and <cite>var2</cite>.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#gnomad.variant_qc.ld.get_r_within_gene_in_gen_anc" title="gnomad.variant_qc.ld.get_r_within_gene_in_gen_anc"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.variant_qc.ld.get_r_within_gene_in_gen_anc</span></code></a>(...)</p></td>
<td><p>Get LD information (<cite>r</cite>) for all pairs of variants within <cite>gene</cite> for a given <cite>gen_anc</cite>.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#gnomad.variant_qc.ld.get_r_within_gene" title="gnomad.variant_qc.ld.get_r_within_gene"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.variant_qc.ld.get_r_within_gene</span></code></a>(bm, ...)</p></td>
<td><p>Get LD information (<cite>r</cite>) for all pairs of variants within <cite>gene</cite>.</p></td>
</tr>
</tbody>
</table>
<span class="target" id="module-gnomad.variant_qc.ld"></span><dl class="py function">
<dt class="sig sig-object py" id="gnomad.variant_qc.ld.get_r_human_readable">
<span class="sig-prename descclassname"><span class="pre">gnomad.variant_qc.ld.</span></span><span class="sig-name descname"><span class="pre">get_r_human_readable</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gen_anc</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">var1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">var2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_genome</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'GRCh37'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/gnomad/variant_qc/ld.html#get_r_human_readable"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gnomad.variant_qc.ld.get_r_human_readable" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>gen_anc</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – </p></li>
<li><p><strong>var1</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – </p></li>
<li><p><strong>var2</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – </p></li>
<li><p><strong>ref_genome</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – </p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gnomad.variant_qc.ld.get_r_for_pair_of_variants">
<span class="sig-prename descclassname"><span class="pre">gnomad.variant_qc.ld.</span></span><span class="sig-name descname"><span class="pre">get_r_for_pair_of_variants</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">bm</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ld_index</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">var1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">var2</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/gnomad/variant_qc/ld.html#get_r_for_pair_of_variants"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gnomad.variant_qc.ld.get_r_for_pair_of_variants" title="Permalink to this definition"></a></dt>
<dd><p>Get <cite>r</cite> value (LD) for pair of variants <cite>var1</cite> and <cite>var2</cite>.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">bm</span> <span class="o">=</span> <span class="n">get_ld_matrix</span><span class="p">(</span><span class="s1">&#39;nfe&#39;</span><span class="p">)</span>
<span class="n">ld_index</span> <span class="o">=</span> <span class="n">get_ld_index</span><span class="p">(</span><span class="s1">&#39;nfe&#39;</span><span class="p">)</span>
<span class="n">var1</span> <span class="o">=</span> <span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">parse_locus</span><span class="p">(</span><span class="s1">&#39;1:10146&#39;</span><span class="p">,</span> <span class="s1">&#39;GRCh37&#39;</span><span class="p">),</span> <span class="p">[</span><span class="s1">&#39;AC&#39;</span><span class="p">,</span> <span class="s1">&#39;A&#39;</span><span class="p">])</span>
<span class="n">var2</span> <span class="o">=</span> <span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">parse_locus</span><span class="p">(</span><span class="s1">&#39;1:10151&#39;</span><span class="p">,</span> <span class="s1">&#39;GRCh37&#39;</span><span class="p">),</span> <span class="p">[</span><span class="s1">&#39;TA&#39;</span><span class="p">,</span> <span class="s1">&#39;T&#39;</span><span class="p">])</span>
<span class="n">get_r_for_pair_of_variants</span><span class="p">(</span><span class="n">bm</span><span class="p">,</span> <span class="n">ld_index</span><span class="p">,</span> <span class="n">var1</span><span class="p">,</span> <span class="n">var2</span><span class="p">)</span>
<span class="c1"># 0.01789767935482124</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>bm</strong> (<a class="reference external" href="https://hail.is/docs/0.2/linalg/hail.linalg.BlockMatrix.html#hail.linalg.BlockMatrix" title="(in Hail v0.2.136)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BlockMatrix</span></code></a>) – Input BlockMatrix</p></li>
<li><p><strong>ld_index</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2.136)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a>) – Corresponding index table</p></li>
<li><p><strong>var1</strong> ((<a class="reference external" href="https://hail.is/docs/0.2/types.html#hail.expr.types.tlocus" title="(in Hail v0.2.136)"><code class="xref py py-class docutils literal notranslate"><span class="pre">tlocus</span></code></a>, <a class="reference external" href="https://hail.is/docs/0.2/types.html#hail.expr.types.tarray" title="(in Hail v0.2.136)"><code class="xref py py-class docutils literal notranslate"><span class="pre">tarray</span></code></a>)) – Tuple of locus and alleles</p></li>
<li><p><strong>var2</strong> ((<a class="reference external" href="https://hail.is/docs/0.2/types.html#hail.expr.types.tlocus" title="(in Hail v0.2.136)"><code class="xref py py-class docutils literal notranslate"><span class="pre">tlocus</span></code></a>, <a class="reference external" href="https://hail.is/docs/0.2/types.html#hail.expr.types.tarray" title="(in Hail v0.2.136)"><code class="xref py py-class docutils literal notranslate"><span class="pre">tarray</span></code></a>)) – Tuple of locus and alleles</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Correlation (r) between two variants</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gnomad.variant_qc.ld.get_r_within_gene_in_gen_anc">
<span class="sig-prename descclassname"><span class="pre">gnomad.variant_qc.ld.</span></span><span class="sig-name descname"><span class="pre">get_r_within_gene_in_gen_anc</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gen_anc</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/gnomad/variant_qc/ld.html#get_r_within_gene_in_gen_anc"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gnomad.variant_qc.ld.get_r_within_gene_in_gen_anc" title="Permalink to this definition"></a></dt>
<dd><p>Get LD information (<cite>r</cite>) for all pairs of variants within <cite>gene</cite> for a given <cite>gen_anc</cite>.</p>
<p>Warning: this returns a table quadratic in number of variants. Exercise caution with large genes.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>gen_anc</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Genetic ancestry group for which to get LD information</p></li>
<li><p><strong>gene</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Gene symbol as string</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Table with pairs of variants</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gnomad.variant_qc.ld.get_r_within_gene">
<span class="sig-prename descclassname"><span class="pre">gnomad.variant_qc.ld.</span></span><span class="sig-name descname"><span class="pre">get_r_within_gene</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">bm</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ld_index</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vep_ht</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference_genome</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/gnomad/variant_qc/ld.html#get_r_within_gene"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gnomad.variant_qc.ld.get_r_within_gene" title="Permalink to this definition"></a></dt>
<dd><p>Get LD information (<cite>r</cite>) for all pairs of variants within <cite>gene</cite>.</p>
<p>Warning: this returns a table quadratic in number of variants. Exercise caution with large genes.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>bm</strong> (<a class="reference external" href="https://hail.is/docs/0.2/linalg/hail.linalg.BlockMatrix.html#hail.linalg.BlockMatrix" title="(in Hail v0.2.136)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BlockMatrix</span></code></a>) – Input Block Matrix</p></li>
<li><p><strong>ld_index</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2.136)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a>) – Corresponding index table</p></li>
<li><p><strong>gene</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Gene symbol as string</p></li>
<li><p><strong>vep_ht</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2.136)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a>) – Table with VEP annotations (if None, gets from get_gnomad_public_data())</p></li>
<li><p><strong>reference_genome</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Reference genome to pass to get_gene_intervals for fast filtering to gene</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Table with pairs of variants</p>
</dd>
</dl>
</dd></dl>

</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="evaluation.html" class="btn btn-neutral float-left" title="gnomad.variant_qc.evaluation" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="pipeline.html" class="btn btn-neutral float-right" title="gnomad.variant_qc.pipeline" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p></p>
  </div>

   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>