<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>gnomad.resources.grch38.gnomad &mdash; gnomad master documentation</title>
      <link rel="stylesheet" href="../../../../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../../../../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../../../../_static/theme_overrides.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../../../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../../../../_static/jquery.js"></script>
        <script src="../../../../_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script data-url_root="../../../../" id="documentation_options" src="../../../../_static/documentation_options.js"></script>
        <script src="../../../../_static/doctools.js"></script>
        <script src="../../../../_static/sphinx_highlight.js"></script>
    <script src="../../../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../../../genindex.html" />
    <link rel="search" title="Search" href="../../../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../../../index.html" class="icon icon-home">
            gnomad
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../../../../getting_started.html">Getting Started</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../examples/index.html">Examples</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../api_reference/index.html">API Reference</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../resource_sources.html">Resource Sources</a></li>
<li class="toctree-l1"><a class="reference external" href="https://github.com/broadinstitute/gnomad_methods/releases">Change Log</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../../index.html">gnomad</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../../../index.html">Module code</a></li>
      <li class="breadcrumb-item active">gnomad.resources.grch38.gnomad</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for gnomad.resources.grch38.gnomad</h1><div class="highlight"><pre>
<span></span><span class="c1"># noqa: D100</span>

<span class="kn">import</span> <span class="nn">logging</span>
<span class="kn">from</span> <span class="nn">typing</span> <span class="kn">import</span> <span class="n">Optional</span>

<span class="kn">import</span> <span class="nn">hail</span> <span class="k">as</span> <span class="nn">hl</span>

<span class="kn">from</span> <span class="nn">gnomad.resources.resource_utils</span> <span class="kn">import</span> <span class="p">(</span>
    <span class="n">DataException</span><span class="p">,</span>
    <span class="n">GnomadPublicMatrixTableResource</span><span class="p">,</span>
    <span class="n">GnomadPublicTableResource</span><span class="p">,</span>
    <span class="n">VersionedMatrixTableResource</span><span class="p">,</span>
    <span class="n">VersionedTableResource</span><span class="p">,</span>
<span class="p">)</span>
<span class="kn">from</span> <span class="nn">gnomad.sample_qc.ancestry</span> <span class="kn">import</span> <span class="n">GEN_ANC_NAMES</span>
<span class="kn">from</span> <span class="nn">gnomad.utils.annotations</span> <span class="kn">import</span> <span class="n">add_gks_va</span><span class="p">,</span> <span class="n">add_gks_vrs</span>

<span class="n">logging</span><span class="o">.</span><span class="n">basicConfig</span><span class="p">(</span>
    <span class="nb">format</span><span class="o">=</span><span class="s2">&quot;</span><span class="si">%(asctime)s</span><span class="s2"> (</span><span class="si">%(name)s</span><span class="s2"> </span><span class="si">%(lineno)s</span><span class="s2">): </span><span class="si">%(message)s</span><span class="s2">&quot;</span><span class="p">,</span>
    <span class="n">datefmt</span><span class="o">=</span><span class="s2">&quot;%m/</span><span class="si">%d</span><span class="s2">/%Y %I:%M:%S %p&quot;</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">logger</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">(</span><span class="vm">__name__</span><span class="p">)</span>
<span class="n">logger</span><span class="o">.</span><span class="n">setLevel</span><span class="p">(</span><span class="n">logging</span><span class="o">.</span><span class="n">INFO</span><span class="p">)</span>

<span class="n">CURRENT_EXOME_RELEASE</span> <span class="o">=</span> <span class="s2">&quot;4.1&quot;</span>
<span class="n">CURRENT_GENOME_RELEASE</span> <span class="o">=</span> <span class="s2">&quot;4.1&quot;</span>
<span class="n">CURRENT_JOINT_RELEASE</span> <span class="o">=</span> <span class="s2">&quot;4.1&quot;</span>

<span class="n">CURRENT_BROWSER_RELEASE</span> <span class="o">=</span> <span class="s2">&quot;4.1&quot;</span>

<span class="n">CURRENT_EXOME_COVERAGE_RELEASE</span> <span class="o">=</span> <span class="s2">&quot;4.0&quot;</span>
<span class="n">CURRENT_GENOME_COVERAGE_RELEASE</span> <span class="o">=</span> <span class="s2">&quot;3.0.1&quot;</span>

<span class="n">CURRENT_EXOME_AN_RELEASE</span> <span class="o">=</span> <span class="s2">&quot;4.1&quot;</span>
<span class="n">CURRENT_GENOME_AN_RELEASE</span> <span class="o">=</span> <span class="s2">&quot;4.1&quot;</span>

<span class="n">EXOME_RELEASES</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;4.0&quot;</span><span class="p">,</span> <span class="s2">&quot;4.1&quot;</span><span class="p">]</span>
<span class="n">GENOME_RELEASES</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;3.0&quot;</span><span class="p">,</span> <span class="s2">&quot;3.1&quot;</span><span class="p">,</span> <span class="s2">&quot;3.1.1&quot;</span><span class="p">,</span> <span class="s2">&quot;3.1.2&quot;</span><span class="p">,</span> <span class="s2">&quot;4.0&quot;</span><span class="p">,</span> <span class="s2">&quot;4.1&quot;</span><span class="p">]</span>
<span class="n">JOINT_RELEASES</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;4.1&quot;</span><span class="p">]</span>
<span class="n">BROWSER_RELEASES</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;4.1&quot;</span><span class="p">]</span>

<span class="n">EXOME_COVERAGE_RELEASES</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;4.0&quot;</span><span class="p">]</span>
<span class="n">GENOME_COVERAGE_RELEASES</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;3.0&quot;</span><span class="p">,</span> <span class="s2">&quot;3.0.1&quot;</span><span class="p">]</span>

<span class="n">EXOME_AN_RELEASES</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;4.1&quot;</span><span class="p">]</span>
<span class="n">GENOME_AN_RELEASES</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;4.1&quot;</span><span class="p">]</span>

<span class="n">DATA_TYPES</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;exomes&quot;</span><span class="p">,</span> <span class="s2">&quot;genomes&quot;</span><span class="p">,</span> <span class="s2">&quot;joint&quot;</span><span class="p">]</span>
<span class="n">MAJOR_RELEASES</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;v3&quot;</span><span class="p">,</span> <span class="s2">&quot;v4&quot;</span><span class="p">]</span>
<span class="n">CURRENT_MAJOR_RELEASE</span> <span class="o">=</span> <span class="n">MAJOR_RELEASES</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>

<span class="n">GENOME_GEN_ANCS</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;AFR&quot;</span><span class="p">,</span> <span class="s2">&quot;AMI&quot;</span><span class="p">,</span> <span class="s2">&quot;AMR&quot;</span><span class="p">,</span> <span class="s2">&quot;ASJ&quot;</span><span class="p">,</span> <span class="s2">&quot;EAS&quot;</span><span class="p">,</span> <span class="s2">&quot;FIN&quot;</span><span class="p">,</span> <span class="s2">&quot;NFE&quot;</span><span class="p">,</span> <span class="s2">&quot;SAS&quot;</span><span class="p">,</span> <span class="s2">&quot;OTH&quot;</span><span class="p">]</span>
<span class="n">SUBSETS</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;v3&quot;</span><span class="p">:</span> <span class="p">[</span>
        <span class="s2">&quot;non_v2&quot;</span><span class="p">,</span>
        <span class="s2">&quot;non_topmed&quot;</span><span class="p">,</span>
        <span class="s2">&quot;non_cancer&quot;</span><span class="p">,</span>
        <span class="s2">&quot;controls_and_biobanks&quot;</span><span class="p">,</span>
        <span class="s2">&quot;non_neuro&quot;</span><span class="p">,</span>
        <span class="s2">&quot;tgp&quot;</span><span class="p">,</span>
        <span class="s2">&quot;hgdp&quot;</span><span class="p">,</span>
    <span class="p">],</span>
    <span class="s2">&quot;v4&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;non_ukb&quot;</span><span class="p">],</span>
<span class="p">}</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">Order to sort subgroupings during VCF export by version.</span>

<span class="sd">Ensures that INFO labels in VCF are in desired order (e.g., tgp_raw_AC_esn_XX).</span>
<span class="sd">&quot;&quot;&quot;</span>

<span class="n">GROUPS</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;adj&quot;</span><span class="p">,</span> <span class="s2">&quot;raw&quot;</span><span class="p">]</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">Group names used to generate labels for high quality genotypes and all raw genotypes.</span>

<span class="sd">Used in VCF export.</span>
<span class="sd">&quot;&quot;&quot;</span>

<span class="n">SEXES</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;XX&quot;</span><span class="p">,</span> <span class="s2">&quot;XY&quot;</span><span class="p">]</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">Sample sexes used in VCF export.</span>

<span class="sd">Used to stratify frequency annotations (AC, AN, AF) for each sex.</span>
<span class="sd">&quot;&quot;&quot;</span>

<span class="n">GEN_ANC_GROUPS</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;v3&quot;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s2">&quot;genomes&quot;</span><span class="p">:</span> <span class="p">[</span>
            <span class="s2">&quot;afr&quot;</span><span class="p">,</span>
            <span class="s2">&quot;ami&quot;</span><span class="p">,</span>
            <span class="s2">&quot;amr&quot;</span><span class="p">,</span>
            <span class="s2">&quot;asj&quot;</span><span class="p">,</span>
            <span class="s2">&quot;eas&quot;</span><span class="p">,</span>
            <span class="s2">&quot;fin&quot;</span><span class="p">,</span>
            <span class="s2">&quot;nfe&quot;</span><span class="p">,</span>
            <span class="s2">&quot;oth&quot;</span><span class="p">,</span>
            <span class="s2">&quot;sas&quot;</span><span class="p">,</span>
            <span class="s2">&quot;mid&quot;</span><span class="p">,</span>
        <span class="p">]</span>
    <span class="p">},</span>
    <span class="s2">&quot;v4&quot;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s2">&quot;exomes&quot;</span><span class="p">:</span> <span class="p">[</span>
            <span class="s2">&quot;afr&quot;</span><span class="p">,</span>
            <span class="s2">&quot;amr&quot;</span><span class="p">,</span>
            <span class="s2">&quot;asj&quot;</span><span class="p">,</span>
            <span class="s2">&quot;eas&quot;</span><span class="p">,</span>
            <span class="s2">&quot;fin&quot;</span><span class="p">,</span>
            <span class="s2">&quot;mid&quot;</span><span class="p">,</span>
            <span class="s2">&quot;nfe&quot;</span><span class="p">,</span>
            <span class="s2">&quot;remaining&quot;</span><span class="p">,</span>
            <span class="s2">&quot;sas&quot;</span><span class="p">,</span>
        <span class="p">],</span>
        <span class="s2">&quot;genomes&quot;</span><span class="p">:</span> <span class="p">[</span>
            <span class="s2">&quot;afr&quot;</span><span class="p">,</span>
            <span class="s2">&quot;ami&quot;</span><span class="p">,</span>
            <span class="s2">&quot;amr&quot;</span><span class="p">,</span>
            <span class="s2">&quot;asj&quot;</span><span class="p">,</span>
            <span class="s2">&quot;eas&quot;</span><span class="p">,</span>
            <span class="s2">&quot;fin&quot;</span><span class="p">,</span>
            <span class="s2">&quot;mid&quot;</span><span class="p">,</span>
            <span class="s2">&quot;nfe&quot;</span><span class="p">,</span>
            <span class="s2">&quot;remaining&quot;</span><span class="p">,</span>
            <span class="s2">&quot;sas&quot;</span><span class="p">,</span>
        <span class="p">],</span>
    <span class="p">},</span>
<span class="p">}</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">Genetic ancestry groups in gnomAD by version.</span>
<span class="sd">&quot;&quot;&quot;</span>

<span class="n">COHORTS_WITH_GEN_ANC_STORED_AS_SUBGRP</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;tgp&quot;</span><span class="p">,</span> <span class="s2">&quot;hgdp&quot;</span><span class="p">]</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">Subsets in gnomAD v3.1 that are broken down by their known genetic ancestry subgroups instead of groups in the frequency struct.</span>
<span class="sd">&quot;&quot;&quot;</span>

<span class="n">TGP_GEN_ANC_GROUPS</span> <span class="o">=</span> <span class="p">[</span>
    <span class="s2">&quot;esn&quot;</span><span class="p">,</span>
    <span class="s2">&quot;pur&quot;</span><span class="p">,</span>
    <span class="s2">&quot;pjl&quot;</span><span class="p">,</span>
    <span class="s2">&quot;clm&quot;</span><span class="p">,</span>
    <span class="s2">&quot;jpt&quot;</span><span class="p">,</span>
    <span class="s2">&quot;chb&quot;</span><span class="p">,</span>
    <span class="s2">&quot;stu&quot;</span><span class="p">,</span>
    <span class="s2">&quot;itu&quot;</span><span class="p">,</span>
    <span class="s2">&quot;tsi&quot;</span><span class="p">,</span>
    <span class="s2">&quot;mxl&quot;</span><span class="p">,</span>
    <span class="s2">&quot;ceu&quot;</span><span class="p">,</span>
    <span class="s2">&quot;msl&quot;</span><span class="p">,</span>
    <span class="s2">&quot;yri&quot;</span><span class="p">,</span>
    <span class="s2">&quot;beb&quot;</span><span class="p">,</span>
    <span class="s2">&quot;fin&quot;</span><span class="p">,</span>
    <span class="s2">&quot;khv&quot;</span><span class="p">,</span>
    <span class="s2">&quot;cdx&quot;</span><span class="p">,</span>
    <span class="s2">&quot;lwk&quot;</span><span class="p">,</span>
    <span class="s2">&quot;acb&quot;</span><span class="p">,</span>
    <span class="s2">&quot;asw&quot;</span><span class="p">,</span>
    <span class="s2">&quot;ibs&quot;</span><span class="p">,</span>
    <span class="s2">&quot;gbr&quot;</span><span class="p">,</span>
    <span class="s2">&quot;pel&quot;</span><span class="p">,</span>
    <span class="s2">&quot;gih&quot;</span><span class="p">,</span>
    <span class="s2">&quot;chs&quot;</span><span class="p">,</span>
    <span class="s2">&quot;gwd&quot;</span><span class="p">,</span>
<span class="p">]</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">1000 Genomes Project (1KG/TGP) genetic ancestry subgroups.</span>
<span class="sd">&quot;&quot;&quot;</span>

<span class="n">HGDP_GEN_ANC_GROUPS</span> <span class="o">=</span> <span class="p">[</span>
    <span class="s2">&quot;japanese&quot;</span><span class="p">,</span>
    <span class="s2">&quot;papuanhighlands&quot;</span><span class="p">,</span>
    <span class="s2">&quot;papuansepik&quot;</span><span class="p">,</span>
    <span class="s2">&quot;adygei&quot;</span><span class="p">,</span>
    <span class="s2">&quot;orcadian&quot;</span><span class="p">,</span>
    <span class="s2">&quot;biaka&quot;</span><span class="p">,</span>
    <span class="s2">&quot;yakut&quot;</span><span class="p">,</span>
    <span class="s2">&quot;han&quot;</span><span class="p">,</span>
    <span class="s2">&quot;northernhan&quot;</span><span class="p">,</span>
    <span class="s2">&quot;uygur&quot;</span><span class="p">,</span>
    <span class="s2">&quot;miao&quot;</span><span class="p">,</span>
    <span class="s2">&quot;mongolian&quot;</span><span class="p">,</span>
    <span class="s2">&quot;balochi&quot;</span><span class="p">,</span>
    <span class="s2">&quot;bedouin&quot;</span><span class="p">,</span>
    <span class="s2">&quot;russian&quot;</span><span class="p">,</span>
    <span class="s2">&quot;daur&quot;</span><span class="p">,</span>
    <span class="s2">&quot;pima&quot;</span><span class="p">,</span>
    <span class="s2">&quot;hezhen&quot;</span><span class="p">,</span>
    <span class="s2">&quot;sindhi&quot;</span><span class="p">,</span>
    <span class="s2">&quot;yi&quot;</span><span class="p">,</span>
    <span class="s2">&quot;oroqen&quot;</span><span class="p">,</span>
    <span class="s2">&quot;san&quot;</span><span class="p">,</span>
    <span class="s2">&quot;tuscan&quot;</span><span class="p">,</span>
    <span class="s2">&quot;tu&quot;</span><span class="p">,</span>
    <span class="s2">&quot;palestinian&quot;</span><span class="p">,</span>
    <span class="s2">&quot;tujia&quot;</span><span class="p">,</span>
    <span class="s2">&quot;druze&quot;</span><span class="p">,</span>
    <span class="s2">&quot;pathan&quot;</span><span class="p">,</span>
    <span class="s2">&quot;basque&quot;</span><span class="p">,</span>
    <span class="s2">&quot;makrani&quot;</span><span class="p">,</span>
    <span class="s2">&quot;bergamoitalian&quot;</span><span class="p">,</span>
    <span class="s2">&quot;naxi&quot;</span><span class="p">,</span>
    <span class="s2">&quot;karitiana&quot;</span><span class="p">,</span>
    <span class="s2">&quot;sardinian&quot;</span><span class="p">,</span>
    <span class="s2">&quot;mbuti&quot;</span><span class="p">,</span>
    <span class="s2">&quot;mozabite&quot;</span><span class="p">,</span>
    <span class="s2">&quot;yoruba&quot;</span><span class="p">,</span>
    <span class="s2">&quot;lahu&quot;</span><span class="p">,</span>
    <span class="s2">&quot;dai&quot;</span><span class="p">,</span>
    <span class="s2">&quot;cambodian&quot;</span><span class="p">,</span>
    <span class="s2">&quot;bougainville&quot;</span><span class="p">,</span>
    <span class="s2">&quot;french&quot;</span><span class="p">,</span>
    <span class="s2">&quot;brahui&quot;</span><span class="p">,</span>
    <span class="s2">&quot;hazara&quot;</span><span class="p">,</span>
    <span class="s2">&quot;bantusouthafrica&quot;</span><span class="p">,</span>
    <span class="s2">&quot;surui&quot;</span><span class="p">,</span>
    <span class="s2">&quot;mandenka&quot;</span><span class="p">,</span>
    <span class="s2">&quot;kalash&quot;</span><span class="p">,</span>
    <span class="s2">&quot;xibo&quot;</span><span class="p">,</span>
    <span class="s2">&quot;colombian&quot;</span><span class="p">,</span>
    <span class="s2">&quot;bantukenya&quot;</span><span class="p">,</span>
    <span class="s2">&quot;she&quot;</span><span class="p">,</span>
    <span class="s2">&quot;burusho&quot;</span><span class="p">,</span>
    <span class="s2">&quot;maya&quot;</span><span class="p">,</span>
<span class="p">]</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">Human Genome Diversity Project (HGDP) genetic ancestry subgroups.</span>
<span class="sd">&quot;&quot;&quot;</span>

<span class="n">TGP_GEN_ANC_GROUP_NAMES</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;chb&quot;</span><span class="p">:</span> <span class="s2">&quot;Han Chinese&quot;</span><span class="p">,</span>
    <span class="s2">&quot;jpt&quot;</span><span class="p">:</span> <span class="s2">&quot;Japanese&quot;</span><span class="p">,</span>
    <span class="s2">&quot;chs&quot;</span><span class="p">:</span> <span class="s2">&quot;Southern Han Chinese&quot;</span><span class="p">,</span>
    <span class="s2">&quot;cdx&quot;</span><span class="p">:</span> <span class="s2">&quot;Chinese Dai&quot;</span><span class="p">,</span>
    <span class="s2">&quot;khv&quot;</span><span class="p">:</span> <span class="s2">&quot;Kinh&quot;</span><span class="p">,</span>
    <span class="s2">&quot;ceu&quot;</span><span class="p">:</span> <span class="s2">&quot;Utah Residents (European Ancestry)&quot;</span><span class="p">,</span>
    <span class="s2">&quot;tsi&quot;</span><span class="p">:</span> <span class="s2">&quot;Toscani&quot;</span><span class="p">,</span>
    <span class="s2">&quot;fin&quot;</span><span class="p">:</span> <span class="s2">&quot;Finnish&quot;</span><span class="p">,</span>
    <span class="s2">&quot;gbr&quot;</span><span class="p">:</span> <span class="s2">&quot;British&quot;</span><span class="p">,</span>
    <span class="s2">&quot;ibs&quot;</span><span class="p">:</span> <span class="s2">&quot;Iberian&quot;</span><span class="p">,</span>
    <span class="s2">&quot;yri&quot;</span><span class="p">:</span> <span class="s2">&quot;Yoruba&quot;</span><span class="p">,</span>
    <span class="s2">&quot;lwk&quot;</span><span class="p">:</span> <span class="s2">&quot;Luhya&quot;</span><span class="p">,</span>
    <span class="s2">&quot;gwd&quot;</span><span class="p">:</span> <span class="s2">&quot;Gambian&quot;</span><span class="p">,</span>
    <span class="s2">&quot;msl&quot;</span><span class="p">:</span> <span class="s2">&quot;Mende&quot;</span><span class="p">,</span>
    <span class="s2">&quot;esn&quot;</span><span class="p">:</span> <span class="s2">&quot;Esan&quot;</span><span class="p">,</span>
    <span class="s2">&quot;asw&quot;</span><span class="p">:</span> <span class="s2">&quot;African-American&quot;</span><span class="p">,</span>
    <span class="s2">&quot;acb&quot;</span><span class="p">:</span> <span class="s2">&quot;African Caribbean&quot;</span><span class="p">,</span>
    <span class="s2">&quot;mxl&quot;</span><span class="p">:</span> <span class="s2">&quot;Mexican-American&quot;</span><span class="p">,</span>
    <span class="s2">&quot;pur&quot;</span><span class="p">:</span> <span class="s2">&quot;Puerto Rican&quot;</span><span class="p">,</span>
    <span class="s2">&quot;clm&quot;</span><span class="p">:</span> <span class="s2">&quot;Colombian&quot;</span><span class="p">,</span>
    <span class="s2">&quot;pel&quot;</span><span class="p">:</span> <span class="s2">&quot;Peruvian&quot;</span><span class="p">,</span>
    <span class="s2">&quot;gih&quot;</span><span class="p">:</span> <span class="s2">&quot;Gujarati&quot;</span><span class="p">,</span>
    <span class="s2">&quot;pjl&quot;</span><span class="p">:</span> <span class="s2">&quot;Punjabi&quot;</span><span class="p">,</span>
    <span class="s2">&quot;beb&quot;</span><span class="p">:</span> <span class="s2">&quot;Bengali&quot;</span><span class="p">,</span>
    <span class="s2">&quot;stu&quot;</span><span class="p">:</span> <span class="s2">&quot;Sri Lankan Tamil&quot;</span><span class="p">,</span>
    <span class="s2">&quot;itu&quot;</span><span class="p">:</span> <span class="s2">&quot;Indian Telugu&quot;</span><span class="p">,</span>
<span class="p">}</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">1000 Genomes Project (1KG/TGP) genetic ancestry group label map.</span>
<span class="sd">&quot;&quot;&quot;</span>

<span class="n">GEN_ANC_GROUPS_STORED_AS_SUBGRPS</span> <span class="o">=</span> <span class="n">TGP_GEN_ANC_GROUPS</span> <span class="o">+</span> <span class="n">HGDP_GEN_ANC_GROUPS</span>
<span class="n">GEN_ANC_GROUPS_TO_REMOVE_FOR_GRPMAX</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;v3&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;asj&quot;</span><span class="p">,</span> <span class="s2">&quot;fin&quot;</span><span class="p">,</span> <span class="s2">&quot;mid&quot;</span><span class="p">,</span> <span class="s2">&quot;oth&quot;</span><span class="p">,</span> <span class="s2">&quot;ami&quot;</span><span class="p">,</span> <span class="s2">&quot;remaining&quot;</span><span class="p">},</span>
    <span class="s2">&quot;v4&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;asj&quot;</span><span class="p">,</span> <span class="s2">&quot;fin&quot;</span><span class="p">,</span> <span class="s2">&quot;oth&quot;</span><span class="p">,</span> <span class="s2">&quot;ami&quot;</span><span class="p">,</span> <span class="s2">&quot;remaining&quot;</span><span class="p">},</span>
<span class="p">}</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">Genetic ancestry groups that are removed before genetic ancestry group max calculations.</span>
<span class="sd">&quot;&quot;&quot;</span>

<span class="n">DOWNSAMPLINGS</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;v3&quot;</span><span class="p">:</span> <span class="p">[</span>
        <span class="mi">10</span><span class="p">,</span>
        <span class="mi">20</span><span class="p">,</span>
        <span class="mi">50</span><span class="p">,</span>
        <span class="mi">100</span><span class="p">,</span>
        <span class="mi">200</span><span class="p">,</span>
        <span class="mi">500</span><span class="p">,</span>
        <span class="mi">1000</span><span class="p">,</span>
        <span class="mi">2000</span><span class="p">,</span>
        <span class="mi">5000</span><span class="p">,</span>
        <span class="mi">10000</span><span class="p">,</span>
        <span class="mi">15000</span><span class="p">,</span>
        <span class="mi">20000</span><span class="p">,</span>
        <span class="mi">25000</span><span class="p">,</span>
        <span class="mi">30000</span><span class="p">,</span>
        <span class="mi">40000</span><span class="p">,</span>
        <span class="mi">50000</span><span class="p">,</span>
        <span class="mi">60000</span><span class="p">,</span>
        <span class="mi">70000</span><span class="p">,</span>
        <span class="mi">75000</span><span class="p">,</span>
        <span class="mi">80000</span><span class="p">,</span>
        <span class="mi">85000</span><span class="p">,</span>
        <span class="mi">90000</span><span class="p">,</span>
        <span class="mi">95000</span><span class="p">,</span>
        <span class="mi">100000</span><span class="p">,</span>
        <span class="mi">110000</span><span class="p">,</span>
        <span class="mi">120000</span><span class="p">,</span>
    <span class="p">],</span>
    <span class="s2">&quot;v4&quot;</span><span class="p">:</span> <span class="p">[</span>
        <span class="mi">10</span><span class="p">,</span>
        <span class="mi">100</span><span class="p">,</span>
        <span class="mi">500</span><span class="p">,</span>
        <span class="mi">1000</span><span class="p">,</span>
        <span class="mi">2000</span><span class="p">,</span>
        <span class="mi">5000</span><span class="p">,</span>
        <span class="mi">10000</span><span class="p">,</span>
        <span class="mi">20000</span><span class="p">,</span>
        <span class="mi">50000</span><span class="p">,</span>
        <span class="mi">100000</span><span class="p">,</span>
        <span class="mi">200000</span><span class="p">,</span>
        <span class="mi">500000</span><span class="p">,</span>
    <span class="p">],</span>
<span class="p">}</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">List of the downsampling numbers to use for frequency calculations by version.</span>
<span class="sd">&quot;&quot;&quot;</span>

<span class="n">gnomad_syndip</span> <span class="o">=</span> <span class="n">VersionedMatrixTableResource</span><span class="p">(</span>
    <span class="n">default_version</span><span class="o">=</span><span class="s2">&quot;3.0&quot;</span><span class="p">,</span>
    <span class="n">versions</span><span class="o">=</span><span class="p">{</span>
        <span class="s2">&quot;3.0&quot;</span><span class="p">:</span> <span class="n">GnomadPublicMatrixTableResource</span><span class="p">(</span>
            <span class="n">path</span><span class="o">=</span><span class="s2">&quot;gs://gnomad-public-requester-pays/truth-sets/hail-0.2/gnomad_v3_syndip.b38.mt&quot;</span>
        <span class="p">)</span>
    <span class="p">},</span>
<span class="p">)</span>

<span class="n">na12878</span> <span class="o">=</span> <span class="n">VersionedMatrixTableResource</span><span class="p">(</span>
    <span class="n">default_version</span><span class="o">=</span><span class="s2">&quot;3.0&quot;</span><span class="p">,</span>
    <span class="n">versions</span><span class="o">=</span><span class="p">{</span>
        <span class="s2">&quot;3.0&quot;</span><span class="p">:</span> <span class="n">GnomadPublicMatrixTableResource</span><span class="p">(</span>
            <span class="n">path</span><span class="o">=</span><span class="s2">&quot;gs://gnomad-public-requester-pays/truth-sets/hail-0.2/gnomad_v3_na12878.mt&quot;</span>
        <span class="p">)</span>
    <span class="p">},</span>
<span class="p">)</span>


<span class="k">def</span> <span class="nf">_public_release_ht_path</span><span class="p">(</span><span class="n">data_type</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">version</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Get public release table path.</span>

<span class="sd">    :param data_type: One of &quot;exomes&quot;, &quot;genomes&quot; or &quot;joint&quot;.</span>
<span class="sd">    :param version: One of the release versions of gnomAD on GRCh38</span>
<span class="sd">    :return: Path to release Table</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">version_prefix</span> <span class="o">=</span> <span class="s2">&quot;r&quot;</span> <span class="k">if</span> <span class="n">version</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;3.0&quot;</span><span class="p">)</span> <span class="k">else</span> <span class="s2">&quot;v&quot;</span>
    <span class="k">return</span> <span class="sa">f</span><span class="s2">&quot;gs://gnomad-public-requester-pays/release/</span><span class="si">{</span><span class="n">version</span><span class="si">}</span><span class="s2">/ht/</span><span class="si">{</span><span class="n">data_type</span><span class="si">}</span><span class="s2">/gnomad.</span><span class="si">{</span><span class="n">data_type</span><span class="si">}</span><span class="s2">.</span><span class="si">{</span><span class="n">version_prefix</span><span class="si">}{</span><span class="n">version</span><span class="si">}</span><span class="s2">.sites.ht&quot;</span>


<span class="k">def</span> <span class="nf">_public_coverage_ht_path</span><span class="p">(</span><span class="n">data_type</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">version</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Get public coverage hail table.</span>

<span class="sd">    :param data_type: One of &quot;exomes&quot; or &quot;genomes&quot;</span>
<span class="sd">    :param version: One of the release versions of gnomAD on GRCh38</span>
<span class="sd">    :return: Path to coverage Table</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">version_prefix</span> <span class="o">=</span> <span class="s2">&quot;r&quot;</span> <span class="k">if</span> <span class="n">version</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;3.0&quot;</span><span class="p">)</span> <span class="k">else</span> <span class="s2">&quot;v&quot;</span>
    <span class="k">return</span> <span class="sa">f</span><span class="s2">&quot;gs://gnomad-public-requester-pays/release/</span><span class="si">{</span><span class="n">version</span><span class="si">}</span><span class="s2">/coverage/</span><span class="si">{</span><span class="n">data_type</span><span class="si">}</span><span class="s2">/gnomad.</span><span class="si">{</span><span class="n">data_type</span><span class="si">}</span><span class="s2">.</span><span class="si">{</span><span class="n">version_prefix</span><span class="si">}{</span><span class="n">version</span><span class="si">}</span><span class="s2">.coverage.ht&quot;</span>


<span class="k">def</span> <span class="nf">_public_an_ht_path</span><span class="p">(</span><span class="n">data_type</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">version</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Get public allele number hail table.</span>

<span class="sd">    :param data_type: One of &quot;exomes&quot; or &quot;genomes&quot;</span>
<span class="sd">    :param version: One of the release versions of gnomAD on GRCh38</span>
<span class="sd">    :return: Path to allle number Table</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="sa">f</span><span class="s2">&quot;gs://gnomad-public-requester-pays/release/</span><span class="si">{</span><span class="n">version</span><span class="si">}</span><span class="s2">/ht/</span><span class="si">{</span><span class="n">data_type</span><span class="si">}</span><span class="s2">/gnomad.</span><span class="si">{</span><span class="n">data_type</span><span class="si">}</span><span class="s2">.v</span><span class="si">{</span><span class="n">version</span><span class="si">}</span><span class="s2">.allele_number_all_sites.ht&quot;</span>


<span class="k">def</span> <span class="nf">_public_pext_ht_path</span><span class="p">(</span><span class="n">pext_type</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;base_level&quot;</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Get public proportion expressed across transcripts (pext) data.</span>

<span class="sd">    :param pext_type: One of &quot;base_level&quot; or &quot;annotation_level&quot;. Default is &quot;base_level&quot;.</span>
<span class="sd">    :return: Path to pext Table.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">valid_types</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;base_level&quot;</span><span class="p">,</span> <span class="s2">&quot;annotation_level&quot;</span><span class="p">]</span>

    <span class="k">if</span> <span class="n">pext_type</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">valid_types</span><span class="p">:</span>
        <span class="k">raise</span> <span class="n">DataException</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;Invalid pext_type: &#39;</span><span class="si">{</span><span class="n">pext_type</span><span class="si">}</span><span class="s2">&#39;. Valid options are </span><span class="si">{</span><span class="n">valid_types</span><span class="si">}</span><span class="s2">.&quot;</span>
        <span class="p">)</span>

    <span class="k">return</span> <span class="sa">f</span><span class="s2">&quot;gs://gnomad-public-requester-pays/release/4.1/pext/gnomad.pext.gtex_v10.</span><span class="si">{</span><span class="n">pext_type</span><span class="si">}</span><span class="s2">.ht&quot;</span>


<span class="k">def</span> <span class="nf">_public_constraint_ht_path</span><span class="p">(</span><span class="n">version</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Get public constraint table path.</span>

<span class="sd">    :param version: One of the release versions of gnomAD on GRCh38.</span>
<span class="sd">    :return: Path to gene constraint Table.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="sa">f</span><span class="s2">&quot;gs://gnomad-public-requester-pays/release/</span><span class="si">{</span><span class="n">version</span><span class="si">}</span><span class="s2">/constraint/gnomad.v</span><span class="si">{</span><span class="n">version</span><span class="si">}</span><span class="s2">.constraint_metrics.ht&quot;</span>


<span class="k">def</span> <span class="nf">_public_browser_variant_ht_path</span><span class="p">(</span><span class="n">version</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Get public browser variant table path.</span>

<span class="sd">    :param version: One of the release versions of gnomAD on GRCh38.</span>
<span class="sd">    :return: Path to browser variant Table.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="sa">f</span><span class="s2">&quot;gs://gnomad-public-requester-pays/release/</span><span class="si">{</span><span class="n">version</span><span class="si">}</span><span class="s2">/ht/browser/gnomad.browser.v</span><span class="si">{</span><span class="n">version</span><span class="si">}</span><span class="s2">.sites.ht&quot;</span>


<span class="k">def</span> <span class="nf">_public_browser_gene_ht_path</span><span class="p">()</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Get public browser gene table path.</span>

<span class="sd">    .. note::</span>

<span class="sd">       This table has smaller number of partitions (n=100) for faster computation and</span>
<span class="sd">       contains pext data compared to gnomad.genes.GRCh38.GENCODEv39.ht (which was</span>
<span class="sd">       used by the browser for ES export) under the same path.</span>

<span class="sd">    :return: Path to browser gene Table.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="s2">&quot;gs://gnomad-public-requester-pays/resources/grch38/browser/gnomad.genes.GRCh38.GENCODEv39.pext.ht&quot;</span>


<div class="viewcode-block" id="public_release"><a class="viewcode-back" href="../../../../api_reference/resources/grch38/gnomad.html#gnomad.resources.grch38.gnomad.public_release">[docs]</a><span class="k">def</span> <span class="nf">public_release</span><span class="p">(</span><span class="n">data_type</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">VersionedTableResource</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Retrieve publicly released versioned table resource.</span>

<span class="sd">    :param data_type: One of &quot;exomes&quot;, &quot;genomes&quot; or &quot;joint&quot;.</span>
<span class="sd">    :return: Release Table</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">data_type</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">DATA_TYPES</span><span class="p">:</span>
        <span class="k">raise</span> <span class="n">DataException</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">data_type</span><span class="si">}</span><span class="s2"> not in </span><span class="si">{</span><span class="n">DATA_TYPES</span><span class="si">}</span><span class="s2">, please select a data type from&quot;</span>
            <span class="sa">f</span><span class="s2">&quot; </span><span class="si">{</span><span class="n">DATA_TYPES</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="p">)</span>

    <span class="k">if</span> <span class="n">data_type</span> <span class="o">==</span> <span class="s2">&quot;exomes&quot;</span><span class="p">:</span>
        <span class="n">current_release</span> <span class="o">=</span> <span class="n">CURRENT_EXOME_RELEASE</span>
        <span class="n">releases</span> <span class="o">=</span> <span class="n">EXOME_RELEASES</span>
    <span class="k">elif</span> <span class="n">data_type</span> <span class="o">==</span> <span class="s2">&quot;genomes&quot;</span><span class="p">:</span>
        <span class="n">current_release</span> <span class="o">=</span> <span class="n">CURRENT_GENOME_RELEASE</span>
        <span class="n">releases</span> <span class="o">=</span> <span class="n">GENOME_RELEASES</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">current_release</span> <span class="o">=</span> <span class="n">CURRENT_JOINT_RELEASE</span>
        <span class="n">releases</span> <span class="o">=</span> <span class="n">JOINT_RELEASES</span>

    <span class="k">return</span> <span class="n">VersionedTableResource</span><span class="p">(</span>
        <span class="n">current_release</span><span class="p">,</span>
        <span class="p">{</span>
            <span class="n">release</span><span class="p">:</span> <span class="n">GnomadPublicTableResource</span><span class="p">(</span>
                <span class="n">path</span><span class="o">=</span><span class="n">_public_release_ht_path</span><span class="p">(</span><span class="n">data_type</span><span class="p">,</span> <span class="n">release</span><span class="p">)</span>
            <span class="p">)</span>
            <span class="k">for</span> <span class="n">release</span> <span class="ow">in</span> <span class="n">releases</span>
        <span class="p">},</span>
    <span class="p">)</span></div>


<div class="viewcode-block" id="coverage"><a class="viewcode-back" href="../../../../api_reference/resources/grch38/gnomad.html#gnomad.resources.grch38.gnomad.coverage">[docs]</a><span class="k">def</span> <span class="nf">coverage</span><span class="p">(</span><span class="n">data_type</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">VersionedTableResource</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Retrieve gnomAD&#39;s coverage table by data_type.</span>

<span class="sd">    :param data_type: One of &quot;exomes&quot; or &quot;genomes&quot;</span>
<span class="sd">    :return: Coverage Table</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">data_type</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">DATA_TYPES</span><span class="p">:</span>
        <span class="k">raise</span> <span class="n">DataException</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">data_type</span><span class="si">}</span><span class="s2"> not in </span><span class="si">{</span><span class="n">DATA_TYPES</span><span class="si">}</span><span class="s2">, please select a data type from&quot;</span>
            <span class="sa">f</span><span class="s2">&quot; </span><span class="si">{</span><span class="n">DATA_TYPES</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="p">)</span>

    <span class="k">if</span> <span class="n">data_type</span> <span class="o">==</span> <span class="s2">&quot;exomes&quot;</span><span class="p">:</span>
        <span class="n">current_release</span> <span class="o">=</span> <span class="n">CURRENT_EXOME_COVERAGE_RELEASE</span>
        <span class="n">releases</span> <span class="o">=</span> <span class="n">EXOME_COVERAGE_RELEASES</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">current_release</span> <span class="o">=</span> <span class="n">CURRENT_GENOME_COVERAGE_RELEASE</span>
        <span class="n">releases</span> <span class="o">=</span> <span class="n">GENOME_COVERAGE_RELEASES</span>

    <span class="k">return</span> <span class="n">VersionedTableResource</span><span class="p">(</span>
        <span class="n">current_release</span><span class="p">,</span>
        <span class="p">{</span>
            <span class="n">release</span><span class="p">:</span> <span class="n">GnomadPublicTableResource</span><span class="p">(</span>
                <span class="n">path</span><span class="o">=</span><span class="n">_public_coverage_ht_path</span><span class="p">(</span><span class="n">data_type</span><span class="p">,</span> <span class="n">release</span><span class="p">)</span>
            <span class="p">)</span>
            <span class="k">for</span> <span class="n">release</span> <span class="ow">in</span> <span class="n">releases</span>
        <span class="p">},</span>
    <span class="p">)</span></div>


<div class="viewcode-block" id="all_sites_an"><a class="viewcode-back" href="../../../../api_reference/resources/grch38/gnomad.html#gnomad.resources.grch38.gnomad.all_sites_an">[docs]</a><span class="k">def</span> <span class="nf">all_sites_an</span><span class="p">(</span><span class="n">data_type</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">VersionedTableResource</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Retrieve gnomAD&#39;s all sites allele number table by data_type.</span>

<span class="sd">    :param data_type: One of &quot;exomes&quot; or &quot;genomes&quot;</span>
<span class="sd">    :return: All sites allele number VersionedTableResource</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">data_type</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">DATA_TYPES</span><span class="p">:</span>
        <span class="k">raise</span> <span class="n">DataException</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">data_type</span><span class="si">}</span><span class="s2"> not in </span><span class="si">{</span><span class="n">DATA_TYPES</span><span class="si">}</span><span class="s2">, please select a data type from&quot;</span>
            <span class="sa">f</span><span class="s2">&quot; </span><span class="si">{</span><span class="n">DATA_TYPES</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="p">)</span>

    <span class="k">if</span> <span class="n">data_type</span> <span class="o">==</span> <span class="s2">&quot;exomes&quot;</span><span class="p">:</span>
        <span class="n">current_release</span> <span class="o">=</span> <span class="n">CURRENT_EXOME_AN_RELEASE</span>
        <span class="n">releases</span> <span class="o">=</span> <span class="n">EXOME_AN_RELEASES</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">current_release</span> <span class="o">=</span> <span class="n">CURRENT_GENOME_AN_RELEASE</span>
        <span class="n">releases</span> <span class="o">=</span> <span class="n">GENOME_AN_RELEASES</span>

    <span class="k">return</span> <span class="n">VersionedTableResource</span><span class="p">(</span>
        <span class="n">current_release</span><span class="p">,</span>
        <span class="p">{</span>
            <span class="n">release</span><span class="p">:</span> <span class="n">GnomadPublicTableResource</span><span class="p">(</span>
                <span class="n">path</span><span class="o">=</span><span class="n">_public_an_ht_path</span><span class="p">(</span><span class="n">data_type</span><span class="p">,</span> <span class="n">release</span><span class="p">)</span>
            <span class="p">)</span>
            <span class="k">for</span> <span class="n">release</span> <span class="ow">in</span> <span class="n">releases</span>
        <span class="p">},</span>
    <span class="p">)</span></div>


<div class="viewcode-block" id="coverage_tsv_path"><a class="viewcode-back" href="../../../../api_reference/resources/grch38/gnomad.html#gnomad.resources.grch38.gnomad.coverage_tsv_path">[docs]</a><span class="k">def</span> <span class="nf">coverage_tsv_path</span><span class="p">(</span><span class="n">data_type</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">version</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Retrieve gnomAD&#39;s coverage table by data_type.</span>

<span class="sd">    :param data_type: One of &quot;exomes&quot; or &quot;genomes&quot;</span>
<span class="sd">    :return: Coverage Table</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">data_type</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">DATA_TYPES</span><span class="p">:</span>
        <span class="k">raise</span> <span class="n">DataException</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">data_type</span><span class="si">}</span><span class="s2"> not in </span><span class="si">{</span><span class="n">DATA_TYPES</span><span class="si">}</span><span class="s2">, please select a data type from&quot;</span>
            <span class="sa">f</span><span class="s2">&quot; </span><span class="si">{</span><span class="n">DATA_TYPES</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="p">)</span>

    <span class="k">if</span> <span class="n">data_type</span> <span class="o">==</span> <span class="s2">&quot;exomes&quot;</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">version</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">version</span> <span class="o">=</span> <span class="n">CURRENT_EXOME_COVERAGE_RELEASE</span>
        <span class="k">elif</span> <span class="n">version</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">EXOME_COVERAGE_RELEASES</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">DataException</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;Version </span><span class="si">{</span><span class="n">version</span><span class="si">}</span><span class="s2"> of gnomAD exomes for GRCh38 does not exist&quot;</span>
            <span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">version</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">version</span> <span class="o">=</span> <span class="n">CURRENT_GENOME_COVERAGE_RELEASE</span>
        <span class="k">elif</span> <span class="n">version</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">GENOME_COVERAGE_RELEASES</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">DataException</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;Version </span><span class="si">{</span><span class="n">version</span><span class="si">}</span><span class="s2"> of gnomAD genomes for GRCh38 does not exist&quot;</span>
            <span class="p">)</span>

    <span class="n">version_prefix</span> <span class="o">=</span> <span class="s2">&quot;r&quot;</span> <span class="k">if</span> <span class="n">version</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;3.0&quot;</span><span class="p">)</span> <span class="k">else</span> <span class="s2">&quot;v&quot;</span>
    <span class="k">return</span> <span class="sa">f</span><span class="s2">&quot;gs://gcp-public-data--gnomad/release/</span><span class="si">{</span><span class="n">version</span><span class="si">}</span><span class="s2">/coverage/</span><span class="si">{</span><span class="n">data_type</span><span class="si">}</span><span class="s2">/gnomad.</span><span class="si">{</span><span class="n">data_type</span><span class="si">}</span><span class="s2">.</span><span class="si">{</span><span class="n">version_prefix</span><span class="si">}{</span><span class="n">version</span><span class="si">}</span><span class="s2">.coverage.summary.tsv.bgz&quot;</span></div>


<div class="viewcode-block" id="release_vcf_path"><a class="viewcode-back" href="../../../../api_reference/resources/grch38/gnomad.html#gnomad.resources.grch38.gnomad.release_vcf_path">[docs]</a><span class="k">def</span> <span class="nf">release_vcf_path</span><span class="p">(</span><span class="n">data_type</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">version</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">contig</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Publically released VCF. Provide specific contig, i.e. &quot;chr20&quot;, to retrieve contig specific VCF.</span>

<span class="sd">    :param data_type: One of &quot;exomes&quot; or &quot;genomes&quot;</span>
<span class="sd">    :param version: One of the release versions of gnomAD on GRCh37</span>
<span class="sd">    :param contig: Single contig &quot;chr1&quot; to &quot;chrY&quot;</span>
<span class="sd">    :return: Path to VCF</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">version</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;2&quot;</span><span class="p">):</span>
        <span class="k">raise</span> <span class="n">DataException</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;gnomAD version </span><span class="si">{</span><span class="n">version</span><span class="si">}</span><span class="s2"> is not available on reference genome GRCh38&quot;</span>
        <span class="p">)</span>

    <span class="n">contig</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;.</span><span class="si">{</span><span class="n">contig</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">if</span> <span class="n">contig</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span>
    <span class="n">version_prefix</span> <span class="o">=</span> <span class="s2">&quot;r&quot;</span> <span class="k">if</span> <span class="n">version</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;3.0&quot;</span><span class="p">)</span> <span class="k">else</span> <span class="s2">&quot;v&quot;</span>
    <span class="k">return</span> <span class="sa">f</span><span class="s2">&quot;gs://gcp-public-data--gnomad/release/</span><span class="si">{</span><span class="n">version</span><span class="si">}</span><span class="s2">/vcf/</span><span class="si">{</span><span class="n">data_type</span><span class="si">}</span><span class="s2">/gnomad.</span><span class="si">{</span><span class="n">data_type</span><span class="si">}</span><span class="s2">.</span><span class="si">{</span><span class="n">version_prefix</span><span class="si">}{</span><span class="n">version</span><span class="si">}</span><span class="s2">.sites</span><span class="si">{</span><span class="n">contig</span><span class="si">}</span><span class="s2">.vcf.bgz&quot;</span></div>


<div class="viewcode-block" id="add_grpMaxFAF95_v4"><a class="viewcode-back" href="../../../../api_reference/resources/grch38/gnomad.html#gnomad.resources.grch38.gnomad.add_grpMaxFAF95_v4">[docs]</a><span class="k">def</span> <span class="nf">add_grpMaxFAF95_v4</span><span class="p">(</span><span class="n">ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Add a grpMaxFAF95 struct with &#39;grpmax&#39; and &#39;grpmax_gen_anc&#39;.</span>

<span class="sd">    Also includes a jointGrpMaxFAF95 annotation using the v4 fafmax and joint_fafmax structures.</span>

<span class="sd">    :param ht: Input hail table.</span>
<span class="sd">    :return: Annotated hail table.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="s2">&quot;gnomad&quot;</span> <span class="ow">in</span> <span class="n">ht</span><span class="o">.</span><span class="n">fafmax</span><span class="p">:</span>
        <span class="n">fafmax_field</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">fafmax</span><span class="o">.</span><span class="n">gnomad</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">fafmax_field</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">fafmax</span>
    <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span>
        <span class="n">grpMaxFAF95</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span>
            <span class="n">grpmax</span><span class="o">=</span><span class="n">fafmax_field</span><span class="o">.</span><span class="n">faf95_max</span><span class="p">,</span>
            <span class="n">grpmax_gen_anc</span><span class="o">=</span><span class="n">fafmax_field</span><span class="o">.</span><span class="n">faf95_max_gen_anc</span><span class="p">,</span>
        <span class="p">),</span>
        <span class="n">jointGrpMaxFAF95</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span>
            <span class="n">grpmax</span><span class="o">=</span><span class="n">ht</span><span class="o">.</span><span class="n">joint_fafmax</span><span class="o">.</span><span class="n">faf95_max</span><span class="p">,</span>
            <span class="n">grpmax_gen_anc</span><span class="o">=</span><span class="n">ht</span><span class="o">.</span><span class="n">joint_fafmax</span><span class="o">.</span><span class="n">faf95_max_gen_anc</span><span class="p">,</span>
        <span class="p">),</span>
    <span class="p">)</span>
    <span class="k">return</span> <span class="n">ht</span></div>


<div class="viewcode-block" id="gnomad_gks"><a class="viewcode-back" href="../../../../api_reference/resources/grch38/gnomad.html#gnomad.resources.grch38.gnomad.gnomad_gks">[docs]</a><span class="k">def</span> <span class="nf">gnomad_gks</span><span class="p">(</span>
    <span class="n">locus_interval</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">IntervalExpression</span><span class="p">,</span>
    <span class="n">version</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">data_type</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;genomes&quot;</span><span class="p">,</span>
    <span class="n">by_gen_anc_group</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="n">by_sex</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="n">vrs_only</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="n">custom_ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">skip_checkpoint</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="n">skip_coverage</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="n">custom_coverage_ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Perform gnomad GKS annotations on a range of variants at once.</span>

<span class="sd">    :param locus_interval: Hail IntervalExpression of locus&lt;reference_genome&gt;.</span>
<span class="sd">        e.g. hl.locus_interval(&#39;chr1&#39;, 6424776, 6461367, reference_genome=&quot;GRCh38&quot;)</span>
<span class="sd">    :param version: String of version of gnomAD release to use.</span>
<span class="sd">    :param data_type: String of either &quot;exomes&quot; or &quot;genomes&quot; for the type of reads that are desired.</span>
<span class="sd">    :param by_gen_anc_group: Boolean to pass to obtain frequency information for each cohort.</span>
<span class="sd">    :param by_sex: Boolean to pass to return frequency information for each cohort split by chromosomal sex.</span>
<span class="sd">    :param vrs_only: Boolean to pass for only VRS info to be returned</span>
<span class="sd">        (will not include allele frequency information).</span>
<span class="sd">    :param custom_ht: Table to use instead of what public_release() method would return for the version.</span>
<span class="sd">    :param skip_checkpoint: Bool to pass to skip checkpointing selected fields</span>
<span class="sd">        (checkpointing may be desirable for large datasets by reducing data copies across the cluster).</span>
<span class="sd">    :param skip_coverage: Bool to pass to skip adding coverage statistics.</span>
<span class="sd">    :param custom_coverage_ht: Custom table to use for coverage statistics instead of the release coverage table.</span>
<span class="sd">    :return: List of dictionaries containing VRS information</span>
<span class="sd">        (and freq info split by ancestry groups and sex if desired) for specified variant.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Obtain the high level version number and verify that it is 4.</span>
    <span class="n">high_level_version</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;v</span><span class="si">{</span><span class="n">version</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span>
    <span class="k">if</span> <span class="n">high_level_version</span> <span class="o">!=</span> <span class="s2">&quot;v4&quot;</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span>
            <span class="s2">&quot;gnomad_gks() is currently only implemented for gnomAD v4.&quot;</span>
        <span class="p">)</span>

    <span class="c1"># Read public_release table if no custom table provided.</span>
    <span class="k">if</span> <span class="n">custom_ht</span><span class="p">:</span>
        <span class="n">ht</span> <span class="o">=</span> <span class="n">custom_ht</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">ht</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">read_table</span><span class="p">(</span><span class="n">public_release</span><span class="p">(</span><span class="n">data_type</span><span class="p">)</span><span class="o">.</span><span class="n">versions</span><span class="p">[</span><span class="n">version</span><span class="p">]</span><span class="o">.</span><span class="n">path</span><span class="p">)</span>

    <span class="c1"># Read coverage statistics if requested.</span>
    <span class="n">coverage_version_3_0_1</span> <span class="o">=</span> <span class="s2">&quot;3.0.1&quot;</span>  <span class="c1"># v4 genomes coverage</span>
    <span class="n">coverage_version_4_0</span> <span class="o">=</span> <span class="s2">&quot;4.0&quot;</span>  <span class="c1"># v4 exomes coverage</span>

    <span class="c1"># In v4, exomes have coverage in v4 coverage table,</span>
    <span class="c1">#   genomes have coverage in v3 coverage table.</span>
    <span class="k">if</span> <span class="n">data_type</span> <span class="o">==</span> <span class="s2">&quot;genomes&quot;</span><span class="p">:</span>
        <span class="n">coverage_version</span> <span class="o">=</span> <span class="n">coverage_version_3_0_1</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">coverage_version</span> <span class="o">=</span> <span class="n">coverage_version_4_0</span>

    <span class="n">coverage_ht</span> <span class="o">=</span> <span class="kc">None</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">skip_coverage</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">custom_coverage_ht</span><span class="p">:</span>
            <span class="n">coverage_ht</span> <span class="o">=</span> <span class="n">custom_coverage_ht</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">coverage_ht</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">read_table</span><span class="p">(</span>
                <span class="n">coverage</span><span class="p">(</span><span class="n">data_type</span><span class="p">)</span><span class="o">.</span><span class="n">versions</span><span class="p">[</span><span class="n">coverage_version</span><span class="p">]</span><span class="o">.</span><span class="n">path</span>
            <span class="p">)</span>
        <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span><span class="n">mean_depth</span><span class="o">=</span><span class="n">coverage_ht</span><span class="p">[</span><span class="n">ht</span><span class="o">.</span><span class="n">locus</span><span class="p">]</span><span class="o">.</span><span class="n">mean</span><span class="p">)</span>
        <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span><span class="n">fraction_cov_over_20</span><span class="o">=</span><span class="n">coverage_ht</span><span class="p">[</span><span class="n">ht</span><span class="o">.</span><span class="n">locus</span><span class="p">]</span><span class="o">.</span><span class="n">over_20</span><span class="p">)</span>

    <span class="c1"># Retrieve genetic ancestry groups from the imported GEN_ANC_NAMES dictionary.</span>
    <span class="n">grps_list</span> <span class="o">=</span> <span class="p">(</span>
        <span class="nb">list</span><span class="p">(</span><span class="n">GEN_ANC_NAMES</span><span class="p">[</span><span class="n">high_level_version</span><span class="p">][</span><span class="n">data_type</span><span class="p">])</span> <span class="k">if</span> <span class="n">by_gen_anc_group</span> <span class="k">else</span> <span class="kc">None</span>
    <span class="p">)</span>

    <span class="c1"># Throw warnings if contradictory arguments are passed.</span>
    <span class="k">if</span> <span class="n">by_gen_anc_group</span> <span class="ow">and</span> <span class="n">vrs_only</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
            <span class="s2">&quot;Both &#39;vrs_only&#39; and &#39;by_gen_anc_groups&#39; have been specified. Ignoring&quot;</span>
            <span class="s2">&quot; &#39;by_gen_anc_groups&#39; list and returning only VRS information.&quot;</span>
        <span class="p">)</span>
    <span class="k">elif</span> <span class="n">by_sex</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">by_gen_anc_group</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
            <span class="s2">&quot;Splitting whole database by sex is not yet supported. If using &#39;by_sex&#39;,&quot;</span>
            <span class="s2">&quot; please also specify &#39;by_ancestry_group&#39; to stratify by.&quot;</span>
        <span class="p">)</span>

    <span class="c1"># Select relevant fields, checkpoint, and filter to interval before adding</span>
    <span class="c1"># annotations.</span>

    <span class="c1"># Pull up LCR flag and make referrable in the same field.</span>
    <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span><span class="n">lcr</span><span class="o">=</span><span class="n">ht</span><span class="o">.</span><span class="n">region_flags</span><span class="o">.</span><span class="n">lcr</span><span class="p">)</span>

    <span class="c1"># Pull up allele balance histogram arrays.</span>
    <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span><span class="n">ab_hist_alt</span><span class="o">=</span><span class="n">ht</span><span class="o">.</span><span class="n">histograms</span><span class="o">.</span><span class="n">qual_hists</span><span class="o">.</span><span class="n">ab_hist_alt</span><span class="p">)</span>

    <span class="n">ht</span> <span class="o">=</span> <span class="n">add_grpMaxFAF95_v4</span><span class="p">(</span><span class="n">ht</span><span class="p">)</span>

    <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span><span class="n">in_autosome_or_par</span><span class="o">=</span><span class="n">ht</span><span class="o">.</span><span class="n">locus</span><span class="o">.</span><span class="n">in_autosome_or_par</span><span class="p">())</span>

    <span class="n">keep_fields</span> <span class="o">=</span> <span class="p">[</span>
        <span class="n">ht</span><span class="o">.</span><span class="n">freq</span><span class="p">,</span>
        <span class="n">ht</span><span class="o">.</span><span class="n">info</span><span class="o">.</span><span class="n">vrs</span><span class="p">,</span>
        <span class="n">ht</span><span class="o">.</span><span class="n">info</span><span class="o">.</span><span class="n">monoallelic</span><span class="p">,</span>
        <span class="n">ht</span><span class="o">.</span><span class="n">grpMaxFAF95</span><span class="p">,</span>
        <span class="n">ht</span><span class="o">.</span><span class="n">filters</span><span class="p">,</span>
        <span class="n">ht</span><span class="o">.</span><span class="n">lcr</span><span class="p">,</span>
        <span class="n">ht</span><span class="o">.</span><span class="n">ab_hist_alt</span><span class="p">,</span>
        <span class="n">ht</span><span class="o">.</span><span class="n">in_autosome_or_par</span><span class="p">,</span>
    <span class="p">]</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">skip_coverage</span><span class="p">:</span>
        <span class="n">keep_fields</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">ht</span><span class="o">.</span><span class="n">mean_depth</span><span class="p">)</span>
        <span class="n">keep_fields</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">ht</span><span class="o">.</span><span class="n">fraction_cov_over_20</span><span class="p">)</span>

    <span class="k">if</span> <span class="s2">&quot;jointGrpMaxFAF95&quot;</span> <span class="ow">in</span> <span class="n">ht</span><span class="o">.</span><span class="n">row</span><span class="p">:</span>
        <span class="n">keep_fields</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">ht</span><span class="o">.</span><span class="n">jointGrpMaxFAF95</span><span class="p">)</span>

    <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">select</span><span class="p">(</span><span class="o">*</span><span class="n">keep_fields</span><span class="p">)</span>

    <span class="c1"># Checkpoint narrower set of columns if not skipped.</span>
    <span class="n">ht</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">filter_intervals</span><span class="p">(</span><span class="n">ht</span><span class="p">,</span> <span class="p">[</span><span class="n">locus_interval</span><span class="p">])</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">skip_checkpoint</span><span class="p">:</span>
        <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">checkpoint</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">new_temp_file</span><span class="p">(</span><span class="s2">&quot;vrs_checkpoint&quot;</span><span class="p">,</span> <span class="n">extension</span><span class="o">=</span><span class="s2">&quot;ht&quot;</span><span class="p">))</span>

    <span class="c1"># Collect all variants as structs, so all dictionary construction can be</span>
    <span class="c1"># done in native Python.</span>
    <span class="n">variant_list</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">collect</span><span class="p">()</span>
    <span class="n">ht_freq_index_dict</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">freq_index_dict</span><span class="o">.</span><span class="n">collect</span><span class="p">()[</span><span class="mi">0</span><span class="p">]</span>
    <span class="c1"># gnomad v4 renamed freq_index_dict keys to use underscores instead of dashes.</span>
    <span class="c1"># Use underscores for v3 as well.</span>
    <span class="n">ht_freq_index_dict</span> <span class="o">=</span> <span class="p">{</span><span class="n">k</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;-&quot;</span><span class="p">,</span> <span class="s2">&quot;_&quot;</span><span class="p">):</span> <span class="n">v</span> <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">ht_freq_index_dict</span><span class="o">.</span><span class="n">items</span><span class="p">()}</span>

    <span class="c1"># Assemble output dicts with VRS and optionally frequency, append to list,</span>
    <span class="c1"># then return list</span>
    <span class="n">outputs</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">variant</span> <span class="ow">in</span> <span class="n">variant_list</span><span class="p">:</span>
        <span class="n">vrs_variant</span> <span class="o">=</span> <span class="n">add_gks_vrs</span><span class="p">(</span><span class="n">variant</span><span class="o">.</span><span class="n">locus</span><span class="p">,</span> <span class="n">variant</span><span class="o">.</span><span class="n">vrs</span><span class="p">)</span>

        <span class="n">out</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;locus&quot;</span><span class="p">:</span> <span class="p">{</span>
                <span class="s2">&quot;contig&quot;</span><span class="p">:</span> <span class="n">variant</span><span class="o">.</span><span class="n">locus</span><span class="o">.</span><span class="n">contig</span><span class="p">,</span>
                <span class="s2">&quot;position&quot;</span><span class="p">:</span> <span class="n">variant</span><span class="o">.</span><span class="n">locus</span><span class="o">.</span><span class="n">position</span><span class="p">,</span>
                <span class="s2">&quot;reference_genome&quot;</span><span class="p">:</span> <span class="n">variant</span><span class="o">.</span><span class="n">locus</span><span class="o">.</span><span class="n">reference_genome</span><span class="o">.</span><span class="n">name</span><span class="p">,</span>
            <span class="p">},</span>
            <span class="s2">&quot;alleles&quot;</span><span class="p">:</span> <span class="n">variant</span><span class="o">.</span><span class="n">alleles</span><span class="p">,</span>
            <span class="s2">&quot;gks_vrs_variant&quot;</span><span class="p">:</span> <span class="n">vrs_variant</span><span class="p">,</span>
        <span class="p">}</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">vrs_only</span><span class="p">:</span>
            <span class="n">va_freq_dict</span> <span class="o">=</span> <span class="n">add_gks_va</span><span class="p">(</span>
                <span class="n">input_struct</span><span class="o">=</span><span class="n">variant</span><span class="p">,</span>
                <span class="n">label_name</span><span class="o">=</span><span class="s2">&quot;gnomAD&quot;</span><span class="p">,</span>
                <span class="n">label_version</span><span class="o">=</span><span class="n">version</span><span class="p">,</span>
                <span class="n">gen_anc_groups</span><span class="o">=</span><span class="n">grps_list</span><span class="p">,</span>
                <span class="n">gen_anc_groups_dict</span><span class="o">=</span><span class="n">GEN_ANC_NAMES</span><span class="p">,</span>
                <span class="n">by_sex</span><span class="o">=</span><span class="n">by_sex</span><span class="p">,</span>
                <span class="n">freq_index_dict</span><span class="o">=</span><span class="n">ht_freq_index_dict</span><span class="p">,</span>
            <span class="p">)</span>

            <span class="c1"># Assign existing VRS information to &quot;focusAllele&quot; key</span>
            <span class="n">va_freq_dict</span><span class="p">[</span><span class="s2">&quot;focusAllele&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">vrs_variant</span>
            <span class="n">out</span><span class="p">[</span><span class="s2">&quot;gks_va_freq&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">va_freq_dict</span>

        <span class="c1"># Append variant dictionary to list of outputs</span>
        <span class="n">outputs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">out</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">outputs</span></div>


<div class="viewcode-block" id="pext"><a class="viewcode-back" href="../../../../api_reference/resources/grch38/gnomad.html#gnomad.resources.grch38.gnomad.pext">[docs]</a><span class="k">def</span> <span class="nf">pext</span><span class="p">(</span><span class="n">pext_type</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;base_level&quot;</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">GnomadPublicTableResource</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Retrieve pext table by type.</span>

<span class="sd">    :param pext_type: One of &quot;base_level&quot; or &quot;annotation_level&quot;. Default is &quot;base_level&quot;.</span>
<span class="sd">    :return: Pext Table.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">GnomadPublicTableResource</span><span class="p">(</span><span class="n">path</span><span class="o">=</span><span class="n">_public_pext_ht_path</span><span class="p">(</span><span class="n">pext_type</span><span class="p">))</span></div>


<div class="viewcode-block" id="constraint"><a class="viewcode-back" href="../../../../api_reference/resources/grch38/gnomad.html#gnomad.resources.grch38.gnomad.constraint">[docs]</a><span class="k">def</span> <span class="nf">constraint</span><span class="p">()</span> <span class="o">-&gt;</span> <span class="n">VersionedTableResource</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Retrieve gene constraint Table.</span>

<span class="sd">    :return: Gene constraint Table.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">VersionedTableResource</span><span class="p">(</span>
        <span class="n">CURRENT_EXOME_RELEASE</span><span class="p">,</span>
        <span class="p">{</span>
            <span class="n">release</span><span class="p">:</span> <span class="n">GnomadPublicTableResource</span><span class="p">(</span><span class="n">path</span><span class="o">=</span><span class="n">_public_constraint_ht_path</span><span class="p">(</span><span class="n">release</span><span class="p">))</span>
            <span class="k">for</span> <span class="n">release</span> <span class="ow">in</span> <span class="n">EXOME_RELEASES</span>
        <span class="p">},</span>
    <span class="p">)</span></div>


<div class="viewcode-block" id="browser_variant"><a class="viewcode-back" href="../../../../api_reference/resources/grch38/gnomad.html#gnomad.resources.grch38.gnomad.browser_variant">[docs]</a><span class="k">def</span> <span class="nf">browser_variant</span><span class="p">()</span> <span class="o">-&gt;</span> <span class="n">VersionedTableResource</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Retrieve browser variant table.</span>

<span class="sd">    :return: Browser variant Table.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">VersionedTableResource</span><span class="p">(</span>
        <span class="n">CURRENT_BROWSER_RELEASE</span><span class="p">,</span>
        <span class="p">{</span>
            <span class="n">release</span><span class="p">:</span> <span class="n">GnomadPublicTableResource</span><span class="p">(</span>
                <span class="n">path</span><span class="o">=</span><span class="n">_public_browser_variant_ht_path</span><span class="p">(</span><span class="n">release</span><span class="p">)</span>
            <span class="p">)</span>
            <span class="k">for</span> <span class="n">release</span> <span class="ow">in</span> <span class="n">BROWSER_RELEASES</span>
        <span class="p">},</span>
    <span class="p">)</span></div>


<div class="viewcode-block" id="browser_gene"><a class="viewcode-back" href="../../../../api_reference/resources/grch38/gnomad.html#gnomad.resources.grch38.gnomad.browser_gene">[docs]</a><span class="k">def</span> <span class="nf">browser_gene</span><span class="p">()</span> <span class="o">-&gt;</span> <span class="n">GnomadPublicTableResource</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Retrieve browser gene table.</span>

<span class="sd">    :return: Browser gene Table.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">GnomadPublicTableResource</span><span class="p">(</span><span class="n">path</span><span class="o">=</span><span class="n">_public_browser_gene_ht_path</span><span class="p">())</span></div>
</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p></p>
  </div>

   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>